{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "053c54f2-b3b1-4d6a-82d7-49228f368010",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run w aln output in both modes\n",
    "import sys\n",
    "sys.path.append('../src/')\n",
    "\n",
    "#matcAFDB_tools up positions\n",
    "\n",
    "\n",
    "#calc fident with cross alphabet comparison\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4e2ee53e-d0d7-4150-84e5-080d1ef7f5fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess ,shlex\n",
    "\n",
    "\n",
    "def runargs(args):\n",
    "    '''run a command line command\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    args : str\n",
    "        command line command\n",
    "    '''\n",
    "\n",
    "    args = shlex.split( args)\n",
    "    p = subprocess.run( args )\n",
    "    return p\n",
    "\n",
    "def runFoldseek_allvall( structfolder , outfolder , foldseekpath = '../foldseek/foldseek' , mode = 2):\n",
    "    '''\n",
    "    run foldseek search and createtsv\n",
    "\n",
    "    parameters\n",
    "    ----------\n",
    "    dbpath : str\n",
    "        path to foldseek database\n",
    "    outfolder : str \n",
    "        path to output folder\n",
    "    maxseqs : int   \n",
    "        maximum number of sequences to compare to\n",
    "\n",
    "    '''\n",
    "\n",
    "    args = foldseekpath + \" easy-search \" + structfolder + \" \"+ structfolder +\" \"+ outfolder+\"aln_\"+str(mode)+\"_allvall.csv tmp\" + \" --format-output 'query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,lddt,qaln,taln,cigar,alntmscore' --exhaustive-search --alignment-type \"+ str(mode) \n",
    "    p = runargs(args)\n",
    "    return outfolder+\"aln_\"+str(mode)+\"_allvall.csv\"\n",
    "\n",
    "def read_dbfiles3di(folder, name = 'outdb'):\n",
    "    #find positions \n",
    "    threeDidb = folder+ name +'_ss'\n",
    "    threeDiseq = [ l.strip().replace('\\x00','') for l in open(threeDidb)]\n",
    "    \n",
    "    \n",
    "    lookup = folder+name+'.lookup'\n",
    "    ids = [ l.split()[1].strip() for l in open(lookup)]\n",
    "    \n",
    "    AADB = folder+name\n",
    "    AAs = [ l.strip().replace('\\x00','') for l in open(AADB)]\n",
    "\n",
    "    mapper3di = dict(zip(ids,threeDiseq))\n",
    "    mapperAA = dict(zip(ids,AAs))\n",
    "    \n",
    "    return mapper3di, mapperAA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "2e4a4f0a-91c7-4264-b2e2-80ae82894084",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapper3di, mapperAA = read_dbfiles3di('../testdata/')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "6156c98f-1d30-4f39-924d-82475162ea03",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../testdata/structsaln_0_allvall.csv exists and will be overwritten\n",
      "easy-search ../testdata/structs ../testdata/structs ../testdata/structsaln_0_allvall.csv tmp --format-output query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,lddt,qaln,taln,cigar,alntmscore --exhaustive-search --alignment-type 0 \n",
      "\n",
      "MMseqs Version:              \t30fdcac78217579fa25d59bc271bd4f3767d3ebb\n",
      "Seq. id. threshold           \t0\n",
      "Coverage threshold           \t0\n",
      "Coverage mode                \t0\n",
      "Max reject                   \t2147483647\n",
      "Max accept                   \t2147483647\n",
      "Add backtrace                \tfalse\n",
      "TMscore threshold            \t0\n",
      "TMalign hit order            \t0\n",
      "TMalign fast                 \t1\n",
      "Preload mode                 \t0\n",
      "Threads                      \t48\n",
      "Verbosity                    \t3\n",
      "LDDT threshold               \t0\n",
      "Sort by structure bit score  \t1\n",
      "Substitution matrix          \taa:3di.out,nucl:3di.out\n",
      "Alignment mode               \t3\n",
      "Alignment mode               \t0\n",
      "E-value threshold            \t10\n",
      "Min alignment length         \t0\n",
      "Seq. id. mode                \t0\n",
      "Alternative alignments       \t0\n",
      "Max sequence length          \t65535\n",
      "Compositional bias           \t1\n",
      "Compositional bias           \t1\n",
      "Gap open cost                \taa:10,nucl:10\n",
      "Gap extension cost           \taa:1,nucl:1\n",
      "Compressed                   \t0\n",
      "Seed substitution matrix     \taa:3di.out,nucl:3di.out\n",
      "Sensitivity                  \t9.5\n",
      "k-mer length                 \t6\n",
      "k-score                      \tseq:2147483647,prof:2147483647\n",
      "Max results per query        \t1000\n",
      "Split database               \t0\n",
      "Split mode                   \t2\n",
      "Split memory limit           \t0\n",
      "Diagonal scoring             \ttrue\n",
      "Exact k-mer matching         \t0\n",
      "Mask residues                \t0\n",
      "Mask residues probability    \t0.99995\n",
      "Mask lower case residues     \t1\n",
      "Minimum diagonal score       \t15\n",
      "Selected taxa                \t\n",
      "Spaced k-mers                \t1\n",
      "Spaced k-mer pattern         \t\n",
      "Local temporary path         \t\n",
      "Exhaustive search mode       \ttrue\n",
      "Search iterations            \t1\n",
      "Alignment type               \t0\n",
      "Remove temporary files       \ttrue\n",
      "MPI runner                   \t\n",
      "Force restart with latest tmp\tfalse\n",
      "Chain name mode              \t0\n",
      "Mask b-factor threshold      \t0\n",
      "Coord store mode             \t2\n",
      "Write lookup file            \t1\n",
      "Tar Inclusion Regex          \t.*\n",
      "Tar Exclusion Regex          \t^$\n",
      "Alignment format             \t0\n",
      "Format alignment output      \tquery,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,lddt,qaln,taln,cigar,alntmscore\n",
      "Database output              \tfalse\n",
      "\n",
      "Alignment backtraces will be computed, since they were requested by output format.\n",
      "createdb ../testdata/structs tmp/9833431157172569924/query --chain-name-mode 0 --mask-bfactor-threshold 0 --coord-store-mode 2 --write-lookup 1 --tar-include '.*' --tar-exclude '^$' --threads 48 -v 3 \n",
      "\n",
      "Output file: tmp/9833431157172569924/query\n",
      "[=================================================================] 23 0s 3ms\n",
      "Time for merging to query_ss: 0h 0m 0s 47ms\n",
      "Time for merging to query_h: 0h 0m 0s 9ms\n",
      "Time for merging to query_ca: 0h 0m 0s 11ms\n",
      "Time for merging to query: 0h 0m 0s 11ms\n",
      "Ignore 0 out of 23.\n",
      "Too short: 0, incorrect  0.\n",
      "Time for processing: 0h 0m 0s 287ms\n",
      "createdb ../testdata/structs tmp/9833431157172569924/target --chain-name-mode 0 --mask-bfactor-threshold 0 --coord-store-mode 2 --write-lookup 1 --tar-include '.*' --tar-exclude '^$' --threads 48 -v 3 \n",
      "\n",
      "Output file: tmp/9833431157172569924/target\n",
      "[=================================================================] 23 0s 84ms\n",
      "Time for merging to target_ss: 0h 0m 0s 7ms\n",
      "Time for merging to target_h: 0h 0m 0s 5ms\n",
      "Time for merging to target_ca: 0h 0m 0s 6ms\n",
      "Time for merging to target: 0h 0m 0s 6ms\n",
      "Ignore 0 out of 23.\n",
      "Too short: 0, incorrect  0.\n",
      "Time for processing: 0h 0m 0s 142ms\n",
      "Create directory tmp/9833431157172569924/search_tmp\n",
      "search tmp/9833431157172569924/query tmp/9833431157172569924/target tmp/9833431157172569924/result tmp/9833431157172569924/search_tmp -a 1 --alignment-mode 3 --comp-bias-corr 1 --gap-open aa:10,nucl:10 --gap-extend aa:1,nucl:1 -s 9.5 -k 6 --mask 0 --mask-prob 0.99995 --exhaustive-search 1 --alignment-type 0 --remove-tmp-files 1 \n",
      "\n",
      "align tmp/9833431157172569924/query_ss tmp/9833431157172569924/target_ss tmp/9833431157172569924/search_tmp/381844110943503795/pref tmp/9833431157172569924/search_tmp/381844110943503795/aln --sub-mat 'aa:3di.out,nucl:3di.out' -a 1 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 10 --min-seq-id 0 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 0.5 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:10,nucl:10 --gap-extend aa:1,nucl:1 --zdrop 40 --threads 48 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 23 type: Aminoacid\n",
      "Target database size: 23 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 23 0s 359ms\n",
      "Time for merging to aln: 0h 0m 0s 3ms\n",
      "529 alignments calculated\n",
      "529 sequence pairs passed the thresholds (1.000000 of overall calculated)\n",
      "23.000000 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 510ms\n",
      "mvdb tmp/9833431157172569924/search_tmp/381844110943503795/aln tmp/9833431157172569924/result -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 2ms\n",
      "Removing temporary files\n",
      "rmdb tmp/9833431157172569924/search_tmp/381844110943503795/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "../testdata/structsaln_0_allvall.csv exists and will be overwritten\n",
      "convertalis tmp/9833431157172569924/query tmp/9833431157172569924/target tmp/9833431157172569924/result ../testdata/structsaln_0_allvall.csv --sub-mat 'aa:3di.out,nucl:3di.out' --format-mode 0 --format-output query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,lddt,qaln,taln,cigar,alntmscore --translation-table 1 --gap-open aa:10,nucl:10 --gap-extend aa:1,nucl:1 --db-output 0 --db-load-mode 0 --search-type 0 --threads 48 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 23 0s 868ms\n",
      "Time for merging to structsaln_0_allvall.csv: 0h 0m 0s 6ms\n",
      "Time for processing: 0h 0m 1s 24ms\n",
      "rmdb tmp/9833431157172569924/result -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb tmp/9833431157172569924/target -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/9833431157172569924/target_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/9833431157172569924/target_ca -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/9833431157172569924/target_ss -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/9833431157172569924/query -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/9833431157172569924/query_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/9833431157172569924/query_ca -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/9833431157172569924/query_ss -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "../testdata/structsaln_2_allvall.csv exists and will be overwritten\n",
      "easy-search ../testdata/structs ../testdata/structs ../testdata/structsaln_2_allvall.csv tmp --format-output query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,lddt,qaln,taln,cigar,alntmscore --exhaustive-search --alignment-type 2 \n",
      "\n",
      "MMseqs Version:              \t30fdcac78217579fa25d59bc271bd4f3767d3ebb\n",
      "Seq. id. threshold           \t0\n",
      "Coverage threshold           \t0\n",
      "Coverage mode                \t0\n",
      "Max reject                   \t2147483647\n",
      "Max accept                   \t2147483647\n",
      "Add backtrace                \tfalse\n",
      "TMscore threshold            \t0\n",
      "TMalign hit order            \t0\n",
      "TMalign fast                 \t1\n",
      "Preload mode                 \t0\n",
      "Threads                      \t48\n",
      "Verbosity                    \t3\n",
      "LDDT threshold               \t0\n",
      "Sort by structure bit score  \t1\n",
      "Substitution matrix          \taa:3di.out,nucl:3di.out\n",
      "Alignment mode               \t3\n",
      "Alignment mode               \t0\n",
      "E-value threshold            \t10\n",
      "Min alignment length         \t0\n",
      "Seq. id. mode                \t0\n",
      "Alternative alignments       \t0\n",
      "Max sequence length          \t65535\n",
      "Compositional bias           \t1\n",
      "Compositional bias           \t1\n",
      "Gap open cost                \taa:10,nucl:10\n",
      "Gap extension cost           \taa:1,nucl:1\n",
      "Compressed                   \t0\n",
      "Seed substitution matrix     \taa:3di.out,nucl:3di.out\n",
      "Sensitivity                  \t9.5\n",
      "k-mer length                 \t6\n",
      "k-score                      \tseq:2147483647,prof:2147483647\n",
      "Max results per query        \t1000\n",
      "Split database               \t0\n",
      "Split mode                   \t2\n",
      "Split memory limit           \t0\n",
      "Diagonal scoring             \ttrue\n",
      "Exact k-mer matching         \t0\n",
      "Mask residues                \t0\n",
      "Mask residues probability    \t0.99995\n",
      "Mask lower case residues     \t1\n",
      "Minimum diagonal score       \t15\n",
      "Selected taxa                \t\n",
      "Spaced k-mers                \t1\n",
      "Spaced k-mer pattern         \t\n",
      "Local temporary path         \t\n",
      "Exhaustive search mode       \ttrue\n",
      "Search iterations            \t1\n",
      "Alignment type               \t2\n",
      "Remove temporary files       \ttrue\n",
      "MPI runner                   \t\n",
      "Force restart with latest tmp\tfalse\n",
      "Chain name mode              \t0\n",
      "Mask b-factor threshold      \t0\n",
      "Coord store mode             \t2\n",
      "Write lookup file            \t1\n",
      "Tar Inclusion Regex          \t.*\n",
      "Tar Exclusion Regex          \t^$\n",
      "Alignment format             \t0\n",
      "Format alignment output      \tquery,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,lddt,qaln,taln,cigar,alntmscore\n",
      "Database output              \tfalse\n",
      "\n",
      "Alignment backtraces will be computed, since they were requested by output format.\n",
      "createdb ../testdata/structs tmp/8202076179875779332/query --chain-name-mode 0 --mask-bfactor-threshold 0 --coord-store-mode 2 --write-lookup 1 --tar-include '.*' --tar-exclude '^$' --threads 48 -v 3 \n",
      "\n",
      "Output file: tmp/8202076179875779332/query\n",
      "[=================================================================] 23 0s 84ms\n",
      "Time for merging to query_ss: 0h 0m 0s 11ms\n",
      "Time for merging to query_h: 0h 0m 0s 11ms\n",
      "Time for merging to query_ca: 0h 0m 0s 11ms\n",
      "Time for merging to query: 0h 0m 0s 14ms\n",
      "Ignore 0 out of 23.\n",
      "Too short: 0, incorrect  0.\n",
      "Time for processing: 0h 0m 0s 182ms\n",
      "createdb ../testdata/structs tmp/8202076179875779332/target --chain-name-mode 0 --mask-bfactor-threshold 0 --coord-store-mode 2 --write-lookup 1 --tar-include '.*' --tar-exclude '^$' --threads 48 -v 3 \n",
      "\n",
      "Output file: tmp/8202076179875779332/target\n",
      "[=================================================================] 23 0s 85ms\n",
      "Time for merging to target_ss: 0h 0m 0s 8ms\n",
      "Time for merging to target_h: 0h 0m 0s 5ms\n",
      "Time for merging to target_ca: 0h 0m 0s 6ms\n",
      "Time for merging to target: 0h 0m 0s 6ms\n",
      "Ignore 0 out of 23.\n",
      "Too short: 0, incorrect  0.\n",
      "Time for processing: 0h 0m 0s 142ms\n",
      "Create directory tmp/8202076179875779332/search_tmp\n",
      "search tmp/8202076179875779332/query tmp/8202076179875779332/target tmp/8202076179875779332/result tmp/8202076179875779332/search_tmp -a 1 --alignment-mode 3 --comp-bias-corr 1 --gap-open aa:10,nucl:10 --gap-extend aa:1,nucl:1 -s 9.5 -k 6 --mask 0 --mask-prob 0.99995 --exhaustive-search 1 --alignment-type 2 --remove-tmp-files 1 \n",
      "\n",
      "structurealign tmp/8202076179875779332/query tmp/8202076179875779332/target tmp/8202076179875779332/search_tmp/11360897984993583988/pref tmp/8202076179875779332/search_tmp/11360897984993583988/aln --tmscore-threshold 0 --lddt-threshold 0 --sort-by-structure-bits 1 --sub-mat 'aa:3di.out,nucl:3di.out' -a 1 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 10 --min-seq-id 0 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 0.5 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:10,nucl:10 --gap-extend aa:1,nucl:1 --zdrop 40 --threads 48 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 23 0s 243ms\n",
      "Time for merging to aln: 0h 0m 0s 25ms\n",
      "Time for processing: 0h 0m 1s 759ms\n",
      "mvdb tmp/8202076179875779332/search_tmp/11360897984993583988/aln tmp/8202076179875779332/result -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 9ms\n",
      "Removing temporary files\n",
      "rmdb tmp/8202076179875779332/search_tmp/11360897984993583988/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "../testdata/structsaln_2_allvall.csv exists and will be overwritten\n",
      "convertalis tmp/8202076179875779332/query tmp/8202076179875779332/target tmp/8202076179875779332/result ../testdata/structsaln_2_allvall.csv --sub-mat 'aa:3di.out,nucl:3di.out' --format-mode 0 --format-output query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,lddt,qaln,taln,cigar,alntmscore --translation-table 1 --gap-open aa:10,nucl:10 --gap-extend aa:1,nucl:1 --db-output 0 --db-load-mode 0 --search-type 0 --threads 48 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 23 0s 867ms\n",
      "Time for merging to structsaln_2_allvall.csv: 0h 0m 0s 8ms\n",
      "Time for processing: 0h 0m 1s 15ms\n",
      "rmdb tmp/8202076179875779332/result -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb tmp/8202076179875779332/target -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 9ms\n",
      "rmdb tmp/8202076179875779332/target_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb tmp/8202076179875779332/target_ca -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/8202076179875779332/target_ss -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/8202076179875779332/query -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/8202076179875779332/query_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/8202076179875779332/query_ca -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/8202076179875779332/query_ss -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'../testdata/structsaln_2_allvall.csv'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "structfolder = '../testdata/structs'\n",
    "runFoldseek_allvall(structfolder , outfolder = structfolder, mode = 0)\n",
    "runFoldseek_allvall(structfolder , outfolder = structfolder, mode = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "46c3aa97-a047-4771-a70c-9f8b6e31fd72",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "291fe30c-e2e7-44af-a59a-cd2a7f6fee71",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "f7d26066-22b4-4d55-9d65-2a8a437eea0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "alndf0 = pd.read_table('../testdata/structsaln_0_allvall.csv' , header = None)\n",
    "alndf2 = pd.read_table('../testdata/structsaln_2_allvall.csv', header = None)\n",
    "columns = 'query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,lddt,qaln,taln,cigar,lntmscore'.split(',')\n",
    "alndf0.columns = columns\n",
    "alndf2.columns = columns\n",
    "\n",
    "alndf0['3diq']= alndf0['query'].map(mapper3di)\n",
    "alndf0['3dit']= alndf0['target'].map(mapper3di)\n",
    "alndf0['AAq']= alndf0['query'].map(mapperAA)\n",
    "alndf0['AAt']= alndf0['target'].map(mapperAA)\n",
    "\n",
    "\n",
    "alndf0 = alndf0.set_index(['query', 'target'])\n",
    "alndf2 = alndf2.set_index(['query', 'target'])\n",
    "\n",
    "alndf = alndf0.merge(alndf2, left_index=True, right_on=['query', 'target'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "bf29915e-0b20-45f6-a727-3ca5ff72c102",
   "metadata": {},
   "outputs": [],
   "source": [
    "#use the qaln and taln of x and y to derive the other side fident\n",
    "import numpy as np\n",
    "def Fident(str1,str2 , verbose = False):\n",
    "    #minlen= min( (len(str1),len(str2))  )\n",
    "    #str1 = str1[:minlen]\n",
    "    #str2 = str2[:minlen]\n",
    "    \n",
    "    str1 = np.array(list(str1))\n",
    "    str2 = np.array(list(str2))            \n",
    " \n",
    "    return len(np.where( (str1 == str2 ) & (str1 != '-' ) & (str2 != '-')  )[0]) / len(str1)\n",
    "\n",
    "\n",
    "def copyaln( aln, seq):\n",
    "    seqiter = iter(seq)\n",
    "    newaln = ''\n",
    "    \n",
    "    minlen= min( (len(aln.replace('-','')),len(seq))  )\n",
    "    for i,char in enumerate(aln):\n",
    "        if i > minlen-1:\n",
    "            break\n",
    "        if char == '-':\n",
    "            newaln += '-'\n",
    "        else:\n",
    "            newaln+=next(seqiter)\n",
    "        \n",
    "    return newaln\n",
    "\n",
    "def calc_fident_crossaln(row , verbose = False):\n",
    "    #amino acid representations of alns using AAand3di or just 3di\n",
    "    qaln_0,taln_0, qaln_2, taln_2 = row.qaln_x, row.qaln_y, row.taln_x, row.taln_y\n",
    "    #start and stop of aln\n",
    "    qstart_0,qend_0, tstart_0 , tend_0, qstart_2, qend_2, tstart_2 , tend_2 = row.qstart_x, row.qend_x , row.tstart_x , row.tend_x , row.qstart_y , row.qend_y , row.tstart_y , row.tend_y\n",
    "    #indexing starts at 1...\n",
    "    \n",
    "    #3di of the query and target\n",
    "    structQ, structT = row['3diq'], row['3dit']\n",
    "    AAq, AAt = row['AAq'], row['AAt'] \n",
    "\n",
    "    #add gaps\n",
    "    qAA_newgaps = copyaln(qaln_0, structQ[qstart_0-1:qend_0-1])\n",
    "    q3di_newgaps = copyaln(qaln_2, structQ[qstart_2-1:qend_2-1])\n",
    "    tAA_newgaps = copyaln(taln_0, structT[tstart_0-1:tend_0-1])\n",
    "    t3di_newgaps = copyaln(taln_2, structT[tstart_2-1:tend_2-1]) \n",
    "    \n",
    "    if verbose == True:\n",
    "        print('q3di->AA', qAA_newgaps )\n",
    "        print('t3di->AA', tAA_newgaps )\n",
    "        print('qAA3di ->3di', q3di_newgaps )\n",
    "        print('tAA3di ->3di', t3di_newgaps )\n",
    "\n",
    "    #return columns\n",
    "    return [ Fident(q3di_newgaps , t3di_newgaps , verbose = verbose) , Fident(qAA_newgaps , tAA_newgaps, verbose = verbose)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "36161816-63bd-46ee-99e1-4d4518f25d1d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fident_x                                                   0.836\n",
      "alnlen_x                                                     773\n",
      "mismatch_x                                                   121\n",
      "gapopen_x                                                      3\n",
      "qstart_x                                                       1\n",
      "qend_x                                                       770\n",
      "tstart_x                                                       1\n",
      "tend_x                                                       770\n",
      "evalue_x                                                     0.0\n",
      "bits_x                                                       813\n",
      "lddt_x                                                    0.8428\n",
      "qaln_x         MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...\n",
      "taln_x         MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...\n",
      "cigar_x                                      603M2D84M3I52M1D28M\n",
      "lntmscore_x                                                0.342\n",
      "3diq           DDDDDDDDDDDDDDDDDDDDDPDAPPDFPDFWWKWADAQFFIWIQP...\n",
      "3dit           DDDPDDPDPPPDDPDPDPPPPPPPPPQFPDFWWKWADAQFFIWIQD...\n",
      "AAq            MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...\n",
      "AAt            MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...\n",
      "fident_y                                                   0.998\n",
      "alnlen_y                                                     770\n",
      "mismatch_y                                                     2\n",
      "gapopen_y                                                      0\n",
      "qstart_y                                                       1\n",
      "qend_y                                                       770\n",
      "tstart_y                                                       1\n",
      "tend_y                                                       770\n",
      "evalue_y                                                     0.0\n",
      "bits_y                                                      2576\n",
      "lddt_y                                                    0.8469\n",
      "qaln_y         MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...\n",
      "taln_y         MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...\n",
      "cigar_y                                                     770M\n",
      "lntmscore_y                                                0.342\n",
      "Name: (P05067.pdb, Q5IS80.pdb), dtype: object\n",
      "q3di->AA DDDDDDDDDDDDDDDDDDDDDPDAPPDFPDFWWKWADAQFFIWIQPRNPRAIDGDPVLAHHDDDDLVVVLVSSCVVCVVFAQDAKAWDPDWDWDAQTHGRPDPGRHDHIDTTGIIGRDGDDDDKDPDDEDPQKDKDKDFDPVDKDAPVVQVVVQQVVLVLVQWGWDDKIAADHDDHRIGGMMITITHHDPDPDDDDDDDDDDDDDDPDDPQDDDPPPDDPPPPPDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDPVVLLVVLVVLLPDDWDQDDAQDQQWWWFHDPVVLAIDIDGHRPDDTGNRTHNDRVVNCVSHNVNVVVVVPPPDDDDDDDDDDDDDPPDDDPQALLRVLVPDDLDPCNSVSLVSSVVRLVVVLVVQVVVLVVVLVVQCVVCPPPDPVVNVVSVVVSVVVVVVSVVVSVVSVVSSVVVSVVSVVVVLVVQLVVLLVQLVVLVVDVVRDLVSNVVSLLSNVVSLLVVLVVLLVVLVVCCVVPVVVSVVCVVVSVVVNVVSVVVNVVSVVVVVVVVVSCVVCVVVVVVSVVVSVVVPVVSVCVVVPDDDPPPDPPPDDDDDDDDDDD--DDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDPPPPDDPPDDDDDDDDDVVVVVVVVVVVVVVVVVVVVVVVVVVVPPDDDDDDDDDDDPPD-DPVNVVVVVCVVVPPDPVVVVVVV\n",
      "t3di->AA DDDPDDPDPPPDDPDPDPPPPPPPPPQFPDFWWKWADAQFFIWIQDRNPRAIDGDPVLPDHDDDDLVVVLVSSCVVCVVFAQDAKAWDPDWDWDAQTHGRPDPGRHDDTDTTTIMGRDGDPDDKDDDDADPQKDKDKDFDPVDKDAPVVQVVVQQVVLVVVQWGWPDKIAADHPDHRIGGMMMTITHHDPPPPDDDDDDDDDDPDPDPPDDDDPPPDPPPPPPPPPPPPPDDDPPDDDDDDDDDDDDDDDDDDDDDDDDDDDDDPDDDDPPPDPPDDDDPVVLLVVLQVQLDDDWDQDPAQDFQWWWFHDLVVLAIDIDGHRPDDTGNRTHNDRVVNCVSNNVSSVVVVVVVPDDDDDDDDPDPPPPPPPDQQLLNVLVPDADDPCNVVSLVVNVVSLVVVLVVLVVVLVVVLVVVVVVCVPDDPVVVVVSVVVSVVVVVVSVVVSVVSVVVSVVVSVVNVVVVLVVQLVVLLVQLVVLVPDVVRDLVSNVVSLLSNVVSLLSVLVVLLVVLVVCCVVPVPVSVVCVVVSVVVNVVSVVVNVVSVVVVVVPVVSCVVCVVVVVVSVVVSVVVVVVSVVVVVPDPPPDPPPPPPDDDPDDPPPDPVPDDDDDDDDPPPPDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDPPPPDDDPDDDPDPVVVVVVVVVVVVVVVVVVVVVVVVVVVPPPPDDDDDDDPPPPDPDPVNVVVVVCVVVPDDDVVVVVVVVVV\n",
      "qAA3di ->3di DDDDDDDDDDDDDDDDDDDDDPDAPPDFPDFWWKWADAQFFIWIQPRNPRAIDGDPVLAHHDDDDLVVVLVSSCVVCVVFAQDAKAWDPDWDWDAQTHGRPDPGRHDHIDTTGIIGRDGDDDDKDPDDEDPQKDKDKDFDPVDKDAPVVQVVVQQVVLVLVQWGWDDKIAADHDDHRIGGMMITITHHDPDPDDDDDDDDDDDDDDPDDPQDDDPPPDDPPPPPDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDPVVLLVVLVVLLPDDWDQDDAQDQQWWWFHDPVVLAIDIDGHRPDDTGNRTHNDRVVNCVSHNVNVVVVVPPPDDDDDDDDDDDDDPPDDDPQALLRVLVPDDLDPCNSVSLVSSVVRLVVVLVVQVVVLVVVLVVQCVVCPPPDPVVNVVSVVVSVVVVVVSVVVSVVSVVSSVVVSVVSVVVVLVVQLVVLLVQLVVLVVDVVRDLVSNVVSLLSNVVSLLVVLVVLLVVLVVCCVVPVVVSVVCVVVSVVVNVVSVVVNVVSVVVVVVVVVSCVVCVVVVVVSVVVSVVVPVVSVCVVVPDDDPPPDPPPDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDPPPPDDP---PDDDDDDDDDVVVVVVVVVVVVVVVVVVVVVVVVVVVPPDDDDDDDDDDDPPDDPVNVVVVVCVVVPPDPVVVVVVV\n",
      "tAA3di ->3di DDDPDDPDPPPDDPDPDPPPPPPPPPQFPDFWWKWADAQFFIWIQDRNPRAIDGDPVLPDHDDDDLVVVLVSSCVVCVVFAQDAKAWDPDWDWDAQTHGRPDPGRHDDTDTTTIMGRDGDPDDKDDDDADPQKDKDKDFDPVDKDAPVVQVVVQQVVLVVVQWGWPDKIAADHPDHRIGGMMMTITHHDPPPPDDDDDDDDDDPDPDPPDDDDPPPDPPPPPPPPPPPPPDDDPPDDDDDDDDDDDDDDDDDDDDDDDDDDDDDPDDDDPPPDPPDDDDPVVLLVVLQVQLDDDWDQDPAQDFQWWWFHDLVVLAIDIDGHRPDDTGNRTHNDRVVNCVSNNVSSVVVVVVVPDDDDDDDDPDPPPPPPPDQQLLNVLVPDADDPCNVVSLVVNVVSLVVVLVVLVVVLVVVLVVVVVVCVPDDPVVVVVSVVVSVVVVVVSVVVSVVSVVVSVVVSVVNVVVVLVVQLVVLLVQLVVLVPDVVRDLVSNVVSLLSNVVSLLSVLVVLLVVLVVCCVVPVPVSVVCVVVSVVVNVVSVVVNVVSVVVVVVPVVSCVVCVVVVVVSVVVSVVVVVVSVVVVVPDPPPDPPPPPPDDDPDDPPPDPVPDDDDDDDDPPPPDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDPPPPDDDPDDDPDPVVVVVVVVVVVVVVVVVVVVVVVVVVVPPPPDDDDDDDPPPPDPDPVNVVVVVCVVVPDDDVVVVVVVVVV\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[0.8010403120936281, 0.8153446033810143]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "testrow = alndf.iloc[121]\n",
    "print(testrow)\n",
    "calc_fident_crossaln(testrow, verbose = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "bb38b06f-8320-45d3-af27-1835f3f2a7fe",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "shape mismatch: objects cannot be broadcast to a single shape.  Mismatch is between arg 0 with shape (768,) and arg 1 with shape (769,).",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[37], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m alndf[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mFidentx\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[43malndf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcalc_fident_crossaln\u001b[49m\u001b[43m \u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28mprint\u001b[39m(alndf)\n",
      "File \u001b[0;32m/work/FAC/FBM/DBC/cdessim2/default/dmoi/miniconda3/envs/ML2/lib/python3.10/site-packages/pandas/core/frame.py:8740\u001b[0m, in \u001b[0;36mDataFrame.apply\u001b[0;34m(self, func, axis, raw, result_type, args, **kwargs)\u001b[0m\n\u001b[1;32m   8729\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mpandas\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mcore\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mapply\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m frame_apply\n\u001b[1;32m   8731\u001b[0m op \u001b[38;5;241m=\u001b[39m frame_apply(\n\u001b[1;32m   8732\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m   8733\u001b[0m     func\u001b[38;5;241m=\u001b[39mfunc,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   8738\u001b[0m     kwargs\u001b[38;5;241m=\u001b[39mkwargs,\n\u001b[1;32m   8739\u001b[0m )\n\u001b[0;32m-> 8740\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mop\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/work/FAC/FBM/DBC/cdessim2/default/dmoi/miniconda3/envs/ML2/lib/python3.10/site-packages/pandas/core/apply.py:688\u001b[0m, in \u001b[0;36mFrameApply.apply\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    685\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mraw:\n\u001b[1;32m    686\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mapply_raw()\n\u001b[0;32m--> 688\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply_standard\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/work/FAC/FBM/DBC/cdessim2/default/dmoi/miniconda3/envs/ML2/lib/python3.10/site-packages/pandas/core/apply.py:812\u001b[0m, in \u001b[0;36mFrameApply.apply_standard\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    811\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mapply_standard\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[0;32m--> 812\u001b[0m     results, res_index \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply_series_generator\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    814\u001b[0m     \u001b[38;5;66;03m# wrap results\u001b[39;00m\n\u001b[1;32m    815\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mwrap_results(results, res_index)\n",
      "File \u001b[0;32m/work/FAC/FBM/DBC/cdessim2/default/dmoi/miniconda3/envs/ML2/lib/python3.10/site-packages/pandas/core/apply.py:828\u001b[0m, in \u001b[0;36mFrameApply.apply_series_generator\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    825\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m option_context(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmode.chained_assignment\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[1;32m    826\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m i, v \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(series_gen):\n\u001b[1;32m    827\u001b[0m         \u001b[38;5;66;03m# ignore SettingWithCopy here in case the user mutates\u001b[39;00m\n\u001b[0;32m--> 828\u001b[0m         results[i] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mf\u001b[49m\u001b[43m(\u001b[49m\u001b[43mv\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    829\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(results[i], ABCSeries):\n\u001b[1;32m    830\u001b[0m             \u001b[38;5;66;03m# If we have a view on v, we need to make a copy because\u001b[39;00m\n\u001b[1;32m    831\u001b[0m             \u001b[38;5;66;03m#  series_generator will swap out the underlying data\u001b[39;00m\n\u001b[1;32m    832\u001b[0m             results[i] \u001b[38;5;241m=\u001b[39m results[i]\u001b[38;5;241m.\u001b[39mcopy(deep\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n",
      "Cell \u001b[0;32mIn[34], line 53\u001b[0m, in \u001b[0;36mcalc_fident_crossaln\u001b[0;34m(row, verbose)\u001b[0m\n\u001b[1;32m     50\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtAA3di ->3di\u001b[39m\u001b[38;5;124m'\u001b[39m, t3di_newgaps )\n\u001b[1;32m     52\u001b[0m \u001b[38;5;66;03m#return columns\u001b[39;00m\n\u001b[0;32m---> 53\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m [ Fident(q3di_newgaps , t3di_newgaps , verbose \u001b[38;5;241m=\u001b[39m verbose) , \u001b[43mFident\u001b[49m\u001b[43m(\u001b[49m\u001b[43mqAA_newgaps\u001b[49m\u001b[43m \u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtAA_newgaps\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mverbose\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mverbose\u001b[49m\u001b[43m)\u001b[49m]\n",
      "Cell \u001b[0;32mIn[34], line 11\u001b[0m, in \u001b[0;36mFident\u001b[0;34m(str1, str2, verbose)\u001b[0m\n\u001b[1;32m      8\u001b[0m str1 \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39marray(\u001b[38;5;28mlist\u001b[39m(str1))\n\u001b[1;32m      9\u001b[0m str2 \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39marray(\u001b[38;5;28mlist\u001b[39m(str2))            \n\u001b[0;32m---> 11\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(np\u001b[38;5;241m.\u001b[39mwhere( (\u001b[43mstr1\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m==\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mstr2\u001b[49m ) \u001b[38;5;241m&\u001b[39m (str1 \u001b[38;5;241m!=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m-\u001b[39m\u001b[38;5;124m'\u001b[39m ) \u001b[38;5;241m&\u001b[39m (str2 \u001b[38;5;241m!=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m-\u001b[39m\u001b[38;5;124m'\u001b[39m)  )[\u001b[38;5;241m0\u001b[39m]) \u001b[38;5;241m/\u001b[39m \u001b[38;5;28mlen\u001b[39m(str1)\n",
      "\u001b[0;31mValueError\u001b[0m: shape mismatch: objects cannot be broadcast to a single shape.  Mismatch is between arg 0 with shape (768,) and arg 1 with shape (769,)."
     ]
    }
   ],
   "source": [
    "alndf['Fidentx'] = alndf.apply(calc_fident_crossaln , axis =1 )\n",
    "print(alndf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5b7b694-6f78-48f1-9a71-b49448ed4e43",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
