{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "053c54f2-b3b1-4d6a-82d7-49228f368010",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run w aln output in both modes\n",
    "import sys\n",
    "sys.path.append('../src/')\n",
    "#matcAFDB_tools up positions\n",
    "#calc fident with cross alphabet comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4e2ee53e-d0d7-4150-84e5-080d1ef7f5fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess ,shlex\n",
    "def runargs(args):\n",
    "    '''run a command line command\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    args : str\n",
    "        command line command\n",
    "    '''\n",
    "\n",
    "    args = shlex.split( args)\n",
    "    p = subprocess.run( args )\n",
    "    return p\n",
    "\n",
    "def runFoldseek_allvall( structfolder , outfolder , foldseekpath = '../foldseek/foldseek' , mode = 2):\n",
    "    '''\n",
    "    run foldseek search and createtsv\n",
    "\n",
    "    parameters\n",
    "    ----------\n",
    "    dbpath : str\n",
    "        path to foldseek database\n",
    "    outfolder : str \n",
    "        path to output folder\n",
    "    maxseqs : int   \n",
    "        maximum number of sequences to compare to\n",
    "\n",
    "    '''\n",
    "\n",
    "    args = foldseekpath + \" easy-search \" + structfolder + \" \"+ structfolder +\" \"+ outfolder+\"aln_\"+str(mode)+\"_allvall.csv tmp\" + \" --format-output 'query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,lddt,qaln,taln,cigar,alntmscore' --exhaustive-search --alignment-type \"+ str(mode) \n",
    "    p = runargs(args)\n",
    "    return outfolder+\"aln_\"+str(mode)+\"_allvall.csv\"\n",
    "\n",
    "def read_dbfiles3di(folder, name = 'outdb'):\n",
    "    #find positions \n",
    "    threeDidb = folder+ name +'_ss'\n",
    "    threeDiseq = [ l.strip().replace('\\x00','') for l in open(threeDidb)]\n",
    "    \n",
    "    \n",
    "    lookup = folder+name+'.lookup'\n",
    "    ids = [ l.split()[1].strip() for l in open(lookup)]\n",
    "    \n",
    "    AADB = folder+name\n",
    "    AAs = [ l.strip().replace('\\x00','') for l in open(AADB)]\n",
    "\n",
    "    mapper3di = dict(zip(ids,threeDiseq))\n",
    "    mapperAA = dict(zip(ids,AAs))\n",
    "    \n",
    "    return mapper3di, mapperAA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2e4a4f0a-91c7-4264-b2e2-80ae82894084",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapper3di, mapperAA = read_dbfiles3di('../testdata/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6156c98f-1d30-4f39-924d-82475162ea03",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../testdata/structsaln_0_allvall.csv exists and will be overwritten\n",
      "easy-search ../testdata/structs ../testdata/structs ../testdata/structsaln_0_allvall.csv tmp --format-output query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,lddt,qaln,taln,cigar,alntmscore --exhaustive-search --alignment-type 0 \n",
      "\n",
      "MMseqs Version:              \t30fdcac78217579fa25d59bc271bd4f3767d3ebb\n",
      "Seq. id. threshold           \t0\n",
      "Coverage threshold           \t0\n",
      "Coverage mode                \t0\n",
      "Max reject                   \t2147483647\n",
      "Max accept                   \t2147483647\n",
      "Add backtrace                \tfalse\n",
      "TMscore threshold            \t0\n",
      "TMalign hit order            \t0\n",
      "TMalign fast                 \t1\n",
      "Preload mode                 \t0\n",
      "Threads                      \t48\n",
      "Verbosity                    \t3\n",
      "LDDT threshold               \t0\n",
      "Sort by structure bit score  \t1\n",
      "Substitution matrix          \taa:3di.out,nucl:3di.out\n",
      "Alignment mode               \t3\n",
      "Alignment mode               \t0\n",
      "E-value threshold            \t10\n",
      "Min alignment length         \t0\n",
      "Seq. id. mode                \t0\n",
      "Alternative alignments       \t0\n",
      "Max sequence length          \t65535\n",
      "Compositional bias           \t1\n",
      "Compositional bias           \t1\n",
      "Gap open cost                \taa:10,nucl:10\n",
      "Gap extension cost           \taa:1,nucl:1\n",
      "Compressed                   \t0\n",
      "Seed substitution matrix     \taa:3di.out,nucl:3di.out\n",
      "Sensitivity                  \t9.5\n",
      "k-mer length                 \t6\n",
      "k-score                      \tseq:2147483647,prof:2147483647\n",
      "Max results per query        \t1000\n",
      "Split database               \t0\n",
      "Split mode                   \t2\n",
      "Split memory limit           \t0\n",
      "Diagonal scoring             \ttrue\n",
      "Exact k-mer matching         \t0\n",
      "Mask residues                \t0\n",
      "Mask residues probability    \t0.99995\n",
      "Mask lower case residues     \t1\n",
      "Minimum diagonal score       \t15\n",
      "Selected taxa                \t\n",
      "Spaced k-mers                \t1\n",
      "Spaced k-mer pattern         \t\n",
      "Local temporary path         \t\n",
      "Exhaustive search mode       \ttrue\n",
      "Search iterations            \t1\n",
      "Alignment type               \t0\n",
      "Remove temporary files       \ttrue\n",
      "MPI runner                   \t\n",
      "Force restart with latest tmp\tfalse\n",
      "Chain name mode              \t0\n",
      "Mask b-factor threshold      \t0\n",
      "Coord store mode             \t2\n",
      "Write lookup file            \t1\n",
      "Tar Inclusion Regex          \t.*\n",
      "Tar Exclusion Regex          \t^$\n",
      "Alignment format             \t0\n",
      "Format alignment output      \tquery,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,lddt,qaln,taln,cigar,alntmscore\n",
      "Database output              \tfalse\n",
      "\n",
      "Alignment backtraces will be computed, since they were requested by output format.\n",
      "createdb ../testdata/structs tmp/9833431157172569924/query --chain-name-mode 0 --mask-bfactor-threshold 0 --coord-store-mode 2 --write-lookup 1 --tar-include '.*' --tar-exclude '^$' --threads 48 -v 3 \n",
      "\n",
      "Output file: tmp/9833431157172569924/query\n",
      "[=================================================================] 23 0s 3ms\n",
      "Time for merging to query_ss: 0h 0m 0s 164ms\n",
      "Time for merging to query_h: 0h 0m 0s 63ms\n",
      "Time for merging to query_ca: 0h 0m 0s 73ms\n",
      "Time for merging to query: 0h 0m 0s 58ms\n",
      "Ignore 0 out of 23.\n",
      "Too short: 0, incorrect  0.\n",
      "Time for processing: 0h 0m 0s 688ms\n",
      "createdb ../testdata/structs tmp/9833431157172569924/target --chain-name-mode 0 --mask-bfactor-threshold 0 --coord-store-mode 2 --write-lookup 1 --tar-include '.*' --tar-exclude '^$' --threads 48 -v 3 \n",
      "\n",
      "Output file: tmp/9833431157172569924/target\n",
      "[=================================================================] 23 0s 60ms\n",
      "Time for merging to target_ss: 0h 0m 0s 9ms\n",
      "Time for merging to target_h: 0h 0m 0s 8ms\n",
      "Time for merging to target_ca: 0h 0m 0s 9ms\n",
      "Time for merging to target: 0h 0m 0s 14ms\n",
      "Ignore 0 out of 23.\n",
      "Too short: 0, incorrect  0.\n",
      "Time for processing: 0h 0m 0s 168ms\n",
      "Create directory tmp/9833431157172569924/search_tmp\n",
      "search tmp/9833431157172569924/query tmp/9833431157172569924/target tmp/9833431157172569924/result tmp/9833431157172569924/search_tmp -a 1 --alignment-mode 3 --comp-bias-corr 1 --gap-open aa:10,nucl:10 --gap-extend aa:1,nucl:1 -s 9.5 -k 6 --mask 0 --mask-prob 0.99995 --exhaustive-search 1 --alignment-type 0 --remove-tmp-files 1 \n",
      "\n",
      "align tmp/9833431157172569924/query_ss tmp/9833431157172569924/target_ss tmp/9833431157172569924/search_tmp/11885433207302564274/pref tmp/9833431157172569924/search_tmp/11885433207302564274/aln --sub-mat 'aa:3di.out,nucl:3di.out' -a 1 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 10 --min-seq-id 0 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 0.5 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:10,nucl:10 --gap-extend aa:1,nucl:1 --zdrop 40 --threads 48 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 23 type: Aminoacid\n",
      "Target database size: 23 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 23 0s 321ms\n",
      "Time for merging to aln: 0h 0m 0s 13ms\n",
      "529 alignments calculated\n",
      "529 sequence pairs passed the thresholds (1.000000 of overall calculated)\n",
      "23.000000 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 466ms\n",
      "mvdb tmp/9833431157172569924/search_tmp/11885433207302564274/aln tmp/9833431157172569924/result -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "Removing temporary files\n",
      "rmdb tmp/9833431157172569924/search_tmp/11885433207302564274/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "../testdata/structsaln_0_allvall.csv exists and will be overwritten\n",
      "convertalis tmp/9833431157172569924/query tmp/9833431157172569924/target tmp/9833431157172569924/result ../testdata/structsaln_0_allvall.csv --sub-mat 'aa:3di.out,nucl:3di.out' --format-mode 0 --format-output query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,lddt,qaln,taln,cigar,alntmscore --translation-table 1 --gap-open aa:10,nucl:10 --gap-extend aa:1,nucl:1 --db-output 0 --db-load-mode 0 --search-type 0 --threads 48 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 23 0s 764ms\n",
      "Time for merging to structsaln_0_allvall.csv: 0h 0m 0s 19ms\n",
      "Time for processing: 0h 0m 0s 909ms\n",
      "rmdb tmp/9833431157172569924/result -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 2ms\n",
      "rmdb tmp/9833431157172569924/target -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/9833431157172569924/target_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/9833431157172569924/target_ca -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/9833431157172569924/target_ss -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/9833431157172569924/query -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/9833431157172569924/query_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/9833431157172569924/query_ca -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/9833431157172569924/query_ss -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "../testdata/structsaln_2_allvall.csv exists and will be overwritten\n",
      "easy-search ../testdata/structs ../testdata/structs ../testdata/structsaln_2_allvall.csv tmp --format-output query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,lddt,qaln,taln,cigar,alntmscore --exhaustive-search --alignment-type 2 \n",
      "\n",
      "MMseqs Version:              \t30fdcac78217579fa25d59bc271bd4f3767d3ebb\n",
      "Seq. id. threshold           \t0\n",
      "Coverage threshold           \t0\n",
      "Coverage mode                \t0\n",
      "Max reject                   \t2147483647\n",
      "Max accept                   \t2147483647\n",
      "Add backtrace                \tfalse\n",
      "TMscore threshold            \t0\n",
      "TMalign hit order            \t0\n",
      "TMalign fast                 \t1\n",
      "Preload mode                 \t0\n",
      "Threads                      \t48\n",
      "Verbosity                    \t3\n",
      "LDDT threshold               \t0\n",
      "Sort by structure bit score  \t1\n",
      "Substitution matrix          \taa:3di.out,nucl:3di.out\n",
      "Alignment mode               \t3\n",
      "Alignment mode               \t0\n",
      "E-value threshold            \t10\n",
      "Min alignment length         \t0\n",
      "Seq. id. mode                \t0\n",
      "Alternative alignments       \t0\n",
      "Max sequence length          \t65535\n",
      "Compositional bias           \t1\n",
      "Compositional bias           \t1\n",
      "Gap open cost                \taa:10,nucl:10\n",
      "Gap extension cost           \taa:1,nucl:1\n",
      "Compressed                   \t0\n",
      "Seed substitution matrix     \taa:3di.out,nucl:3di.out\n",
      "Sensitivity                  \t9.5\n",
      "k-mer length                 \t6\n",
      "k-score                      \tseq:2147483647,prof:2147483647\n",
      "Max results per query        \t1000\n",
      "Split database               \t0\n",
      "Split mode                   \t2\n",
      "Split memory limit           \t0\n",
      "Diagonal scoring             \ttrue\n",
      "Exact k-mer matching         \t0\n",
      "Mask residues                \t0\n",
      "Mask residues probability    \t0.99995\n",
      "Mask lower case residues     \t1\n",
      "Minimum diagonal score       \t15\n",
      "Selected taxa                \t\n",
      "Spaced k-mers                \t1\n",
      "Spaced k-mer pattern         \t\n",
      "Local temporary path         \t\n",
      "Exhaustive search mode       \ttrue\n",
      "Search iterations            \t1\n",
      "Alignment type               \t2\n",
      "Remove temporary files       \ttrue\n",
      "MPI runner                   \t\n",
      "Force restart with latest tmp\tfalse\n",
      "Chain name mode              \t0\n",
      "Mask b-factor threshold      \t0\n",
      "Coord store mode             \t2\n",
      "Write lookup file            \t1\n",
      "Tar Inclusion Regex          \t.*\n",
      "Tar Exclusion Regex          \t^$\n",
      "Alignment format             \t0\n",
      "Format alignment output      \tquery,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,lddt,qaln,taln,cigar,alntmscore\n",
      "Database output              \tfalse\n",
      "\n",
      "Alignment backtraces will be computed, since they were requested by output format.\n",
      "createdb ../testdata/structs tmp/8202076179875779332/query --chain-name-mode 0 --mask-bfactor-threshold 0 --coord-store-mode 2 --write-lookup 1 --tar-include '.*' --tar-exclude '^$' --threads 48 -v 3 \n",
      "\n",
      "Output file: tmp/8202076179875779332/query\n",
      "[=================================================================] 23 0s 71ms\n",
      "Time for merging to query_ss: 0h 0m 0s 12ms\n",
      "Time for merging to query_h: 0h 0m 0s 10ms\n",
      "Time for merging to query_ca: 0h 0m 0s 14ms\n",
      "Time for merging to query: 0h 0m 0s 10ms\n",
      "Ignore 0 out of 23.\n",
      "Too short: 0, incorrect  0.\n",
      "Time for processing: 0h 0m 0s 163ms\n",
      "createdb ../testdata/structs tmp/8202076179875779332/target --chain-name-mode 0 --mask-bfactor-threshold 0 --coord-store-mode 2 --write-lookup 1 --tar-include '.*' --tar-exclude '^$' --threads 48 -v 3 \n",
      "\n",
      "Output file: tmp/8202076179875779332/target\n",
      "[=================================================================] 23 0s 71ms\n",
      "Time for merging to target_ss: 0h 0m 0s 13ms\n",
      "Time for merging to target_h: 0h 0m 0s 10ms\n",
      "Time for merging to target_ca: 0h 0m 0s 12ms\n",
      "Time for merging to target: 0h 0m 0s 9ms\n",
      "Ignore 0 out of 23.\n",
      "Too short: 0, incorrect  0.\n",
      "Time for processing: 0h 0m 0s 156ms\n",
      "Create directory tmp/8202076179875779332/search_tmp\n",
      "search tmp/8202076179875779332/query tmp/8202076179875779332/target tmp/8202076179875779332/result tmp/8202076179875779332/search_tmp -a 1 --alignment-mode 3 --comp-bias-corr 1 --gap-open aa:10,nucl:10 --gap-extend aa:1,nucl:1 -s 9.5 -k 6 --mask 0 --mask-prob 0.99995 --exhaustive-search 1 --alignment-type 2 --remove-tmp-files 1 \n",
      "\n",
      "structurealign tmp/8202076179875779332/query tmp/8202076179875779332/target tmp/8202076179875779332/search_tmp/3496790260524239220/pref tmp/8202076179875779332/search_tmp/3496790260524239220/aln --tmscore-threshold 0 --lddt-threshold 0 --sort-by-structure-bits 1 --sub-mat 'aa:3di.out,nucl:3di.out' -a 1 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 10 --min-seq-id 0 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 0.5 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:10,nucl:10 --gap-extend aa:1,nucl:1 --zdrop 40 --threads 48 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 23 0s 223ms\n",
      "Time for merging to aln: 0h 0m 0s 7ms\n",
      "Time for processing: 0h 0m 1s 529ms\n",
      "mvdb tmp/8202076179875779332/search_tmp/3496790260524239220/aln tmp/8202076179875779332/result -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "Removing temporary files\n",
      "rmdb tmp/8202076179875779332/search_tmp/3496790260524239220/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "../testdata/structsaln_2_allvall.csv exists and will be overwritten\n",
      "convertalis tmp/8202076179875779332/query tmp/8202076179875779332/target tmp/8202076179875779332/result ../testdata/structsaln_2_allvall.csv --sub-mat 'aa:3di.out,nucl:3di.out' --format-mode 0 --format-output query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,lddt,qaln,taln,cigar,alntmscore --translation-table 1 --gap-open aa:10,nucl:10 --gap-extend aa:1,nucl:1 --db-output 0 --db-load-mode 0 --search-type 0 --threads 48 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 23 0s 776ms\n",
      "Time for merging to structsaln_2_allvall.csv: 0h 0m 0s 28ms\n",
      "Time for processing: 0h 0m 0s 909ms\n",
      "rmdb tmp/8202076179875779332/result -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb tmp/8202076179875779332/target -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/8202076179875779332/target_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/8202076179875779332/target_ca -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/8202076179875779332/target_ss -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/8202076179875779332/query -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/8202076179875779332/query_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/8202076179875779332/query_ca -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/8202076179875779332/query_ss -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'../testdata/structsaln_2_allvall.csv'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "structfolder = '../testdata/structs'\n",
    "runFoldseek_allvall(structfolder , outfolder = structfolder, mode = 0)\n",
    "runFoldseek_allvall(structfolder , outfolder = structfolder, mode = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "591671ab-578d-49c3-887d-4a2e7422e262",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          query      target  fident  alnlen  mismatch  gapopen  qstart  qend  \\\n",
      "0    Q06481.pdb  Q06481.pdb   1.000     763         0        0       1   763   \n",
      "1    Q06481.pdb  P15943.pdb   0.926     767        51        2       1   763   \n",
      "2    Q06481.pdb  Q06335.pdb   0.853     732        52        1      32   763   \n",
      "3    Q06481.pdb  Q95241.pdb   0.499     779       334       12      12   763   \n",
      "4    Q06481.pdb  P08592.pdb   0.490     789       345       11      12   763   \n",
      "..          ...         ...     ...     ...       ...      ...     ...   ...   \n",
      "504  Q95241.pdb  Q28757.pdb   1.000      57         0        0     648   704   \n",
      "505  Q95241.pdb  Q28053.pdb   1.000      58         0        0     647   704   \n",
      "506  Q95241.pdb  Q28748.pdb   1.000      57         0        0     648   704   \n",
      "507  Q95241.pdb  Q28280.pdb   1.000      57         0        0     647   703   \n",
      "508  Q95241.pdb  Q29149.pdb   1.000      56         0        0     648   703   \n",
      "\n",
      "     tstart  tend        evalue  bits    lddt  \\\n",
      "0         1   763  0.000000e+00  5206  1.0000   \n",
      "1         1   765  0.000000e+00  3413  0.8783   \n",
      "2        32   707  8.279000e-86  1923  0.8111   \n",
      "3         2   751  3.540000e-73  1579  0.8113   \n",
      "4         2   770  1.571000e-73  1539  0.8009   \n",
      "..      ...   ...           ...   ...     ...   \n",
      "504       1    57  9.810000e-04   104  0.6976   \n",
      "505       1    58  9.810000e-04   102  0.6886   \n",
      "506       1    57  1.587000e-03    98  0.6934   \n",
      "507       1    57  2.435000e-03    92  0.6925   \n",
      "508       1    56  2.841000e-01    40  0.5746   \n",
      "\n",
      "                                                  qaln  \\\n",
      "0    MAATGTAAAAATGRLLLLLLVGLTAPALALAGYIEALAANAGTGFA...   \n",
      "1    MAATGTAAAAATGRLLLLLLVGLTAPALALAGYIEALAANAGTGFA...   \n",
      "2    GYIEALAANAGTGFAVAEPQIAMFCGKLNMHVNIQTGKWEPDPTGT...   \n",
      "3    TGRLLLLLLVGLTAPALALAGYIEALAANAGTGFAVAEPQIAMFCG...   \n",
      "4    TGRLLLLLLVGLTAPALALAGYIEALAANAGTGFAVAEPQIAMFCG...   \n",
      "..                                                 ...   \n",
      "504  SEVKMDAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVVI...   \n",
      "505  ISEVKMDAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVV...   \n",
      "506  SEVKMDAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVVI...   \n",
      "507  ISEVKMDAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVV...   \n",
      "508  SEVKMDAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVVI...   \n",
      "\n",
      "                                                  taln  \\\n",
      "0    MAATGTAAAAATGRLLLLLLVGLTAPALALAGYIEALAANAGTGFA...   \n",
      "1    MAATGTAAAAATGKLLVLLLLGLTAPAAALAGYIEALAANAGTGFA...   \n",
      "2    GYIEALAANAGTGFAVAEPQIAMFCGKLNMHVNIQTGKWEPDPTGT...   \n",
      "3    LPGLALLLLAAWTARALEVPTD-----GNAG---LLAEPQIAMFCG...   \n",
      "4    LPSLALLLLAAWTVRALEVP--------TDGNAGLLAEPQIAMFCG...   \n",
      "..                                                 ...   \n",
      "504  SEVKMDAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVVI...   \n",
      "505  ISEVKMDAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVV...   \n",
      "506  SEVKMDAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVVI...   \n",
      "507  ISEVKMDAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVV...   \n",
      "508  SEVKMDAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVVI...   \n",
      "\n",
      "                                                 cigar  alntmscore  \n",
      "0                                                 763M      1.0000  \n",
      "1                                      218M2I37M4D506M      0.6142  \n",
      "2                                          276M56I400M      0.3385  \n",
      "3    22M5I4M3I79M2D96M11I7M2I45M10D82M1D216M5D16M7I...      0.3129  \n",
      "4    20M8I85M2D96M7D29M4I26M6I78M14D8M6D234M2I9M3D2...      0.2984  \n",
      "..                                                 ...         ...  \n",
      "504                                                57M      0.4598  \n",
      "505                                                58M      0.4498  \n",
      "506                                                57M      0.4593  \n",
      "507                                                57M      0.4402  \n",
      "508                                                56M      0.4594  \n",
      "\n",
      "[509 rows x 17 columns]\n",
      "[[0.     0.032  0.     0.5195 0.5105 0.     0.     0.6805 0.762  0.\n",
      "  0.345  0.024  0.327  0.6725 0.03   0.032  0.5585 0.006  0.7535 0.004\n",
      "  0.     0.034  0.7715]\n",
      " [0.032  0.     0.0515 0.5155 0.5105 0.0525 0.0525 0.6825 0.7645 0.054\n",
      "  0.341  0.02   0.3315 0.675  0.051  0.032  0.549  0.032  0.755  0.03\n",
      "  0.053  0.008  0.7695]\n",
      " [0.     0.0515 0.     0.587  0.604  0.     0.     0.812  0.76   0.\n",
      "  0.225  0.     0.19   0.793  0.     0.     0.587  0.     0.8705 0.\n",
      "  0.     0.0515 1.    ]\n",
      " [0.5195 0.5155 0.587  0.     0.074  0.6185 0.6005 0.6455 0.7605 0.305\n",
      "  0.535  0.517  0.523  0.648  0.506  0.5215 0.107  0.516  0.7765 0.521\n",
      "  0.62   0.5145 0.783 ]\n",
      " [0.5105 0.5105 0.604  0.074  0.     0.5965 0.579  0.6475 0.757  0.305\n",
      "  0.528  0.5145 0.53   0.6465 0.5025 0.516  0.1465 0.515  0.7605 0.5125\n",
      "  0.6235 0.509  0.7795]\n",
      " [0.     0.0525 0.     0.6185 0.5965 0.     0.     0.812  0.7485 0.\n",
      "  0.204  0.     0.176  0.793  0.     0.     0.622  0.     0.9355 0.\n",
      "  0.     0.0525 0.9485]\n",
      " [0.     0.0525 0.     0.6005 0.579  0.     0.     0.808  0.756  0.\n",
      "  0.204  0.     0.176  0.7745 0.     0.     0.6005 0.     0.9375 0.\n",
      "  0.     0.0525 1.    ]\n",
      " [0.6805 0.6825 0.812  0.6455 0.6475 0.812  0.808  0.     0.7555 1.\n",
      "  0.6585 0.6755 0.687  0.112  0.689  0.681  0.5955 0.693  0.769  0.6805\n",
      "  0.818  0.6885 0.7925]\n",
      " [0.762  0.7645 0.76   0.7605 0.757  0.7485 0.756  0.7555 0.     1.\n",
      "  0.744  0.7595 0.765  0.762  0.746  0.7565 0.749  0.7625 0.7185 0.7605\n",
      "  0.769  0.7615 0.778 ]\n",
      " [0.     0.054  0.     0.305  0.305  0.     0.     1.     1.     0.\n",
      "  0.2075 0.     0.179  1.     0.     0.     0.5375 0.     1.     0.\n",
      "  0.     0.054  1.    ]\n",
      " [0.345  0.341  0.225  0.535  0.528  0.204  0.204  0.6585 0.744  0.2075\n",
      "  0.     0.343  0.3055 0.657  0.3235 0.346  0.5705 0.3425 0.753  0.3435\n",
      "  0.229  0.336  0.777 ]\n",
      " [0.024  0.02   0.     0.517  0.5145 0.     0.     0.6755 0.7595 0.\n",
      "  0.343  0.     0.3225 0.6795 0.046  0.026  0.5435 0.023  0.7495 0.021\n",
      "  0.     0.028  0.7805]\n",
      " [0.327  0.3315 0.19   0.523  0.53   0.176  0.176  0.687  0.765  0.179\n",
      "  0.3055 0.3225 0.     0.676  0.313  0.3235 0.5665 0.3275 0.766  0.3315\n",
      "  0.193  0.3305 0.7765]\n",
      " [0.6725 0.675  0.793  0.648  0.6465 0.793  0.7745 0.112  0.762  1.\n",
      "  0.657  0.6795 0.676  0.     0.6655 0.677  0.607  0.6735 0.7705 0.675\n",
      "  0.809  0.6755 0.7845]\n",
      " [0.03   0.051  0.     0.506  0.5025 0.     0.     0.689  0.746  0.\n",
      "  0.3235 0.046  0.313  0.6655 0.     0.05   0.5405 0.032  0.7465 0.03\n",
      "  0.     0.051  0.7735]\n",
      " [0.032  0.032  0.     0.5215 0.516  0.     0.     0.681  0.7565 0.\n",
      "  0.346  0.026  0.3235 0.677  0.05   0.     0.5495 0.032  0.756  0.03\n",
      "  0.     0.037  0.7705]\n",
      " [0.5585 0.549  0.587  0.107  0.1465 0.622  0.6005 0.5955 0.749  0.5375\n",
      "  0.5705 0.5435 0.5665 0.607  0.5405 0.5495 0.     0.556  0.747  0.5555\n",
      "  0.6665 0.548  0.78  ]\n",
      " [0.006  0.032  0.     0.516  0.515  0.     0.     0.693  0.7625 0.\n",
      "  0.3425 0.023  0.3275 0.6735 0.032  0.032  0.556  0.     0.749  0.002\n",
      "  0.     0.034  0.7765]\n",
      " [0.7535 0.755  0.8705 0.7765 0.7605 0.9355 0.9375 0.769  0.7185 1.\n",
      "  0.753  0.7495 0.766  0.7705 0.7465 0.756  0.747  0.749  0.     0.752\n",
      "  0.891  0.7505 0.768 ]\n",
      " [0.004  0.03   0.     0.521  0.5125 0.     0.     0.6805 0.7605 0.\n",
      "  0.3435 0.021  0.3315 0.675  0.03   0.03   0.5555 0.002  0.752  0.\n",
      "  0.     0.033  0.768 ]\n",
      " [0.     0.053  0.     0.62   0.6235 0.     0.     0.818  0.769  0.\n",
      "  0.229  0.     0.193  0.809  0.     0.     0.6665 0.     0.891  0.\n",
      "  0.     0.053  1.    ]\n",
      " [0.034  0.008  0.0515 0.5145 0.509  0.0525 0.0525 0.6885 0.7615 0.054\n",
      "  0.336  0.028  0.3305 0.6755 0.051  0.037  0.548  0.034  0.7505 0.033\n",
      "  0.053  0.     0.769 ]\n",
      " [0.7715 0.7695 1.     0.783  0.7795 0.9485 1.     0.7925 0.778  1.\n",
      "  0.777  0.7805 0.7765 0.7845 0.7735 0.7705 0.78   0.7765 0.768  0.768\n",
      "  1.     0.769  0.    ]] 1.0 0.0\n",
      "\n",
      " . This analysis will run on 1 threads.\n",
      "\n",
      "#  Analysing dataset 1\n",
      "\n",
      " . Computing tree...\n",
      "\n",
      " . Performing NNI...\n",
      "\n",
      " . Time used 0h00m00s\n",
      "\n",
      "   /-P14599.pdb\n",
      "  |\n",
      "  |   /-Q10651.pdb\n",
      "  |--|\n",
      "  |   \\-Q06BR2.pdb\n",
      "--|\n",
      "  |      /-Q03157.pdb\n",
      "  |   /-|\n",
      "  |  |   \\-P51693.pdb\n",
      "  |  |\n",
      "   \\-|      /-Q06335.pdb\n",
      "     |   /-|\n",
      "     |  |  |   /-Q06481.pdb\n",
      "     |  |   \\-|\n",
      "      \\-|      \\-P15943.pdb\n",
      "        |\n",
      "        |   /-O93279.pdb\n",
      "        |  |\n",
      "         \\-|   /-O73683.pdb\n",
      "           |  |\n",
      "            \\-|   /-Q29149.pdb\n",
      "              |  |\n",
      "               \\-|   /-Q28748.pdb\n",
      "                 |  |\n",
      "                 |  |      /-Q95241.pdb\n",
      "                  \\-|   /-|\n",
      "                    |  |   \\-Q28053.pdb\n",
      "                    |  |\n",
      "                     \\-|   /-Q28757.pdb\n",
      "                       |  |\n",
      "                       |  |      /-Q28280.pdb\n",
      "                        \\-|   /-|\n",
      "                          |  |  |   /-P53601.pdb\n",
      "                          |  |   \\-|\n",
      "                          |  |     |   /-Q5IS80.pdb\n",
      "                           \\-|      \\-|\n",
      "                             |         \\-P05067.pdb\n",
      "                             |\n",
      "                             |   /-Q60495.pdb\n",
      "                              \\-|\n",
      "                                |   /-P79307.pdb\n",
      "                                 \\-|\n",
      "                                   |   /-P12023.pdb\n",
      "                                    \\-|\n",
      "                                       \\-P08592.pdb\n",
      "\n",
      "MAD phylogenetic rooting\n",
      "\n",
      "Analyzing file '../testdata/fidentstructblob_tree.nwk'...\n",
      ">> Warning: 5 tiny branch lengths (<10^-6) were converted to 0. (Override with '-t'.)\n",
      ">> [MAD=0.453_AI=0.997_CCV=11.2%_N=1/1]\n",
      "\n",
      "Minimal ancestor deviation, MAD = 0.453\n",
      "           Ambiguity index,  AI = 0.997\n",
      "                  Clock CV, CCV = 11.2%\n",
      "Rooted tree written to '../testdata/fidentstructblob_tree.nwk.rooted'\n",
      "\n",
      "\n",
      "    - Please cite DOI:10.1038/s41559-017-0193\n",
      "\n",
      "[[0.      0.6901  0.5712  0.7236  0.7249  0.55965 0.56035 0.6969  0.68435\n",
      "  0.5595  0.72225 0.709   0.7166  0.6984  0.6808  0.5967  0.54235 0.5636\n",
      "  0.72535 0.5001  0.5829  0.6143  0.75655]\n",
      " [0.6901  0.      0.56255 0.71415 0.7012  0.5507  0.55135 0.68625 0.6748\n",
      "  0.5462  0.73075 0.6959  0.7057  0.66865 0.5966  0.6711  0.6896  0.6501\n",
      "  0.71505 0.6964  0.5703  0.586   0.7319 ]\n",
      " [0.5712  0.56255 0.      0.5791  0.5734  0.1237  0.1241  0.5396  0.63195\n",
      "  0.5282  0.5539  0.54175 0.5699  0.5347  0.5456  0.58095 0.5761  0.57805\n",
      "  0.52545 0.56025 0.363   0.57415 1.     ]\n",
      " [0.7236  0.71415 0.5791  0.      0.3858  0.57465 0.57555 0.61145 0.6606\n",
      "  0.1599  0.6975  0.69215 0.7032  0.63775 0.6783  0.679   0.6779  0.70575\n",
      "  0.5842  0.7122  0.58365 0.72395 0.72475]\n",
      " [0.7249  0.7012  0.5734  0.3858  0.      0.54395 0.5633  0.6293  0.6609\n",
      "  0.1373  0.69155 0.68235 0.69145 0.63795 0.68655 0.68485 0.6619  0.7044\n",
      "  0.5242  0.718   0.5825  0.7193  0.7027 ]\n",
      " [0.55965 0.5507  0.1237  0.57465 0.54395 0.      0.0399  0.5365  0.60155\n",
      "  0.5276  0.5418  0.5309  0.5582  0.5311  0.53545 0.56895 0.5767  0.5664\n",
      "  0.7842  0.54885 0.3428  0.56335 0.77595]\n",
      " [0.56035 0.55135 0.1241  0.57555 0.5633  0.0399  0.      0.5289  0.6246\n",
      "  0.5277  0.5419  0.5315  0.5592  0.5315  0.53595 0.56995 0.57175 0.5673\n",
      "  0.8011  0.5491  0.329   0.5642  1.     ]\n",
      " [0.6969  0.68625 0.5396  0.61145 0.6293  0.5365  0.5289  0.      0.6797\n",
      "  1.      0.6433  0.64695 0.6468  0.405   0.6836  0.6951  0.64385 0.67845\n",
      "  0.69155 0.68895 0.573   0.6752  0.7108 ]\n",
      " [0.68435 0.6748  0.63195 0.6606  0.6609  0.60155 0.6246  0.6797  0.\n",
      "  1.      0.68935 0.6907  0.68355 0.67625 0.67195 0.66255 0.66395 0.6694\n",
      "  0.655   0.68045 0.6202  0.6836  0.684  ]\n",
      " [0.5595  0.5462  0.5282  0.1599  0.1373  0.5276  0.5277  1.      1.\n",
      "  0.      0.52155 0.5126  0.5442  1.      0.5406  0.5656  0.40625 0.5583\n",
      "  1.      0.5489  0.519   0.5711  1.     ]\n",
      " [0.72225 0.73075 0.5539  0.6975  0.69155 0.5418  0.5419  0.6433  0.68935\n",
      "  0.52155 0.      0.69175 0.6051  0.6178  0.72705 0.72915 0.6833  0.73895\n",
      "  0.7175  0.71765 0.5675  0.7111  0.7411 ]\n",
      " [0.709   0.6959  0.54175 0.69215 0.68235 0.5309  0.5315  0.64695 0.6907\n",
      "  0.5126  0.69175 0.      0.6182  0.63875 0.66445 0.6795  0.66    0.7019\n",
      "  0.56895 0.6772  0.5573  0.6624  0.7565 ]\n",
      " [0.7166  0.7057  0.5699  0.7032  0.69145 0.5582  0.5592  0.6468  0.68355\n",
      "  0.5442  0.6051  0.6182  0.      0.6496  0.7046  0.6904  0.6747  0.7169\n",
      "  0.71975 0.68575 0.5793  0.69035 0.7599 ]\n",
      " [0.6984  0.66865 0.5347  0.63775 0.63795 0.5311  0.5315  0.405   0.67625\n",
      "  1.      0.6178  0.63875 0.6496  0.      0.68235 0.69575 0.6474  0.69065\n",
      "  0.6959  0.67775 0.56745 0.6493  0.7173 ]\n",
      " [0.6808  0.5966  0.5456  0.6783  0.68655 0.53545 0.53595 0.6836  0.67195\n",
      "  0.5406  0.72705 0.66445 0.7046  0.68235 0.      0.6843  0.6947  0.6468\n",
      "  0.6891  0.6467  0.5598  0.6004  0.74375]\n",
      " [0.5967  0.6711  0.58095 0.679   0.68485 0.56895 0.56995 0.6951  0.66255\n",
      "  0.5656  0.72915 0.6795  0.6904  0.69575 0.6843  0.      0.6921  0.636\n",
      "  0.7005  0.5785  0.5906  0.6622  0.753  ]\n",
      " [0.54235 0.6896  0.5761  0.6779  0.6619  0.5767  0.57175 0.64385 0.66395\n",
      "  0.40625 0.6833  0.66    0.6747  0.6474  0.6947  0.6921  0.      0.6497\n",
      "  0.71095 0.61825 0.6069  0.6615  0.7321 ]\n",
      " [0.5636  0.6501  0.57805 0.70575 0.7044  0.5664  0.5673  0.67845 0.6694\n",
      "  0.5583  0.73895 0.7019  0.7169  0.69065 0.6468  0.636   0.6497  0.\n",
      "  0.7251  0.658   0.59    0.6129  0.7554 ]\n",
      " [0.72535 0.71505 0.52545 0.5842  0.5242  0.7842  0.8011  0.69155 0.655\n",
      "  1.      0.7175  0.56895 0.71975 0.6959  0.6891  0.7005  0.71095 0.7251\n",
      "  0.      0.72085 0.87175 0.71345 0.71115]\n",
      " [0.5001  0.6964  0.56025 0.7122  0.718   0.54885 0.5491  0.68895 0.68045\n",
      "  0.5489  0.71765 0.6772  0.68575 0.67775 0.6467  0.5785  0.61825 0.658\n",
      "  0.72085 0.      0.574   0.6242  0.7537 ]\n",
      " [0.5829  0.5703  0.363   0.58365 0.5825  0.3428  0.329   0.573   0.6202\n",
      "  0.519   0.5675  0.5573  0.5793  0.56745 0.5598  0.5906  0.6069  0.59\n",
      "  0.87175 0.574   0.      0.5871  1.     ]\n",
      " [0.6143  0.586   0.57415 0.72395 0.7193  0.56335 0.5642  0.6752  0.6836\n",
      "  0.5711  0.7111  0.6624  0.69035 0.6493  0.6004  0.6622  0.6615  0.6129\n",
      "  0.71345 0.6242  0.5871  0.      0.74615]\n",
      " [0.75655 0.7319  1.      0.72475 0.7027  0.77595 1.      0.7108  0.684\n",
      "  1.      0.7411  0.7565  0.7599  0.7173  0.74375 0.753   0.7321  0.7554\n",
      "  0.71115 0.7537  1.      0.74615 0.     ]] 1.0 0.0\n",
      "\n",
      " . This analysis will run on 1 threads.\n",
      "\n",
      "#  Analysing dataset 1\n",
      "\n",
      " . Computing tree...\n",
      "\n",
      " . Performing NNI...\n",
      "\n",
      " . Warning: \n",
      "\tExplained variance = 0.850 (<0.90).\n",
      "\tCheck your input data.\n",
      "\n",
      " . Time used 0h00m00s\n",
      "\n",
      "   /-Q06BR2.pdb\n",
      "  |\n",
      "  |--P14599.pdb\n",
      "  |\n",
      "  |      /-Q10651.pdb\n",
      "--|   /-|\n",
      "  |  |   \\-P79307.pdb\n",
      "  |  |\n",
      "  |  |      /-P12023.pdb\n",
      "  |  |   /-|\n",
      "  |  |  |  |   /-Q95241.pdb\n",
      "   \\-|  |   \\-|\n",
      "     |  |      \\-P08592.pdb\n",
      "     |  |\n",
      "     |  |         /-Q28280.pdb\n",
      "     |  |      /-|\n",
      "     |  |     |  |   /-Q28053.pdb\n",
      "     |  |     |   \\-|\n",
      "      \\-|     |     |   /-Q28748.pdb\n",
      "        |     |      \\-|\n",
      "        |   /-|         \\-Q28757.pdb\n",
      "        |  |  |\n",
      "        |  |  |      /-Q03157.pdb\n",
      "        |  |  |   /-|\n",
      "        |  |  |  |   \\-P51693.pdb\n",
      "        |  |   \\-|\n",
      "        |  |     |   /-O73683.pdb\n",
      "         \\-|      \\-|\n",
      "           |         \\-O93279.pdb\n",
      "           |\n",
      "           |      /-P05067.pdb\n",
      "           |   /-|\n",
      "           |  |  |   /-Q60495.pdb\n",
      "           |  |   \\-|\n",
      "           |  |     |   /-Q5IS80.pdb\n",
      "            \\-|      \\-|\n",
      "              |         \\-P53601.pdb\n",
      "              |\n",
      "              |   /-Q06335.pdb\n",
      "               \\-|\n",
      "                 |   /-P15943.pdb\n",
      "                  \\-|\n",
      "                    |   /-Q29149.pdb\n",
      "                     \\-|\n",
      "                        \\-Q06481.pdb\n",
      "\n",
      "MAD phylogenetic rooting\n",
      "\n",
      "Analyzing file '../testdata/alntmscorestructblob_tree.nwk'...\n",
      ">> [MAD=0.087_AI=0.989_CCV=9.97%_N=1/1]\n",
      "\n",
      "Minimal ancestor deviation, MAD = 0.087\n",
      "           Ambiguity index,  AI = 0.989\n",
      "                  Clock CV, CCV = 9.97%\n",
      "Rooted tree written to '../testdata/alntmscorestructblob_tree.nwk.rooted'\n",
      "\n",
      "\n",
      "    - Please cite DOI:10.1038/s41559-017-0193\n",
      "\n",
      "[[0.      0.1586  0.1938  0.1827  0.1998  0.17185 0.1911  0.2437  0.2841\n",
      "  0.35835 0.13755 0.09415 0.15565 0.2246  0.14375 0.0838  0.1759  0.1402\n",
      "  0.2676  0.0633  0.1974  0.1617  0.30365]\n",
      " [0.1586  0.      0.34305 0.1938  0.19945 0.3253  0.3397  0.27765 0.3\n",
      "  0.446   0.1846  0.1541  0.1983  0.2574  0.09195 0.1459  0.22015 0.10875\n",
      "  0.2836  0.1527  0.33195 0.0996  0.311  ]\n",
      " [0.1938  0.34305 0.      0.341   0.31725 0.0304  0.0324  0.2494  0.3157\n",
      "  0.35515 0.2024  0.1962  0.20615 0.2361  0.30855 0.19895 0.27385 0.267\n",
      "  0.34625 0.20665 0.09565 0.295   1.     ]\n",
      " [0.1827  0.1938  0.341   0.      0.1226  0.3336  0.3461  0.2582  0.29775\n",
      "  0.1822  0.1859  0.1841  0.2144  0.24705 0.1862  0.187   0.1981  0.1894\n",
      "  0.2691  0.1897  0.3122  0.1948  0.31945]\n",
      " [0.1998  0.19945 0.31725 0.1226  0.      0.3172  0.33485 0.2609  0.3041\n",
      "  0.1868  0.1938  0.18735 0.2146  0.24915 0.1851  0.187   0.20105 0.187\n",
      "  0.2788  0.2013  0.3082  0.2063  0.3036 ]\n",
      " [0.17185 0.3253  0.0304  0.3336  0.3172  0.      0.03295 0.23775 0.396\n",
      "  0.35525 0.1838  0.18115 0.1846  0.22395 0.2958  0.1809  0.30135 0.24915\n",
      "  0.69275 0.1898  0.09085 0.28115 0.6367 ]\n",
      " [0.1911  0.3397  0.0324  0.3461  0.33485 0.03295 0.      0.26615 0.414\n",
      "  0.35345 0.205   0.19855 0.20365 0.2569  0.3065  0.19415 0.3008  0.2657\n",
      "  0.7091  0.2107  0.0918  0.30015 1.     ]\n",
      " [0.2437  0.27765 0.2494  0.2582  0.2609  0.23775 0.26615 0.      0.2899\n",
      "  1.      0.23    0.2384  0.2527  0.0813  0.25335 0.2478  0.25265 0.27\n",
      "  0.31265 0.2387  0.27685 0.2664  0.32855]\n",
      " [0.2841  0.3     0.3157  0.29775 0.3041  0.396   0.414   0.2899  0.\n",
      "  1.      0.27545 0.2749  0.2849  0.27835 0.29015 0.27235 0.26945 0.29685\n",
      "  0.3032  0.2729  0.387   0.3283  0.3371 ]\n",
      " [0.35835 0.446   0.35515 0.1822  0.1868  0.35525 0.35345 1.      1.\n",
      "  0.      0.36715 0.36625 0.3677  1.      0.43845 0.3653  0.2488  0.4193\n",
      "  1.      0.3682  0.323   0.4255  1.     ]\n",
      " [0.13755 0.1846  0.2024  0.1859  0.1938  0.1838  0.205   0.23    0.27545\n",
      "  0.36715 0.      0.1275  0.1474  0.20165 0.1781  0.14095 0.1905  0.1752\n",
      "  0.26315 0.1306  0.20695 0.19215 0.2926 ]\n",
      " [0.09415 0.1541  0.1962  0.1841  0.18735 0.18115 0.19855 0.2384  0.2749\n",
      "  0.36625 0.1275  0.      0.14575 0.23195 0.15365 0.0778  0.19025 0.1477\n",
      "  0.2773  0.07715 0.2125  0.16605 0.3074 ]\n",
      " [0.15565 0.1983  0.20615 0.2144  0.2146  0.1846  0.20365 0.2527  0.2849\n",
      "  0.3677  0.1474  0.14575 0.      0.2396  0.18645 0.1433  0.2068  0.1931\n",
      "  0.27955 0.1442  0.1994  0.20905 0.31105]\n",
      " [0.2246  0.2574  0.2361  0.24705 0.24915 0.22395 0.2569  0.0813  0.27835\n",
      "  1.      0.20165 0.23195 0.2396  0.      0.24365 0.23945 0.23625 0.257\n",
      "  0.3051  0.21555 0.2703  0.2602  0.315  ]\n",
      " [0.14375 0.09195 0.30855 0.1862  0.1851  0.2958  0.3065  0.25335 0.29015\n",
      "  0.43845 0.1781  0.15365 0.18645 0.24365 0.      0.1391  0.23635 0.10325\n",
      "  0.277   0.13695 0.30315 0.109   0.30465]\n",
      " [0.0838  0.1459  0.19895 0.187   0.187   0.1809  0.19415 0.2478  0.27235\n",
      "  0.3653  0.14095 0.0778  0.1433  0.23945 0.1391  0.      0.18525 0.13565\n",
      "  0.2828  0.0707  0.19325 0.1583  0.31175]\n",
      " [0.1759  0.22015 0.27385 0.1981  0.20105 0.30135 0.3008  0.25265 0.26945\n",
      "  0.2488  0.1905  0.19025 0.2068  0.23625 0.23635 0.18525 0.      0.22095\n",
      "  0.2624  0.18755 0.30225 0.23565 0.30715]\n",
      " [0.1402  0.10875 0.267   0.1894  0.187   0.24915 0.2657  0.27    0.29685\n",
      "  0.4193  0.1752  0.1477  0.1931  0.257   0.10325 0.13565 0.22095 0.\n",
      "  0.26915 0.1459  0.2712  0.11225 0.2996 ]\n",
      " [0.2676  0.2836  0.34625 0.2691  0.2788  0.69275 0.7091  0.31265 0.3032\n",
      "  1.      0.26315 0.2773  0.27955 0.3051  0.277   0.2828  0.2624  0.26915\n",
      "  0.      0.2847  0.7296  0.28745 0.26315]\n",
      " [0.0633  0.1527  0.20665 0.1897  0.2013  0.1898  0.2107  0.2387  0.2729\n",
      "  0.3682  0.1306  0.07715 0.1442  0.21555 0.13695 0.0707  0.18755 0.1459\n",
      "  0.2847  0.      0.2134  0.16345 0.31715]\n",
      " [0.1974  0.33195 0.09565 0.3122  0.3082  0.09085 0.0918  0.27685 0.387\n",
      "  0.323   0.20695 0.2125  0.1994  0.2703  0.30315 0.19325 0.30225 0.2712\n",
      "  0.7296  0.2134  0.      0.30325 1.     ]\n",
      " [0.1617  0.0996  0.295   0.1948  0.2063  0.28115 0.30015 0.2664  0.3283\n",
      "  0.4255  0.19215 0.16605 0.20905 0.2602  0.109   0.1583  0.23565 0.11225\n",
      "  0.28745 0.16345 0.30325 0.      0.3244 ]\n",
      " [0.30365 0.311   1.      0.31945 0.3036  0.6367  1.      0.32855 0.3371\n",
      "  1.      0.2926  0.3074  0.31105 0.315   0.30465 0.31175 0.30715 0.2996\n",
      "  0.26315 0.31715 1.      0.3244  0.     ]] 1.0 0.0\n",
      "\n",
      " . This analysis will run on 1 threads.\n",
      "\n",
      "#  Analysing dataset 1\n",
      "\n",
      " . Computing tree...\n",
      "\n",
      " . Performing NNI...\n",
      "\n",
      " . Warning: \n",
      "\tExplained variance = 0.591 (<0.90).\n",
      "\tCheck your input data.\n",
      "\n",
      " . Time used 0h00m00s\n",
      "\n",
      "   /-Q10651.pdb\n",
      "  |\n",
      "  |--P14599.pdb\n",
      "  |\n",
      "--|   /-Q06BR2.pdb\n",
      "  |  |\n",
      "  |  |      /-P05067.pdb\n",
      "  |  |   /-|\n",
      "  |  |  |  |   /-Q95241.pdb\n",
      "   \\-|  |   \\-|\n",
      "     |  |     |   /-P12023.pdb\n",
      "     |  |      \\-|\n",
      "     |  |         \\-P08592.pdb\n",
      "     |  |\n",
      "      \\-|      /-P79307.pdb\n",
      "        |   /-|\n",
      "        |  |  |   /-Q60495.pdb\n",
      "        |  |   \\-|\n",
      "        |  |     |   /-Q5IS80.pdb\n",
      "        |  |      \\-|\n",
      "        |  |         \\-P53601.pdb\n",
      "         \\-|\n",
      "           |      /-Q06335.pdb\n",
      "           |   /-|\n",
      "           |  |  |   /-Q06481.pdb\n",
      "           |  |   \\-|\n",
      "           |  |     |   /-Q29149.pdb\n",
      "            \\-|      \\-|\n",
      "              |         \\-P15943.pdb\n",
      "              |\n",
      "              |   /-O93279.pdb\n",
      "              |  |\n",
      "               \\-|   /-O73683.pdb\n",
      "                 |  |\n",
      "                  \\-|      /-Q03157.pdb\n",
      "                    |   /-|\n",
      "                    |  |   \\-P51693.pdb\n",
      "                     \\-|\n",
      "                       |   /-Q28280.pdb\n",
      "                        \\-|\n",
      "                          |   /-Q28757.pdb\n",
      "                           \\-|\n",
      "                             |   /-Q28748.pdb\n",
      "                              \\-|\n",
      "                                 \\-Q28053.pdb\n",
      "\n",
      "MAD phylogenetic rooting\n",
      "\n",
      "Analyzing file '../testdata/lddtstructblob_tree.nwk'...\n",
      ">> [MAD=0.299_AI=0.999_CCV=32.0%_N=1/1]\n",
      "\n",
      "Minimal ancestor deviation, MAD = 0.299\n",
      "           Ambiguity index,  AI = 0.999\n",
      "                  Clock CV, CCV = 32.0%\n",
      "Rooted tree written to '../testdata/lddtstructblob_tree.nwk.rooted'\n",
      "\n",
      "\n",
      "    - Please cite DOI:10.1038/s41559-017-0193\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#make a foldtree with distmat\n",
    "import foldseek2tree\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "outfolder = '../testdata/'\n",
    "res = pd.read_table('../testdata/structsaln_2_allvall.csv', header = None)\n",
    "columns = 'query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,lddt,qaln,taln,cigar,alntmscore'.split(',')\n",
    "res.columns= columns\n",
    "ids = list( set(list(res['query'].unique()) + list(res['target'].unique())))\n",
    "pos = { protid : i for i,protid in enumerate(ids)}\n",
    "kernels = ['fident', 'alntmscore', 'lddt']\n",
    "matrices = { k:np.zeros((len(pos), len(pos))) for k in kernels }\n",
    "print(res)\n",
    "\n",
    "#calc kernel for tm, aln score, lddt\n",
    "for idx,row in res.iterrows():\n",
    "    for k in matrices:\n",
    "        matrices[k][pos[row['query']] , pos[row['target']]] += row[k]\n",
    "        matrices[k][pos[row['target']] , pos[row['query']]] += row[k]\n",
    "\n",
    "trees = {}\n",
    "for i,k in enumerate(matrices):\n",
    "    matrices[k] /= 2\n",
    "    matrices[k] = 1-matrices[k]\n",
    "    print(matrices[k], np.amax(matrices[k]), np.amin(matrices[k]) )\n",
    "    np.save( outfolder + k + '_distmat.npy' , matrices[k])\n",
    "    distmat_txt = foldseek2tree.distmat_to_txt( ids , matrices[k] , outfolder + k + '_distmat.txt' )\n",
    "    out_tree = foldseek2tree.runFastme(  fastmepath = 'fastme' , clusterfile = distmat_txt )\n",
    "    out_tree = foldseek2tree.postprocess(out_tree, outfolder + k + 'structblob_tree.nwk' , delta = 10**-10)\n",
    "    out_tree = foldseek2tree.runargs('../madroot/mad ' + out_tree )\n",
    "    trees[k] = out_tree\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46c3aa97-a047-4771-a70c-9f8b6e31fd72",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f7d26066-22b4-4d55-9d65-2a8a437eea0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "alndf0 = pd.read_table('../testdata/structsaln_0_allvall.csv' , header = None)\n",
    "alndf2 = pd.read_table('../testdata/structsaln_2_allvall.csv', header = None)\n",
    "columns = 'query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,lddt,qaln,taln,cigar,lntmscore'.split(',')\n",
    "alndf0.columns = columns\n",
    "alndf2.columns = columns\n",
    "\n",
    "alndf0['3diq']= alndf0['query'].map(mapper3di)\n",
    "alndf0['3dit']= alndf0['target'].map(mapper3di)\n",
    "\n",
    "\n",
    "alndf0['AAq']= alndf0['query'].map(mapperAA)\n",
    "alndf0['AAt']= alndf0['target'].map(mapperAA)\n",
    "\n",
    "\n",
    "alndf0 = alndf0.set_index(['query', 'target'])\n",
    "alndf2 = alndf2.set_index(['query', 'target'])\n",
    "\n",
    "alndf = alndf0.merge(alndf2, left_index=True, right_on=['query', 'target'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bf29915e-0b20-45f6-a727-3ca5ff72c102",
   "metadata": {},
   "outputs": [],
   "source": [
    "#use the qaln and taln of x and y to derive the other side fident\n",
    "import numpy as np\n",
    "\n",
    "def Fident(str1,str2 , verbose = False):\n",
    "    #minlen= min( (len(str1),len(str2))  )\n",
    "    #str1 = str1[:minlen]\n",
    "    #str2 = str2[:minlen]\n",
    "    str1 = np.array(list(str1))\n",
    "    str2 = np.array(list(str2))            \n",
    "    return len(np.where( (str1 == str2 ) & (str1 != '-' ) & (str2 != '-')  )[0]) / len(str1)\n",
    "\n",
    "def copyaln( aln, seq):\n",
    "    seqiter = iter(seq)\n",
    "    newaln = ''\n",
    "    for i,char in enumerate(aln):\n",
    "        if char == '-':\n",
    "            newaln += '-'\n",
    "        else:\n",
    "            newaln+=next(seqiter)\n",
    "    return newaln\n",
    "\n",
    "def calc_fident_crossaln(row , verbose = False):\n",
    "    #amino acid representations of alns using AAand3di or just 3di\n",
    "    qaln_0,qaln_2, taln_0, taln_2 = row.qaln_x, row.qaln_y, row.taln_x, row.taln_y\n",
    "    #start and stop of aln\n",
    "    \n",
    "    qstart_0,qend_0, tstart_0 , tend_0, qstart_2, qend_2, tstart_2 , tend_2 = row.qstart_x, row.qend_x , row.tstart_x , row.tend_x , row.qstart_y , row.qend_y , row.tstart_y , row.tend_y\n",
    "    #indexing starts at 1...\n",
    "    \n",
    "    #3di of the query and target\n",
    "    structQ, structT = row['3diq'], row['3dit']\n",
    "    AAq, AAt = row['AAq'], row['AAt']\n",
    "\n",
    "    #add gaps\n",
    "    q3di_newgaps = copyaln(qaln_0, structQ[qstart_0-1:qend_0])\n",
    "    t3di_newgaps = copyaln(taln_0, structT[tstart_0-1:tend_0])\n",
    "    t3diAA_newgaps = copyaln(taln_2, structT[tstart_2-1:tend_2]) \n",
    "    q3diAA_newgaps = copyaln(qaln_2, structQ[qstart_2-1:qend_2])\n",
    "    row = pd.Series({ 'fident_3di_mode2' :Fident(q3di_newgaps , t3di_newgaps , verbose = verbose) ,  'fident_3di_mode0' :Fident(q3di_newgaps , t3di_newgaps, verbose = verbose) , \n",
    "     '3di_qaln_mode0':q3di_newgaps , '3di_taln_mode0':t3di_newgaps , '3di_qaln_mode2':q3diAA_newgaps , '3di_taln_mode2':t3diAA_newgaps })\n",
    "    #return columns\n",
    "    return row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "36161816-63bd-46ee-99e1-4d4518f25d1d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fident_x                                                   0.836\n",
      "alnlen_x                                                     773\n",
      "mismatch_x                                                   121\n",
      "gapopen_x                                                      3\n",
      "qstart_x                                                       1\n",
      "qend_x                                                       770\n",
      "tstart_x                                                       1\n",
      "tend_x                                                       770\n",
      "evalue_x                                                     0.0\n",
      "bits_x                                                       813\n",
      "lddt_x                                                    0.8428\n",
      "qaln_x         MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...\n",
      "taln_x         MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...\n",
      "cigar_x                                      603M2D84M3I52M1D28M\n",
      "lntmscore_x                                                0.342\n",
      "3diq           DDDDDDDDDDDDDDDDDDDDDPDAPPDFPDFWWKWADAQFFIWIQP...\n",
      "3dit           DDDPDDPDPPPDDPDPDPPPPPPPPPQFPDFWWKWADAQFFIWIQD...\n",
      "AAq            MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...\n",
      "AAt            MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...\n",
      "fident_y                                                   0.998\n",
      "alnlen_y                                                     770\n",
      "mismatch_y                                                     2\n",
      "gapopen_y                                                      0\n",
      "qstart_y                                                       1\n",
      "qend_y                                                       770\n",
      "tstart_y                                                       1\n",
      "tend_y                                                       770\n",
      "evalue_y                                                     0.0\n",
      "bits_y                                                      2576\n",
      "lddt_y                                                    0.8469\n",
      "qaln_y         MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...\n",
      "taln_y         MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...\n",
      "cigar_y                                                     770M\n",
      "lntmscore_y                                                0.342\n",
      "Name: (P05067.pdb, Q5IS80.pdb), dtype: object\n",
      "MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMNVQNGKWDSDPSGTKTCIDTKEGILQYCQEVYPELQITNVVEANQPVTIQNWCKRGRKQCKTHPHFVIPYRCLVGEFVSDALLVPDKCKFLHQERMDVCETHLHWHTVAKETCSEKSTNLHDYGMLLPCGIDKFRGVEFVCCPLAEESDNVDSADAEEDDSDVWWGGADTDYADGSEDKVVEVAEEEEVAEVEEEEADDDEDDEDGDEVEEEAEEPYEEATERTTSIATTTTTTTESVEEVVREVCSEQAETGPCRAMISRWYFDVTEGKCAPFFYGGCGGNRNNFDTEEYCMAVCGSAMSQSLLKTTQEPLARDPVKLPTTAASTPDAVDKYLETPGDENEHAHFQKAKERLEAKHRERMSQVMREWEEAERQAKNLPKADKKAVIQHFQEKVESLEQEAANERQQLVETHMARVEAMLNDRRRLALENYITALQAVPPRPRHVFNMLKKYVRAEQKDRQHTLKHFEHVRMVDPKKAAQIRSQVMTHLRVIYERMNQSLSLLYNVPAVAEEIQDEVDELLQKEQNYSDDVLANMISEPRISYGNDALMPSLTETKTTVELLPVNGEFSLDDLQPWHSFGADSVPANTENEVEPVDARPAADRGLTTRPGSGLTNIKTEEISEVKMDAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVVIATVIVITLVMLKKKQYTSIHHGVVEVDAAVTPEERHLSKMQQNGYENPTYKFFEQMQN\n",
      "MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMNVQNGKWDSDPSGTKTCIDTKEGILQYCQEVYPELQITNVVEANQPVTIQNWCKRGRKQCKTHPHFVIPYRCLVGEFVSDALLVPDKCKFLHQERMDVCETHLHWHTVAKETCSEKSTNLHDYGMLLPCGIDKFRGVEFVCCPLAEESDNVDSADAEEDDSDVWWGGADTDYADGSEDKVVEVAEEEEVAEVEEEEADDDEDDEDGDEVEEEAEEPYEEATERTTSIATTTTTTTESVEEVVREVCSEQAETGPCRAMISRWYFDVTEGKCAPFFYGGCGGNRNNFDTEEYCMAVCGSAMSQSLLKTTQEPLARDPVKLPTTAASTPDAVDKYLETPGDENEHAHFQKAKERLEAKHRERMSQVMREWEEAERQAKNLPKADKKAVIQHFQEKVESLEQEAANERQQLVETHMARVEAMLNDRRRLALENYITALQAVPPRPRHVFNMLKKYVRAEQKDRQHTLKHFEHVRMVDPKKAAQIRSQVMTHLRVIYERMNQSLSLLYNVPAVAEEIQDEVDELLQKEQNYSDDVLANMISEPRISYGNDALMPSLTETKTT--VELLPVNGEFSLDDLQPWHSFGADSVPANTENEVEPVDARPAADRGLTTRPGSGLTNIKTEEISEVKMDAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVVIATVIVITLVMLKKKQYTSIHHGVVEVDAAV-TPEERHLSKMQQNGYENPTYKFFEQMQN\n",
      "                       fident_x  alnlen_x  mismatch_x  gapopen_x  qstart_x  \\\n",
      "query      target                                                            \n",
      "Q5IS80.pdb Q5IS80.pdb     1.000       770           0          0         1   \n",
      "           P53601.pdb     0.910       770          68          1         2   \n",
      "           Q60495.pdb     0.915       772          61          3         2   \n",
      "           P79307.pdb     0.863       770         105          0         1   \n",
      "           P05067.pdb     0.836       773         121          3         1   \n",
      "...                         ...       ...         ...        ...       ...   \n",
      "Q29149.pdb P05067.pdb     0.846        13           2          0         1   \n",
      "           Q06481.pdb     0.785        14           3          0         1   \n",
      "           P12023.pdb     0.846        13           2          0         1   \n",
      "           Q28053.pdb     1.000        10           0          0         1   \n",
      "           P08592.pdb     0.687        16           5          0         1   \n",
      "\n",
      "                       qend_x  tstart_x  tend_x       evalue_x  bits_x  ...  \\\n",
      "query      target                                                       ...   \n",
      "Q5IS80.pdb Q5IS80.pdb     770         1     770   0.000000e+00    1049  ...   \n",
      "           P53601.pdb     770         1     770  2.038000e-288     941  ...   \n",
      "           Q60495.pdb     770         1     770  2.038000e-288     941  ...   \n",
      "           P79307.pdb     770         1     770  1.123000e-268     882  ...   \n",
      "           P05067.pdb     770         1     770  1.398000e-256     845  ...   \n",
      "...                       ...       ...     ...            ...     ...  ...   \n",
      "Q29149.pdb P05067.pdb      13       678     690   1.977000e+00      18  ...   \n",
      "           Q06481.pdb      14       241     254   2.757000e+00      18  ...   \n",
      "           P12023.pdb      13       200     212   2.757000e+00      18  ...   \n",
      "           Q28053.pdb      10        12      21   5.356000e+00      17  ...   \n",
      "           P08592.pdb      16       602     617   7.463000e+00      16  ...   \n",
      "\n",
      "                                                                  qaln_y  \\\n",
      "query      target                                                          \n",
      "Q5IS80.pdb Q5IS80.pdb  MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...   \n",
      "           P53601.pdb  MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...   \n",
      "           Q60495.pdb  MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...   \n",
      "           P79307.pdb  MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...   \n",
      "           P05067.pdb  MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...   \n",
      "...                                                                  ...   \n",
      "Q29149.pdb P05067.pdb  SEVKMDAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVVI...   \n",
      "           Q06481.pdb                            GAIIGLMVGGVVIATVIVITLVM   \n",
      "           P12023.pdb  SEVKMDAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVVI...   \n",
      "           Q28053.pdb  SEVKMDAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVVI...   \n",
      "           P08592.pdb  SEVKMDAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVVI...   \n",
      "\n",
      "                                                                  taln_y  \\\n",
      "query      target                                                          \n",
      "Q5IS80.pdb Q5IS80.pdb  MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...   \n",
      "           P53601.pdb  MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...   \n",
      "           Q60495.pdb  MLPSLALLLLTTWTARALEVPTDGNAGLLAEPQIAMFCGKLNMHMN...   \n",
      "           P79307.pdb  MLPGLALVLLAAWTARALEVPTDGNAGLLAEPQVAMFCGKLNMHMN...   \n",
      "           P05067.pdb  MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...   \n",
      "...                                                                  ...   \n",
      "Q29149.pdb P05067.pdb  SEVKMDAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVVI...   \n",
      "           Q06481.pdb                            SALIGLLVIAVAIATVIVISLVM   \n",
      "           P12023.pdb  SEVKMDAEFGHDSGFEVRHQKLVFFAEDVGSNKGAIIGLMVGGVVI...   \n",
      "           Q28053.pdb  SEVKMDAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVVI...   \n",
      "           P08592.pdb  SEVKMDAEFGHDSGFEVRHQKLVFFAEDVGSNKGAIIGLMVGGVVI...   \n",
      "\n",
      "                      cigar_y lntmscore_y  fident_3di_mode2 fident_3di_mode0  \\\n",
      "query      target                                                              \n",
      "Q5IS80.pdb Q5IS80.pdb    770M      1.0000          1.000000         1.000000   \n",
      "           P53601.pdb    770M      0.4999          0.910390         0.910390   \n",
      "           Q60495.pdb    770M      0.4215          0.915803         0.915803   \n",
      "           P79307.pdb    770M      0.3228          0.863636         0.863636   \n",
      "           P05067.pdb    770M      0.3420          0.836999         0.836999   \n",
      "...                       ...         ...               ...              ...   \n",
      "Q29149.pdb P05067.pdb     56M      0.4417          0.846154         0.846154   \n",
      "           Q06481.pdb     23M      0.8627          0.785714         0.785714   \n",
      "           P12023.pdb     56M      0.4289          0.846154         0.846154   \n",
      "           Q28053.pdb     56M      0.4718          1.000000         1.000000   \n",
      "           P08592.pdb     56M      0.4538          0.687500         0.687500   \n",
      "\n",
      "                                                          3di_qaln_mode0  \\\n",
      "query      target                                                          \n",
      "Q5IS80.pdb Q5IS80.pdb  DDDPDDPDPPPDDPDPDPPPPPPPPPQFPDFWWKWADAQFFIWIQD...   \n",
      "           P53601.pdb  DDPDDPDPPPDDPDPDP-PPPPPPPPQFPDFWWKWADAQFFIWIQD...   \n",
      "           Q60495.pdb  DDPDDPDPPPDDPDPDP-PPPPPPPPQFPDFWWKWADAQFFIWIQD...   \n",
      "           P79307.pdb  DDDPDDPDPPPDDPDPDPPPPPPPPPQFPDFWWKWADAQFFIWIQD...   \n",
      "           P05067.pdb  DDDPDDPDPPPDDPDPDPPPPPPPPPQFPDFWWKWADAQFFIWIQD...   \n",
      "...                                                                  ...   \n",
      "Q29149.pdb P05067.pdb                                      DDDDDPPPPPPPP   \n",
      "           Q06481.pdb                                     DDDDDPPPPPPPPP   \n",
      "           P12023.pdb                                      DDDDDPPPPPPPP   \n",
      "           Q28053.pdb                                         DDDDDPPPPP   \n",
      "           P08592.pdb                                   DDDDDPPPPPPPPPVV   \n",
      "\n",
      "                                                          3di_taln_mode0  \\\n",
      "query      target                                                          \n",
      "Q5IS80.pdb Q5IS80.pdb  DDDPDDPDPPPDDPDPDPPPPPPPPPQFPDFWWKWADAQFFIWIQD...   \n",
      "           P53601.pdb  DDPDPPPPPPPPPPPDPDPPPDPPPPQFPDFWWWWADAQFFIWIQD...   \n",
      "           Q60495.pdb  DDDDDPDDDDDDDDDDPDPDPDPPPPQFPDFWWKWADAQFFIWIQD...   \n",
      "           P79307.pdb  DDDDDDDDDDDDPPPPPPPPPDPPPPQFPDFWWKWADAQFFIWIQP...   \n",
      "           P05067.pdb  DDDDDDDDDDDDDDDDDDDDDPDAPPDFPDFWWKWADAQFFIWIQP...   \n",
      "...                                                                  ...   \n",
      "Q29149.pdb P05067.pdb                                      DDDDDPPPPDDPP   \n",
      "           Q06481.pdb                                     DDDDDPPDDPPVPP   \n",
      "           P12023.pdb                                      DDDDDPPPPDPQP   \n",
      "           Q28053.pdb                                         DDDDDPPPPP   \n",
      "           P08592.pdb                                   DDDDDDDDDPPPDPVV   \n",
      "\n",
      "                                                          3di_qaln_mode2  \\\n",
      "query      target                                                          \n",
      "Q5IS80.pdb Q5IS80.pdb  DDDPDDPDPPPDDPDPDPPPPPPPPPQFPDFWWKWADAQFFIWIQD...   \n",
      "           P53601.pdb  DDDPDDPDPPPDDPDPDPPPPPPPPPQFPDFWWKWADAQFFIWIQD...   \n",
      "           Q60495.pdb  DDDPDDPDPPPDDPDPDPPPPPPPPPQFPDFWWKWADAQFFIWIQD...   \n",
      "           P79307.pdb  DDDPDDPDPPPDDPDPDPPPPPPPPPQFPDFWWKWADAQFFIWIQD...   \n",
      "           P05067.pdb  DDDPDDPDPPPDDPDPDPPPPPPPPPQFPDFWWKWADAQFFIWIQD...   \n",
      "...                                                                  ...   \n",
      "Q29149.pdb P05067.pdb  DDDDDPPPPPPPPPVVVVVVVVVVVCVVVVCPVVVVVCCVVVVVVV...   \n",
      "           Q06481.pdb                            VVVVCCVVVVVVVVVVVVVVVVV   \n",
      "           P12023.pdb  DDDDDPPPPPPPPPVVVVVVVVVVVCVVVVCPVVVVVCCVVVVVVV...   \n",
      "           Q28053.pdb  DDDDDPPPPPPPPPVVVVVVVVVVVCVVVVCPVVVVVCCVVVVVVV...   \n",
      "           P08592.pdb  DDDDDPPPPPPPPPVVVVVVVVVVVCVVVVCPVVVVVCCVVVVVVV...   \n",
      "\n",
      "                                                          3di_taln_mode2  \n",
      "query      target                                                         \n",
      "Q5IS80.pdb Q5IS80.pdb  DDDPDDPDPPPDDPDPDPPPPPPPPPQFPDFWWKWADAQFFIWIQD...  \n",
      "           P53601.pdb  DDPDPPPPPPPPPPPDPDPPPDPPPPQFPDFWWWWADAQFFIWIQD...  \n",
      "           Q60495.pdb  DDDDDPDDDDDDDDDDPDPDPDPPPPQFPDFWWKWADAQFFIWIQD...  \n",
      "           P79307.pdb  DDDDDDDDDDDDPPPPPPPPPDPPPPQFPDFWWKWADAQFFIWIQP...  \n",
      "           P05067.pdb  DDDDDDDDDDDDDDDDDDDDDPDAPPDFPDFWWKWADAQFFIWIQP...  \n",
      "...                                                                  ...  \n",
      "Q29149.pdb P05067.pdb  DDDDDDDDDDDDDDDDPPPPDDPPDDDDDDDDDVVVVVVVVVVVVV...  \n",
      "           Q06481.pdb                            VNVVVVVVVVVVVVVVVVVVVVV  \n",
      "           P12023.pdb  DDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDPVVVVVVVVVVVVV...  \n",
      "           Q28053.pdb  DPPDPPDDPPDDDDDPPPPPDDDDDPDPDPDPVVSVVCCVVVVVVV...  \n",
      "           P08592.pdb  DDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDPDVVVVVVVVVVVVV...  \n",
      "\n",
      "[509 rows x 40 columns]\n"
     ]
    }
   ],
   "source": [
    "testrow = alndf.iloc[121]\n",
    "print(testrow)\n",
    "print(testrow.qaln_y)\n",
    "print(testrow.qaln_x)\n",
    "calc_fident_crossaln(testrow, verbose = True)\n",
    "res = alndf.apply(calc_fident_crossaln , axis = 1)\n",
    "alndf = pd.concat([alndf,res] , axis = 1)\n",
    "print(alndf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "33ed78cd-d0eb-4410-b8ab-80d9c8b05506",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          query      target  fident_x  alnlen_x  mismatch_x  gapopen_x  \\\n",
      "0    Q5IS80.pdb  Q5IS80.pdb     1.000       770           0          0   \n",
      "1    Q5IS80.pdb  P53601.pdb     0.910       770          68          1   \n",
      "2    Q5IS80.pdb  Q60495.pdb     0.915       772          61          3   \n",
      "3    Q5IS80.pdb  P79307.pdb     0.863       770         105          0   \n",
      "4    Q5IS80.pdb  P05067.pdb     0.836       773         121          3   \n",
      "..          ...         ...       ...       ...         ...        ...   \n",
      "504  Q29149.pdb  P05067.pdb     0.846        13           2          0   \n",
      "505  Q29149.pdb  Q06481.pdb     0.785        14           3          0   \n",
      "506  Q29149.pdb  P12023.pdb     0.846        13           2          0   \n",
      "507  Q29149.pdb  Q28053.pdb     1.000        10           0          0   \n",
      "508  Q29149.pdb  P08592.pdb     0.687        16           5          0   \n",
      "\n",
      "     qstart_x  qend_x  tstart_x  tend_x  ...  \\\n",
      "0           1     770         1     770  ...   \n",
      "1           2     770         1     770  ...   \n",
      "2           2     770         1     770  ...   \n",
      "3           1     770         1     770  ...   \n",
      "4           1     770         1     770  ...   \n",
      "..        ...     ...       ...     ...  ...   \n",
      "504         1      13       678     690  ...   \n",
      "505         1      14       241     254  ...   \n",
      "506         1      13       200     212  ...   \n",
      "507         1      10        12      21  ...   \n",
      "508         1      16       602     617  ...   \n",
      "\n",
      "                                                qaln_y  \\\n",
      "0    MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...   \n",
      "1    MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...   \n",
      "2    MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...   \n",
      "3    MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...   \n",
      "4    MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...   \n",
      "..                                                 ...   \n",
      "504  SEVKMDAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVVI...   \n",
      "505                            GAIIGLMVGGVVIATVIVITLVM   \n",
      "506  SEVKMDAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVVI...   \n",
      "507  SEVKMDAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVVI...   \n",
      "508  SEVKMDAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVVI...   \n",
      "\n",
      "                                                taln_y  cigar_y lntmscore_y  \\\n",
      "0    MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...     770M      1.0000   \n",
      "1    MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...     770M      0.4999   \n",
      "2    MLPSLALLLLTTWTARALEVPTDGNAGLLAEPQIAMFCGKLNMHMN...     770M      0.4215   \n",
      "3    MLPGLALVLLAAWTARALEVPTDGNAGLLAEPQVAMFCGKLNMHMN...     770M      0.3228   \n",
      "4    MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...     770M      0.3420   \n",
      "..                                                 ...      ...         ...   \n",
      "504  SEVKMDAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVVI...      56M      0.4417   \n",
      "505                            SALIGLLVIAVAIATVIVISLVM      23M      0.8627   \n",
      "506  SEVKMDAEFGHDSGFEVRHQKLVFFAEDVGSNKGAIIGLMVGGVVI...      56M      0.4289   \n",
      "507  SEVKMDAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVVI...      56M      0.4718   \n",
      "508  SEVKMDAEFGHDSGFEVRHQKLVFFAEDVGSNKGAIIGLMVGGVVI...      56M      0.4538   \n",
      "\n",
      "    fident_3di_mode2 fident_3di_mode0  \\\n",
      "0           1.000000         1.000000   \n",
      "1           0.910390         0.910390   \n",
      "2           0.915803         0.915803   \n",
      "3           0.863636         0.863636   \n",
      "4           0.836999         0.836999   \n",
      "..               ...              ...   \n",
      "504         0.846154         0.846154   \n",
      "505         0.785714         0.785714   \n",
      "506         0.846154         0.846154   \n",
      "507         1.000000         1.000000   \n",
      "508         0.687500         0.687500   \n",
      "\n",
      "                                        3di_qaln_mode0  \\\n",
      "0    DDDPDDPDPPPDDPDPDPPPPPPPPPQFPDFWWKWADAQFFIWIQD...   \n",
      "1    DDPDDPDPPPDDPDPDP-PPPPPPPPQFPDFWWKWADAQFFIWIQD...   \n",
      "2    DDPDDPDPPPDDPDPDP-PPPPPPPPQFPDFWWKWADAQFFIWIQD...   \n",
      "3    DDDPDDPDPPPDDPDPDPPPPPPPPPQFPDFWWKWADAQFFIWIQD...   \n",
      "4    DDDPDDPDPPPDDPDPDPPPPPPPPPQFPDFWWKWADAQFFIWIQD...   \n",
      "..                                                 ...   \n",
      "504                                      DDDDDPPPPPPPP   \n",
      "505                                     DDDDDPPPPPPPPP   \n",
      "506                                      DDDDDPPPPPPPP   \n",
      "507                                         DDDDDPPPPP   \n",
      "508                                   DDDDDPPPPPPPPPVV   \n",
      "\n",
      "                                        3di_taln_mode0  \\\n",
      "0    DDDPDDPDPPPDDPDPDPPPPPPPPPQFPDFWWKWADAQFFIWIQD...   \n",
      "1    DDPDPPPPPPPPPPPDPDPPPDPPPPQFPDFWWWWADAQFFIWIQD...   \n",
      "2    DDDDDPDDDDDDDDDDPDPDPDPPPPQFPDFWWKWADAQFFIWIQD...   \n",
      "3    DDDDDDDDDDDDPPPPPPPPPDPPPPQFPDFWWKWADAQFFIWIQP...   \n",
      "4    DDDDDDDDDDDDDDDDDDDDDPDAPPDFPDFWWKWADAQFFIWIQP...   \n",
      "..                                                 ...   \n",
      "504                                      DDDDDPPPPDDPP   \n",
      "505                                     DDDDDPPDDPPVPP   \n",
      "506                                      DDDDDPPPPDPQP   \n",
      "507                                         DDDDDPPPPP   \n",
      "508                                   DDDDDDDDDPPPDPVV   \n",
      "\n",
      "                                        3di_qaln_mode2  \\\n",
      "0    DDDPDDPDPPPDDPDPDPPPPPPPPPQFPDFWWKWADAQFFIWIQD...   \n",
      "1    DDDPDDPDPPPDDPDPDPPPPPPPPPQFPDFWWKWADAQFFIWIQD...   \n",
      "2    DDDPDDPDPPPDDPDPDPPPPPPPPPQFPDFWWKWADAQFFIWIQD...   \n",
      "3    DDDPDDPDPPPDDPDPDPPPPPPPPPQFPDFWWKWADAQFFIWIQD...   \n",
      "4    DDDPDDPDPPPDDPDPDPPPPPPPPPQFPDFWWKWADAQFFIWIQD...   \n",
      "..                                                 ...   \n",
      "504  DDDDDPPPPPPPPPVVVVVVVVVVVCVVVVCPVVVVVCCVVVVVVV...   \n",
      "505                            VVVVCCVVVVVVVVVVVVVVVVV   \n",
      "506  DDDDDPPPPPPPPPVVVVVVVVVVVCVVVVCPVVVVVCCVVVVVVV...   \n",
      "507  DDDDDPPPPPPPPPVVVVVVVVVVVCVVVVCPVVVVVCCVVVVVVV...   \n",
      "508  DDDDDPPPPPPPPPVVVVVVVVVVVCVVVVCPVVVVVCCVVVVVVV...   \n",
      "\n",
      "                                        3di_taln_mode2  \n",
      "0    DDDPDDPDPPPDDPDPDPPPPPPPPPQFPDFWWKWADAQFFIWIQD...  \n",
      "1    DDPDPPPPPPPPPPPDPDPPPDPPPPQFPDFWWWWADAQFFIWIQD...  \n",
      "2    DDDDDPDDDDDDDDDDPDPDPDPPPPQFPDFWWKWADAQFFIWIQD...  \n",
      "3    DDDDDDDDDDDDPPPPPPPPPDPPPPQFPDFWWKWADAQFFIWIQP...  \n",
      "4    DDDDDDDDDDDDDDDDDDDDDPDAPPDFPDFWWKWADAQFFIWIQP...  \n",
      "..                                                 ...  \n",
      "504  DDDDDDDDDDDDDDDDPPPPDDPPDDDDDDDDDVVVVVVVVVVVVV...  \n",
      "505                            VNVVVVVVVVVVVVVVVVVVVVV  \n",
      "506  DDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDPVVVVVVVVVVVVV...  \n",
      "507  DPPDPPDDPPDDDDDPPPPPDDDDDPDPDPDPVVSVVCCVVVVVVV...  \n",
      "508  DDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDPDVVVVVVVVVVVVV...  \n",
      "\n",
      "[509 rows x 42 columns]\n"
     ]
    }
   ],
   "source": [
    "alndf.reset_index(inplace=True)\n",
    "print(alndf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0c6148ea-3540-4e21-b5e9-32a7652a9789",
   "metadata": {},
   "outputs": [],
   "source": [
    "#write a small nexus file\n",
    "with open('template.nx', 'w') as nexout:\n",
    "    nxstr= \"\"\"\n",
    "    #nexus\n",
    "    begin sets;\n",
    "        charset part1 = aln1.fasta: *;\n",
    "        charset part2 = aln2.fasta: *;\n",
    "        charpartition mine = LG+I+G:part1, 3di_substmat.txt+I+F:part2;\n",
    "    end;\n",
    "    \"\"\"\n",
    "    nexout.write(nxstr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0af11bdf-d6c9-4fd8-a1ed-a9a639aa1ca3",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileExistsError",
     "evalue": "[Errno 17] File exists: 'alnscratch'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileExistsError\u001b[0m                           Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[11], line 4\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mos\u001b[39;00m\n\u001b[1;32m      3\u001b[0m alnfolder \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124malnscratch/\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m----> 4\u001b[0m \u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmkdir\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43malnscratch\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mFileExistsError\u001b[0m: [Errno 17] File exists: 'alnscratch'"
     ]
    }
   ],
   "source": [
    "#alnscratch \n",
    "import os\n",
    "alnfolder = 'alnscratch/'\n",
    "os.mkdir('alnscratch')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f1e81ee2-901a-41df-9888-9f962748a689",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "      /-P14599.pdb\n",
      "   /-|\n",
      "  |  |   /-Q10651.pdb\n",
      "  |   \\-|\n",
      "  |      \\-Q06BR2.pdb\n",
      "--|\n",
      "  |      /-Q03157.pdb\n",
      "  |   /-|\n",
      "  |  |   \\-P51693.pdb\n",
      "  |  |\n",
      "   \\-|      /-Q06335.pdb\n",
      "     |   /-|\n",
      "     |  |  |   /-Q06481.pdb\n",
      "     |  |   \\-|\n",
      "      \\-|      \\-P15943.pdb\n",
      "        |\n",
      "        |   /-O93279.pdb\n",
      "        |  |\n",
      "         \\-|   /-O73683.pdb\n",
      "           |  |\n",
      "            \\-|   /-Q29149.pdb\n",
      "              |  |\n",
      "               \\-|   /-Q28748.pdb\n",
      "                 |  |\n",
      "                 |  |      /-Q95241.pdb\n",
      "                  \\-|   /-|\n",
      "                    |  |   \\-Q28053.pdb\n",
      "                    |  |\n",
      "                     \\-|   /-Q28757.pdb\n",
      "                       |  |\n",
      "                       |  |      /-Q28280.pdb\n",
      "                        \\-|   /-|\n",
      "                          |  |  |   /-P53601.pdb\n",
      "                          |  |   \\-|\n",
      "                          |  |     |   /-Q5IS80.pdb\n",
      "                           \\-|      \\-|\n",
      "                             |         \\-P05067.pdb\n",
      "                             |\n",
      "                             |   /-Q60495.pdb\n",
      "                              \\-|\n",
      "                                |   /-P79307.pdb\n",
      "                                 \\-|\n",
      "                                   |   /-P12023.pdb\n",
      "                                    \\-|\n",
      "                                       \\-P08592.pdb\n"
     ]
    }
   ],
   "source": [
    "import toytree\n",
    "guide_tree = '../testdata/fidentstructblob_tree.nwk.rooted'\n",
    "#madroot the tree\n",
    "tre = toytree.tree(guide_tree)\n",
    "print(tre)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "6d8dee4f-f923-43b5-b212-4fa3f2f76d2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import os\n",
    "import glob\n",
    "import toytree\n",
    "import tqdm\n",
    "import pandas as pd\n",
    "from subprocess import PIPE, Popen\n",
    "import shlex\n",
    "from Bio import SeqIO\n",
    "\n",
    "def get_leafset( treenode ):\n",
    "    \"\"\"\n",
    "    this function returns the leafset of a node\n",
    "    \"\"\"\n",
    "    if treenode.is_leaf():\n",
    "        return [treenode.name]\n",
    "    else:\n",
    "        return treenode.get_leaf_names()\n",
    "\n",
    "def mergeAlign( fasta1, fasta2, outfile):\n",
    "    print(fasta1,fasta2)\n",
    "    #args =  'clustalo --p1 ' + fasta1 + ' --p2 ' + fasta2 + ' --force -o ' + outfile\n",
    "    print('merge' , fasta1 , fasta2 )\n",
    "    \n",
    "    \n",
    "    args = 'mafft --add ' + fasta1 + ' ' + fasta2 \n",
    "    args = shlex.split(args)\n",
    "    p = Popen( args, stdin=PIPE, stdout=PIPE)\n",
    "    aln = p.communicate()[0].decode()\n",
    "    print('mafft aln',aln)\n",
    "    with open( outfile , 'w') as alnout:\n",
    "        alnout.write(aln)\n",
    "    \n",
    "    return outfile\n",
    "\n",
    "def sub2fasta( sub, outfile , fastacol1='qaln_y' , fastacol2='taln_y' ):\n",
    "    with open(outfile, 'w') as f:\n",
    "        f.write('>' + sub['query'] + '\\n')\n",
    "        f.write(sub[fastacol1] + '\\n')\n",
    "        f.write('>' + sub['target'] + '\\n')\n",
    "        f.write(sub[fastacol2] + '\\n')    \n",
    "    return outfile\n",
    "\n",
    "def retalns(allvall, leafname,leafset):\n",
    "    sub = allvall[allvall['query'] == leafname]\n",
    "    sub = sub[sub['target'].isin(leafset)]\n",
    "    sub = sub[sub['query'] != sub['target']]\n",
    "    return sub.iloc[0]\n",
    "\n",
    "#traverse tree from root to leaves recursively\n",
    "def traverse_tree_merge( treenode, topleafset, allvall ):\n",
    "    \"\"\"\n",
    "    this function traverses a tree from root to leaves recursively\n",
    "    it returns a dictionary with the iteratively built alignment\n",
    "    \"\"\"\n",
    "    \n",
    "    if treenode.is_leaf():  \n",
    "        print(treenode)\n",
    "        #if the node is a leaf, then we need to add it to the alignment with one of the pivots in the current leafset\n",
    "        sub = retalns(allvall, treenode.name , topleafset )\n",
    "        treenode.aln = sub2fasta(sub, alnfolder + treenode.name + '_inter.fasta')\n",
    "        treenode.aln3di = sub2fasta(sub, alnfolder + treenode.name + '_inter.3di.fasta' , fastacol1='3di_qaln_mode2' , fastacol2='3di_taln_mode2')\n",
    "        return treenode.aln, treenode.aln3di\n",
    "    \n",
    "    else:\n",
    "        childalns3di = []\n",
    "        childalnsAA = []\n",
    "        treenode.leafset = get_leafset(treenode) \n",
    "        for c in treenode.get_children():\n",
    "            if not c.aln:\n",
    "                c.aln,c.aln3di = traverse_tree_merge(c , treenode.leafset , allvall)\n",
    "            childalnsAA.append(c.aln)\n",
    "            childalns3di.append(c.aln3di)\n",
    "        treenode.aln = mergeAlign( childalnsAA[0], childalnsAA[1] ,  alnfolder + treenode.name + '_inter.fasta')            \n",
    "        treenode.aln3di = mergeAlign( childalns3di[0], childalns3di[1] , alnfolder + treenode.name + '_inter.3di.fasta' )\n",
    "        return treenode.aln, treenode.aln3di\n",
    "\n",
    "def remove_redundant( alignment ):\n",
    "    \"\"\"\n",
    "    this function removes redundant sequences from an alignment\n",
    "    \"\"\"\n",
    "    aln = SeqIO.parse(alignment, 'fasta')\n",
    "    seqs = []\n",
    "    ids = []\n",
    "    for s in aln:\n",
    "        if s.id not in ids:\n",
    "            seqs.append(s)\n",
    "            ids.append(s.id)\n",
    "    \n",
    "    with open(alignment, 'w') as f:\n",
    "        for s in seqs:\n",
    "            f.write('>' + s.id + '\\n')\n",
    "            f.write(str(s.seq) + '\\n')\n",
    "    return alignment\n",
    "\n",
    "#remove all alns except the final merged one\n",
    "def cleanup( filedir ):\n",
    "    \"\"\"\n",
    "    this function removes all alns except the final merged one\n",
    "    \"\"\"\n",
    "    for f in glob.glob(filedir + '*inter.fasta'):\n",
    "        os.remove(f)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "155c007a-6785-44db-83c3-bd589b89e6c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#prepare tree attributes\n",
    "for i,n in enumerate(tre.treenode.traverse()):\n",
    "    n.aln = None\n",
    "    n.aln3di = None\n",
    "    n.leafset = None\n",
    "    if len(n.name) == 0:\n",
    "        n.name = 'internal_'+str(i)\n",
    "finalaln, finalaln3di = traverse_tree_merge( tre.treenode.get_tree_root(), get_leafset(tre.treenode.get_tree_root())     , alndf )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "07a475b9-25ab-45e8-9852-4b9fcd85f5e8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "alnscratch/44_inter.fasta alnscratch/44_inter.3di.fasta\n"
     ]
    }
   ],
   "source": [
    "print(finalaln, finalaln3di)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "d066205a-6de6-4d25-9441-a095934a88dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#copy the final fastas to final fasta\n",
    "\n",
    "for fasta in [finalaln, finalaln3di]:\n",
    "    remove_redundant(fasta)\n",
    "\n",
    "for fasta,data in {'aa':finalaln, '3di':finalaln3di}.items():\n",
    "    with open(fasta+'.final.fasta' , 'w') as fastout:\n",
    "        with open( data ) as fastin:\n",
    "            fastout.write( fastin.read())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "d5b7b694-6f78-48f1-9a71-b49448ed4e43",
   "metadata": {},
   "outputs": [],
   "source": [
    "#run iqtree\n",
    "#make partition\n",
    "def runIQparitioned( fasta1 = 'aa.final.fasta', fasta2 = '3di.final.fasta', nexus= 'template.nx' ):\n",
    "    print(fasta1,fasta2)\n",
    "    args = 'iqtree --spp -mdef template.nx'\n",
    "    args = shlex.split(args)\n",
    "    p = Popen( args, stdin=PIPE, stdout=PIPE)\n",
    "    aln = p.communicate()[0].decode()\n",
    "    print('mafft aln',aln)\n",
    "    with open( outfile , 'w') as alnout:\n",
    "        alnout.write(aln)\n",
    "    \n",
    "    return outfile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fb533cb-9fb7-4f70-b977-54fc461df14f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4dc3eae-1481-4251-a7ad-d27519fca5a1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
