{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/projects/snake_tree/notebooks\n"
     ]
    }
   ],
   "source": [
    "cd projects/snake_tree/notebooks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "import glob\n",
    "import os\n",
    "import numpy as np\n",
    "import sys\n",
    "sys.path.append( '../src/')\n",
    "from matplotlib import pyplot as plt\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from compile_treeset import compile_folder , compare_treesets\n",
    "import plotly.express as px\n",
    "import plotly.figure_factory as ff\n",
    "import plotly.express as px\n",
    "from scipy.stats import wilcoxon, ks_2samp , describe\n",
    "import treescore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_hogs = glob.glob('../OMA_data_unfiltered/OMA_data/LUCA/*/')\n",
    "sample_hogs = sample_hogs[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['../OMA_data_unfiltered/OMA_data/LUCA/HOG1077/', '../OMA_data_unfiltered/OMA_data/LUCA/HOG748/', '../OMA_data_unfiltered/OMA_data/LUCA/HOG458/', '../OMA_data_unfiltered/OMA_data/LUCA/HOG411/', '../OMA_data_unfiltered/OMA_data/LUCA/HOG569/', '../OMA_data_unfiltered/OMA_data/LUCA/HOG1118/', '../OMA_data_unfiltered/OMA_data/LUCA/HOG423/', '../OMA_data_unfiltered/OMA_data/LUCA/HOG600/', '../OMA_data_unfiltered/OMA_data/LUCA/HOG475/', '../OMA_data_unfiltered/OMA_data/LUCA/HOG1231/']\n",
      "copied files\n"
     ]
    }
   ],
   "source": [
    "print(sample_hogs)\n",
    "#make a test directory and copy the files there\n",
    "\n",
    "if not os.path.exists('test'):\n",
    "    os.makedirs('test')\n",
    "    for hog in sample_hogs:\n",
    "        os.system('cp -r '+hog+' test/')\n",
    "    print('copied files')\n",
    "else:\n",
    "    #clear files\n",
    "    os.system('rm -r test')\n",
    "    os.makedirs('test')\n",
    "    for hog in sample_hogs:\n",
    "        os.system('cp -r '+hog+' test/')\n",
    "    print('copied files')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['test/HOG1077/', 'test/HOG748/', 'test/HOG458/', 'test/HOG411/', 'test/HOG569/', 'test/HOG1118/', 'test/HOG423/', 'test/HOG600/', 'test/HOG475/', 'test/HOG1231/']\n",
      "['test/HOG1231/3di_xaln_tree.nwk', 'test/HOG1231/fident_1_raw_struct_tree.PP.nwk.rooted', 'test/HOG1231/alntmscore_1_raw_struct_tree.PP.nwk', 'test/HOG1231/alntmscore_0_raw_treescores_struct_tree.json', 'test/HOG1231/alntmscore_0_exp_treescores_struct_tree.json', 'test/HOG1231/alnfident_dist_1_raw_.json', 'test/HOG1231/templateXfoldtree.nx.treefile.rooted', 'test/HOG1231/lddt_distmat.npy', 'test/HOG1231/templateXAA.nx.log', 'test/HOG1231/3di_xaln_tree.PP.nwk.rooted', 'test/HOG1231/logs', 'test/HOG1231/templateXfoldtree.nx.treefile.rooted.final', 'test/HOG1231/3di_xaln_tree.PP.nwk.rooted.final.treescore', 'test/HOG1231/templateX3di.nx.iqtree', 'test/HOG1231/alnfident_dist_3di_AA.json', 'test/HOG1231/lddt_1_exp_struct_tree.PP.nwk', 'test/HOG1231/sequences.aln.muscle.fst.log', 'test/HOG1231/rejected', 'test/HOG1231/fident_0_raw_struct_tree.PP.nwk', 'test/HOG1231/alntmscore_1_raw_struct_tree.PP.nwk.rooted', 'test/HOG1231/lddt_1_exp_struct_tree.nwk', 'test/HOG1231/sequences.aln.clustalo.fst', 'test/HOG1231/lddt_0_raw_struct_tree.PP.nwk', 'test/HOG1231/lddt_0_raw_struct_tree.nwk', 'test/HOG1231/alnfident_dist_AA_AA.json', 'test/HOG1231/plddt.json', 'test/HOG1231/alntmscore_0_raw_struct_tree.PP.nwk.rooted', 'test/HOG1231/sequences.aln.clustalo.fst.nwk', 'test/HOG1231/alntmscore_0_exp_struct_tree.nwk', 'test/HOG1231/allvall_1.csv', 'test/HOG1231/outdb_ss', 'test/HOG1231/fident_0_exp_struct_tree.PP.nwk', 'test/HOG1231/alntmscore_1_exp_treescores_struct_tree.json', 'test/HOG1231/alnAA_3di.fasta.treefile.rooted.final.treescore', 'test/HOG1231/templateX3di.nx', 'test/HOG1231/templateXAA.nx.treefile', 'test/HOG1231/fident_1_raw_struct_tree.nwk', 'test/HOG1231/AA_xaln_fastmemat.txt', 'test/HOG1231/alnAA_foldtree.fasta', 'test/HOG1231/sequences.fst', 'test/HOG1231/alnAA_3di.fasta.treefile.rooted', 'test/HOG1231/sequences.aln.clustalo.fst.treefile.rooted', 'test/HOG1231/alnAA_3di.fasta.ckp.gz', 'test/HOG1231/fident_0_raw_fastmemat.txt', 'test/HOG1231/tmp', 'test/HOG1231/alntmscore_0_raw_struct_tree.PP.nwk', 'test/HOG1231/lddt_0_raw_fastmemat.txt', 'test/HOG1231/sequences.aln.clustalo.fst.iqtree', 'test/HOG1231/alntmscore_1_exp_struct_tree.nwk', 'test/HOG1231/aln3di_3di.fasta', 'test/HOG1231/alnfident_dist_AA_3dcoffee.json', 'test/HOG1231/outdb_ss.index', 'test/HOG1231/alnscratch', 'test/HOG1231/alnAA_3di.fasta.iqtree', 'test/HOG1231/AA_xaln_tree.PP.nwk.rooted', 'test/HOG1231/alntmscore_distmat.npy', 'test/HOG1231/alnfident_dist_0_raw_.json', 'test/HOG1231/fident_1_exp_struct_tree.nwk', 'test/HOG1231/sequences.aln.muscle.fst', 'test/HOG1231/outdb.dbtype', 'test/HOG1231/sequence_dataset.csv', 'test/HOG1231/templateXAA.nx.bionj', 'test/HOG1231/alnfident_dist_AA_foldtree.json', 'test/HOG1231/alnAA_3di.fasta.log', 'test/HOG1231/fident_0_raw_treescores_struct_tree.json', 'test/HOG1231/lddt_1_exp_struct_tree.PP.nwk.rooted', 'test/HOG1231/templateXAA.nx.mldist', 'test/HOG1231/AA_xaln_tree.PP.nwk.rooted.final', 'test/HOG1231/alnAA_3di.fasta.treefile.rooted.final', 'test/HOG1231/outdb', 'test/HOG1231/outdb.lookup', 'test/HOG1231/lddt_0_exp_treescores_struct_tree.json', 'test/HOG1231/outdb.index', 'test/HOG1231/fident_1_exp_fastmemat.txt', 'test/HOG1231/templateXfoldtree.nx.best_model.nex', 'test/HOG1231/templateX3di.nx.treefile.rooted', 'test/HOG1231/sequences.aln.muscle.fst.ckp.gz', 'test/HOG1231/templateXAA.nx.best_model.nex', 'test/HOG1231/templateXAA.nx.ckp.gz', 'test/HOG1231/templateX3di.nx.log', 'test/HOG1231/alntmscore_1_exp_struct_tree.PP.nwk', 'test/HOG1231/aln3di_foldtree.fasta', 'test/HOG1231/treescores_sequences.clustalo.json', 'test/HOG1231/sequences.aln.clustalo.fst.ckp.gz', 'test/HOG1231/treescores_sequences.muscle.json', 'test/HOG1231/3di_xaln_fastmemat.txt', 'test/HOG1231/outdb_ca.dbtype', 'test/HOG1231/fident_0_exp_fastmemat.txt', 'test/HOG1231/lddt_0_raw_struct_tree.PP.nwk.rooted', 'test/HOG1231/lddt_1_exp_treescores_struct_tree.json', 'test/HOG1231/AA_xaln_tree.PP.nwk.rooted.final.treescore', 'test/HOG1231/fident_0_exp_treescores_struct_tree.json', 'test/HOG1231/lddt_0_exp_struct_tree.PP.nwk.rooted', 'test/HOG1231/alntmscore_0_raw_struct_tree.nwk', 'test/HOG1231/sequences.aln.muscle.fst.treefile.rooted', 'test/HOG1231/templateXfoldtree.nx', 'test/HOG1231/templateX3di.nx.bionj', 'test/HOG1231/3di_xaln_tree.PP.nwk.rooted.final', 'test/HOG1231/outdb_h.dbtype', 'test/HOG1231/templateX3di.nx.treefile.rooted.final.treescore', 'test/HOG1231/lddt_1_raw_fastmemat.txt', 'test/HOG1231/lddt_1_raw_struct_tree.nwk', 'test/HOG1231/templateXfoldtree.nx.mldist', 'test/HOG1231/sequences.aln.clustalo.fst.treefile', 'test/HOG1231/alnAA_3di.fasta.treefile', 'test/HOG1231/lddt_0_exp_struct_tree.nwk', 'test/HOG1231/outdb_ss.dbtype', 'test/HOG1231/templateX3di.nx.treefile.rooted.final', 'test/HOG1231/alntmscore_0_raw_fastmemat.txt', 'test/HOG1231/alntmscore_1_raw_fastmemat.txt', 'test/HOG1231/alnfident_dist_3di_foldtree.json', 'test/HOG1231/fident_0_raw_struct_tree.PP.nwk.rooted', 'test/HOG1231/lddt_1_raw_struct_tree.PP.nwk', 'test/HOG1231/templateXfoldtree.nx.treefile.rooted.final.treescore', 'test/HOG1231/outdb_h.index', 'test/HOG1231/templateXAA.nx.treefile.rooted', 'test/HOG1231/sequences.aln.clustalo.fst.log', 'test/HOG1231/outdb_h', 'test/HOG1231/lddt_1_raw_treescores_struct_tree.json', 'test/HOG1231/templateXfoldtree.nx.treefile', 'test/HOG1231/finalset.csv', 'test/HOG1231/allvall_0.csv', 'test/HOG1231/alnfident_dist_3di_3di.json', 'test/HOG1231/lddt_1_exp_fastmemat.txt', 'test/HOG1231/tcoffee_templates.txt', 'test/HOG1231/alnAA_3di.fasta.mldist', 'test/HOG1231/alntmscore_1_exp_fastmemat.txt', 'test/HOG1231/sequences.aln.clustalo.fst.nwk.rooted', 'test/HOG1231/outdb_ca.index', 'test/HOG1231/sequences.aln.muscle.fst.mldist', 'test/HOG1231/fident_1_raw_treescores_struct_tree.json', 'test/HOG1231/templateXfoldtree.nx.log', 'test/HOG1231/fident_0_raw_struct_tree.nwk', 'test/HOG1231/sequences.aln.muscle.fst.nwk.rooted', 'test/HOG1231/lddt_0_exp_fastmemat.txt', 'test/HOG1231/fident_1_exp_treescores_struct_tree.json', 'test/HOG1231/templateXAA.nx.treefile.rooted.final', 'test/HOG1231/lddt_1_raw_struct_tree.PP.nwk.rooted', 'test/HOG1231/templateX3di.nx.mldist', 'test/HOG1231/sequences.aln.muscle.fst.treefile', 'test/HOG1231/sequences.aln.muscle.fst.iqtree', 'test/HOG1231/outdb_ca', 'test/HOG1231/templateXfoldtree.nx.ckp.gz', 'test/HOG1231/fident_0_exp_struct_tree.nwk', 'test/HOG1231/alntmscore_1_exp_struct_tree.PP.nwk.rooted', 'test/HOG1231/alnfident_dist_AA_3di.json', 'test/HOG1231/3di_xaln_tree.PP.nwk', 'test/HOG1231/lddt_0_exp_struct_tree.PP.nwk', 'test/HOG1231/alntmscore_0_exp_struct_tree.PP.nwk.rooted', 'test/HOG1231/alnAA_AA.fasta', 'test/HOG1231/alntmscore_1_raw_struct_tree.nwk', 'test/HOG1231/templateXAA.nx.iqtree', 'test/HOG1231/treescores_sequences_iq.muscle.json', 'test/HOG1231/treescores_sequences_iq.clustalo.json', 'test/HOG1231/AA_xaln_tree.PP.nwk', 'test/HOG1231/alntmscore_1_raw_treescores_struct_tree.json', 'test/HOG1231/seq3di.fasta', 'test/HOG1231/templateX3di.nx.treefile', 'test/HOG1231/templateXAA.nx.treefile.rooted.final.treescore', 'test/HOG1231/alntmscore_0_exp_struct_tree.PP.nwk', 'test/HOG1231/fident_1_raw_struct_tree.PP.nwk', 'test/HOG1231/templateX3di.nx.best_model.nex', 'test/HOG1231/outdb.source', 'test/HOG1231/templateXAA.nx', 'test/HOG1231/fident_1_raw_fastmemat.txt', 'test/HOG1231/aln3di_AA.fasta', 'test/HOG1231/sequences.aln.muscle.fst.nwk', 'test/HOG1231/alnAA_3di.fasta', 'test/HOG1231/fident_distmat.npy', 'test/HOG1231/finalset_ncbi_tree.nwk', 'test/HOG1231/sequences.aln.muscle.fst.bionj', 'test/HOG1231/fident_1_exp_struct_tree.PP.nwk.rooted', 'test/HOG1231/identifiers.txt', 'test/HOG1231/fident_0_exp_struct_tree.PP.nwk.rooted', 'test/HOG1231/templateXfoldtree.nx.iqtree', 'test/HOG1231/sequences.aln.clustalo.fst.bionj', 'test/HOG1231/fident_1_exp_struct_tree.PP.nwk', 'test/HOG1231/sequences.aln.clustalo.fst.mldist', 'test/HOG1231/alntmscore_0_exp_fastmemat.txt', 'test/HOG1231/AA_xaln_tree.nwk', 'test/HOG1231/alnAA_3di.fasta.bionj', 'test/HOG1231/templateX3di.nx.ckp.gz', 'test/HOG1231/structs', 'test/HOG1231/templateXfoldtree.nx.bionj', 'test/HOG1231/lddt_0_raw_treescores_struct_tree.json']\n"
     ]
    }
   ],
   "source": [
    "testhogs = glob.glob('test/*/')\n",
    "print(testhogs)\n",
    "sample1 = testhogs[9]\n",
    "samplefiles = glob.glob(sample1+'*')\n",
    "print(samplefiles)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample1 = sample_hogs[0]\n",
    "treeML = sample1+'sequences.aln.muscle.fst.treefile.rooted'\n",
    "treeFT = sample1+'fident_1_raw_struct_tree.PP.nwk.rooted'\n",
    "finalset = sample1 +'finalset.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import toytree\n",
    "import ete3\n",
    "from ete3 import PhyloTree\n",
    "from ete3 import NCBITaxa\n",
    "\n",
    "ncbi = NCBITaxa()\n",
    "#ncbi.update_taxonomy_database()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import copy\n",
    "\n",
    "def calc_scores(t , uniprot_csv):\n",
    "    uniprot_df = pd.read_csv(uniprot_csv)\n",
    "    tree = toytree.tree(t)\n",
    "    lineages = treescore.make_lineages(uniprot_df)\n",
    "    species = treescore.get_species(uniprot_df)\n",
    "    tree = treescore.label_leaves( tree , lineages , species)\n",
    "    #tree = treescore.labelwRED(tree.treenode)\n",
    "    overlap = treescore.getTaxOverlap(tree.treenode)\n",
    "    taxscore = tree.treenode.score\n",
    "\n",
    "    #calc descriptive stats on normalized branch lens to see if trees are balanced  \n",
    "    lengths = np.array([node.dist for node in tree.treenode.traverse()])\n",
    "    lengths /= np.sum(lengths)\n",
    "    #calc the root first taxscore\n",
    "\n",
    "    tree4 = copy.deepcopy(tree)\n",
    "    treescore.getTaxOverlap_root(tree4.treenode)\n",
    "    root_score , root_score_nr = treescore.sum_rootscore(tree4.treenode)\n",
    "\n",
    "\n",
    "    species_set = set()\n",
    "    #label the leaves with species and number\n",
    "    for l in tree.treenode.get_leaves():\n",
    "        if l.sp_num:\n",
    "            l.name = l.sp_num\n",
    "            species_set.add(l.sp_num.split('_')[0])\n",
    "\n",
    "\n",
    "    #get ncbi tree of species set\n",
    "    species_set = list(species_set)\n",
    "    species_set = [int(s) for s in species_set]\n",
    "    species_set = ncbi.get_topology(species_set)\n",
    "\n",
    "    #change to phylo tree\n",
    "    ncbitree = PhyloTree(species_set.write() , sp_naming_function=None)\n",
    "\n",
    "    ncbitree.write(outfile=uniprot_csv.replace('.csv','_ncbi_tree.nwk' ) , format=1)\n",
    "    etetree = PhyloTree(tree.write() , sp_naming_function=None)\n",
    "\n",
    "    for l in etetree.get_leaves():\n",
    "        l.species   = l.name.split('_')[0]\n",
    "    \n",
    "    recon_tree, events = etetree.reconcile(ncbitree)\n",
    "    recon_dups = recon_tree.search_nodes(evoltype=\"D\")\n",
    "    recon_losses = recon_tree.search_nodes(evoltype=\"L\")\n",
    "    recon_speciations = recon_tree.search_nodes(evoltype=\"S\")\n",
    "    print( 'algo 1' )\n",
    "    print('dups:', len(recon_dups))\n",
    "    print('losses:', len(recon_losses))\n",
    "    print('speciations:', len(recon_speciations))\n",
    "\n",
    "    print( 'algo 2' )\n",
    "    events = etetree.get_descendant_evol_events()\n",
    "    dups = etetree.search_nodes(evoltype=\"D\")\n",
    "    losses = etetree.search_nodes(evoltype=\"L\")\n",
    "    speciations = etetree.search_nodes(evoltype=\"S\")    \n",
    "    print('dups:', len(dups))\n",
    "    print('losses:', len(losses))\n",
    "    print('speciations:', len(speciations))\n",
    "\n",
    "    scores = {}\n",
    "    #measure the distances of leaves to root\n",
    "    distances = np.array([ node.get_distance(tree.treenode) for node in tree.treenode.get_leaves() ])\n",
    "    distances_norm = distances / np.mean(distances)\n",
    "    scores[t] = {'score': taxscore, 'stats': describe(lengths) , 'ultrametricity':  describe(distances), \n",
    "                    'ultrametricity_norm':  describe(distances_norm) , 'root_score': root_score , 'root_score_nr': root_score_nr  , \n",
    "                    'SO_speciations': len(recon_speciations) , 'SO_dups': len(recon_dups) , 'SO_losses': len(recon_losses) ,\n",
    "                    'RECON_speciations':speciations ,'RECON_dups': len(dups) , 'RECON_losses': len(losses)  }\n",
    "    return scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../OMA_data_unfiltered/OMA_data/LUCA/HOG1077/\n",
      "ML\n",
      "root scoring\n",
      "root score 4623\n",
      "root score_nr 3909\n",
      "algo 1\n",
      "dups: 23\n",
      "losses: 654\n",
      "speciations: 35\n",
      "algo 2\n",
      "dups: 12\n",
      "losses: 0\n",
      "speciations: 46\n",
      "{'../OMA_data_unfiltered/OMA_data/LUCA/HOG1077/sequences.aln.muscle.fst.treefile.rooted': {'score': 626, 'stats': DescribeResult(nobs=117, minmax=(2.991111384783379e-08, 0.0726053954572658), mean=0.00854700854700855, variance=0.00017982310920735504, skewness=2.671951856873734, kurtosis=7.301067678776121), 'ultrametricity': DescribeResult(nobs=59, minmax=(1.707441299, 4.3434943268), mean=2.139202316576271, variance=0.16713058024887115, skewness=3.168209235115583, kurtosis=13.302032524293164), 'ultrametricity_norm': DescribeResult(nobs=59, minmax=(0.7981672821543633, 2.030427086369106), mean=1.0, variance=0.03652180224239135, skewness=3.1682092351155826, kurtosis=13.302032524293164), 'root_score': 4623, 'root_score_nr': 3909, 'SO_speciations': 35, 'SO_dups': 23, 'SO_losses': 654, 'RECON_speciations': [PhyloTree node '' (0x7fc4efbd1be), PhyloTree node '' (0x7fc4ef504ad), PhyloTree node '' (0x7fc4ef5048f), PhyloTree node '' (0x7fc4ef50573), PhyloTree node '' (0x7fc4ef50780), PhyloTree node '' (0x7fc4ef505c1), PhyloTree node '' (0x7fc4ef5056d), PhyloTree node '' (0x7fc4ef507bc), PhyloTree node '' (0x7fc4ef507bf), PhyloTree node '' (0x7fc4ef50561), PhyloTree node '' (0x7fc4efe1901), PhyloTree node '' (0x7fc4ef5059d), PhyloTree node '' (0x7fc4ef50444), PhyloTree node '' (0x7fc4efe1b14), PhyloTree node '' (0x7fc4ef50567), PhyloTree node '' (0x7fc4ef506e1), PhyloTree node '' (0x7fc4efe19ac), PhyloTree node '' (0x7fc4ef505df), PhyloTree node '' (0x7fc4ef505c7), PhyloTree node '' (0x7fc4ef504b3), PhyloTree node '' (0x7fc4ef50408), PhyloTree node '' (0x7fc4efe1b0b), PhyloTree node '' (0x7fc4ef5045c), PhyloTree node '' (0x7fc4ef9269f), PhyloTree node '' (0x7fc4efe188f), PhyloTree node '' (0x7fc4eebcfc8), PhyloTree node '' (0x7fc4efe1a45), PhyloTree node '' (0x7fc4efe1b50), PhyloTree node '' (0x7fc4ed82982), PhyloTree node '' (0x7fc4efe1b1a), PhyloTree node '' (0x7fc4efe1b23), PhyloTree node '' (0x7fc4efe191f), PhyloTree node '' (0x7fc4efe190a), PhyloTree node '' (0x7fc4efe18d4), PhyloTree node '' (0x7fc4efe18b3), PhyloTree node '' (0x7fc4efe1aba), PhyloTree node '' (0x7fc4efe1ad8), PhyloTree node '' (0x7fc4efe1a75), PhyloTree node '' (0x7fc4efe19d9), PhyloTree node '' (0x7fc4efe19a6), PhyloTree node '' (0x7fc4efe196a), PhyloTree node '' (0x7fc4efe19dc), PhyloTree node '' (0x7fc4efe1958), PhyloTree node '' (0x7fc4efe18aa), PhyloTree node '' (0x7fc4efe19f7), PhyloTree node '' (0x7fc4efe186e)], 'RECON_dups': 12, 'RECON_losses': 0}}\n",
      "FT\n",
      "root scoring\n",
      "root score 4673\n",
      "root score_nr 3871\n",
      "algo 1\n",
      "dups: 24\n",
      "losses: 769\n",
      "speciations: 34\n",
      "algo 2\n",
      "dups: 11\n",
      "losses: 0\n",
      "speciations: 47\n",
      "{'../OMA_data_unfiltered/OMA_data/LUCA/HOG1077/fident_1_exp_struct_tree.PP.nwk.rooted': {'score': 598, 'stats': DescribeResult(nobs=117, minmax=(0.0, 0.07133576512676505), mean=0.008547008547008546, variance=0.00014689123631819967, skewness=2.5645555439636056, kurtosis=7.619932273280028), 'ultrametricity': DescribeResult(nobs=59, minmax=(0.4569437, 0.9896036999999999), mean=0.5861682576271187, variance=0.010348762225819373, skewness=2.2766300381634377, kurtosis=6.112437106906205), 'ultrametricity_norm': DescribeResult(nobs=59, minmax=(0.7795435765317017, 1.6882587672113765), mean=0.9999999999999998, variance=0.03011922655978452, skewness=2.2766300381634417, kurtosis=6.112437106906217), 'root_score': 4673, 'root_score_nr': 3871, 'SO_speciations': 34, 'SO_dups': 24, 'SO_losses': 769, 'RECON_speciations': [PhyloTree node '' (0x7fc4ef1928d), PhyloTree node '' (0x7fc4f4785a9), PhyloTree node '' (0x7fc4f478480), PhyloTree node '' (0x7fc4f478768), PhyloTree node '' (0x7fc4f478477), PhyloTree node '' (0x7fc4f47840b), PhyloTree node '' (0x7fc4f478657), PhyloTree node '' (0x7fc4f4787a7), PhyloTree node '' (0x7fc4f4786a2), PhyloTree node '' (0x7fc4f47869f), PhyloTree node '' (0x7fc4f47873e), PhyloTree node '' (0x7fc4f4787a4), PhyloTree node '' (0x7fc4f478777), PhyloTree node '' (0x7fc4f478441), PhyloTree node '' (0x7fc4f4786ae), PhyloTree node '' (0x7fc4f47864b), PhyloTree node '' (0x7fc4f47876e), PhyloTree node '' (0x7fc4f47850a), PhyloTree node '' (0x7fc4f478471), PhyloTree node '' (0x7fc4f478687), PhyloTree node '' (0x7fc4f4784fb), PhyloTree node '' (0x7fc4f47858e), PhyloTree node '' (0x7fc4f478501), PhyloTree node '' (0x7fc4f47878f), PhyloTree node '' (0x7fc4f4787f5), PhyloTree node '' (0x7fc4f478453), PhyloTree node '' (0x7fc4f4784f8), PhyloTree node '' (0x7fc4f478588), PhyloTree node '' (0x7fc4f4785e5), PhyloTree node '' (0x7fc4f4787e3), PhyloTree node '' (0x7fc4f478543), PhyloTree node '' (0x7fc4f4785fd), PhyloTree node '' (0x7fc4f4786c9), PhyloTree node '' (0x7fc4f478522), PhyloTree node '' (0x7fc4f47860c), PhyloTree node '' (0x7fc4f4785f7), PhyloTree node '' (0x7fc4f478423), PhyloTree node '' (0x7fc4f478420), PhyloTree node '' (0x7fc4f478603), PhyloTree node '' (0x7fc4f478489), PhyloTree node '' (0x7fc4f4786e7), PhyloTree node '' (0x7fc4f47870e), PhyloTree node '' (0x7fc4f478600), PhyloTree node '' (0x7fc4f47866f), PhyloTree node '' (0x7fc4f478789), PhyloTree node '' (0x7fc4f478660), PhyloTree node '' (0x7fc4f4785c7)], 'RECON_dups': 11, 'RECON_losses': 0}}\n",
      "../OMA_data_unfiltered/OMA_data/LUCA/HOG748/\n",
      "ML\n",
      "root scoring\n",
      "root score 6616\n",
      "root score_nr 5772\n",
      "algo 1\n",
      "dups: 54\n",
      "losses: 712\n",
      "speciations: 31\n",
      "algo 2\n",
      "dups: 34\n",
      "losses: 0\n",
      "speciations: 51\n",
      "{'../OMA_data_unfiltered/OMA_data/LUCA/HOG748/sequences.aln.muscle.fst.treefile.rooted': {'score': 1003, 'stats': DescribeResult(nobs=171, minmax=(1.9125109876945515e-08, 0.09134175578051827), mean=0.005847953216374269, variance=0.00010888839006397293, skewness=5.206253235378204, kurtosis=33.602964484779775), 'ultrametricity': DescribeResult(nobs=86, minmax=(0.9525909739999999, 7.9462350232), mean=1.998825808355814, variance=0.8719684500642272, skewness=4.20270040591009, kurtosis=21.945812908236356), 'ultrametricity_norm': DescribeResult(nobs=86, minmax=(0.47657528235718466, 3.9754514825563425), mean=1.0, variance=0.2182483026232015, skewness=4.202700405910089, kurtosis=21.94581290823635), 'root_score': 6616, 'root_score_nr': 5772, 'SO_speciations': 31, 'SO_dups': 54, 'SO_losses': 712, 'RECON_speciations': [PhyloTree node '' (0x7fc4ef1916a), PhyloTree node '' (0x7fc4ef5feae), PhyloTree node '' (0x7fc4ef5ff98), PhyloTree node '' (0x7fc4ef5ff71), PhyloTree node '' (0x7fc4ef5fc98), PhyloTree node '' (0x7fc4eeeac1a), PhyloTree node '' (0x7fc4ef5fdf1), PhyloTree node '' (0x7fc4ef5fdac), PhyloTree node '' (0x7fc4ef5fc50), PhyloTree node '' (0x7fc4eeeacb0), PhyloTree node '' (0x7fc4ef5fd73), PhyloTree node '' (0x7fc4ef5fc8f), PhyloTree node '' (0x7fc4ef5fd28), PhyloTree node '' (0x7fc4eeaab7a), PhyloTree node '' (0x7fc4ef5ff05), PhyloTree node '' (0x7fc4f49f104), PhyloTree node '' (0x7fc4eeaabe6), PhyloTree node '' (0x7fc4eeaaa99), PhyloTree node '' (0x7fc4eeaaaa8), PhyloTree node '' (0x7fc4ef5fe18), PhyloTree node '' (0x7fc4f414fe0), PhyloTree node '' (0x7fc4eeaa832), PhyloTree node '' (0x7fc4eeaa9be), PhyloTree node '' (0x7fc4ef5fe5a), PhyloTree node '' (0x7fc4ef5fc11), PhyloTree node '' (0x7fc4eeaabd7), PhyloTree node '' (0x7fc4eeaa93d), PhyloTree node '' (0x7fc4eeaab6e), PhyloTree node '' (0x7fc4eeaa8b0), PhyloTree node '' (0x7fc4eeaab14), PhyloTree node '' (0x7fc4ef50435), PhyloTree node '' (0x7fc4ef50471), PhyloTree node '' (0x7fc4eeaaae7), PhyloTree node '' (0x7fc4eeaaba4), PhyloTree node '' (0x7fc4eeaa88f), PhyloTree node '' (0x7fc4eecc2b4), PhyloTree node '' (0x7fc4edc25b5), PhyloTree node '' (0x7fc4eeaa80e), PhyloTree node '' (0x7fc4eeaa985), PhyloTree node '' (0x7fc4eeaa820), PhyloTree node '' (0x7fc4eecc140), PhyloTree node '' (0x7fc4edc2411), PhyloTree node '' (0x7fc4ec58bcb), PhyloTree node '' (0x7fc4eeaabf5), PhyloTree node '' (0x7fc4eecc3b3), PhyloTree node '' (0x7fc4eecc011), PhyloTree node '' (0x7fc4ec58aab), PhyloTree node '' (0x7fc4ec588cb), PhyloTree node '' (0x7fc4ec58b44), PhyloTree node '' (0x7fc4ec58a66), PhyloTree node '' (0x7fc4ec58a9f)], 'RECON_dups': 34, 'RECON_losses': 0}}\n",
      "FT\n",
      "root scoring\n",
      "root score 7953\n",
      "root score_nr 6787\n",
      "algo 1\n",
      "dups: 54\n",
      "losses: 650\n",
      "speciations: 31\n",
      "algo 2\n",
      "dups: 34\n",
      "losses: 0\n",
      "speciations: 51\n",
      "{'../OMA_data_unfiltered/OMA_data/LUCA/HOG748/fident_1_exp_struct_tree.PP.nwk.rooted': {'score': 958, 'stats': DescribeResult(nobs=171, minmax=(0.0, 0.057140310317597275), mean=0.005847953216374269, variance=6.69726033637575e-05, skewness=3.6643521602802407, kurtosis=17.65390190804322), 'ultrametricity': DescribeResult(nobs=86, minmax=(0.73846, 1.2600900000000004), mean=0.8871412790697676, variance=0.007445972764227089, skewness=1.5781169293579047, kurtosis=5.008626918398431), 'ultrametricity_norm': DescribeResult(nobs=86, minmax=(0.8324040571918029, 1.420393830981799), mean=1.0, variance=0.00946097441583391, skewness=1.5781169293579047, kurtosis=5.008626918398427), 'root_score': 7953, 'root_score_nr': 6787, 'SO_speciations': 31, 'SO_dups': 54, 'SO_losses': 650, 'RECON_speciations': [PhyloTree node '' (0x7fc4eea4a96), PhyloTree node '' (0x7fc4eea4a27), PhyloTree node '' (0x7fc4eea4ac0), PhyloTree node '' (0x7fc4eea497c), PhyloTree node '' (0x7fc4eea49a0), PhyloTree node '' (0x7fc4eea4a93), PhyloTree node '' (0x7fc4eea4af6), PhyloTree node '' (0x7fc4eea4aab), PhyloTree node '' (0x7fc4eea4841), PhyloTree node '' (0x7fc4eea4a48), PhyloTree node '' (0x7fc4eea4b89), PhyloTree node '' (0x7fc4eea4a78), PhyloTree node '' (0x7fc4eea4856), PhyloTree node '' (0x7fc4eea49ac), PhyloTree node '' (0x7fc4eea4ac6), PhyloTree node '' (0x7fc4eea4bc5), PhyloTree node '' (0x7fc4eea4883), PhyloTree node '' (0x7fc4eea48c2), PhyloTree node '' (0x7fc4eea499a), PhyloTree node '' (0x7fc4eea4b92), PhyloTree node '' (0x7fc4eea48dd), PhyloTree node '' (0x7fc4eea4a54), PhyloTree node '' (0x7fc4eea4925), PhyloTree node '' (0x7fc4eea4aea), PhyloTree node '' (0x7fc4eea48ad), PhyloTree node '' (0x7fc4eea4b71), PhyloTree node '' (0x7fc4eea4abd), PhyloTree node '' (0x7fc4eea4808), PhyloTree node '' (0x7fc4eea493d), PhyloTree node '' (0x7fc4eea4bd4), PhyloTree node '' (0x7fc4eea4bef), PhyloTree node '' (0x7fc4eea4b50), PhyloTree node '' (0x7fc4eea494c), PhyloTree node '' (0x7fc4eea4949), PhyloTree node '' (0x7fc4eea4b8f), PhyloTree node '' (0x7fc4eea4892), PhyloTree node '' (0x7fc4eea4a9f), PhyloTree node '' (0x7fc4eea4958), PhyloTree node '' (0x7fc4eea49df), PhyloTree node '' (0x7fc4eea4a21), PhyloTree node '' (0x7fc4eea4955), PhyloTree node '' (0x7fc4eea49d0), PhyloTree node '' (0x7fc4eea48d7), PhyloTree node '' (0x7fc4eea4b2f), PhyloTree node '' (0x7fc4eea4b6e), PhyloTree node '' (0x7fc4eea4a3f), PhyloTree node '' (0x7fc4eea4bfb), PhyloTree node '' (0x7fc4eea4a09), PhyloTree node '' (0x7fc4eea4847), PhyloTree node '' (0x7fc4eea498b), PhyloTree node '' (0x7fc4eea4af3)], 'RECON_dups': 34, 'RECON_losses': 0}}\n",
      "../OMA_data_unfiltered/OMA_data/LUCA/HOG458/\n",
      "ML\n",
      "root scoring\n",
      "root score 210\n",
      "root score_nr 137\n",
      "algo 1\n",
      "dups: 8\n",
      "losses: 91\n",
      "speciations: 5\n",
      "algo 2\n",
      "dups: 2\n",
      "losses: 0\n",
      "speciations: 11\n",
      "{'../OMA_data_unfiltered/OMA_data/LUCA/HOG458/sequences.aln.muscle.fst.treefile.rooted': {'score': 21, 'stats': DescribeResult(nobs=27, minmax=(0.0059972505397157955, 0.12917520785904119), mean=0.037037037037037035, variance=0.0010250521549931163, skewness=1.387734270543774, kurtosis=1.2422408747653684), 'ultrametricity': DescribeResult(nobs=14, minmax=(0.604292819, 1.5848094895), mean=1.0697201607857143, variance=0.1183996493911888, skewness=0.12495370583822032, kurtosis=-1.42843304737579), 'ultrametricity_norm': DescribeResult(nobs=14, minmax=(0.5649073852699421, 1.4815178283037596), mean=1.0, variance=0.10346895299451978, skewness=0.12495370583822002, kurtosis=-1.4284330473757896), 'root_score': 210, 'root_score_nr': 137, 'SO_speciations': 5, 'SO_dups': 8, 'SO_losses': 91, 'RECON_speciations': [PhyloTree node '' (0x7fc4edc2576), PhyloTree node '' (0x7fc4f56fd9a), PhyloTree node '' (0x7fc4f56ff1d), PhyloTree node '' (0x7fc4f56fef0), PhyloTree node '' (0x7fc4f56fffb), PhyloTree node '' (0x7fc4f56ff5f), PhyloTree node '' (0x7fc4f56fd16), PhyloTree node '' (0x7fc4f56fca1), PhyloTree node '' (0x7fc4f56ffd7), PhyloTree node '' (0x7fc4f56fd76), PhyloTree node '' (0x7fc4f56ff77)], 'RECON_dups': 2, 'RECON_losses': 0}}\n",
      "FT\n",
      "root scoring\n",
      "root score 212\n",
      "root score_nr 139\n",
      "algo 1\n",
      "dups: 7\n",
      "losses: 76\n",
      "speciations: 6\n",
      "algo 2\n",
      "dups: 2\n",
      "losses: 0\n",
      "speciations: 11\n",
      "{'../OMA_data_unfiltered/OMA_data/LUCA/HOG458/fident_1_exp_struct_tree.PP.nwk.rooted': {'score': 22, 'stats': DescribeResult(nobs=27, minmax=(0.0020370674052665864, 0.15044810969472575), mean=0.037037037037037035, variance=0.0011947176526748778, skewness=1.480904809914163, kurtosis=2.412827715279657), 'ultrametricity': DescribeResult(nobs=14, minmax=(0.361757, 0.749123), mean=0.5451411428571429, variance=0.01493732250982417, skewness=0.08642599732840896, kurtosis=-0.9460798378033166), 'ultrametricity_norm': DescribeResult(nobs=14, minmax=(0.6636024536764791, 1.3741817322276695), mean=0.9999999999999999, variance=0.050263744287054025, skewness=0.08642599732841093, kurtosis=-0.946079837803318), 'root_score': 212, 'root_score_nr': 139, 'SO_speciations': 6, 'SO_dups': 7, 'SO_losses': 76, 'RECON_speciations': [PhyloTree node '' (0x7fc4ef92552), PhyloTree node '' (0x7fc4ef92714), PhyloTree node '' (0x7fc4ef92597), PhyloTree node '' (0x7fc4ef927f8), PhyloTree node '' (0x7fc4ef925dc), PhyloTree node '' (0x7fc4ef927a4), PhyloTree node '' (0x7fc4ef924d7), PhyloTree node '' (0x7fc4ef9250d), PhyloTree node '' (0x7fc4ef924fe), PhyloTree node '' (0x7fc4ef92729), PhyloTree node '' (0x7fc4ef926a8)], 'RECON_dups': 2, 'RECON_losses': 0}}\n",
      "../OMA_data_unfiltered/OMA_data/LUCA/HOG411/\n",
      "ML\n",
      "root scoring\n",
      "root score 336\n",
      "root score_nr 251\n",
      "algo 1\n",
      "dups: 5\n",
      "losses: 74\n",
      "speciations: 8\n",
      "algo 2\n",
      "dups: 2\n",
      "losses: 0\n",
      "speciations: 11\n",
      "{'../OMA_data_unfiltered/OMA_data/LUCA/HOG411/sequences.aln.muscle.fst.treefile.rooted': {'score': 54, 'stats': DescribeResult(nobs=27, minmax=(0.005678702670629255, 0.16112529017674118), mean=0.037037037037037035, variance=0.0011975955114128232, skewness=2.054807242386695, kurtosis=4.335583884812022), 'ultrametricity': DescribeResult(nobs=14, minmax=(1.075574145, 2.309023222), mean=1.6438420627714285, variance=0.12742360012560752, skewness=-0.01053943993687412, kurtosis=-0.9213459652548708), 'ultrametricity_norm': DescribeResult(nobs=14, minmax=(0.6543050390051709, 1.404650285020151), mean=1.0, variance=0.04715521106673023, skewness=-0.01053943993687383, kurtosis=-0.9213459652548717), 'root_score': 336, 'root_score_nr': 251, 'SO_speciations': 8, 'SO_dups': 5, 'SO_losses': 74, 'RECON_speciations': [PhyloTree node '' (0x7fc4ec2bdbb), PhyloTree node '' (0x7fc4ec2be6f), PhyloTree node '' (0x7fc4ec2bf47), PhyloTree node '' (0x7fc4ec2bf0b), PhyloTree node '' (0x7fc4ec2bcf2), PhyloTree node '' (0x7fc4ec2bee1), PhyloTree node '' (0x7fc4ec2bc32), PhyloTree node '' (0x7fc4ec2be51), PhyloTree node '' (0x7fc4ec2bca7), PhyloTree node '' (0x7fc4ec2bf89), PhyloTree node '' (0x7fc4ec2bcbc)], 'RECON_dups': 2, 'RECON_losses': 0}}\n",
      "FT\n",
      "root scoring\n",
      "root score 349\n",
      "root score_nr 250\n",
      "algo 1\n",
      "dups: 6\n",
      "losses: 94\n",
      "speciations: 7\n",
      "algo 2\n",
      "dups: 3\n",
      "losses: 0\n",
      "speciations: 10\n",
      "{'../OMA_data_unfiltered/OMA_data/LUCA/HOG411/fident_1_exp_struct_tree.PP.nwk.rooted': {'score': 53, 'stats': DescribeResult(nobs=27, minmax=(0.003051918200287693, 0.14829534500912014), mean=0.03703703703703703, variance=0.0012416591125058737, skewness=1.6687288961414521, kurtosis=2.4332409649753775), 'ultrametricity': DescribeResult(nobs=14, minmax=(0.6298699999999999, 0.9568599999999999), mean=0.8092278571428569, variance=0.0072481254027472524, skewness=-0.44485648130286065, kurtosis=-0.010259685885915193), 'ultrametricity_norm': DescribeResult(nobs=14, minmax=(0.7783592648724226, 1.1824358140343663), mean=1.0000000000000002, variance=0.011068379703795931, skewness=-0.44485648130286815, kurtosis=-0.010259685885909864), 'root_score': 349, 'root_score_nr': 250, 'SO_speciations': 7, 'SO_dups': 6, 'SO_losses': 94, 'RECON_speciations': [PhyloTree node '' (0x7fc4ee0e9df), PhyloTree node '' (0x7fc4ee0eb14), PhyloTree node '' (0x7fc4ee0e9ca), PhyloTree node '' (0x7fc4ee0ea4e), PhyloTree node '' (0x7fc4ee0e9a6), PhyloTree node '' (0x7fc4ee0e9d0), PhyloTree node '' (0x7fc4ee0ea2d), PhyloTree node '' (0x7fc4ee0ea09), PhyloTree node '' (0x7fc4ee0ea2a), PhyloTree node '' (0x7fc4ee0eb92)], 'RECON_dups': 3, 'RECON_losses': 0}}\n",
      "../OMA_data_unfiltered/OMA_data/LUCA/HOG569/\n",
      "ML\n",
      "root scoring\n",
      "root score 425\n",
      "root score_nr 308\n",
      "algo 1\n",
      "dups: 8\n",
      "losses: 138\n",
      "speciations: 8\n",
      "algo 2\n",
      "dups: 4\n",
      "losses: 0\n",
      "speciations: 12\n",
      "{'../OMA_data_unfiltered/OMA_data/LUCA/HOG569/sequences.aln.muscle.fst.treefile.rooted': {'score': 52, 'stats': DescribeResult(nobs=33, minmax=(0.003361628522695759, 0.09593448630499905), mean=0.030303030303030304, variance=0.0006593417285188652, skewness=0.9201777951803927, kurtosis=-0.06713012741668489), 'ultrametricity': DescribeResult(nobs=17, minmax=(1.6127939939999998, 3.0933647998000002), mean=2.0386776204, variance=0.20256399627188992, skewness=1.469604517496269, kurtosis=1.0863596465705472), 'ultrametricity_norm': DescribeResult(nobs=17, minmax=(0.7910981009756513, 1.5173388714558336), mean=1.0, variance=0.04873771291201753, skewness=1.4696045174962689, kurtosis=1.0863596465705472), 'root_score': 425, 'root_score_nr': 308, 'SO_speciations': 8, 'SO_dups': 8, 'SO_losses': 138, 'RECON_speciations': [PhyloTree node '' (0x7fc4ec58b3b), PhyloTree node '' (0x7fc4ec58904), PhyloTree node '' (0x7fc4ec58a81), PhyloTree node '' (0x7fc4ec5891c), PhyloTree node '' (0x7fc4ec58a72), PhyloTree node '' (0x7fc4ec58ba1), PhyloTree node '' (0x7fc4ec58856), PhyloTree node '' (0x7fc4ec58a3c), PhyloTree node '' (0x7fc4ec58bc2), PhyloTree node '' (0x7fc4ec589f4), PhyloTree node '' (0x7fc4ec58985), PhyloTree node '' (0x7fc4ec58bfe)], 'RECON_dups': 4, 'RECON_losses': 0}}\n",
      "FT\n",
      "root scoring\n",
      "root score 433\n",
      "root score_nr 318\n",
      "algo 1\n",
      "dups: 7\n",
      "losses: 118\n",
      "speciations: 9\n",
      "algo 2\n",
      "dups: 4\n",
      "losses: 0\n",
      "speciations: 12\n",
      "{'../OMA_data_unfiltered/OMA_data/LUCA/HOG569/fident_1_exp_struct_tree.PP.nwk.rooted': {'score': 57, 'stats': DescribeResult(nobs=33, minmax=(0.002255773138403261, 0.09854841146366365), mean=0.030303030303030297, variance=0.0007575236243871292, skewness=0.7012775717788547, kurtosis=-0.6900859142596749), 'ultrametricity': DescribeResult(nobs=17, minmax=(0.678257, 0.942907), mean=0.7847736470588235, variance=0.0068310408586176475, skewness=0.4516655632836373, kurtosis=-0.9400370504198956), 'ultrametricity_norm': DescribeResult(nobs=17, minmax=(0.864270866563847, 1.2015018643067708), mean=1.0, variance=0.011091698637581651, skewness=0.4516655632836379, kurtosis=-0.9400370504198943), 'root_score': 433, 'root_score_nr': 318, 'SO_speciations': 9, 'SO_dups': 7, 'SO_losses': 118, 'RECON_speciations': [PhyloTree node '' (0x7fc4f524432), PhyloTree node '' (0x7fc4f524675), PhyloTree node '' (0x7fc4f52472f), PhyloTree node '' (0x7fc4f524510), PhyloTree node '' (0x7fc4f5244fb), PhyloTree node '' (0x7fc4f52469c), PhyloTree node '' (0x7fc4f524630), PhyloTree node '' (0x7fc4f524411), PhyloTree node '' (0x7fc4f5245f4), PhyloTree node '' (0x7fc4f52479b), PhyloTree node '' (0x7fc4f5245d6), PhyloTree node '' (0x7fc4f52444d)], 'RECON_dups': 4, 'RECON_losses': 0}}\n",
      "../OMA_data_unfiltered/OMA_data/LUCA/HOG1118/\n",
      "ML\n",
      "root scoring\n",
      "root score 836\n",
      "root score_nr 790\n",
      "algo 1\n",
      "dups: 1\n",
      "losses: 4\n",
      "speciations: 7\n",
      "algo 2\n",
      "dups: 0\n",
      "losses: 0\n",
      "speciations: 8\n",
      "{'../OMA_data_unfiltered/OMA_data/LUCA/HOG1118/sequences.aln.muscle.fst.treefile.rooted': {'score': 162, 'stats': DescribeResult(nobs=17, minmax=(1.5390224572589244e-05, 0.23155828229766695), mean=0.058823529411764705, variance=0.004419010525093272, skewness=1.2785122221755565, kurtosis=0.7359287320910739), 'ultrametricity': DescribeResult(nobs=9, minmax=(1.334570371, 1.669072913), mean=1.458943301811111, variance=0.011166778933836965, skewness=0.8470046051674676, kurtosis=-0.31431205732288126), 'ultrametricity_norm': DescribeResult(nobs=9, minmax=(0.9147513610318397, 1.1440286342368733), mean=0.9999999999999998, variance=0.005246275458214987, skewness=0.8470046051674766, kurtosis=-0.3143120573228728), 'root_score': 836, 'root_score_nr': 790, 'SO_speciations': 7, 'SO_dups': 1, 'SO_losses': 4, 'RECON_speciations': [PhyloTree node '' (0x7fc4ef8e507), PhyloTree node '' (0x7fc4ec2beae), PhyloTree node '' (0x7fc4ec2bf1a), PhyloTree node '' (0x7fc4ec2bd94), PhyloTree node '' (0x7fc4ec2bd7c), PhyloTree node '' (0x7fc4ec2be24), PhyloTree node '' (0x7fc4ec2be33), PhyloTree node '' (0x7fc4ec2bda9)], 'RECON_dups': 0, 'RECON_losses': 0}}\n",
      "FT\n",
      "root scoring\n",
      "root score 836\n",
      "root score_nr 790\n",
      "algo 1\n",
      "dups: 1\n",
      "losses: 4\n",
      "speciations: 7\n",
      "algo 2\n",
      "dups: 0\n",
      "losses: 0\n",
      "speciations: 8\n",
      "{'../OMA_data_unfiltered/OMA_data/LUCA/HOG1118/fident_1_exp_struct_tree.PP.nwk.rooted': {'score': 162, 'stats': DescribeResult(nobs=17, minmax=(0.002060253184566543, 0.26345948651746076), mean=0.0588235294117647, variance=0.0048551412992872515, skewness=1.808971750676322, kurtosis=2.703781973743851), 'ultrametricity': DescribeResult(nobs=9, minmax=(0.60451, 0.73939), mean=0.6889555555555555, variance=0.0017057356777777782, skewness=-0.7988485888645112, kurtosis=0.10350518422868404), 'ultrametricity_norm': DescribeResult(nobs=9, minmax=(0.8774296035867497, 1.0732042060445763), mean=1.0, variance=0.0035935965725235586, skewness=-0.798848588864512, kurtosis=0.10350518422868404), 'root_score': 836, 'root_score_nr': 790, 'SO_speciations': 7, 'SO_dups': 1, 'SO_losses': 4, 'RECON_speciations': [PhyloTree node '' (0x7fc4eeb1612), PhyloTree node '' (0x7fc4eeb1654), PhyloTree node '' (0x7fc4eeb17e3), PhyloTree node '' (0x7fc4eeb14aa), PhyloTree node '' (0x7fc4eeb1657), PhyloTree node '' (0x7fc4eeb1567), PhyloTree node '' (0x7fc4eeb16ed), PhyloTree node '' (0x7fc4ed7cb0e)], 'RECON_dups': 0, 'RECON_losses': 0}}\n",
      "../OMA_data_unfiltered/OMA_data/LUCA/HOG423/\n",
      "ML\n",
      "root scoring\n",
      "root score 240\n",
      "root score_nr 179\n",
      "algo 1\n",
      "dups: 8\n",
      "losses: 33\n",
      "speciations: 3\n",
      "algo 2\n",
      "dups: 6\n",
      "losses: 0\n",
      "speciations: 5\n",
      "{'../OMA_data_unfiltered/OMA_data/LUCA/HOG423/sequences.aln.muscle.fst.treefile.rooted': {'score': 43, 'stats': DescribeResult(nobs=23, minmax=(0.0004167458616684485, 0.16473503273103732), mean=0.043478260869565216, variance=0.0023578561748345372, skewness=1.320286326394552, kurtosis=0.6294874007240647), 'ultrametricity': DescribeResult(nobs=12, minmax=(0.5381114336, 1.2939798146), mean=0.8063970663750001, variance=0.05577957634881248, skewness=1.0904986394378136, kurtosis=0.20001420017812466), 'ultrametricity_norm': DescribeResult(nobs=12, minmax=(0.6673033125219248, 1.6046435044919403), mean=0.9999999999999999, variance=0.08577827989177653, skewness=1.0904986394378147, kurtosis=0.2000142001781251), 'root_score': 240, 'root_score_nr': 179, 'SO_speciations': 3, 'SO_dups': 8, 'SO_losses': 33, 'RECON_speciations': [PhyloTree node '' (0x7fc4eecc1c7), PhyloTree node '' (0x7fc4ec2558e), PhyloTree node '' (0x7fc4eecc37d), PhyloTree node '' (0x7fc4eecc293), PhyloTree node '' (0x7fc4eecc188)], 'RECON_dups': 6, 'RECON_losses': 0}}\n",
      "FT\n",
      "root scoring\n",
      "root score 256\n",
      "root score_nr 181\n",
      "algo 1\n",
      "dups: 8\n",
      "losses: 36\n",
      "speciations: 3\n",
      "algo 2\n",
      "dups: 6\n",
      "losses: 0\n",
      "speciations: 5\n",
      "{'../OMA_data_unfiltered/OMA_data/LUCA/HOG423/fident_1_exp_struct_tree.PP.nwk.rooted': {'score': 45, 'stats': DescribeResult(nobs=23, minmax=(0.0009906195772604292, 0.22565366224611144), mean=0.043478260869565216, variance=0.0028129716542458004, skewness=1.9740377141085685, kurtosis=4.120324865590082), 'ultrametricity': DescribeResult(nobs=12, minmax=(0.48671999999999993, 0.72573), mean=0.5704983333333334, variance=0.004241401033333332, skewness=0.9737180997321918, kurtosis=0.7280437510250191), 'ultrametricity_norm': DescribeResult(nobs=12, minmax=(0.8531488552405935, 1.2720983701383874), mean=0.9999999999999999, variance=0.013031685123965523, skewness=0.9737180997321944, kurtosis=0.7280437510250235), 'root_score': 256, 'root_score_nr': 181, 'SO_speciations': 3, 'SO_dups': 8, 'SO_losses': 36, 'RECON_speciations': [PhyloTree node '' (0x7fc4eeb15cd), PhyloTree node '' (0x7fc4eeb14d7), PhyloTree node '' (0x7fc4eeb1597), PhyloTree node '' (0x7fc4eeb16a8), PhyloTree node '' (0x7fc4eeb1774)], 'RECON_dups': 6, 'RECON_losses': 0}}\n",
      "../OMA_data_unfiltered/OMA_data/LUCA/HOG600/\n",
      "ML\n",
      "root scoring\n",
      "root score 13827\n",
      "root score_nr 12348\n",
      "algo 1\n",
      "dups: 56\n",
      "losses: 857\n",
      "speciations: 73\n",
      "algo 2\n",
      "dups: 36\n",
      "losses: 0\n",
      "speciations: 93\n",
      "{'../OMA_data_unfiltered/OMA_data/LUCA/HOG600/sequences.aln.muscle.fst.treefile.rooted': {'score': 1895, 'stats': DescribeResult(nobs=259, minmax=(1.2916577498312821e-08, 0.020572011462258736), mean=0.003861003861003861, variance=2.3078508839278938e-05, skewness=1.5880757869388469, kurtosis=1.6831168098554787), 'ultrametricity': DescribeResult(nobs=130, minmax=(1.275871982, 3.1839201508999997), mean=1.7774851215307692, variance=0.1478616734515561, skewness=1.3781026127479368, kurtosis=1.3896034599152651), 'ultrametricity_norm': DescribeResult(nobs=130, minmax=(0.7177961528596198, 1.7912499589071156), mean=1.0000000000000002, variance=0.04679976462119233, skewness=1.3781026127479332, kurtosis=1.389603459915258), 'root_score': 13827, 'root_score_nr': 12348, 'SO_speciations': 73, 'SO_dups': 56, 'SO_losses': 857, 'RECON_speciations': [PhyloTree node '' (0x7fc4ec2be72), PhyloTree node '' (0x7fc4ec2bd37), PhyloTree node '' (0x7fc4ec2bf41), PhyloTree node '' (0x7fc4ec2bc26), PhyloTree node '' (0x7fc4ec2bdb8), PhyloTree node '' (0x7fc4eecc39b), PhyloTree node '' (0x7fc4eecc128), PhyloTree node '' (0x7fc4ec2bc62), PhyloTree node '' (0x7fc4ec2bdc4), PhyloTree node '' (0x7fc4f49f074), PhyloTree node '' (0x7fc4eecc0e0), PhyloTree node '' (0x7fc4eecc2f3), PhyloTree node '' (0x7fc4eecc113), PhyloTree node '' (0x7fc4eecc1b8), PhyloTree node '' (0x7fc4ec2bf68), PhyloTree node '' (0x7fc4ec2bf05), PhyloTree node '' (0x7fc4f49f389), PhyloTree node '' (0x7fc4eecc2ea), PhyloTree node '' (0x7fc4eecc0a1), PhyloTree node '' (0x7fc4eecc047), PhyloTree node '' (0x7fc4eecc302), PhyloTree node '' (0x7fc4eecc3f2), PhyloTree node '' (0x7fc4eecc395), PhyloTree node '' (0x7fc4ec2bd3d), PhyloTree node '' (0x7fc4f49f24b), PhyloTree node '' (0x7fc4eecc1b5), PhyloTree node '' (0x7fc4eecc209), PhyloTree node '' (0x7fc4eecc137), PhyloTree node '' (0x7fc4eecc2e4), PhyloTree node '' (0x7fc4eecc0ef), PhyloTree node '' (0x7fc4eecc03b), PhyloTree node '' (0x7fc4eecc3ad), PhyloTree node '' (0x7fc4eecc0b6), PhyloTree node '' (0x7fc4eecc176), PhyloTree node '' (0x7fc4eecc092), PhyloTree node '' (0x7fc4ec2bf0e), PhyloTree node '' (0x7fc4ef8e435), PhyloTree node '' (0x7fc4ef8e7fb), PhyloTree node '' (0x7fc4eecc14c), PhyloTree node '' (0x7fc4eecc095), PhyloTree node '' (0x7fc4eecc062), PhyloTree node '' (0x7fc4eecc185), PhyloTree node '' (0x7fc4eecc263), PhyloTree node '' (0x7fc4eecc161), PhyloTree node '' (0x7fc4eecc1fd), PhyloTree node '' (0x7fc4eecc1a0), PhyloTree node '' (0x7fc4eecc1eb), PhyloTree node '' (0x7fc4ec2bc7a), PhyloTree node '' (0x7fc4ef8e5dc), PhyloTree node '' (0x7fc4f729baa), PhyloTree node '' (0x7fc4eecc233), PhyloTree node '' (0x7fc4eecc29c), PhyloTree node '' (0x7fc4eecc1e5), PhyloTree node '' (0x7fc4eecc3b0), PhyloTree node '' (0x7fc4eecc10d), PhyloTree node '' (0x7fc4eecc335), PhyloTree node '' (0x7fc4eecc2c0), PhyloTree node '' (0x7fc4eecc26c), PhyloTree node '' (0x7fc4f7299b8), PhyloTree node '' (0x7fc4eecc1f4), PhyloTree node '' (0x7fc4eecc164), PhyloTree node '' (0x7fc4eecc341), PhyloTree node '' (0x7fc4eecc16d), PhyloTree node '' (0x7fc4eecc3aa), PhyloTree node '' (0x7fc4eecc31a), PhyloTree node '' (0x7fc4eecc1ee), PhyloTree node '' (0x7fc4eecc194), PhyloTree node '' (0x7fc4f729916), PhyloTree node '' (0x7fc4eecc34a), PhyloTree node '' (0x7fc4eecc344), PhyloTree node '' (0x7fc4eecc053), PhyloTree node '' (0x7fc4eecc18e), PhyloTree node '' (0x7fc4eecc3a7), PhyloTree node '' (0x7fc4eecc0e6), PhyloTree node '' (0x7fc4eecc0fe), PhyloTree node '' (0x7fc4f729acf), PhyloTree node '' (0x7fc4f729b50), PhyloTree node '' (0x7fc4eecc3d7), PhyloTree node '' (0x7fc4eecc00b), PhyloTree node '' (0x7fc4eecc0ad), PhyloTree node '' (0x7fc4eecc3ef), PhyloTree node '' (0x7fc4eecc31d), PhyloTree node '' (0x7fc4eecc33b), PhyloTree node '' (0x7fc4eecc374), PhyloTree node '' (0x7fc4ee0e80e), PhyloTree node '' (0x7fc4eecc080), PhyloTree node '' (0x7fc4eecc059), PhyloTree node '' (0x7fc4ee0ea63), PhyloTree node '' (0x7fc4ee0ea72), PhyloTree node '' (0x7fc4eecc215), PhyloTree node '' (0x7fc4ee0ead2), PhyloTree node '' (0x7fc4eecc22d), PhyloTree node '' (0x7fc4ee0e811)], 'RECON_dups': 36, 'RECON_losses': 0}}\n",
      "FT\n",
      "root scoring\n",
      "root score 15405\n",
      "root score_nr 13779\n",
      "algo 1\n",
      "dups: 64\n",
      "losses: 865\n",
      "speciations: 65\n",
      "algo 2\n",
      "dups: 36\n",
      "losses: 0\n",
      "speciations: 93\n",
      "{'../OMA_data_unfiltered/OMA_data/LUCA/HOG600/fident_1_exp_struct_tree.PP.nwk.rooted': {'score': 1928, 'stats': DescribeResult(nobs=259, minmax=(0.0, 0.024571624442377697), mean=0.003861003861003861, variance=2.370881365937881e-05, skewness=1.5544723817295334, kurtosis=1.6468058728302681), 'ultrametricity': DescribeResult(nobs=130, minmax=(0.583806, 1.098896), mean=0.739332, variance=0.00900331998108527, skewness=1.3539509133377343, kurtosis=2.1316342960849974), 'ultrametricity_norm': DescribeResult(nobs=130, minmax=(0.7896398370420867, 1.4863363144027313), mean=1.0, variance=0.01647114072284735, skewness=1.3539509133377339, kurtosis=2.1316342960849957), 'root_score': 15405, 'root_score_nr': 13779, 'SO_speciations': 65, 'SO_dups': 64, 'SO_losses': 865, 'RECON_speciations': [PhyloTree node '' (0x7fc4ef49b41), PhyloTree node '' (0x7fc4f414e7e), PhyloTree node '' (0x7fc4f414cec), PhyloTree node '' (0x7fc4f414fc2), PhyloTree node '' (0x7fc4f414c23), PhyloTree node '' (0x7fc4f414f4a), PhyloTree node '' (0x7fc4f414f5f), PhyloTree node '' (0x7fc4f414df4), PhyloTree node '' (0x7fc4f414dc7), PhyloTree node '' (0x7fc4f414c35), PhyloTree node '' (0x7fc4f414d13), PhyloTree node '' (0x7fc4f414fb6), PhyloTree node '' (0x7fc4f414f20), PhyloTree node '' (0x7fc4f414c08), PhyloTree node '' (0x7fc4f414d40), PhyloTree node '' (0x7fc4f414e9c), PhyloTree node '' (0x7fc4f414d97), PhyloTree node '' (0x7fc4f414d10), PhyloTree node '' (0x7fc4f414e33), PhyloTree node '' (0x7fc4f414c20), PhyloTree node '' (0x7fc4f414f8c), PhyloTree node '' (0x7fc4f414e81), PhyloTree node '' (0x7fc4f414fa1), PhyloTree node '' (0x7fc4f414df1), PhyloTree node '' (0x7fc4f414cad), PhyloTree node '' (0x7fc4f414dee), PhyloTree node '' (0x7fc4f414f80), PhyloTree node '' (0x7fc4f414f02), PhyloTree node '' (0x7fc4f414c53), PhyloTree node '' (0x7fc4f414e6c), PhyloTree node '' (0x7fc4f414e72), PhyloTree node '' (0x7fc4f414d88), PhyloTree node '' (0x7fc4f414f8f), PhyloTree node '' (0x7fc4f414d70), PhyloTree node '' (0x7fc4f414c77), PhyloTree node '' (0x7fc4f414e09), PhyloTree node '' (0x7fc4ee00ff8), PhyloTree node '' (0x7fc4ee00e48), PhyloTree node '' (0x7fc4f414f9b), PhyloTree node '' (0x7fc4f414e03), PhyloTree node '' (0x7fc4f414eff), PhyloTree node '' (0x7fc4f414e15), PhyloTree node '' (0x7fc4f414ffe), PhyloTree node '' (0x7fc4f414cce), PhyloTree node '' (0x7fc4f414d67), PhyloTree node '' (0x7fc4ee00e69), PhyloTree node '' (0x7fc4ee00ede), PhyloTree node '' (0x7fc4ee00eb4), PhyloTree node '' (0x7fc4f414ee1), PhyloTree node '' (0x7fc4f414ddf), PhyloTree node '' (0x7fc4f414cb6), PhyloTree node '' (0x7fc4f414e39), PhyloTree node '' (0x7fc4f414cfb), PhyloTree node '' (0x7fc4f414fc5), PhyloTree node '' (0x7fc4f414cef), PhyloTree node '' (0x7fc4ee00e1b), PhyloTree node '' (0x7fc4ee00c47), PhyloTree node '' (0x7fc4ee00c56), PhyloTree node '' (0x7fc4ee00e63), PhyloTree node '' (0x7fc4f414f71), PhyloTree node '' (0x7fc4f414d2e), PhyloTree node '' (0x7fc4f414c71), PhyloTree node '' (0x7fc4f414e90), PhyloTree node '' (0x7fc4f414ecf), PhyloTree node '' (0x7fc4f414ce6), PhyloTree node '' (0x7fc4f414ede), PhyloTree node '' (0x7fc4ee00f5f), PhyloTree node '' (0x7fc4ee00c44), PhyloTree node '' (0x7fc4ee00d07), PhyloTree node '' (0x7fc4f414ea5), PhyloTree node '' (0x7fc4f414f1d), PhyloTree node '' (0x7fc4f414dfa), PhyloTree node '' (0x7fc4f414cb3), PhyloTree node '' (0x7fc4f414d7f), PhyloTree node '' (0x7fc4f414c0e), PhyloTree node '' (0x7fc4ee00c05), PhyloTree node '' (0x7fc4ee00f32), PhyloTree node '' (0x7fc4f414c1d), PhyloTree node '' (0x7fc4f414e54), PhyloTree node '' (0x7fc4f414d19), PhyloTree node '' (0x7fc4ee00c6b), PhyloTree node '' (0x7fc4ee00f08), PhyloTree node '' (0x7fc4f414d9d), PhyloTree node '' (0x7fc4f414f0e), PhyloTree node '' (0x7fc4f414f4d), PhyloTree node '' (0x7fc4f414dca), PhyloTree node '' (0x7fc4ee00e15), PhyloTree node '' (0x7fc4ee00c50), PhyloTree node '' (0x7fc4f414e78), PhyloTree node '' (0x7fc4f414f53), PhyloTree node '' (0x7fc4f414cbf), PhyloTree node '' (0x7fc4f414ed5), PhyloTree node '' (0x7fc4f414cfe)], 'RECON_dups': 36, 'RECON_losses': 0}}\n",
      "../OMA_data_unfiltered/OMA_data/LUCA/HOG475/\n",
      "ML\n",
      "root scoring\n",
      "root score 3021\n",
      "root score_nr 2558\n",
      "algo 1\n",
      "dups: 21\n",
      "losses: 801\n",
      "speciations: 25\n",
      "algo 2\n",
      "dups: 9\n",
      "losses: 0\n",
      "speciations: 37\n",
      "{'../OMA_data_unfiltered/OMA_data/LUCA/HOG475/sequences.aln.muscle.fst.treefile.rooted': {'score': 436, 'stats': DescribeResult(nobs=93, minmax=(2.83314223416679e-08, 0.06526093844657502), mean=0.010752688172043013, variance=0.0002025415621498512, skewness=2.039858919600559, kurtosis=3.7144237617457776), 'ultrametricity': DescribeResult(nobs=47, minmax=(1.2635825576, 3.6384618846000003), mean=2.1773751068319154, variance=0.22456956710107748, skewness=0.9664233186756652, kurtosis=1.1969229565548734), 'ultrametricity_norm': DescribeResult(nobs=47, minmax=(0.5803237823539348, 1.6710312675035441), mean=0.9999999999999999, variance=0.04736792875478408, skewness=0.9664233186756679, kurtosis=1.1969229565548796), 'root_score': 3021, 'root_score_nr': 2558, 'SO_speciations': 25, 'SO_dups': 21, 'SO_losses': 801, 'RECON_speciations': [PhyloTree node '' (0x7fc4eec29d6), PhyloTree node '' (0x7fc4eec2ba4), PhyloTree node '' (0x7fc4eec288f), PhyloTree node '' (0x7fc4eec2a4e), PhyloTree node '' (0x7fc4eec2850), PhyloTree node '' (0x7fc4eec2949), PhyloTree node '' (0x7fc4eec28e6), PhyloTree node '' (0x7fc4eec2b9b), PhyloTree node '' (0x7fc4eec29be), PhyloTree node '' (0x7fc4eec28da), PhyloTree node '' (0x7fc4eec298e), PhyloTree node '' (0x7fc4eec287a), PhyloTree node '' (0x7fc4eec2b23), PhyloTree node '' (0x7fc4eec2955), PhyloTree node '' (0x7fc4eec2a54), PhyloTree node '' (0x7fc4eec2862), PhyloTree node '' (0x7fc4eec2985), PhyloTree node '' (0x7fc4eec2b77), PhyloTree node '' (0x7fc4eec2aa2), PhyloTree node '' (0x7fc4eec2a42), PhyloTree node '' (0x7fc4eec2859), PhyloTree node '' (0x7fc4eec2bc8), PhyloTree node '' (0x7fc4eec2b1d), PhyloTree node '' (0x7fc4eec29eb), PhyloTree node '' (0x7fc4eec286b), PhyloTree node '' (0x7fc4eec2ad2), PhyloTree node '' (0x7fc4ef3e504), PhyloTree node '' (0x7fc4eec28b3), PhyloTree node '' (0x7fc4eec2baa), PhyloTree node '' (0x7fc4eec2bf5), PhyloTree node '' (0x7fc4ef3e44a), PhyloTree node '' (0x7fc4eec291f), PhyloTree node '' (0x7fc4ef3e744), PhyloTree node '' (0x7fc4f78a717), PhyloTree node '' (0x7fc4f78a7c2), PhyloTree node '' (0x7fc4f78a7b0), PhyloTree node '' (0x7fc4ee71aff)], 'RECON_dups': 9, 'RECON_losses': 0}}\n",
      "FT\n",
      "root scoring\n",
      "root score 2813\n",
      "root score_nr 2311\n",
      "algo 1\n",
      "dups: 24\n",
      "losses: 910\n",
      "speciations: 22\n",
      "algo 2\n",
      "dups: 10\n",
      "losses: 0\n",
      "speciations: 36\n",
      "{'../OMA_data_unfiltered/OMA_data/LUCA/HOG475/fident_1_exp_struct_tree.PP.nwk.rooted': {'score': 410, 'stats': DescribeResult(nobs=93, minmax=(0.0, 0.06337203886226911), mean=0.010752688172043012, variance=0.00017913048749718363, skewness=1.7112878741330697, kurtosis=2.5118921591565115), 'ultrametricity': DescribeResult(nobs=47, minmax=(0.550548, 0.977178), mean=0.7271045957446808, variance=0.0097543836359852, skewness=0.6826055302582074, kurtosis=-0.18999809100181642), 'ultrametricity_norm': DescribeResult(nobs=47, minmax=(0.7571785451804822, 1.3439304409831172), mean=1.0000000000000002, variance=0.018450411339301725, skewness=0.682605530258202, kurtosis=-0.18999809100182086), 'root_score': 2813, 'root_score_nr': 2311, 'SO_speciations': 22, 'SO_dups': 24, 'SO_losses': 910, 'RECON_speciations': [PhyloTree node '' (0x7fc4f524603), PhyloTree node '' (0x7fc4f52449e), PhyloTree node '' (0x7fc4ec589fd), PhyloTree node '' (0x7fc4ef193d4), PhyloTree node '' (0x7fc4ef192e7), PhyloTree node '' (0x7fc4ec58892), PhyloTree node '' (0x7fc4ef19302), PhyloTree node '' (0x7fc4ef19188), PhyloTree node '' (0x7fc4ef190f5), PhyloTree node '' (0x7fc4ef1911f), PhyloTree node '' (0x7fc4ef19293), PhyloTree node '' (0x7fc4ef1939e), PhyloTree node '' (0x7fc4ef191e8), PhyloTree node '' (0x7fc4ef1907a), PhyloTree node '' (0x7fc4ef190a4), PhyloTree node '' (0x7fc4ef190bc), PhyloTree node '' (0x7fc4ef193aa), PhyloTree node '' (0x7fc4ef19305), PhyloTree node '' (0x7fc4ef191f7), PhyloTree node '' (0x7fc4ef19032), PhyloTree node '' (0x7fc4ef19059), PhyloTree node '' (0x7fc4ef19266), PhyloTree node '' (0x7fc4ef190c5), PhyloTree node '' (0x7fc4ef19365), PhyloTree node '' (0x7fc4ef19362), PhyloTree node '' (0x7fc4ef1911c), PhyloTree node '' (0x7fc4ef19053), PhyloTree node '' (0x7fc4ef1920c), PhyloTree node '' (0x7fc4ef19131), PhyloTree node '' (0x7fc4ef1908c), PhyloTree node '' (0x7fc4ef19374), PhyloTree node '' (0x7fc4ef19395), PhyloTree node '' (0x7fc4ef190fe), PhyloTree node '' (0x7fc4ef19227), PhyloTree node '' (0x7fc4ef19020), PhyloTree node '' (0x7fc4ef1907d)], 'RECON_dups': 10, 'RECON_losses': 0}}\n",
      "../OMA_data_unfiltered/OMA_data/LUCA/HOG1231/\n",
      "ML\n",
      "root scoring\n",
      "root score 744\n",
      "root score_nr 639\n",
      "algo 1\n",
      "dups: 11\n",
      "losses: 78\n",
      "speciations: 6\n",
      "algo 2\n",
      "dups: 9\n",
      "losses: 0\n",
      "speciations: 8\n",
      "{'../OMA_data_unfiltered/OMA_data/LUCA/HOG1231/sequences.aln.muscle.fst.treefile.rooted': {'score': 143, 'stats': DescribeResult(nobs=35, minmax=(0.0034049105971058024, 0.11241875978306243), mean=0.028571428571428567, variance=0.0006097868621712291, skewness=1.470273612667945, kurtosis=2.1881883511687947), 'ultrametricity': DescribeResult(nobs=18, minmax=(0.8511997070000001, 2.3324786239999997), mean=1.7005136213111112, variance=0.15804953265854188, skewness=-0.2016442979455516, kurtosis=-0.4860690607399696), 'ultrametricity_norm': DescribeResult(nobs=18, minmax=(0.5005544773841433, 1.3716318380335217), mean=1.0, variance=0.05465538853086497, skewness=-0.20164429794555142, kurtosis=-0.48606906073996825), 'root_score': 744, 'root_score_nr': 639, 'SO_speciations': 6, 'SO_dups': 11, 'SO_losses': 78, 'RECON_speciations': [PhyloTree node '' (0x7fc4ec70f53), PhyloTree node '' (0x7fc4ec70ffe), PhyloTree node '' (0x7fc4ec70eb1), PhyloTree node '' (0x7fc4ec70ea5), PhyloTree node '' (0x7fc4ec70ede), PhyloTree node '' (0x7fc4ec70eff), PhyloTree node '' (0x7fc4ec70eae), PhyloTree node '' (0x7fc4ec70ec9)], 'RECON_dups': 9, 'RECON_losses': 0}}\n",
      "FT\n",
      "root scoring\n",
      "root score 740\n",
      "root score_nr 623\n",
      "algo 1\n",
      "dups: 10\n",
      "losses: 61\n",
      "speciations: 7\n",
      "algo 2\n",
      "dups: 9\n",
      "losses: 0\n",
      "speciations: 8\n",
      "{'../OMA_data_unfiltered/OMA_data/LUCA/HOG1231/fident_1_exp_struct_tree.PP.nwk.rooted': {'score': 139, 'stats': DescribeResult(nobs=35, minmax=(0.0018279944472697708, 0.10511756453535198), mean=0.028571428571428567, variance=0.0005956428217754908, skewness=1.1872276646870867, kurtosis=1.2098931797194448), 'ultrametricity': DescribeResult(nobs=18, minmax=(0.535658, 0.9676979999999998), mean=0.7868008888888888, variance=0.008528468154928102, skewness=-0.75885563116389, kurtosis=1.6255664880170304), 'ultrametricity_norm': DescribeResult(nobs=18, minmax=(0.6808050264870062, 1.2299147264139112), mean=1.0, variance=0.01377657777588684, skewness=-0.7588556311638898, kurtosis=1.6255664880170304), 'root_score': 740, 'root_score_nr': 623, 'SO_speciations': 7, 'SO_dups': 10, 'SO_losses': 61, 'RECON_speciations': [PhyloTree node '' (0x7fc4f4564a4), PhyloTree node '' (0x7fc4eedf7bf), PhyloTree node '' (0x7fc4eedf612), PhyloTree node '' (0x7fc4eedf6ed), PhyloTree node '' (0x7fc4eedf57f), PhyloTree node '' (0x7fc4eedf726), PhyloTree node '' (0x7fc4eedf465), PhyloTree node '' (0x7fc4eedf684)], 'RECON_dups': 9, 'RECON_losses': 0}}\n"
     ]
    }
   ],
   "source": [
    "for sample in sample_hogs:\n",
    "    print( sample)\n",
    "    treeML = sample+'sequences.aln.muscle.fst.treefile.rooted'\n",
    "    treeFT = sample+'fident_1_exp_struct_tree.PP.nwk.rooted'\n",
    "    finalset = sample +'finalset.csv'\n",
    "\n",
    "    print('ML')\n",
    "    scores = calc_scores(treeML , finalset)\n",
    "    print(scores)\n",
    "    print('FT')\n",
    "    scores = calc_scores(treeFT , finalset)\n",
    "    print(scores)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "astralpath = '../A-pro/ASTRAL-MP/astral.1.1.6.jar'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#prepare final set df for astral"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "#running astral\n",
    "\n",
    "\n",
    "def run_astral( phylo , st,  mapping, output ,root  , astralpath = '/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro' ):\n",
    "\n",
    "    print('running astral')\n",
    "    cmd = astralpath + ' -c {st} -a {mapping} -u 2 -i {phylo} -o {output} -C --root {root} 2>{logfile}'.format(phylo = phylo , st=st, mapping = mapping , output = output , root=root,logfile = output+'.log')\n",
    "\n",
    "    print(cmd)\n",
    "    subprocess.run(cmd , shell = True)\n",
    "    return output\n",
    "\n",
    "\n",
    "def prepare_astral_input(uniprot_df , output):\n",
    "    finalset = pd.read_csv(uniprot_df)\n",
    "    finalset['species'] = finalset['Taxonomic lineage (Ids)'].map(lambda x: x.split(',')[-1].split('(')[0].strip())\n",
    "    mapper = dict(zip(finalset['Entry'], finalset['species']))\n",
    "    with open(output , 'w') as f:\n",
    "        for k,v in mapper.items():\n",
    "            f.write('{}\\t{}\\n'.format(k,v))\n",
    "    return mapper , output\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FT\n",
      "running astral\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro -c ../OMA_data_unfiltered/OMA_data/LUCA/HOG1077/finalset_ncbi_tree.nwk -a ../OMA_data_unfiltered/OMA_data/LUCA/HOG1077/speciesmap.txt -u 2 -i ../OMA_data_unfiltered/OMA_data/LUCA/HOG1077/fident_1_exp_struct_tree.PP.nwk.rooted -o ../OMA_data_unfiltered/OMA_data/LUCA/HOG1077/FT_astral_support.tre -C --root root 2>../OMA_data_unfiltered/OMA_data/LUCA/HOG1077/FT_astral_support.tre.log\n",
      "        pp1       pp2       pp3        f1        f2        f3        q1  \\\n",
      "0  0.284784  0.426968  0.288248  0.000000  0.263158  0.008772  0.000000   \n",
      "1  0.357007  0.336824  0.306169  0.100000  0.063158  0.000000  0.612903   \n",
      "2  0.333333  0.333333  0.333333  0.000000  0.000000  0.000000       NaN   \n",
      "3  0.345342  0.328480  0.326178  0.035088  0.004386  0.000000  0.888889   \n",
      "4  0.333333  0.333333  0.333333  0.000000  0.000000  0.000000       NaN   \n",
      "5  0.320829  0.354424  0.324747  0.000000  0.061776  0.007722  0.000000   \n",
      "6  0.333333  0.333333  0.333333  0.000000  0.000000  0.000000       NaN   \n",
      "\n",
      "         q2        q3  \n",
      "0  0.967742  0.032258  \n",
      "1  0.387097  0.000000  \n",
      "2       NaN       NaN  \n",
      "3  0.111111  0.000000  \n",
      "4       NaN       NaN  \n",
      "5  0.888889  0.111111  \n",
      "6       NaN       NaN  \n",
      "ML\n",
      "running astral\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro -c ../OMA_data_unfiltered/OMA_data/LUCA/HOG1077/finalset_ncbi_tree.nwk -a ../OMA_data_unfiltered/OMA_data/LUCA/HOG1077/speciesmap.txt -u 2 -i ../OMA_data_unfiltered/OMA_data/LUCA/HOG1077/sequences.aln.muscle.fst.treefile.rooted -o ../OMA_data_unfiltered/OMA_data/LUCA/HOG1077/ML_astral_support.tre -C --root root 2>../OMA_data_unfiltered/OMA_data/LUCA/HOG1077/ML_astral_support.tre.log\n",
      "        pp1       pp2       pp3        f1        f2        f3        q1  \\\n",
      "0  0.199393  0.596827  0.203780  0.000000  0.779487  0.020513  0.000000   \n",
      "1  0.374951  0.385172  0.239877  0.365385  0.384615  0.000000  0.487179   \n",
      "2  0.371781  0.380914  0.247305  0.324786  0.341880  0.000000  0.487179   \n",
      "3  0.333333  0.333333  0.333333  0.000000  0.000000  0.000000       NaN   \n",
      "4  0.333333  0.333333  0.333333  0.000000  0.000000  0.000000       NaN   \n",
      "5  0.333333  0.333333  0.333333  0.000000  0.000000  0.000000       NaN   \n",
      "6  0.333333  0.333333  0.333333  0.000000  0.000000  0.000000       NaN   \n",
      "\n",
      "         q2        q3  \n",
      "0  0.974359  0.025641  \n",
      "1  0.512821  0.000000  \n",
      "2  0.512821  0.000000  \n",
      "3       NaN       NaN  \n",
      "4       NaN       NaN  \n",
      "5       NaN       NaN  \n",
      "6       NaN       NaN  \n",
      "FT\n",
      "running astral\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro -c ../OMA_data_unfiltered/OMA_data/LUCA/HOG748/finalset_ncbi_tree.nwk -a ../OMA_data_unfiltered/OMA_data/LUCA/HOG748/speciesmap.txt -u 2 -i ../OMA_data_unfiltered/OMA_data/LUCA/HOG748/fident_1_exp_struct_tree.PP.nwk.rooted -o ../OMA_data_unfiltered/OMA_data/LUCA/HOG748/FT_astral_support.tre -C --root root 2>../OMA_data_unfiltered/OMA_data/LUCA/HOG748/FT_astral_support.tre.log\n",
      "         pp1       pp2       pp3        f1        f2        f3        q1  \\\n",
      "0   0.257565  0.369654  0.372781  0.321970  0.681818  0.689394  0.190157   \n",
      "1   0.142012  0.699545  0.158443  0.043750  1.345833  0.172917  0.028000   \n",
      "2   0.227510  0.416602  0.355887  0.134259  0.699074  0.569444  0.095710   \n",
      "3   0.238607  0.534263  0.227131  0.072917  0.648438  0.028646  0.097222   \n",
      "4   0.258644  0.456409  0.284947  0.115476  0.558333  0.201190  0.131973   \n",
      "5   0.219620  0.564851  0.215529  0.043403  0.718750  0.026042  0.055066   \n",
      "6   0.419332  0.324568  0.256100  0.431818  0.246212  0.049242  0.593750   \n",
      "7   0.223465  0.546984  0.229551  0.037946  0.687500  0.062500  0.048159   \n",
      "8   0.294352  0.352824  0.352824  0.128205  0.269231  0.269231  0.192308   \n",
      "9   0.263126  0.486892  0.249981  0.041667  0.458333  0.000000  0.083333   \n",
      "10  0.333333  0.333333  0.333333  0.000000  0.000000  0.000000       NaN   \n",
      "11  0.333333  0.333333  0.333333  0.000000  0.000000  0.000000       NaN   \n",
      "\n",
      "          q2        q3  \n",
      "0   0.402685  0.407159  \n",
      "1   0.861333  0.110667  \n",
      "2   0.498350  0.405941  \n",
      "3   0.864583  0.038194  \n",
      "4   0.638095  0.229932  \n",
      "5   0.911894  0.033040  \n",
      "6   0.338542  0.067708  \n",
      "7   0.872521  0.079320  \n",
      "8   0.403846  0.403846  \n",
      "9   0.916667  0.000000  \n",
      "10       NaN       NaN  \n",
      "11       NaN       NaN  \n",
      "ML\n",
      "running astral\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro -c ../OMA_data_unfiltered/OMA_data/LUCA/HOG748/finalset_ncbi_tree.nwk -a ../OMA_data_unfiltered/OMA_data/LUCA/HOG748/speciesmap.txt -u 2 -i ../OMA_data_unfiltered/OMA_data/LUCA/HOG748/sequences.aln.muscle.fst.treefile.rooted -o ../OMA_data_unfiltered/OMA_data/LUCA/HOG748/ML_astral_support.tre -C --root root 2>../OMA_data_unfiltered/OMA_data/LUCA/HOG748/ML_astral_support.tre.log\n",
      "         pp1       pp2       pp3        f1        f2        f3        q1  \\\n",
      "0   0.275008  0.462655  0.262337  0.166667  0.560606  0.125000  0.195556   \n",
      "1   0.211998  0.549545  0.238457  0.000000  0.704167  0.108333  0.000000   \n",
      "2   0.261253  0.508224  0.230523  0.132716  0.609568  0.021605  0.173737   \n",
      "3   0.285052  0.381943  0.333005  0.117188  0.343750  0.242188  0.166667   \n",
      "4   0.258295  0.502711  0.238994  0.065476  0.523810  0.000000  0.111111   \n",
      "5   0.302240  0.404991  0.292770  0.027778  0.217014  0.005208  0.111111   \n",
      "6   0.323268  0.331121  0.345611  0.003788  0.018939  0.045455  0.055556   \n",
      "7   0.273839  0.443283  0.282878  0.004464  0.325893  0.029018  0.012422   \n",
      "8   0.385854  0.334525  0.279621  0.320513  0.217949  0.076923  0.520833   \n",
      "9   0.291704  0.410028  0.298268  0.050000  0.283333  0.066667  0.125000   \n",
      "10  0.333333  0.333333  0.333333  0.000000  0.000000  0.000000       NaN   \n",
      "11  0.333333  0.333333  0.333333  0.000000  0.000000  0.000000       NaN   \n",
      "\n",
      "          q2        q3  \n",
      "0   0.657778  0.146667  \n",
      "1   0.866667  0.133333  \n",
      "2   0.797980  0.028283  \n",
      "3   0.488889  0.344444  \n",
      "4   0.888889  0.000000  \n",
      "5   0.868056  0.020833  \n",
      "6   0.277778  0.666667  \n",
      "7   0.906832  0.080745  \n",
      "8   0.354167  0.125000  \n",
      "9   0.708333  0.166667  \n",
      "10       NaN       NaN  \n",
      "11       NaN       NaN  \n",
      "FT\n",
      "running astral\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro -c ../OMA_data_unfiltered/OMA_data/LUCA/HOG458/finalset_ncbi_tree.nwk -a ../OMA_data_unfiltered/OMA_data/LUCA/HOG458/speciesmap.txt -u 2 -i ../OMA_data_unfiltered/OMA_data/LUCA/HOG458/fident_1_exp_struct_tree.PP.nwk.rooted -o ../OMA_data_unfiltered/OMA_data/LUCA/HOG458/FT_astral_support.tre -C --root root 2>../OMA_data_unfiltered/OMA_data/LUCA/HOG458/FT_astral_support.tre.log\n",
      "        pp1       pp2       pp3        f1        f2        f3    q1        q2  \\\n",
      "0  0.631172  0.184414  0.184414  0.875000  0.000000  0.000000  1.00  0.000000   \n",
      "1  0.625915  0.187042  0.187042  0.857143  0.000000  0.000000  1.00  0.000000   \n",
      "2  0.618835  0.190582  0.190582  0.833333  0.000000  0.000000  1.00  0.000000   \n",
      "3  0.492723  0.225206  0.282072  0.600000  0.000000  0.200000  0.75  0.000000   \n",
      "4  0.235644  0.448467  0.315889  0.000000  0.500000  0.250000  0.00  0.666667   \n",
      "5  0.233295  0.501282  0.265423  0.000000  0.555556  0.111111  0.00  0.833333   \n",
      "6  0.511966  0.244017  0.244017  0.500000  0.000000  0.000000  1.00  0.000000   \n",
      "7  0.333333  0.333333  0.333333  0.000000  0.000000  0.000000   NaN       NaN   \n",
      "\n",
      "         q3  \n",
      "0  0.000000  \n",
      "1  0.000000  \n",
      "2  0.000000  \n",
      "3  0.250000  \n",
      "4  0.333333  \n",
      "5  0.166667  \n",
      "6  0.000000  \n",
      "7       NaN  \n",
      "ML\n",
      "running astral\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro -c ../OMA_data_unfiltered/OMA_data/LUCA/HOG458/finalset_ncbi_tree.nwk -a ../OMA_data_unfiltered/OMA_data/LUCA/HOG458/speciesmap.txt -u 2 -i ../OMA_data_unfiltered/OMA_data/LUCA/HOG458/sequences.aln.muscle.fst.treefile.rooted -o ../OMA_data_unfiltered/OMA_data/LUCA/HOG458/ML_astral_support.tre -C --root root 2>../OMA_data_unfiltered/OMA_data/LUCA/HOG458/ML_astral_support.tre.log\n",
      "        pp1       pp2       pp3        f1        f2        f3    q1        q2  \\\n",
      "0  0.631172  0.184414  0.184414  0.875000  0.000000  0.000000  1.00  0.000000   \n",
      "1  0.625915  0.187042  0.187042  0.857143  0.000000  0.000000  1.00  0.000000   \n",
      "2  0.618835  0.190582  0.190582  0.833333  0.000000  0.000000  1.00  0.000000   \n",
      "3  0.492723  0.225206  0.282072  0.600000  0.000000  0.200000  0.75  0.000000   \n",
      "4  0.235644  0.448467  0.315889  0.000000  0.500000  0.250000  0.00  0.666667   \n",
      "5  0.233295  0.501282  0.265423  0.000000  0.555556  0.111111  0.00  0.833333   \n",
      "6  0.511966  0.244017  0.244017  0.500000  0.000000  0.000000  1.00  0.000000   \n",
      "7  0.333333  0.333333  0.333333  0.000000  0.000000  0.000000   NaN       NaN   \n",
      "\n",
      "         q3  \n",
      "0  0.000000  \n",
      "1  0.000000  \n",
      "2  0.000000  \n",
      "3  0.250000  \n",
      "4  0.333333  \n",
      "5  0.166667  \n",
      "6  0.000000  \n",
      "7       NaN  \n",
      "FT\n",
      "running astral\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro -c ../OMA_data_unfiltered/OMA_data/LUCA/HOG411/finalset_ncbi_tree.nwk -a ../OMA_data_unfiltered/OMA_data/LUCA/HOG411/speciesmap.txt -u 2 -i ../OMA_data_unfiltered/OMA_data/LUCA/HOG411/fident_1_exp_struct_tree.PP.nwk.rooted -o ../OMA_data_unfiltered/OMA_data/LUCA/HOG411/FT_astral_support.tre -C --root root 2>../OMA_data_unfiltered/OMA_data/LUCA/HOG411/FT_astral_support.tre.log\n",
      "        pp1       pp2       pp3        f1        f2        f3        q1   q2  \\\n",
      "0  0.635229  0.182385  0.182385  0.888889  0.000000  0.000000  1.000000  0.0   \n",
      "1  0.631172  0.184414  0.184414  0.875000  0.000000  0.000000  1.000000  0.0   \n",
      "2  0.618835  0.190582  0.190582  0.833333  0.000000  0.000000  1.000000  0.0   \n",
      "3  0.522362  0.217053  0.260586  0.666667  0.000000  0.166667  0.800000  0.0   \n",
      "4  0.195605  0.195605  0.608789  0.000000  0.000000  0.800000  0.000000  0.0   \n",
      "5  0.495933  0.228191  0.275876  0.583333  0.000000  0.166667  0.777778  0.0   \n",
      "6  0.216444  0.567111  0.216444  0.000000  0.666667  0.000000  0.000000  1.0   \n",
      "7  0.511966  0.244017  0.244017  0.500000  0.000000  0.000000  1.000000  0.0   \n",
      "8  0.333333  0.333333  0.333333  0.000000  0.000000  0.000000       NaN  NaN   \n",
      "\n",
      "         q3  \n",
      "0  0.000000  \n",
      "1  0.000000  \n",
      "2  0.000000  \n",
      "3  0.200000  \n",
      "4  1.000000  \n",
      "5  0.222222  \n",
      "6  0.000000  \n",
      "7  0.000000  \n",
      "8       NaN  \n",
      "ML\n",
      "running astral\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro -c ../OMA_data_unfiltered/OMA_data/LUCA/HOG411/finalset_ncbi_tree.nwk -a ../OMA_data_unfiltered/OMA_data/LUCA/HOG411/speciesmap.txt -u 2 -i ../OMA_data_unfiltered/OMA_data/LUCA/HOG411/sequences.aln.muscle.fst.treefile.rooted -o ../OMA_data_unfiltered/OMA_data/LUCA/HOG411/ML_astral_support.tre -C --root root 2>../OMA_data_unfiltered/OMA_data/LUCA/HOG411/ML_astral_support.tre.log\n",
      "        pp1       pp2       pp3        f1        f2   f3        q1        q2  \\\n",
      "0  0.635229  0.182385  0.182385  0.888889  0.000000  0.0  1.000000  0.000000   \n",
      "1  0.631172  0.184414  0.184414  0.875000  0.000000  0.0  1.000000  0.000000   \n",
      "2  0.618835  0.190582  0.190582  0.833333  0.000000  0.0  1.000000  0.000000   \n",
      "3  0.618835  0.190582  0.190582  0.833333  0.000000  0.0  1.000000  0.000000   \n",
      "4  0.195605  0.195605  0.608789  0.000000  0.000000  0.8  0.000000  0.000000   \n",
      "5  0.315889  0.235644  0.448467  0.250000  0.000000  0.5  0.333333  0.000000   \n",
      "6  0.351232  0.402097  0.246671  0.285714  0.380952  0.0  0.428571  0.571429   \n",
      "7  0.511966  0.244017  0.244017  0.500000  0.000000  0.0  1.000000  0.000000   \n",
      "8  0.333333  0.333333  0.333333  0.000000  0.000000  0.0       NaN       NaN   \n",
      "\n",
      "         q3  \n",
      "0  0.000000  \n",
      "1  0.000000  \n",
      "2  0.000000  \n",
      "3  0.000000  \n",
      "4  1.000000  \n",
      "5  0.666667  \n",
      "6  0.000000  \n",
      "7  0.000000  \n",
      "8       NaN  \n",
      "FT\n",
      "running astral\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro -c ../OMA_data_unfiltered/OMA_data/LUCA/HOG569/finalset_ncbi_tree.nwk -a ../OMA_data_unfiltered/OMA_data/LUCA/HOG569/speciesmap.txt -u 2 -i ../OMA_data_unfiltered/OMA_data/LUCA/HOG569/fident_1_exp_struct_tree.PP.nwk.rooted -o ../OMA_data_unfiltered/OMA_data/LUCA/HOG569/FT_astral_support.tre -C --root root 2>../OMA_data_unfiltered/OMA_data/LUCA/HOG569/FT_astral_support.tre.log\n",
      "        pp1       pp2       pp3      f1   f2      f3    q1   q2    q3\n",
      "0  0.387941  0.292381  0.319678  0.1875  0.0  0.0625  0.75  0.0  0.25\n",
      "1  0.333333  0.333333  0.333333  0.0000  0.0  0.0000   NaN  NaN   NaN\n",
      "2       NaN       NaN       NaN     NaN  NaN     NaN   NaN  NaN   NaN\n",
      "3  0.333333  0.333333  0.333333  0.0000  0.0  0.0000   NaN  NaN   NaN\n",
      "ML\n",
      "running astral\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro -c ../OMA_data_unfiltered/OMA_data/LUCA/HOG569/finalset_ncbi_tree.nwk -a ../OMA_data_unfiltered/OMA_data/LUCA/HOG569/speciesmap.txt -u 2 -i ../OMA_data_unfiltered/OMA_data/LUCA/HOG569/sequences.aln.muscle.fst.treefile.rooted -o ../OMA_data_unfiltered/OMA_data/LUCA/HOG569/ML_astral_support.tre -C --root root 2>../OMA_data_unfiltered/OMA_data/LUCA/HOG569/ML_astral_support.tre.log\n",
      "        pp1       pp2       pp3     f1   f2      f3        q1   q2        q3\n",
      "0  0.365952  0.302310  0.331738  0.125  0.0  0.0625  0.666667  0.0  0.333333\n",
      "1  0.333333  0.333333  0.333333  0.000  0.0  0.0000       NaN  NaN       NaN\n",
      "2       NaN       NaN       NaN    NaN  NaN     NaN       NaN  NaN       NaN\n",
      "3  0.333333  0.333333  0.333333  0.000  0.0  0.0000       NaN  NaN       NaN\n",
      "FT\n",
      "running astral\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro -c ../OMA_data_unfiltered/OMA_data/LUCA/HOG1118/finalset_ncbi_tree.nwk -a ../OMA_data_unfiltered/OMA_data/LUCA/HOG1118/speciesmap.txt -u 2 -i ../OMA_data_unfiltered/OMA_data/LUCA/HOG1118/fident_1_exp_struct_tree.PP.nwk.rooted -o ../OMA_data_unfiltered/OMA_data/LUCA/HOG1118/FT_astral_support.tre -C --root root 2>../OMA_data_unfiltered/OMA_data/LUCA/HOG1118/FT_astral_support.tre.log\n",
      "        pp1       pp2       pp3        f1   f2        f3        q1   q2  \\\n",
      "0  0.314049  0.230986  0.454965  0.266667  0.0  0.533333  0.333333  0.0   \n",
      "1  0.593435  0.203283  0.203283  0.750000  0.0  0.000000  1.000000  0.0   \n",
      "2  0.511966  0.244017  0.244017  0.500000  0.0  0.000000  1.000000  0.0   \n",
      "3  0.324282  0.261584  0.414135  0.166667  0.0  0.333333  0.333333  0.0   \n",
      "4  0.333333  0.333333  0.333333  0.000000  0.0  0.000000       NaN  NaN   \n",
      "5  0.333333  0.333333  0.333333  0.000000  0.0  0.000000       NaN  NaN   \n",
      "\n",
      "         q3  \n",
      "0  0.666667  \n",
      "1  0.000000  \n",
      "2  0.000000  \n",
      "3  0.666667  \n",
      "4       NaN  \n",
      "5       NaN  \n",
      "ML\n",
      "running astral\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro -c ../OMA_data_unfiltered/OMA_data/LUCA/HOG1118/finalset_ncbi_tree.nwk -a ../OMA_data_unfiltered/OMA_data/LUCA/HOG1118/speciesmap.txt -u 2 -i ../OMA_data_unfiltered/OMA_data/LUCA/HOG1118/sequences.aln.muscle.fst.treefile.rooted -o ../OMA_data_unfiltered/OMA_data/LUCA/HOG1118/ML_astral_support.tre -C --root root 2>../OMA_data_unfiltered/OMA_data/LUCA/HOG1118/ML_astral_support.tre.log\n",
      "        pp1       pp2       pp3        f1   f2        f3        q1   q2  \\\n",
      "0  0.314049  0.230986  0.454965  0.266667  0.0  0.533333  0.333333  0.0   \n",
      "1  0.593435  0.203283  0.203283  0.750000  0.0  0.000000  1.000000  0.0   \n",
      "2  0.511966  0.244017  0.244017  0.500000  0.0  0.000000  1.000000  0.0   \n",
      "3  0.324282  0.261584  0.414135  0.166667  0.0  0.333333  0.333333  0.0   \n",
      "4  0.333333  0.333333  0.333333  0.000000  0.0  0.000000       NaN  NaN   \n",
      "5  0.333333  0.333333  0.333333  0.000000  0.0  0.000000       NaN  NaN   \n",
      "\n",
      "         q3  \n",
      "0  0.666667  \n",
      "1  0.000000  \n",
      "2  0.000000  \n",
      "3  0.666667  \n",
      "4       NaN  \n",
      "5       NaN  \n",
      "FT\n",
      "running astral\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro -c ../OMA_data_unfiltered/OMA_data/LUCA/HOG423/finalset_ncbi_tree.nwk -a ../OMA_data_unfiltered/OMA_data/LUCA/HOG423/speciesmap.txt -u 2 -i ../OMA_data_unfiltered/OMA_data/LUCA/HOG423/fident_1_exp_struct_tree.PP.nwk.rooted -o ../OMA_data_unfiltered/OMA_data/LUCA/HOG423/FT_astral_support.tre -C --root root 2>../OMA_data_unfiltered/OMA_data/LUCA/HOG423/FT_astral_support.tre.log\n",
      "        pp1       pp2       pp3    f1   f2        f3   q1   q2   q3\n",
      "0  0.593435  0.203283  0.203283  0.75  0.0  0.000000  1.0  0.0  0.0\n",
      "1  0.216444  0.216444  0.567111  0.00  0.0  0.666667  0.0  0.0  1.0\n",
      "2  0.511966  0.244017  0.244017  0.50  0.0  0.000000  1.0  0.0  0.0\n",
      "3  0.333333  0.333333  0.333333  0.00  0.0  0.000000  NaN  NaN  NaN\n",
      "ML\n",
      "running astral\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro -c ../OMA_data_unfiltered/OMA_data/LUCA/HOG423/finalset_ncbi_tree.nwk -a ../OMA_data_unfiltered/OMA_data/LUCA/HOG423/speciesmap.txt -u 2 -i ../OMA_data_unfiltered/OMA_data/LUCA/HOG423/sequences.aln.muscle.fst.treefile.rooted -o ../OMA_data_unfiltered/OMA_data/LUCA/HOG423/ML_astral_support.tre -C --root root 2>../OMA_data_unfiltered/OMA_data/LUCA/HOG423/ML_astral_support.tre.log\n",
      "        pp1       pp2       pp3        f1   f2        f3        q1   q2  \\\n",
      "0  0.593435  0.203283  0.203283  0.750000  0.0  0.000000  1.000000  0.0   \n",
      "1  0.216444  0.216444  0.567111  0.000000  0.0  0.666667  0.000000  0.0   \n",
      "2  0.324282  0.261584  0.414135  0.166667  0.0  0.333333  0.333333  0.0   \n",
      "3  0.333333  0.333333  0.333333  0.000000  0.0  0.000000       NaN  NaN   \n",
      "\n",
      "         q3  \n",
      "0  0.000000  \n",
      "1  1.000000  \n",
      "2  0.666667  \n",
      "3       NaN  \n",
      "FT\n",
      "running astral\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro -c ../OMA_data_unfiltered/OMA_data/LUCA/HOG600/finalset_ncbi_tree.nwk -a ../OMA_data_unfiltered/OMA_data/LUCA/HOG600/speciesmap.txt -u 2 -i ../OMA_data_unfiltered/OMA_data/LUCA/HOG600/fident_1_exp_struct_tree.PP.nwk.rooted -o ../OMA_data_unfiltered/OMA_data/LUCA/HOG600/FT_astral_support.tre -C --root root 2>../OMA_data_unfiltered/OMA_data/LUCA/HOG600/FT_astral_support.tre.log\n",
      "        pp1       pp2       pp3   f1   f2   f3   q1   q2   q3\n",
      "0  0.244017  0.511966  0.244017  0.0  0.5  0.0  0.0  1.0  0.0\n",
      "1  0.333333  0.333333  0.333333  0.0  0.0  0.0  NaN  NaN  NaN\n",
      "2  0.333333  0.333333  0.333333  0.0  0.0  0.0  NaN  NaN  NaN\n",
      "ML\n",
      "running astral\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro -c ../OMA_data_unfiltered/OMA_data/LUCA/HOG600/finalset_ncbi_tree.nwk -a ../OMA_data_unfiltered/OMA_data/LUCA/HOG600/speciesmap.txt -u 2 -i ../OMA_data_unfiltered/OMA_data/LUCA/HOG600/sequences.aln.muscle.fst.treefile.rooted -o ../OMA_data_unfiltered/OMA_data/LUCA/HOG600/ML_astral_support.tre -C --root root 2>../OMA_data_unfiltered/OMA_data/LUCA/HOG600/ML_astral_support.tre.log\n",
      "        pp1       pp2       pp3   f1   f2   f3   q1   q2   q3\n",
      "0  0.244017  0.511966  0.244017  0.0  0.5  0.0  0.0  1.0  0.0\n",
      "1  0.333333  0.333333  0.333333  0.0  0.0  0.0  NaN  NaN  NaN\n",
      "2  0.333333  0.333333  0.333333  0.0  0.0  0.0  NaN  NaN  NaN\n",
      "FT\n",
      "running astral\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro -c ../OMA_data_unfiltered/OMA_data/LUCA/HOG475/finalset_ncbi_tree.nwk -a ../OMA_data_unfiltered/OMA_data/LUCA/HOG475/speciesmap.txt -u 2 -i ../OMA_data_unfiltered/OMA_data/LUCA/HOG475/fident_1_exp_struct_tree.PP.nwk.rooted -o ../OMA_data_unfiltered/OMA_data/LUCA/HOG475/FT_astral_support.tre -C --root root 2>../OMA_data_unfiltered/OMA_data/LUCA/HOG475/FT_astral_support.tre.log\n",
      "        pp1       pp2       pp3        f1        f2        f3        q1  \\\n",
      "0  0.439680  0.313598  0.246722  0.467742  0.225806  0.024194  0.651685   \n",
      "1  0.262766  0.339088  0.398147  0.028226  0.229839  0.342742  0.046980   \n",
      "2  0.333333  0.333333  0.333333  0.000000  0.000000  0.000000       NaN   \n",
      "3  0.273537  0.269239  0.457224  0.012097  0.000000  0.350806  0.033333   \n",
      "4  0.510711  0.262805  0.226484  0.627778  0.150000  0.016667  0.790210   \n",
      "5  0.271100  0.269671  0.459229  0.004032  0.000000  0.350806  0.011364   \n",
      "6  0.333333  0.333333  0.333333  0.000000  0.000000  0.000000       NaN   \n",
      "\n",
      "         q2        q3  \n",
      "0  0.314607  0.033708  \n",
      "1  0.382550  0.570470  \n",
      "2       NaN       NaN  \n",
      "3  0.000000  0.966667  \n",
      "4  0.188811  0.020979  \n",
      "5  0.000000  0.988636  \n",
      "6       NaN       NaN  \n",
      "ML\n",
      "running astral\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro -c ../OMA_data_unfiltered/OMA_data/LUCA/HOG475/finalset_ncbi_tree.nwk -a ../OMA_data_unfiltered/OMA_data/LUCA/HOG475/speciesmap.txt -u 2 -i ../OMA_data_unfiltered/OMA_data/LUCA/HOG475/sequences.aln.muscle.fst.treefile.rooted -o ../OMA_data_unfiltered/OMA_data/LUCA/HOG475/ML_astral_support.tre -C --root root 2>../OMA_data_unfiltered/OMA_data/LUCA/HOG475/ML_astral_support.tre.log\n",
      "        pp1       pp2       pp3        f1        f2        f3        q1  \\\n",
      "0  0.337420  0.337420  0.325160  0.022989  0.022989  0.000000  0.500000   \n",
      "1  0.337420  0.337420  0.325160  0.022989  0.022989  0.000000  0.500000   \n",
      "2  0.344133  0.333152  0.322715  0.040000  0.020000  0.000000  0.666667   \n",
      "3  0.335407  0.329186  0.335407  0.011494  0.000000  0.011494  0.500000   \n",
      "4  0.363994  0.318569  0.317437  0.085057  0.002299  0.000000  0.973684   \n",
      "5  0.333333  0.333333  0.333333  0.000000  0.000000  0.000000       NaN   \n",
      "6  0.333333  0.333333  0.333333  0.000000  0.000000  0.000000       NaN   \n",
      "7       NaN       NaN       NaN       NaN       NaN       NaN       NaN   \n",
      "8  0.333333  0.333333  0.333333  0.000000  0.000000  0.000000       NaN   \n",
      "\n",
      "         q2   q3  \n",
      "0  0.500000  0.0  \n",
      "1  0.500000  0.0  \n",
      "2  0.333333  0.0  \n",
      "3  0.000000  0.5  \n",
      "4  0.026316  0.0  \n",
      "5       NaN  NaN  \n",
      "6       NaN  NaN  \n",
      "7       NaN  NaN  \n",
      "8       NaN  NaN  \n",
      "FT\n",
      "running astral\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro -c ../OMA_data_unfiltered/OMA_data/LUCA/HOG1231/finalset_ncbi_tree.nwk -a ../OMA_data_unfiltered/OMA_data/LUCA/HOG1231/speciesmap.txt -u 2 -i ../OMA_data_unfiltered/OMA_data/LUCA/HOG1231/fident_1_exp_struct_tree.PP.nwk.rooted -o ../OMA_data_unfiltered/OMA_data/LUCA/HOG1231/FT_astral_support.tre -C --root root 2>../OMA_data_unfiltered/OMA_data/LUCA/HOG1231/FT_astral_support.tre.log\n",
      "        pp1       pp2       pp3        f1        f2   f3   q1   q2   q3\n",
      "0  0.187042  0.625915  0.187042  0.000000  0.857143  0.0  0.0  1.0  0.0\n",
      "1  0.608789  0.195605  0.195605  0.800000  0.000000  0.0  1.0  0.0  0.0\n",
      "2  0.545431  0.227284  0.227284  0.600000  0.000000  0.0  1.0  0.0  0.0\n",
      "3  0.333333  0.333333  0.333333  0.000000  0.000000  0.0  NaN  NaN  NaN\n",
      "4  0.333333  0.333333  0.333333  0.000000  0.000000  0.0  NaN  NaN  NaN\n",
      "5  0.483309  0.258346  0.258346  0.416667  0.000000  0.0  1.0  0.0  0.0\n",
      "6  0.333333  0.333333  0.333333  0.000000  0.000000  0.0  NaN  NaN  NaN\n",
      "ML\n",
      "running astral\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro -c ../OMA_data_unfiltered/OMA_data/LUCA/HOG1231/finalset_ncbi_tree.nwk -a ../OMA_data_unfiltered/OMA_data/LUCA/HOG1231/speciesmap.txt -u 2 -i ../OMA_data_unfiltered/OMA_data/LUCA/HOG1231/sequences.aln.muscle.fst.treefile.rooted -o ../OMA_data_unfiltered/OMA_data/LUCA/HOG1231/ML_astral_support.tre -C --root root 2>../OMA_data_unfiltered/OMA_data/LUCA/HOG1231/ML_astral_support.tre.log\n",
      "        pp1       pp2       pp3        f1   f2   f3   q1   q2   q3\n",
      "0  0.625915  0.187042  0.187042  0.857143  0.0  0.0  1.0  0.0  0.0\n",
      "1  0.608789  0.195605  0.195605  0.800000  0.0  0.0  1.0  0.0  0.0\n",
      "2  0.545431  0.227284  0.227284  0.600000  0.0  0.0  1.0  0.0  0.0\n",
      "3  0.333333  0.333333  0.333333  0.000000  0.0  0.0  NaN  NaN  NaN\n",
      "4  0.333333  0.333333  0.333333  0.000000  0.0  0.0  NaN  NaN  NaN\n",
      "5  0.483309  0.258346  0.258346  0.416667  0.0  0.0  1.0  0.0  0.0\n",
      "6  0.333333  0.333333  0.333333  0.000000  0.0  0.0  NaN  NaN  NaN\n"
     ]
    }
   ],
   "source": [
    "from Bio import Phylo\n",
    "import re\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "q1,q2,q3: these three values show quartet support (as defined in the description of -t 1) for the main topology, the first alternative, and the second alternative, respectively.\n",
    "f1, f2, f3: these three values show the total number of quartet trees in all the gene trees that support the main topology, the first alternative, and the second alternative, respectively.\n",
    "pp1, pp2, pp3: these three show the local posterior probabilities (as defined in the description of -t 4) for the main topology, the first alternative, and the second alternative, respectively.\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "def parse_astral(string , index = 0):   \n",
    "    #'[pp1=0.284784;pp2=0.426968;pp3=0.288248;f1=0.000000;f2=0.263158;f3=0.008772;q1=0.000000;q2=0.967742;q3=0.032258]'\n",
    "    string = string.replace('[','').replace(']','')\n",
    "    string = string.split(';')\n",
    "    string = {s.split('=')[0]:float(s.split('=')[1]) for s in string}\n",
    "    #to dataframe\n",
    "    string = pd.DataFrame(string , index = [index])\n",
    "    return string\n",
    "\n",
    "def retastral_support( astral_file):\n",
    "    with open(astral_file , 'r') as f:\n",
    "        nstring = f.read()\n",
    "    #find all quoted strings with single quotes\n",
    "    quoted = re.findall(r\"'(.*?)'\", nstring)\n",
    "    dfs = [ parse_astral(q,i) for i,q in enumerate(quoted)]\n",
    "    return pd.concat(dfs)\n",
    "\n",
    "for sample in sample_hogs:  \n",
    "    ncbitree = sample+'finalset_ncbi_tree.nwk'\n",
    "    speciesmap, speciesmap_file = prepare_astral_input(sample+'finalset.csv' , sample+'speciesmap.txt')\n",
    "    FT = sample+'fident_1_exp_struct_tree.PP.nwk.rooted'\n",
    "    ML = sample+'sequences.aln.muscle.fst.treefile.rooted'\n",
    "    print('FT')\n",
    "\n",
    "    #open ncbi tree\n",
    "    t = PhyloTree(ncbitree , sp_naming_function=None)\n",
    "    #find root node name\n",
    "    root = t.get_tree_root()\n",
    "    root.name = 'root'\n",
    "    t.write(outfile=ncbitree , quoted_node_names = False, format=0)\n",
    "\n",
    "    output = run_astral(  FT , ncbitree,speciesmap_file , sample+'FT_astral_support.tre' , 'root')\n",
    "    astraldf = retastral_support(output)\n",
    "    print(astraldf)\n",
    "    print('ML')\n",
    "    output = run_astral( ML , ncbitree,speciesmap_file , sample+'ML_astral_support.tre', 'root')\n",
    "    astraldf = retastral_support(output)\n",
    "    print(astraldf)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
