{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import os\n",
    "import glob\n",
    "import toytree\n",
    "import tqdm\n",
    "import pandas as pd\n",
    "from subprocess import PIPE, Popen\n",
    "import shlex\n",
    "from Bio import SeqIO\n",
    "import random\n",
    "import itertools\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "def remove_seeds( alnfile):\n",
    "    \"\"\"\n",
    "    this function removes seeds from an alignment\n",
    "    \"\"\"\n",
    "    aln = SeqIO.parse(alnfile, 'fasta')\n",
    "    sequences = []\n",
    "    for s in aln:\n",
    "        sequences.append(s)\n",
    "    #write new aln\n",
    "    \n",
    "    with open(alnfile, 'w') as f:\n",
    "        for s in sequences:\n",
    "            f.write('>' + str(s.id).replace('seed','').replace('_' , '') + '\\n')\n",
    "            f.write(str(s.seq) + '\\n')\n",
    "    \n",
    "    return alnfile\n",
    "\n",
    "def Fident(str1,str2 , verbose = False):\n",
    "    #minlen= min( (len(str1),len(str2))  )\n",
    "    #str1 = str1[:minlen]\n",
    "    #str2 = str2[:minlen]\n",
    "    str1 = np.array(list(str1))\n",
    "    str2 = np.array(list(str2))            \n",
    "    return len(np.where( (str1 == str2 ) & (str1 != '-' ) & (str2 != '-')  )[0]) / len(str1)\n",
    "\n",
    "def copyaln( aln, seq):\n",
    "    seqiter = iter(seq)\n",
    "    newaln = ''\n",
    "    for i,char in enumerate(aln):\n",
    "        if char == '-':\n",
    "            newaln += '-'\n",
    "        else:\n",
    "            newaln+=next(seqiter)\n",
    "    return newaln\n",
    "\n",
    "def read_dbfiles3di(  AADB , threeDidb):\n",
    "    #find positions \n",
    "    threeDiseq = [ l.strip().replace('\\x00','') for l in open(threeDidb)]\n",
    "    lookup = AADB+'.lookup'\n",
    "    ids = [ l.split()[1].strip() for l in open(lookup)]\n",
    "    AAs = [ l.strip().replace('\\x00','') for l in open(AADB)]\n",
    "\n",
    "    mapper3di = dict(zip(ids,threeDiseq))\n",
    "    mapperAA = dict(zip(ids,AAs))\n",
    "    \n",
    "    return mapper3di, mapperAA\n",
    "\n",
    "def calc_fident_crossaln(row , verbose = False):\n",
    "    #amino acid representations of alns using AAand3di or just 3di\n",
    "    qaln_2, taln_2 = row.qaln , row.taln\n",
    "    #start and stop of aln\n",
    "    \n",
    "    qstart_2, qend_2, tstart_2 , tend_2 = row.qstart, row.qend , row.tstart , row.tend\n",
    "    #indexing starts at 1...\n",
    "    \n",
    "    #3di of the query and target\n",
    "    structQ, structT = row['3diq'], row['3dit']\n",
    "    AAq, AAt = row['AAq'], row['AAt']\n",
    "\n",
    "    #add gaps\n",
    "    t3diAA_newgaps = copyaln(taln_2, structT[tstart_2-1:tend_2]) \n",
    "    q3diAA_newgaps = copyaln(qaln_2, structQ[qstart_2-1:qend_2])\n",
    "    row = pd.Series( { '3di_qaln_mode2':q3diAA_newgaps , '3di_taln_mode2':t3diAA_newgaps })\n",
    "    #return columns\n",
    "    return row\n",
    "\n",
    "def get_leafset( treenode ):\n",
    "    \"\"\"\n",
    "    this function returns the leafset of a node\n",
    "    \"\"\"\n",
    "    if treenode.is_leaf():\n",
    "        return [treenode.name]\n",
    "    else:\n",
    "        return treenode.get_leaf_names()\n",
    "\n",
    "\n",
    "def mafft_profile(aln1,aln2, outprofile , submat = None):\n",
    "    \"\"\"\n",
    "    this function aligns two alignments using MAFFT\n",
    "    \"\"\"\n",
    "    #make profile\n",
    "    if submat:\n",
    "        cmd = 'mafft --textmatrix {} --seed {} {} > {}'.format(submat, aln1,aln2, outprofile)\n",
    "    else:\n",
    "        cmd = 'mafft --seed {} {} > {}'.format(aln1,aln2, outprofile)\n",
    "\n",
    "    print(cmd)\n",
    "    subprocess.run(cmd , shell=True)\n",
    "    return outprofile\n",
    "\n",
    "def mafft_addfull(aln1,aln2, outprofile , submat = None):\n",
    "    \"\"\"\n",
    "    this function aligns two alignments using MAFFT\n",
    "    \"\"\"\n",
    "    #make profile\n",
    "    profile = aln1 + '.profile'\n",
    "    if submat:\n",
    "        cmd = 'mafft --textmatrix {} --addfull {} --keeplength {} > {}'.format(submat, aln1,aln2, outprofile)\n",
    "    else:\n",
    "        cmd = 'mafft --addfull  {} --keeplength {} > {}'.format(aln1,aln2, outprofile)\n",
    "    print(cmd)\n",
    "\n",
    "    subprocess.run(cmd , shell = True)\n",
    "    return outprofile\n",
    "\n",
    "\n",
    "def sub2fasta( sub, outfile , fastacol1='qaln' , fastacol2='taln' ):\n",
    "    with open(outfile, 'w') as f:\n",
    "        f.write('>' + sub['query'] + '\\n')\n",
    "        f.write(sub[fastacol1] + '\\n')\n",
    "        f.write('>' + sub['target'] + '\\n')\n",
    "        f.write(sub[fastacol2] + '\\n')    \n",
    "    return outfile\n",
    "\n",
    "def retalns(allvall, leafname1,leafname2):\n",
    "    sub = allvall[allvall['query'].isin( leafname1)]\n",
    "    sub = sub[sub['target'].isin(leafname2)]\n",
    "    sub = sub[sub['query'] != sub['target']]\n",
    "    #get max prot lenght aligned\n",
    "    sub['alnlen'] = sub.apply(lambda x: max(x['qend'] - x['qstart'] , x['tend'] - x['tstart']) , axis = 1)\n",
    "    sub = sub[sub['alnlen'] == sub['alnlen'].max()]\n",
    "    if len(sub)==0:\n",
    "        print(leafname1, leafname2)\n",
    "        raise Exception('no sub')\n",
    "    return sub.iloc[0]\n",
    "\n",
    "def get_fasta_leafset(fasta):\n",
    "    \"\"\"\n",
    "    this function returns the leafset of a fasta file\n",
    "    \"\"\"\n",
    "    aln = SeqIO.parse(fasta, 'fasta')\n",
    "    leafset = []\n",
    "    for s in aln:\n",
    "        leafset.append(s.id)\n",
    "    return leafset\n",
    "\n",
    "#traverse tree from root to leaves recursively\n",
    "def traverse_tree_merge_mafft( treenode, topleafset, allvall , alnfolder , submat = None , verbose = False):\n",
    "    \"\"\"\n",
    "    this function traverses a tree from root to leaves recursively\n",
    "    it returns a dictionary with the iteratively built alignment\n",
    "    \"\"\"\n",
    "    if verbose == True:\n",
    "        print('traverse', treenode.name , treenode.is_leaf() , treenode.leafset)\n",
    "    \n",
    "    if treenode.is_leaf():\n",
    "        topleafset.remove(treenode.name)\n",
    "        #if the node is a leaf, then we need to add it to the alignment with one of the pivots in the current leafset\n",
    "        #select the alignment of the leaf with itself\n",
    "        sub = allvall[allvall['query'].isin( [treenode.name] )]\n",
    "        sub = sub[sub['target'].isin([treenode.name])]\n",
    "        sub = sub.iloc[0]\n",
    "        print('leaf',sub)\n",
    "        with open(alnfolder + treenode.name + '_inter.fasta', 'w') as f:\n",
    "            f.write('>' + sub['query'] + '\\n')\n",
    "            f.write(sub['qaln'] + '\\n')\n",
    "        with open(alnfolder + treenode.name + '_inter.3di.fasta', 'w') as f:\n",
    "            f.write('>' + sub['query'] + '\\n')\n",
    "            f.write(sub['3di_qaln_mode2'] + '\\n')\n",
    "        treenode.aln = alnfolder + treenode.name + '_inter.fasta'\n",
    "        treenode.aln3di = alnfolder + treenode.name + '_inter.3di.fasta'\n",
    "        return treenode.aln, treenode.aln3di\n",
    "    \n",
    "    else:\n",
    "        childalns3di = {}\n",
    "        childalnsAA = {}\n",
    "       \n",
    "        #treenode.leafset = get_leafset(treenode)\n",
    "        #get the intersection of the child leafsets\n",
    "        treenode.leafset = get_leafset(treenode)\n",
    "        children = treenode.get_children()\n",
    "        \n",
    "        if len(children) == 2 and children[0].is_leaf() and children[1].is_leaf():\n",
    "            #treat the case of a cherry\n",
    "            print('cherry', children[0].name , children[1].name)\n",
    "            treenode.aln = sub2fasta( retalns(allvall, [children[0].name] , [children[1].name]) , alnfolder + treenode.name + '_inter.fasta')\n",
    "            treenode.aln3di = sub2fasta( retalns(allvall, [children[0].name] , [children[1].name]) , alnfolder + treenode.name + '_inter.3di.fasta' , fastacol1='3di_qaln_mode2' , fastacol2='3di_taln_mode2')\n",
    "            return treenode.aln, treenode.aln3di\n",
    "        \n",
    "        else:\n",
    "            #not a cherry. one or both sides is a subtree\n",
    "            print('not cherry', treenode.name  )\n",
    "            print( 'children', [c.name for c in children])\n",
    "            for c in treenode.get_children():\n",
    "                #make sub aln for each child\n",
    "                if verbose == True:\n",
    "                    print('traverse', c.name , c.is_leaf() , c.leafset)\n",
    "                if not c.aln:\n",
    "                    c.aln,c.aln3di = traverse_tree_merge_mafft(c , treenode.leafset , allvall, alnfolder , verbose = verbose)\n",
    "                childalnsAA[c] = { 'fasta': c.aln  }\n",
    "                childalns3di[c] = { 'fasta': c.aln3di  }\n",
    "\n",
    "            c1,c2 = children\n",
    "            if c1.is_leaf():\n",
    "                treenode.aln = mafft_addfull(childalnsAA[c1]['fasta'], childalnsAA[c2]['fasta'], alnfolder + treenode.name + '_inter.fasta' )\n",
    "                treenode.aln3di = mafft_addfull(childalns3di[c1]['fasta'], childalns3di[c2]['fasta'], alnfolder + treenode.name + '_inter3di.fasta' , submat =submat )\n",
    "\n",
    "            elif c2.is_leaf():\n",
    "                treenode.aln = mafft_addfull(childalnsAA[c2]['fasta'], childalnsAA[c1]['fasta'], alnfolder + treenode.name + '_inter.fasta' )\n",
    "                treenode.aln3di = mafft_addfull(childalns3di[c1]['fasta'], childalns3di[c2]['fasta'], alnfolder + treenode.name + '_inter3di.fasta' , submat = submat)\n",
    "\n",
    "            else:                \n",
    "                treenode.aln = mafft_profile(childalnsAA[c1]['fasta'], childalnsAA[c2]['fasta'], alnfolder + treenode.name + '_inter.fasta' )\n",
    "                treenode.aln3di = mafft_profile(childalns3di[c1]['fasta'], childalns3di[c2]['fasta'], alnfolder + treenode.name + '_inter3di.fasta' , submat =  submat)\n",
    "\n",
    "            if verbose == True:\n",
    "                #check if node is root  \n",
    "                if treenode.up == None:\n",
    "                    print('final aln')\n",
    "                    print('childalnsAA', childalnsAA)\n",
    "                    print('childalns3di', childalns3di)\n",
    "            return treenode.aln, treenode.aln3di\n",
    "\n",
    "\n",
    "def remove_redundant( alignment ):\n",
    "    \"\"\"\n",
    "    this function removes redundant sequences from an alignment\n",
    "    \"\"\"\n",
    "    aln = SeqIO.parse(alignment, 'fasta')\n",
    "    seqs = []\n",
    "    ids = []\n",
    "    for s in aln:\n",
    "        if s.id not in ids:\n",
    "            seqs.append(s)\n",
    "            ids.append(s.id)\n",
    "    \n",
    "    with open(alignment, 'w') as f:\n",
    "        for s in seqs:\n",
    "            f.write('>' + s.id + '\\n')\n",
    "            f.write(str(s.seq) + '\\n')\n",
    "    return alignment\n",
    "\n",
    "#remove all alns except the final merged one\n",
    "def cleanup( filedir ):\n",
    "    \"\"\"\n",
    "    this function removes all alns except the final merged one\n",
    "    \"\"\"\n",
    "    for f in glob.glob(filedir + '*inter.fasta'):\n",
    "        os.remove(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#traverse tree from root to leaves recursively\n",
    "def traverse_tree_merge( treenode, topleafset, allvall , alnfolder , verbose = False):\n",
    "    \"\"\"\n",
    "    this function traverses a tree from root to leaves recursively\n",
    "    it returns a dictionary with the iteratively built alignment\n",
    "    \"\"\"\n",
    "    if verbose == True:\n",
    "        print('traverse', treenode.name , treenode.is_leaf() , treenode.leafset)\n",
    "    \n",
    "    if treenode.is_leaf():\n",
    "        topleafset.remove(treenode.name)\n",
    "        #if the node is a leaf, then we need to add it to the alignment with one of the pivots in the current leafset\n",
    "        sub = retalns(allvall, [treenode.name] , topleafset)  \n",
    "        treenode.aln = sub2fasta(sub, alnfolder + treenode.name + '_inter.fasta')\n",
    "        treenode.aln3di = sub2fasta(sub, alnfolder + treenode.name + '_inter.3di.fasta' , fastacol1='3di_qaln_mode2' , fastacol2='3di_taln_mode2')\n",
    "        return treenode.aln, treenode.aln3di\n",
    "    \n",
    "    else:\n",
    "        childalns3di = {}\n",
    "        childalnsAA = {}\n",
    "        bridges3di = {}\n",
    "        bridgesAA = {}\n",
    "        #treenode.leafset = get_leafset(treenode)\n",
    "        #get the intersection of the child leafsets\n",
    "        treenode.leafset = get_leafset(treenode)\n",
    "        children = treenode.get_children()\n",
    "        \n",
    "        if len(children) == 2 and children[0].is_leaf() and children[1].is_leaf():\n",
    "            #treat the case of a cherry\n",
    "            print('cherry', children[0].name , children[1].name)\n",
    "            treenode.aln = sub2fasta( retalns(allvall, [children[0].name] , [children[1].name]) , alnfolder + treenode.name + '_inter.fasta')\n",
    "            treenode.aln3di = sub2fasta( retalns(allvall, [children[0].name] , [children[1].name]) , alnfolder + treenode.name + '_inter.3di.fasta' , fastacol1='3di_qaln_mode2' , fastacol2='3di_taln_mode2')\n",
    "            return treenode.aln, treenode.aln3di\n",
    "        \n",
    "        else:\n",
    "            #not a cherry. one or both sides is a subtree\n",
    "            print('not cherry', treenode.name  )\n",
    "            print( 'children', [c.name for c in children])\n",
    "            for c in treenode.get_children():\n",
    "                #make sub aln for each child\n",
    "                if verbose == True:\n",
    "                    print('traverse', c.name , c.is_leaf() , c.leafset)\n",
    "                if not c.aln:\n",
    "                    c.aln,c.aln3di = traverse_tree_merge(c , treenode.leafset , allvall, alnfolder , verbose = verbose)\n",
    "                childalnsAA[c] = { 'fasta': c.aln , 'protset':set(get_fasta_leafset(c.aln) ) }\n",
    "                childalns3di[c] = { 'fasta': c.aln3di , 'protset':set(get_fasta_leafset(c.aln3di) ) }\n",
    "            \n",
    "            \"\"\"\n",
    "            for c1,c2 in itertools.combinations(treenode.get_children(),2):\n",
    "                bridge = retalns(allvall, childalnsAA[c1]['protset'] , childalnsAA[c2]['protset'] )\n",
    "                bridgesAA[(c1,c2)] = { 'fasta': sub2fasta(bridge, alnfolder + treenode.name + '_bridge.fasta') , 'protset':set([bridge.query , bridge.target]) }\n",
    "                bridges3di[(c1,c2)] = { 'fasta' : sub2fasta(bridge, alnfolder + treenode.name + '_bridge.3di.fasta' , fastacol1='3di_qaln_mode2' , fastacol2='3di_taln_mode2') , 'protset':set([bridge.query, bridge.target]) }\n",
    "            \n",
    "            #successively merge the alignments of the children\n",
    "            for i, c in enumerate(itertools.combinations(treenode.get_children(),2)):\n",
    "                c1,c2 = c\n",
    "                if verbose == True:\n",
    "                    print('merge', c1.name , c2.name)\n",
    "                if i == 0:\n",
    "                    #first merge\n",
    "                    try:\n",
    "                        print('first merge')\n",
    "                        \n",
    "                        alnAA = mergealns( childalnsAA[c1]['fasta'], bridgesAA[(c1,c2)]['fasta'] ,alnfolder + treenode.name + '_inter.fasta' , verbose=verbose)\n",
    "                        aln3di = mergealns( childalns3di[c1]['fasta'], bridges3di[(c1,c2)]['fasta'] ,alnfolder + treenode.name + '_inter3di.fasta', verbose=verbose)\n",
    "\n",
    "                        print('2 merge')\n",
    "\n",
    "                        alnAA = mergealns( childalnsAA[c2]['fasta'], alnAA , alnfolder + treenode.name + '_inter.fasta' , verbose=verbose)\n",
    "                        aln3di = mergealns( childalns3di[c2]['fasta'], aln3di ,alnfolder + treenode.name + '_inter3di.fasta', verbose=verbose)\n",
    "\n",
    "                    except:\n",
    "                        print( treenode , childalnsAA , childalns3di , bridgesAA , bridges3di)\n",
    "                        raise Exception('merge error 1')\n",
    "                else:\n",
    "                    try:\n",
    "                        print('3 merge')\n",
    "\n",
    "                        alnAA = mergealns( childalnsAA[c1]['fasta'], bridgesAA[(c1,c2)]['fasta'] , alnfolder + treenode.name + '_inter.fasta' , verbose=verbose)\n",
    "                        alnAA = mergealns( childalnsAA[c1]['fasta'], alnAA , alnfolder + treenode.name + '_inter.fasta' , verbose=verbose)\n",
    "\n",
    "                        print('4 merge')\n",
    "\n",
    "                        aln3di = mergealns( childalns3di[c2]['fasta'], bridges3di[(c1,c2)]['fasta'] , alnfolder + treenode.name + '_inter3di.fasta'   , verbose=verbose)\n",
    "                        aln3di = mergealns( childalns3di[c2]['fasta'], aln3di , alnfolder + treenode.name + '_inter3di.fasta' , verbose=verbose)\n",
    "\n",
    "                        \n",
    "                    except:\n",
    "                        print( treenode )\n",
    "                        print( childalnsAA , childalns3di , bridgesAA , bridges3di)\n",
    "                        raise Exception('merge error 2') \n",
    "            \"\"\"\n",
    "\n",
    "            treenode.aln = alnAA\n",
    "            treenode.aln3di = aln3di\n",
    "            if verbose == True:\n",
    "                #check if node is root  \n",
    "                if treenode.up == None:\n",
    "                    print('final aln')\n",
    "                    print('childalnsAA', childalnsAA)\n",
    "                    print('childalns3di', childalns3di)\n",
    "            return treenode.aln, treenode.aln3di\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def aln_mapping( s1 , s2, maxaln = 0 , coidx1= 0 , coidx2 = 0 , start1 = 0 , start2=0 ,  verbose = False):\n",
    "    #build a dictionary of the positions of the characters in the string\n",
    "    #convolve the strings\n",
    "    maxaln, maxcount = convolve_strings(s1,s2)\n",
    "\n",
    "    if start1 != 0 or start2 != 0:\n",
    "        print('start', start1, start2)\n",
    "    if verbose == True:\n",
    "        print('maxaln', maxaln, maxcount)\n",
    "    \n",
    "    #find starting points\n",
    "    if len(s1) < len(s2):\n",
    "        if maxaln < 0:\n",
    "            coidx1 = np.abs(maxaln)\n",
    "            coidx2 = 0\n",
    "        else:\n",
    "            coidx1 = 0\n",
    "            coidx2 = maxaln\n",
    "    else:\n",
    "        if maxaln < 0:\n",
    "            coidx1 = 0\n",
    "            coidx2 = np.abs(maxaln)\n",
    "        else:\n",
    "            coidx1 = maxaln\n",
    "            coidx2 = 0\n",
    "    \n",
    "    substr1 = s1[coidx1:]\n",
    "    substr2 = s2[coidx2:]\n",
    "\n",
    "    print('substr1', substr1)\n",
    "    print('substr2', substr2)\n",
    "    print('coidx1', coidx1)\n",
    "    print('coidx2', coidx2)\n",
    "    \n",
    "    #find equivalent positions in the strings\n",
    "    maps1 = {}\n",
    "    maps2 = {}\n",
    "\n",
    "    oppositemap1 = {}\n",
    "    oppositemap2 = {}\n",
    "\n",
    "    for i, char in enumerate(substr1):\n",
    "        if substr2[i] == char:\n",
    "            maps1[i+coidx1] = i + start1\n",
    "            maps2[i+coidx2] = i + start2\n",
    "            \n",
    "            oppositemap1[i+coidx1] = i + start2\n",
    "            oppositemap2[i+coidx2] = i + start1\n",
    "        else:\n",
    "            #if there is mismatch convolve the remaining strings\n",
    "            print('mismatch')\n",
    "            sub1,sub2 , om1 , om2 = aln_mapping( substr1[i:] , substr2[i:], maxaln = 0 , start1= start1+coidx1+i , start2 = start2+coidx2+i , verbose = False)\n",
    "            maps1.update( sub1 )\n",
    "            maps2.update( sub2 )\n",
    "            oppositemap1.update( om1 )\n",
    "            oppositemap2.update( om2 )\n",
    "            break\n",
    "    \n",
    "    return maps1, maps2 , oppositemap1, oppositemap2\n",
    "\n",
    "def aln_mapping_full( s1 , s2, maxaln = 0 , coidx1= 0 , coidx2 = 0 , verbose = False):\n",
    "    maps1, maps2 , om1 , om2 = aln_mapping( s1 , s2, maxaln = 0 , coidx1= 0 , coidx2 = 0 , verbose = False)\n",
    "    #add the reverse mapping\n",
    "    revmap1 = { v:k for k,v in maps1.items()}\n",
    "    revmap2 = { v:k for k,v in maps2.items()}\n",
    "\n",
    "    revmapom1 = { v:k for k,v in om1.items()}   \n",
    "    revmapom2 = { v:k for k,v in om2.items()}\n",
    "    \n",
    "    return maps1, maps2, om1, om2 , revmap1, revmap2 , revmapom1, revmapom2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def convolve_strings(str1, str2):\n",
    "    # Determine the lengths of the strings\n",
    "    len1, len2 = len(str1), len(str2)\n",
    "\n",
    "    if len(str1) < len(str2):\n",
    "        str1, str2 = str2, str1\n",
    "        len1, len2 = len2, len1\n",
    "    max_alignment = 0\n",
    "    max_count = 0\n",
    "    # Slide str2 over str1, starting with one character overlap\n",
    "    # and continue until str2 is again overlapping by just one character\n",
    "    for i in range(-len2 + 1, len1):\n",
    "        count = 0\n",
    "        for j in range(len2):\n",
    "            if 0 <= i + j < len1 and str1[i + j] == str2[j]:\n",
    "                count += 1\n",
    "        if count > max_count:\n",
    "            max_count = count\n",
    "            max_alignment = i\n",
    "    return max_alignment, max_count\n",
    "\n",
    "\n",
    "def alnchop(s1,s2,rawaln1,rawaln2,aln1,aln2,maxaln = 0):\n",
    "\n",
    "    #align the two sequences\n",
    "    coidx1 = 0\n",
    "    coidx2 = 0\n",
    "\n",
    "    if len(s1) < len(s2):\n",
    "        if maxaln < 0:\n",
    "            coidx1 = np.abs(maxaln)\n",
    "            coidx2 = 0\n",
    "        else:\n",
    "            coidx1 = 0\n",
    "            coidx2 = maxaln\n",
    "    else:\n",
    "        if maxaln < 0:\n",
    "            coidx1 = 0\n",
    "            coidx2 = np.abs(maxaln)\n",
    "        else:\n",
    "            coidx1 = maxaln\n",
    "            coidx2 = 0\n",
    "    discardcount = 0 \n",
    "    rawaln1 = iter(rawaln1)\n",
    "    while coidx1 > 0:\n",
    "        rchar1 = next(rawaln1)\n",
    "        discardcount += 1\n",
    "        if rchar1 != '-':\n",
    "            coidx1 -= 1\n",
    "    aln1 = aln1[discardcount:]\n",
    "\n",
    "\n",
    "    discardcount = 0 \n",
    "    rawaln2 = iter(rawaln2)\n",
    "    while coidx1 > 0:\n",
    "        rchar2 = next(rawaln2)\n",
    "        discardcount += 1\n",
    "        if rchar2 != '-':\n",
    "            coidx2 -= 1\n",
    "    aln2 = aln2[discardcount:]\n",
    "\n",
    "\n",
    "    rawaln1 = ''.join([ s for s in iter(rawaln1)])\n",
    "    rawaln2 = ''.join([ s for s in iter(rawaln2)])\n",
    "\n",
    "    return aln1, aln2, rawaln1, rawaln2\n",
    "\n",
    "\n",
    "def mergealns( aln1f, aln2f, outfile , verbose = False):\n",
    "    if set(get_fasta_leafset(aln1f)) == set(get_fasta_leafset(aln2f)):\n",
    "        print('identical')\n",
    "        return aln1f\n",
    "\n",
    "    #find sequences in common between the two alignments\n",
    "    aln1 = SeqIO.parse(aln1f, 'fasta')\n",
    "    aln2 = SeqIO.parse(aln2f, 'fasta')\n",
    "    ids1 = {s.id:str(s.seq) for s in aln1}\n",
    "    ids2 = {s.id:str(s.seq) for s in aln2}\n",
    "    aln1 = SeqIO.parse(aln1f, 'fasta')\n",
    "    aln2 = SeqIO.parse(aln2f, 'fasta')\n",
    "    idlist = [ s.id for s in aln1] + [ s.id for s in aln2]\n",
    "    commonids = set(ids1.keys()).intersection(set(ids2.keys()))\n",
    "    try:\n",
    "        assert len(commonids) > 0\n",
    "    except:\n",
    "        print('no common ids')\n",
    "        print('ids1', ids1)\n",
    "        print('ids2', ids2)\n",
    "        raise Exception('no common ids')\n",
    "    #transform both alignments into numpy matrices\n",
    "    aln1 = SeqIO.parse(aln1f, 'fasta')\n",
    "    aln2 = SeqIO.parse(aln2f, 'fasta')\n",
    "    aln1 = np.array([ list(str(s.seq)) for s in aln1])\n",
    "    aln2 = np.array([ list(str(s.seq)) for s in aln2])\n",
    "    nrows1 = aln1.shape[0]\n",
    "    nrows2 = aln2.shape[0]\n",
    "\n",
    "    #generate a list of column arrays\n",
    "    aln1 = [ aln1[:,i] for i in range(aln1.shape[1])]\n",
    "    aln2 = [ aln2[:,i] for i in range(aln2.shape[1])]\n",
    "    #find the best common sequence\n",
    "    maxconv = 0\n",
    "    maxaln = 0\n",
    "\n",
    "    print(ids1)\n",
    "    print(ids2)\n",
    "    \n",
    "    for commonid in commonids:\n",
    "        s1t = ids1[commonid]\n",
    "        s2t = ids2[commonid]\n",
    "        s1t = s1t.replace('-','')\n",
    "        s2t = s2t.replace('-','')\n",
    "        #if the common subsequence is not found start by removing the first character of the common sequence\n",
    "        #convolution of the two sequences\n",
    "        aln, count = convolve_strings(s1t,s2t)\n",
    "        if count > maxconv:\n",
    "            maxconv = count\n",
    "            maxaln = aln\n",
    "            ID = commonid\n",
    "            s1 = s1t\n",
    "            s2 = s2t\n",
    "\n",
    "    rawaln1 = ids1[ID]\n",
    "    rawaln2 = ids2[ID]\n",
    "    print('pivot' , ID)\n",
    "    print('s1',s1)\n",
    "    print('s2',s2)\n",
    "\n",
    "    print('rawaln1', rawaln1)\n",
    "    print('rawaln2', rawaln2)\n",
    "    \n",
    "    print('maxaln', maxaln)\n",
    "    #use the sequence convolution to align the two alignment arrays\n",
    "    if len(s1) < len(s2):\n",
    "        if maxaln < 0:\n",
    "            coidx1 = np.abs(maxaln)\n",
    "            coidx2 = 0\n",
    "        else:\n",
    "            coidx1 = 0\n",
    "            coidx2 = maxaln\n",
    "    else:\n",
    "        if maxaln < 0:\n",
    "            coidx1 = 0\n",
    "            coidx2 = np.abs(maxaln)\n",
    "        else:\n",
    "            coidx1 = maxaln\n",
    "            coidx2 = 0\n",
    "    print('coidx1', coidx1)\n",
    "    print('coidx2', coidx2)\n",
    "\n",
    "\n",
    "    #remove the leading gaps\n",
    "    for i in range(len(rawaln1)):\n",
    "        if rawaln1 != '-':\n",
    "            break\n",
    "    aln1 = aln1[i:]\n",
    "    rawaln1 = rawaln1[i:]\n",
    "    for i in range(len(rawaln1)):\n",
    "        if rawaln1 != '-':\n",
    "            break\n",
    "    aln2 = aln2[i:]\n",
    "    rawaln2 = rawaln2[i:]\n",
    "\n",
    "    #remove the trailing gaps\n",
    "    for i in range(len(rawaln1)):\n",
    "        if rawaln1[-i] != '-':\n",
    "            break\n",
    "    aln1 = aln1[:len(rawaln1)-i]\n",
    "    rawaln1 = rawaln1[:len(rawaln1)-i]\n",
    "    for i in range(len(rawaln2)):\n",
    "        if rawaln2[-i] != '-':\n",
    "            break\n",
    "    aln2 = aln2[:len(rawaln2)-i]\n",
    "    rawaln2 = rawaln2[:len(rawaln2)-i]\n",
    "    \n",
    "\n",
    "    #construct alignment with common sequence\n",
    "\n",
    "    rawaln1 = iter(rawaln1)\n",
    "    rawaln2 = iter(rawaln2)    \n",
    "    char1 = next(rawaln1)\n",
    "    char2 = next(rawaln2)\n",
    "    i = 1\n",
    "    j = 1 \n",
    "    pchar1 = char1\n",
    "    pchar2 = char2\n",
    "    newaln1 = []\n",
    "    newaln2 = []\n",
    "    convolved = False\n",
    "    while True:\n",
    "        try:\n",
    "            if pchar1 == '-' and char1 != '-':\n",
    "                print('end insertion1')\n",
    "\n",
    "            if pchar2 == '-' and char2 != '-':\n",
    "                print('end insertion2')\n",
    "\n",
    "            if char1 == '-' and char2 != '-':\n",
    "                print('insertion1')\n",
    "                newaln2.append(['-']*nrows2)\n",
    "                newaln1.append(aln1[i])\n",
    "                pchar1 = char1\n",
    "                char1 = next(rawaln1)\n",
    "                i +=1\n",
    "\n",
    "            elif char2 == '-' and char1 != '-':\n",
    "                print('insertion2')\n",
    "                newaln2.append(aln2[j])\n",
    "                newaln1.append(['-']*nrows1)\n",
    "                pchar2 = char2\n",
    "                char2 = next(rawaln2)\n",
    "                j +=1\n",
    "\n",
    "            elif char1 == char2 and char1 != '-' and char2 != '-':\n",
    "                char1 = next(rawaln1)\n",
    "                char2 = next(rawaln2)\n",
    "                newaln2.append(aln2[j])\n",
    "                newaln1.append(aln1[i])\n",
    "                pchar1 = char1\n",
    "                pchar2 = char2\n",
    "                j+= 1\n",
    "                i+= 1\n",
    "\n",
    "            elif char1 != '-' and char2 != '-' and char1 != char2:\n",
    "                convolved = True\n",
    "                print('mismatch')\n",
    "                #mismatch reconvolve remaining strings\n",
    "                rawaln1 = ''.join([ s for s in iter(rawaln1)])\n",
    "                rawaln2 = ''.join([ s for s in iter(rawaln2)])\n",
    "                s1 = rawaln1.replace('-','')\n",
    "                s2 = rawaln2.replace('-','')\n",
    "\n",
    "                maxaln, count = convolve_strings(s1,s2)\n",
    "                if len(s1) < len(s2):\n",
    "                    if maxaln < 0:\n",
    "                        coidx1 = np.abs(maxaln)\n",
    "                        coidx2 = 0\n",
    "                    else:\n",
    "                        coidx1 = 0\n",
    "                        coidx2 = maxaln\n",
    "                else:\n",
    "                    if maxaln < 0:\n",
    "                        coidx1 = 0\n",
    "                        coidx2 = np.abs(maxaln)\n",
    "                    else:\n",
    "                        coidx1 = maxaln\n",
    "                        coidx2 = 0\n",
    "                \n",
    "                rawaln1 = iter(rawaln1)\n",
    "                discardcount1 = 0\n",
    "                count1 = 0\n",
    "                while count1 < coidx1:\n",
    "                    discardcount1 += 1\n",
    "                    char1 = next(rawaln1)\n",
    "                    if char1 != '-':\n",
    "                        count1 += 1\n",
    "                aln1 = aln1[discardcount1:]\n",
    "                rawaln2 = iter(rawaln2) \n",
    "                discardcount2 = 0\n",
    "                count2 = 0\n",
    "                while count2 < coidx2:\n",
    "                    discardcount2 += 1\n",
    "                    char2 = next(rawaln2)\n",
    "                    if char2 != '-':\n",
    "                        count2 += 1\n",
    "                aln2 = aln2[discardcount2:]\n",
    "\n",
    "                char1 = next(rawaln1)\n",
    "                char2 = next(rawaln2)\n",
    "                \n",
    "\n",
    "                i = 0 \n",
    "                j = 0\n",
    "                print('char1', char1)\n",
    "                print('char2', char2)\n",
    "\n",
    "                print( 'newaln1' , newaln1)\n",
    "\n",
    "                print( 'newaln2' , newaln2)\n",
    "                \n",
    "            else:\n",
    "                print('end')\n",
    "                break\n",
    "        \n",
    "        except StopIteration:\n",
    "            break\n",
    "    try:\n",
    "        newaln1 = np.vstack(newaln1).T\n",
    "        newaln2 = np.vstack(newaln2).T\n",
    "        newaln = np.concatenate((newaln1, newaln2), axis = 0)\n",
    "    except:\n",
    "        print( 'aln err')\n",
    "        print( 'char1', char1 )\n",
    "        print( 'char2' , char2 )\n",
    "        print( 'aln1', aln1 ) \n",
    "        print( 'aln2' , aln2 )\n",
    "        print( 'maxaln'  , maxaln)\n",
    "        raise Exception('aln err: newaln not created')\n",
    "    #write out the new alignment\n",
    "    with open(outfile, 'w') as f:\n",
    "        for i in range(newaln.shape[0]):\n",
    "            #print('>' + idlist[i] + '\\n' + ''.join(list(newaln[i,:])) + '\\n')\n",
    "            f.write('>' + idlist[i] + '\\n')\n",
    "            f.write(''.join(list(newaln[i,:])) + '\\n')\n",
    "    remove_redundant( outfile )\n",
    "    with open(outfile) as out:\n",
    "        print(out.read())\n",
    "    return outfile  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import toytree\n",
    "import os\n",
    "import pandas as pd\n",
    "import glob\n",
    "\n",
    "\n",
    "alndf = pd.read_table('../testdata/HOG113/allvall_1.csv', header = None)\n",
    "\n",
    "infolder = '../testdata/HOG113/allvall_1.csv'.split('/')[:-1]\n",
    "infolder = ''.join( [i + '/' for i in infolder])\n",
    "mapper3di, mapperAA = read_dbfiles3di( \"../testdata/HOG113/outdb\" , \"../testdata/HOG113/outdb_ss\")\n",
    "#add the 3di alignment to the dataframe\n",
    "columns = 'query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,lddt,qaln,taln,cigar,lntmscore'.split(',')\n",
    "alndf.columns = columns\n",
    "\n",
    "alndf['3diq']= alndf['query'].map(mapper3di)\n",
    "alndf['3dit']= alndf['target'].map(mapper3di)\n",
    "alndf['AAq']= alndf['query'].map(mapperAA)\n",
    "alndf['AAt']= alndf['target'].map(mapperAA)\n",
    "\n",
    "#output a fasta with the 3di sequences\n",
    "res = alndf.apply(calc_fident_crossaln , axis = 1)\n",
    "alndf = pd.concat([alndf,res] , axis = 1)\n",
    "\n",
    "with open('../testdata/HOG113/3diseqs.fasta' , 'w') as out:\n",
    "    for seq in alndf['query'].unique():\n",
    "        out.write('>'+seq.replace('.pdb', '' )+'\\n')\n",
    "        out.write(mapper3di[seq]+'\\n')\n",
    "\n",
    "\n",
    "alndf['query'] = alndf['query'].map(lambda x :x.replace('.pdb', ''))\n",
    "alndf['target'] = alndf['target'].map(lambda x :x.replace('.pdb', ''))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "traverse 76 False None\n",
      "not cherry 76\n",
      "children ['75', '74']\n",
      "traverse 75 False None\n",
      "traverse 75 False None\n",
      "not cherry 75\n",
      "children ['Q74CJ3', '73']\n",
      "traverse Q74CJ3 True None\n",
      "traverse Q74CJ3 True None\n",
      "leaf query                                                        Q74CJ3\n",
      "target                                                       Q74CJ3\n",
      "fident                                                          1.0\n",
      "alnlen                                                          143\n",
      "mismatch                                                          0\n",
      "gapopen                                                           0\n",
      "qstart                                                            1\n",
      "qend                                                            143\n",
      "tstart                                                            1\n",
      "tend                                                            143\n",
      "evalue                                                          0.0\n",
      "bits                                                           1236\n",
      "lddt                                                            1.0\n",
      "qaln              MPYRYLPDIATADVAFEAWGETREEMFCAAADALTNVMVDDLASVT...\n",
      "taln              MPYRYLPDIATADVAFEAWGETREEMFCAAADALTNVMVDDLASVT...\n",
      "cigar                                                          143M\n",
      "lntmscore                                                       1.0\n",
      "3diq              DPDDDDVPPDDFKDKDKFKDLDLWRRLLRVLLVLLVLFFVDPVLFD...\n",
      "3dit              DPDDDDVPPDDFKDKDKFKDLDLWRRLLRVLLVLLVLFFVDPVLFD...\n",
      "AAq               MPYRYLPDIATADVAFEAWGETREEMFCAAADALTNVMVDDLASVT...\n",
      "AAt               MPYRYLPDIATADVAFEAWGETREEMFCAAADALTNVMVDDLASVT...\n",
      "3di_qaln_mode2    DPDDDDVPPDDFKDKDKFKDLDLWRRLLRVLLVLLVLFFVDPVLFD...\n",
      "3di_taln_mode2    DPDDDDVPPDDFKDKDKFKDLDLWRRLLRVLLVLLVLFFVDPVLFD...\n",
      "Name: 1170, dtype: object\n",
      "traverse 73 False None\n",
      "traverse 73 False None\n",
      "not cherry 73\n",
      "children ['70', '69']\n",
      "traverse 70 False None\n",
      "traverse 70 False None\n",
      "not cherry 70\n",
      "children ['A9A300', '66']\n",
      "traverse A9A300 True None\n",
      "traverse A9A300 True None\n",
      "leaf query                                                        A9A300\n",
      "target                                                       A9A300\n",
      "fident                                                          1.0\n",
      "alnlen                                                          138\n",
      "mismatch                                                          0\n",
      "gapopen                                                           0\n",
      "qstart                                                            1\n",
      "qend                                                            138\n",
      "tstart                                                            1\n",
      "tend                                                            138\n",
      "evalue                                                          0.0\n",
      "bits                                                           1182\n",
      "lddt                                                            1.0\n",
      "qaln              MSYKFVDHATDAIIEVTAKDLQEAFSVTADAVINLTLDQDKVEEKE...\n",
      "taln              MSYKFVDHATDAIIEVTAKDLQEAFSVTADAVINLTLDQDKVEEKE...\n",
      "cigar                                                          138M\n",
      "lntmscore                                                       1.0\n",
      "3diq              DDDDDDDDFPKDKDKDWFQDPQVRLFVVLVVLLVQFWALVLWDFPD...\n",
      "3dit              DDDDDDDDFPKDKDKDWFQDPQVRLFVVLVVLLVQFWALVLWDFPD...\n",
      "AAq               MSYKFVDHATDAIIEVTAKDLQEAFSVTADAVINLTLDQDKVEEKE...\n",
      "AAt               MSYKFVDHATDAIIEVTAKDLQEAFSVTADAVINLTLDQDKVEEKE...\n",
      "3di_qaln_mode2    DDDDDDDDFPKDKDKDWFQDPQVRLFVVLVVLLVQFWALVLWDFPD...\n",
      "3di_taln_mode2    DDDDDDDDFPKDKDKDWFQDPQVRLFVVLVVLLVQFWALVLWDFPD...\n",
      "Name: 273, dtype: object\n",
      "traverse 66 False None\n",
      "traverse 66 False None\n",
      "cherry O67888 B5YH50\n",
      "mafft --addfull  ../testdata/HOG113/alnscratch/A9A300_inter.fasta --keeplength ../testdata/HOG113/alnscratch/66_inter.fasta > ../testdata/HOG113/alnscratch/70_inter.fasta\n",
      "mafft --addfull  ../testdata/HOG113/alnscratch/A9A300_inter.3di.fasta --keeplength ../testdata/HOG113/alnscratch/66_inter.3di.fasta > ../testdata/HOG113/alnscratch/70_inter3di.fasta\n",
      "traverse 69 False None\n",
      "traverse 69 False None\n",
      "not cherry 69\n",
      "children ['65', '64']\n",
      "traverse 65 False None\n",
      "traverse 65 False None\n",
      "cherry Q5JCY8 Q60334\n",
      "traverse 64 False None\n",
      "traverse 64 False None\n",
      "cherry B1L703 Q9UX79\n",
      "mafft --seed ../testdata/HOG113/alnscratch/65_inter.fasta ../testdata/HOG113/alnscratch/64_inter.fasta > ../testdata/HOG113/alnscratch/69_inter.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    1 / 2\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 2\n",
      "njobc = 3\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "To keep the alignment length, 3 letters were DELETED.\n",
      "To know the positions of deleted letters, rerun the same command with the --mapout option.\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    1 / 2\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 2\n",
      "njobc = 3\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "To keep the alignment length, 3 letters were DELETED.\n",
      "To know the positions of deleted letters, rerun the same command with the --mapout option.\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 2\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 400.000000\n",
      "\n",
      "adding 0-1\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 4\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "    0 / 4\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP     3 / 3 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "    0 / 4\n",
      "    0 / 2\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP     3 /3 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "    0 / 4\n",
      "Segment   1/  1    1- 139\n",
      "STEP 002-002-1  identical.   \n",
      "Converged.\n",
      "\n",
      "done\n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --seed ../testdata/HOG113/alnscratch/65_inter.3di.fasta ../testdata/HOG113/alnscratch/64_inter.3di.fasta > ../testdata/HOG113/alnscratch/69_inter3di.fasta\n",
      "mafft --seed ../testdata/HOG113/alnscratch/70_inter.fasta ../testdata/HOG113/alnscratch/69_inter.fasta > ../testdata/HOG113/alnscratch/73_inter.fasta\n",
      "mafft --seed ../testdata/HOG113/alnscratch/70_inter3di.fasta ../testdata/HOG113/alnscratch/69_inter3di.fasta > ../testdata/HOG113/alnscratch/73_inter3di.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "nhomologs = 2\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 400.000000\n",
      "\n",
      "adding 0-1\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 4\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "    0 / 4\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP     3 / 3 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "    0 / 4\n",
      "    0 / 2\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP     3 /3 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "    0 / 4\n",
      "Segment   1/  1    1- 140\n",
      "done 002-001-1  identical.   \n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 4\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 1600.000000\n",
      "\n",
      "adding 1-2\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 7\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "    0 / 7\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP     5 / 6 \n",
      "Reallocating..done. *alloclen = 1278\n",
      "STEP     6 / 6 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "    0 / 7\n",
      "    0 / 3\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP     6 /6 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "    0 / 7\n",
      "Segment   1/  1    1- 144\n",
      "STEP 002-003-1  identical.   \n",
      "Converged.\n",
      "\n",
      "done\n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 4\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 1600.000000\n",
      "\n",
      "adding 1-2\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 7\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "    0 / 7\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP     6 / 6 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "    0 / 7\n",
      "    0 / 3\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP     6 /6 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "    0 / 7\n",
      "Segment   1/  1    1- 144\n",
      "done 002-001-1  identical.   \n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --addfull  ../testdata/HOG113/alnscratch/Q74CJ3_inter.fasta --keeplength ../testdata/HOG113/alnscratch/73_inter.fasta > ../testdata/HOG113/alnscratch/75_inter.fasta\n",
      "mafft --addfull  ../testdata/HOG113/alnscratch/Q74CJ3_inter.3di.fasta --keeplength ../testdata/HOG113/alnscratch/73_inter3di.fasta > ../testdata/HOG113/alnscratch/75_inter3di.fasta\n",
      "traverse 74 False None\n",
      "traverse 74 False None\n",
      "not cherry 74\n",
      "children ['72', '71']\n",
      "traverse 72 False None\n",
      "traverse 72 False None\n",
      "cherry B8E097 Q8TU10\n",
      "traverse 71 False None\n",
      "traverse 71 False None\n",
      "not cherry 71\n",
      "children ['68', '67']\n",
      "traverse 68 False None\n",
      "traverse 68 False None\n",
      "not cherry 68\n",
      "children ['A2F1C4', '63']\n",
      "traverse A2F1C4 True None\n",
      "traverse A2F1C4 True None\n",
      "leaf query                                                        A2F1C4\n",
      "target                                                       A2F1C4\n",
      "fident                                                          1.0\n",
      "alnlen                                                          145\n",
      "mismatch                                                          0\n",
      "gapopen                                                           0\n",
      "qstart                                                            1\n",
      "qend                                                            145\n",
      "tstart                                                            1\n",
      "tend                                                            145\n",
      "evalue                                                          0.0\n",
      "bits                                                           1245\n",
      "lddt                                                            1.0\n",
      "qaln              MYTEAKWEFIDHPADIQLHAWGPQPDVTLEQLGKAFFEVMFETDNF...\n",
      "taln              MYTEAKWEFIDHPADIQLHAWGPQPDVTLEQLGKAFFEVMFETDNF...\n",
      "cigar                                                          145M\n",
      "lntmscore                                                       1.0\n",
      "3diq              DDPDDDDDDDDDPQKDKDKFKFQDPQVRLFVRLLVLLLQFFPQVLW...\n",
      "3dit              DDPDDDDDDDDDPQKDKDKFKFQDPQVRLFVRLLVLLLQFFPQVLW...\n",
      "AAq               MYTEAKWEFIDHPADIQLHAWGPQPDVTLEQLGKAFFEVMFETDNF...\n",
      "AAt               MYTEAKWEFIDHPADIQLHAWGPQPDVTLEQLGKAFFEVMFETDNF...\n",
      "3di_qaln_mode2    DDPDDDDDDDDDPQKDKDKFKFQDPQVRLFVRLLVLLLQFFPQVLW...\n",
      "3di_taln_mode2    DDPDDDDDDDDDPQKDKDKFKFQDPQVRLFVRLLVLLLQFFPQVLW...\n",
      "Name: 234, dtype: object\n",
      "traverse 63 False None\n",
      "traverse 63 False None\n",
      "cherry B8BWG5 Q9TZN1\n",
      "mafft --addfull  ../testdata/HOG113/alnscratch/A2F1C4_inter.fasta --keeplength ../testdata/HOG113/alnscratch/63_inter.fasta > ../testdata/HOG113/alnscratch/68_inter.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    6 / 7\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 7\n",
      "njobc = 8\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "To keep the alignment length, 2 letters were DELETED.\n",
      "To know the positions of deleted letters, rerun the same command with the --mapout option.\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    6 / 7\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 7\n",
      "njobc = 8\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "To keep the alignment length, 3 letters were DELETED.\n",
      "To know the positions of deleted letters, rerun the same command with the --mapout option.\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    1 / 2\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 2\n",
      "njobc = 3\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "To keep the alignment length, 9 letters were DELETED.\n",
      "To know the positions of deleted letters, rerun the same command with the --mapout option.\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --addfull  ../testdata/HOG113/alnscratch/A2F1C4_inter.3di.fasta --keeplength ../testdata/HOG113/alnscratch/63_inter.3di.fasta > ../testdata/HOG113/alnscratch/68_inter3di.fasta\n",
      "traverse 67 False None\n",
      "traverse 67 False None\n",
      "not cherry 67\n",
      "children ['62', '61']\n",
      "traverse 62 False None\n",
      "traverse 62 False None\n",
      "cherry Q54BU9 A9UTI8\n",
      "traverse 61 False None\n",
      "traverse 61 False None\n",
      "not cherry 61\n",
      "children ['60', '59']\n",
      "traverse 60 False None\n",
      "traverse 60 False None\n",
      "not cherry 60\n",
      "children ['58', '57']\n",
      "traverse 58 False None\n",
      "traverse 58 False None\n",
      "not cherry 58\n",
      "children ['Q7PYJ8', '55']\n",
      "traverse Q7PYJ8 True None\n",
      "traverse Q7PYJ8 True None\n",
      "leaf query                                                        Q7PYJ8\n",
      "target                                                       Q7PYJ8\n",
      "fident                                                          1.0\n",
      "alnlen                                                          157\n",
      "mismatch                                                          0\n",
      "gapopen                                                           0\n",
      "qstart                                                            1\n",
      "qend                                                            157\n",
      "tstart                                                            1\n",
      "tend                                                            157\n",
      "evalue                                                          0.0\n",
      "bits                                                           1381\n",
      "lddt                                                            1.0\n",
      "qaln              MEVPLETSDHVVPEVKYEYLDHTADVQLHAWGETLQEAFEQCGMAM...\n",
      "taln              MEVPLETSDHVVPEVKYEYLDHTADVQLHAWGETLQEAFEQCGMAM...\n",
      "cigar                                                          157M\n",
      "lntmscore                                                       1.0\n",
      "3diq              DWDDDDPDDPDQDDQAWDWDDDPQFIKIKGFDQDLQRQLFNSLVSV...\n",
      "3dit              DWDDDDPDDPDQDDQAWDWDDDPQFIKIKGFDQDLQRQLFNSLVSV...\n",
      "AAq               MEVPLETSDHVVPEVKYEYLDHTADVQLHAWGETLQEAFEQCGMAM...\n",
      "AAt               MEVPLETSDHVVPEVKYEYLDHTADVQLHAWGETLQEAFEQCGMAM...\n",
      "3di_qaln_mode2    DWDDDDPDDPDQDDQAWDWDDDPQFIKIKGFDQDLQRQLFNSLVSV...\n",
      "3di_taln_mode2    DWDDDDPDDPDQDDQAWDWDDDPQFIKIKGFDQDLQRQLFNSLVSV...\n",
      "Name: 1209, dtype: object\n",
      "traverse 55 False None\n",
      "traverse 55 False None\n",
      "cherry Q9VD92 D6WKY5\n",
      "mafft --addfull  ../testdata/HOG113/alnscratch/Q7PYJ8_inter.fasta --keeplength ../testdata/HOG113/alnscratch/55_inter.fasta > ../testdata/HOG113/alnscratch/58_inter.fasta\n",
      "mafft --addfull  ../testdata/HOG113/alnscratch/Q7PYJ8_inter.3di.fasta --keeplength ../testdata/HOG113/alnscratch/55_inter.3di.fasta > ../testdata/HOG113/alnscratch/58_inter3di.fasta\n",
      "traverse 57 False None\n",
      "traverse 57 False None\n",
      "not cherry 57\n",
      "children ['C3Y972', '54']\n",
      "traverse C3Y972 True None\n",
      "traverse C3Y972 True None\n",
      "leaf query                                                        C3Y972\n",
      "target                                                       C3Y972\n",
      "fident                                                          1.0\n",
      "alnlen                                                          164\n",
      "mismatch                                                          0\n",
      "gapopen                                                           0\n",
      "qstart                                                            1\n",
      "qend                                                            164\n",
      "tstart                                                            1\n",
      "tend                                                            164\n",
      "evalue                                                          0.0\n",
      "bits                                                           1351\n",
      "lddt                                                            1.0\n",
      "qaln              MEEEGAEYDPEWEGGEEGDGGEYVCKYEYLDHTADVQLHSWGSDLK...\n",
      "taln              MEEEGAEYDPEWEGGEEGDGGEYVCKYEYLDHTADVQLHSWGSDLK...\n",
      "cigar                                                          164M\n",
      "lntmscore                                                       1.0\n",
      "3diq              DDDDDDDDDPPDDDDDDDDDDDPPPPDDDDDDPQKDKDKAKDQDPQ...\n",
      "3dit              DDDDDDDDDPPDDDDDDDDDDDPPPPDDDDDDPQKDKDKAKDQDPQ...\n",
      "AAq               MEEEGAEYDPEWEGGEEGDGGEYVCKYEYLDHTADVQLHSWGSDLK...\n",
      "AAt               MEEEGAEYDPEWEGGEEGDGGEYVCKYEYLDHTADVQLHSWGSDLK...\n",
      "3di_qaln_mode2    DDDDDDDDDPPDDDDDDDDDDDPPPPDDDDDDPQKDKDKAKDQDPQ...\n",
      "3di_taln_mode2    DDDDDDDDDPPDDDDDDDDDDDPPPPDDDDDDPQKDKDKAKDQDPQ...\n",
      "Name: 546, dtype: object\n",
      "traverse 54 False None\n",
      "traverse 54 False None\n",
      "cherry B7PHX9 T1ELH7\n",
      "mafft --addfull  ../testdata/HOG113/alnscratch/C3Y972_inter.fasta --keeplength ../testdata/HOG113/alnscratch/54_inter.fasta > ../testdata/HOG113/alnscratch/57_inter.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    1 / 2\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 2\n",
      "njobc = 3\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "To keep the alignment length, 9 letters were DELETED.\n",
      "To know the positions of deleted letters, rerun the same command with the --mapout option.\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    1 / 2\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 2\n",
      "njobc = 3\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "To keep the alignment length, 3 letters were DELETED.\n",
      "To know the positions of deleted letters, rerun the same command with the --mapout option.\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    1 / 2\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 2\n",
      "njobc = 3\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "To keep the alignment length, 3 letters were DELETED.\n",
      "To know the positions of deleted letters, rerun the same command with the --mapout option.\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    1 / 2\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 2\n",
      "njobc = 3\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "To keep the alignment length, 59 letters were DELETED.\n",
      "To know the positions of deleted letters, rerun the same command with the --mapout option.\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    1 / 2\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 2\n",
      "njobc = 3\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "To keep the alignment length, 59 letters were DELETED.\n",
      "To know the positions of deleted letters, rerun the same command with the --mapout option.\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 3\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 900.000000\n",
      "\n",
      "adding 1-2\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 6\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "    0 / 6\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP     5 / 5 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "    0 / 6\n",
      "    0 / 3\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP     5 /5 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "    0 / 6\n",
      "Segment   1/  1    1- 157\n",
      "STEP 002-004-1  identical.   \n",
      "Converged.\n",
      "\n",
      "done\n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 3\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 900.000000\n",
      "\n",
      "adding 1-2\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 6\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "    0 / 6\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP     5 / 5 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "    0 / 6\n",
      "    0 / 3\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP     5 /5 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --addfull  ../testdata/HOG113/alnscratch/C3Y972_inter.3di.fasta --keeplength ../testdata/HOG113/alnscratch/54_inter.3di.fasta > ../testdata/HOG113/alnscratch/57_inter3di.fasta\n",
      "mafft --seed ../testdata/HOG113/alnscratch/58_inter.fasta ../testdata/HOG113/alnscratch/57_inter.fasta > ../testdata/HOG113/alnscratch/60_inter.fasta\n",
      "mafft --seed ../testdata/HOG113/alnscratch/58_inter3di.fasta ../testdata/HOG113/alnscratch/57_inter3di.fasta > ../testdata/HOG113/alnscratch/60_inter3di.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "    0 / 6\n",
      "Segment   1/  1    1- 157\n",
      "done 002-001-1  rejected..   \n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    1 / 2\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 2\n",
      "njobc = 3\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    1 / 2\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 2\n",
      "njobc = 3\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "traverse 59 False None\n",
      "traverse 59 False None\n",
      "not cherry 59\n",
      "children ['F7AW07', '56']\n",
      "traverse F7AW07 True None\n",
      "traverse F7AW07 True None\n",
      "leaf query                                                        F7AW07\n",
      "target                                                       F7AW07\n",
      "fident                                                          1.0\n",
      "alnlen                                                          160\n",
      "mismatch                                                          0\n",
      "gapopen                                                           0\n",
      "qstart                                                            1\n",
      "qend                                                            160\n",
      "tstart                                                            1\n",
      "tend                                                            160\n",
      "evalue                                                          0.0\n",
      "bits                                                           1325\n",
      "lddt                                                            1.0\n",
      "qaln              MEEKLPAYEENTSGYHNVIKKFEYLDHTADVQIHSWGDNIEETFEQ...\n",
      "taln              MEEKLPAYEENTSGYHNVIKKFEYLDHTADVQIHSWGDNIEETFEQ...\n",
      "cigar                                                          160M\n",
      "lntmscore                                                       1.0\n",
      "3diq              DDDDDDDDPDPPDPDDDDDDDDDDDDDDQKDKDKFKDQFPQVRLQV...\n",
      "3dit              DDDDDDDDPDPPDPDDDDDDDDDDDDDDQKDKDKFKDQFPQVRLQV...\n",
      "AAq               MEEKLPAYEENTSGYHNVIKKFEYLDHTADVQIHSWGDNIEETFEQ...\n",
      "AAt               MEEKLPAYEENTSGYHNVIKKFEYLDHTADVQIHSWGDNIEETFEQ...\n",
      "3di_qaln_mode2    DDDDDDDDPDPPDPDDDDDDDDDDDDDDQKDKDKFKDQFPQVRLQV...\n",
      "3di_taln_mode2    DDDDDDDDPDPPDPDDDDDDDDDDDDDDQKDKDKFKDQFPQVRLQV...\n",
      "Name: 702, dtype: object\n",
      "traverse 56 False None\n",
      "traverse 56 False None\n",
      "not cherry 56\n",
      "children ['53', '52']\n",
      "traverse 53 False None\n",
      "traverse 53 False None\n",
      "not cherry 53\n",
      "children ['H2MRT4', '51']\n",
      "traverse H2MRT4 True None\n",
      "traverse H2MRT4 True None\n",
      "leaf query                                                        H2MRT4\n",
      "target                                                       H2MRT4\n",
      "fident                                                          1.0\n",
      "alnlen                                                          162\n",
      "mismatch                                                          0\n",
      "gapopen                                                           0\n",
      "qstart                                                            1\n",
      "qend                                                            162\n",
      "tstart                                                            1\n",
      "tend                                                            162\n",
      "evalue                                                          0.0\n",
      "bits                                                           1353\n",
      "lddt                                                            1.0\n",
      "qaln              MDDRELDLTNEQKAVKSKYPPINKKYEYLDHTADVQIHSWGDSLEE...\n",
      "taln              MDDRELDLTNEQKAVKSKYPPINKKYEYLDHTADVQIHSWGDSLEE...\n",
      "cigar                                                          162M\n",
      "lntmscore                                                       1.0\n",
      "3diq              DDDPDDDDDPVRVVVCPVDDDDDDDDDDDDDFDKDKDKAWDQAPQR...\n",
      "3dit              DDDPDDDDDPVRVVVCPVDDDDDDDDDDDDDFDKDKDKAWDQAPQR...\n",
      "AAq               MDDRELDLTNEQKAVKSKYPPINKKYEYLDHTADVQIHSWGDSLEE...\n",
      "AAt               MDDRELDLTNEQKAVKSKYPPINKKYEYLDHTADVQIHSWGDSLEE...\n",
      "3di_qaln_mode2    DDDPDDDDDPVRVVVCPVDDDDDDDDDDDDDFDKDKDKAWDQAPQR...\n",
      "3di_taln_mode2    DDDPDDDDDPVRVVVCPVDDDDDDDDDDDDDFDKDKDKAWDQAPQR...\n",
      "Name: 780, dtype: object\n",
      "traverse 51 False None\n",
      "traverse 51 False None\n",
      "cherry W5M2J6 Q566V0\n",
      "mafft --addfull  ../testdata/HOG113/alnscratch/H2MRT4_inter.fasta --keeplength ../testdata/HOG113/alnscratch/51_inter.fasta > ../testdata/HOG113/alnscratch/53_inter.fasta\n",
      "mafft --addfull  ../testdata/HOG113/alnscratch/H2MRT4_inter.3di.fasta --keeplength ../testdata/HOG113/alnscratch/51_inter.3di.fasta > ../testdata/HOG113/alnscratch/53_inter3di.fasta\n",
      "traverse 52 False None\n",
      "traverse 52 False None\n",
      "not cherry 52\n",
      "children ['F6ZCC9', '50']\n",
      "traverse F6ZCC9 True None\n",
      "traverse F6ZCC9 True None\n",
      "leaf query                                                        F6ZCC9\n",
      "target                                                       F6ZCC9\n",
      "fident                                                          1.0\n",
      "alnlen                                                          169\n",
      "mismatch                                                          0\n",
      "gapopen                                                           0\n",
      "qstart                                                            1\n",
      "qend                                                            169\n",
      "tstart                                                            1\n",
      "tend                                                            169\n",
      "evalue                                                          0.0\n",
      "bits                                                           1395\n",
      "lddt                                                            1.0\n",
      "qaln              MDERYYTISEEQVAVKAKYPPVNKNYEYLDHTADVQLHAWGETLEE...\n",
      "taln              MDERYYTISEEQVAVKAKYPPVNKNYEYLDHTADVQLHAWGETLEE...\n",
      "cigar                                                          169M\n",
      "lntmscore                                                       1.0\n",
      "3diq              DDDDDDDDDPVVVVVCVVDDDDDDDDDDDDDFDKDKDKAWDQDQQR...\n",
      "3dit              DDDDDDDDDPVVVVVCVVDDDDDDDDDDDDDFDKDKDKAWDQDQQR...\n",
      "AAq               MDERYYTISEEQVAVKAKYPPVNKNYEYLDHTADVQLHAWGETLEE...\n",
      "AAt               MDERYYTISEEQVAVKAKYPPVNKNYEYLDHTADVQLHAWGETLEE...\n",
      "3di_qaln_mode2    DDDDDDDDDPVVVVVCVVDDDDDDDDDDDDDFDKDKDKAWDQDQQR...\n",
      "3di_taln_mode2    DDDDDDDDDPVVVVVCVVDDDDDDDDDDDDDFDKDKDKAWDQDQQR...\n",
      "Name: 624, dtype: object\n",
      "traverse 50 False None\n",
      "traverse 50 False None\n",
      "not cherry 50\n",
      "children ['49', '48']\n",
      "traverse 49 False None\n",
      "traverse 49 False None\n",
      "cherry A0A5F8GYJ2 F7AIV7\n",
      "traverse 48 False None\n",
      "traverse 48 False None\n",
      "not cherry 48\n",
      "children ['47', '46']\n",
      "traverse 47 False None\n",
      "traverse 47 False None\n",
      "cherry A0A2I3S880 A0A2I2YKY3\n",
      "traverse 46 False None\n",
      "traverse 46 False None\n",
      "not cherry 46\n",
      "children ['Q2YDE7', '45']\n",
      "traverse Q2YDE7 True None\n",
      "traverse Q2YDE7 True None\n",
      "leaf query                                                        Q2YDE7\n",
      "target                                                       Q2YDE7\n",
      "fident                                                          1.0\n",
      "alnlen                                                          167\n",
      "mismatch                                                          0\n",
      "gapopen                                                           0\n",
      "qstart                                                            1\n",
      "qend                                                            167\n",
      "tstart                                                            1\n",
      "tend                                                            167\n",
      "evalue                                                          0.0\n",
      "bits                                                           1389\n",
      "lddt                                                            1.0\n",
      "qaln              MALEGEDVRDYNLTEEQKAIKAKYPPVSRKYEYLDHTADVQLHAWG...\n",
      "taln              MALEGEDVRDYNLTEEQKAIKAKYPPVSRKYEYLDHTADVQLHAWG...\n",
      "cigar                                                          167M\n",
      "lntmscore                                                       1.0\n",
      "3diq              DDPDDAPDDDDPDDPVVVVVCVVDDDDDDDDGDDDDWDKDKDKAWD...\n",
      "3dit              DDPDDAPDDDDPDDPVVVVVCVVDDDDDDDDGDDDDWDKDKDKAWD...\n",
      "AAq               MALEGEDVRDYNLTEEQKAIKAKYPPVSRKYEYLDHTADVQLHAWG...\n",
      "AAt               MALEGEDVRDYNLTEEQKAIKAKYPPVSRKYEYLDHTADVQLHAWG...\n",
      "3di_qaln_mode2    DDPDDAPDDDDPDDPVVVVVCVVDDDDDDDDGDDDDWDKDKDKAWD...\n",
      "3di_taln_mode2    DDPDDAPDDDDPDDPVVVVVCVVDDDDDDDDGDDDDWDKDKDKAWD...\n",
      "Name: 936, dtype: object\n",
      "traverse 45 False None\n",
      "traverse 45 False None\n",
      "not cherry 45\n",
      "children ['44', '43']\n",
      "traverse 44 False None\n",
      "traverse 44 False None\n",
      "not cherry 44\n",
      "children ['Q8IWT0', '42']\n",
      "traverse Q8IWT0 True None\n",
      "traverse Q8IWT0 True None\n",
      "leaf query                                                        Q8IWT0\n",
      "target                                                       Q8IWT0\n",
      "fident                                                          1.0\n",
      "alnlen                                                          167\n",
      "mismatch                                                          0\n",
      "gapopen                                                           0\n",
      "qstart                                                            1\n",
      "qend                                                            167\n",
      "tstart                                                            1\n",
      "tend                                                            167\n",
      "evalue                                                          0.0\n",
      "bits                                                           1470\n",
      "lddt                                                            1.0\n",
      "qaln              MAQEEEDVRDYNLTEEQKAIKAKYPPVNRKYEYLDHTADVQLHAWG...\n",
      "taln              MAQEEEDVRDYNLTEEQKAIKAKYPPVNRKYEYLDHTADVQLHAWG...\n",
      "cigar                                                          167M\n",
      "lntmscore                                                       1.0\n",
      "3diq              DPPPPPPQDPLVDDPLLVVQLVVDDDDDAQWDWDDDDQWIKIKGWD...\n",
      "3dit              DPPPPPPQDPLVDDPLLVVQLVVDDDDDAQWDWDDDDQWIKIKGWD...\n",
      "AAq               MAQEEEDVRDYNLTEEQKAIKAKYPPVNRKYEYLDHTADVQLHAWG...\n",
      "AAt               MAQEEEDVRDYNLTEEQKAIKAKYPPVNRKYEYLDHTADVQLHAWG...\n",
      "3di_qaln_mode2    DPPPPPPQDPLVDDPLLVVQLVVDDDDDAQWDWDDDDQWIKIKGWD...\n",
      "3di_taln_mode2    DPPPPPPQDPLVDDPLLVVQLVVDDDDDAQWDWDDDDQWIKIKGWD...\n",
      "Name: 1248, dtype: object\n",
      "traverse 42 False None\n",
      "traverse 42 False None\n",
      "cherry G3QX27 K7BTY5\n",
      "mafft --addfull  ../testdata/HOG113/alnscratch/Q8IWT0_inter.fasta --keeplength ../testdata/HOG113/alnscratch/42_inter.fasta > ../testdata/HOG113/alnscratch/44_inter.fasta\n",
      "mafft --addfull  ../testdata/HOG113/alnscratch/Q8IWT0_inter.3di.fasta --keeplength ../testdata/HOG113/alnscratch/42_inter.3di.fasta > ../testdata/HOG113/alnscratch/44_inter3di.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    1 / 2\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 2\n",
      "njobc = 3\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    1 / 2\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 2\n",
      "njobc = 3\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    1 / 2\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 2\n",
      "njobc = 3\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    1 / 2\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 2\n",
      "njobc = 3\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "To keep the alignment length, 4 letters were DELETED.\n",
      "To know the positions of deleted letters, rerun the same command with the --mapout option.\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "traverse 43 False None\n",
      "traverse 43 False None\n",
      "not cherry 43\n",
      "children ['41', '40']\n",
      "traverse 41 False None\n",
      "traverse 41 False None\n",
      "cherry M0R5Y9 Q505B7\n",
      "traverse 40 False None\n",
      "traverse 40 False None\n",
      "not cherry 40\n",
      "children ['A0A2I2Y839', '39']\n",
      "traverse A0A2I2Y839 True None\n",
      "traverse A0A2I2Y839 True None\n",
      "leaf query                                                    A0A2I2Y839\n",
      "target                                                   A0A2I2Y839\n",
      "fident                                                          1.0\n",
      "alnlen                                                          137\n",
      "mismatch                                                          0\n",
      "gapopen                                                           0\n",
      "qstart                                                            1\n",
      "qend                                                            137\n",
      "tstart                                                            1\n",
      "tend                                                            137\n",
      "evalue                                                          0.0\n",
      "bits                                                            908\n",
      "lddt                                                            1.0\n",
      "qaln              MAQAEEDVRDYNLTEEQKAIKAKYPPVNRKYEYLDHTADVQLHAWG...\n",
      "taln              MAQAEEDVRDYNLTEEQKAIKAKYPPVNRKYEYLDHTADVQLHAWG...\n",
      "cigar                                                          137M\n",
      "lntmscore                                                       1.0\n",
      "3diq              DDPPDQQDPDDPDDPVVVVVCVVDDDDDDDDGDDDDDDDDDDDADD...\n",
      "3dit              DDPPDQQDPDDPDDPVVVVVCVVDDDDDDDDGDDDDDDDDDDDADD...\n",
      "AAq               MAQAEEDVRDYNLTEEQKAIKAKYPPVNRKYEYLDHTADVQLHAWG...\n",
      "AAt               MAQAEEDVRDYNLTEEQKAIKAKYPPVNRKYEYLDHTADVQLHAWG...\n",
      "3di_qaln_mode2    DDPPDQQDPDDPDDPVVVVVCVVDDDDDDDDGDDDDDDDDDDDADD...\n",
      "3di_taln_mode2    DDPPDQQDPDDPDDPVVVVVCVVDDDDDDDDGDDDDDDDDDDDADD...\n",
      "Name: 0, dtype: object\n",
      "traverse 39 False None\n",
      "traverse 39 False None\n",
      "cherry A0A2I3RJJ2 A0A2I2ZWB6\n",
      "mafft --addfull  ../testdata/HOG113/alnscratch/A0A2I2Y839_inter.fasta --keeplength ../testdata/HOG113/alnscratch/39_inter.fasta > ../testdata/HOG113/alnscratch/40_inter.fasta\n",
      "mafft --addfull  ../testdata/HOG113/alnscratch/A0A2I2Y839_inter.3di.fasta --keeplength ../testdata/HOG113/alnscratch/39_inter.3di.fasta > ../testdata/HOG113/alnscratch/40_inter3di.fasta\n",
      "mafft --seed ../testdata/HOG113/alnscratch/41_inter.fasta ../testdata/HOG113/alnscratch/40_inter.fasta > ../testdata/HOG113/alnscratch/43_inter.fasta\n",
      "mafft --seed ../testdata/HOG113/alnscratch/41_inter.3di.fasta ../testdata/HOG113/alnscratch/40_inter3di.fasta > ../testdata/HOG113/alnscratch/43_inter3di.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 3\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 900.000000\n",
      "\n",
      "adding 0-1\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 5\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "    0 / 5\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP     4 / 4 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "    0 / 5\n",
      "    0 / 2\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP     4 /4 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "    0 / 5\n",
      "Segment   1/  1    1- 180\n",
      "STEP 002-003-1  identical.   \n",
      "Converged.\n",
      "\n",
      "done\n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 3\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 900.000000\n",
      "\n",
      "adding 0-1\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 5\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "    0 / 5\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP     4 / 4 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "    0 / 5\n",
      "    0 / 2\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP     4 /4 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "    0 / 5\n",
      "Segment   1/  1    1- 182\n",
      "done 002-001-1  accepted..   \n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 5\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 2500.000000\n",
      "\n",
      "adding 1-2\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 8\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "    0 / 8\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP     7 / 7 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "    0 / 8\n",
      "    0 / 3\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP     7 /7 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "    0 / 8\n",
      "Segment   1/  1    1- 182\n",
      "done 002-001-1  identical.   \n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --seed ../testdata/HOG113/alnscratch/44_inter.fasta ../testdata/HOG113/alnscratch/43_inter.fasta > ../testdata/HOG113/alnscratch/45_inter.fasta\n",
      "mafft --seed ../testdata/HOG113/alnscratch/44_inter3di.fasta ../testdata/HOG113/alnscratch/43_inter3di.fasta > ../testdata/HOG113/alnscratch/45_inter3di.fasta\n",
      "mafft --addfull  ../testdata/HOG113/alnscratch/Q2YDE7_inter.fasta --keeplength ../testdata/HOG113/alnscratch/45_inter.fasta > ../testdata/HOG113/alnscratch/46_inter.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 5\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 2500.000000\n",
      "\n",
      "adding 1-2\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 8\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "    0 / 8\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP     7 / 7 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "    0 / 8\n",
      "    0 / 3\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP     7 /7 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "    0 / 8\n",
      "Segment   1/  1    1- 190\n",
      "done 002-001-1  identical.   \n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    7 / 8\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 8\n",
      "njobc = 9\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    7 / 8\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 8\n",
      "njobc = 9\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --addfull  ../testdata/HOG113/alnscratch/Q2YDE7_inter.3di.fasta --keeplength ../testdata/HOG113/alnscratch/45_inter3di.fasta > ../testdata/HOG113/alnscratch/46_inter3di.fasta\n",
      "mafft --seed ../testdata/HOG113/alnscratch/47_inter.fasta ../testdata/HOG113/alnscratch/46_inter.fasta > ../testdata/HOG113/alnscratch/48_inter.fasta\n",
      "mafft --seed ../testdata/HOG113/alnscratch/47_inter.3di.fasta ../testdata/HOG113/alnscratch/46_inter3di.fasta > ../testdata/HOG113/alnscratch/48_inter3di.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "nhomologs = 9\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 8100.000000\n",
      "\n",
      "adding 0-1\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 11\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "    0 / 11\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP    10 / 10 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "    0 / 11\n",
      "    0 / 2\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    10 /10 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "    0 / 11\n",
      "Segment   1/  1    1- 182\n",
      "STEP 002-001-1  identical.   \n",
      "Converged.\n",
      "\n",
      "done\n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 9\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 8100.000000\n",
      "\n",
      "adding 0-1\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 11\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "    0 / 11\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP    10 / 10 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "    0 / 11\n",
      "    0 / 2\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    10 /10 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "    0 / 11\n",
      "Segment   1/  1    1- 190\n",
      "done 002-001-1  identical.   \n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 11\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 12100.000000\n",
      "\n",
      "adding 0-1\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 13\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "   10 / 13\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP    12 / 12 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "   10 / 13\n",
      "    0 / 2\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    12 /12 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   10 / 13\n",
      "Segment   1/  1    1- 182\n",
      "STEP 002-006-0  identical.   "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --seed ../testdata/HOG113/alnscratch/49_inter.fasta ../testdata/HOG113/alnscratch/48_inter.fasta > ../testdata/HOG113/alnscratch/50_inter.fasta\n",
      "mafft --seed ../testdata/HOG113/alnscratch/49_inter.3di.fasta ../testdata/HOG113/alnscratch/48_inter3di.fasta > ../testdata/HOG113/alnscratch/50_inter3di.fasta\n",
      "mafft --addfull  ../testdata/HOG113/alnscratch/F6ZCC9_inter.fasta --keeplength ../testdata/HOG113/alnscratch/50_inter.fasta > ../testdata/HOG113/alnscratch/52_inter.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "done 002-001-1  identical.   \n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 11\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 12100.000000\n",
      "\n",
      "adding 0-1\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 13\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "   10 / 13\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP    12 / 12 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "   10 / 13\n",
      "    0 / 2\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    12 /12 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   10 / 13\n",
      "Segment   1/  1    1- 184\n",
      "done 002-001-1  identical.   \n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "   12 / 13\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 13\n",
      "njobc = 14\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "To keep the alignment length, 7 letters were DELETED.\n",
      "To know the positions of deleted letters, rerun the same command with the --mapout option.\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --addfull  ../testdata/HOG113/alnscratch/F6ZCC9_inter.3di.fasta --keeplength ../testdata/HOG113/alnscratch/50_inter3di.fasta > ../testdata/HOG113/alnscratch/52_inter3di.fasta\n",
      "mafft --seed ../testdata/HOG113/alnscratch/53_inter.fasta ../testdata/HOG113/alnscratch/52_inter.fasta > ../testdata/HOG113/alnscratch/56_inter.fasta\n",
      "mafft --seed ../testdata/HOG113/alnscratch/53_inter3di.fasta ../testdata/HOG113/alnscratch/52_inter3di.fasta > ../testdata/HOG113/alnscratch/56_inter3di.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "   12 / 13\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 13\n",
      "njobc = 14\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "To keep the alignment length, 7 letters were DELETED.\n",
      "To know the positions of deleted letters, rerun the same command with the --mapout option.\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 14\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 19600.000000\n",
      "\n",
      "adding 1-2\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 17\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "   10 / 17\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP    16 / 16 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "   10 / 17\n",
      "    0 / 3\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    16 /16 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   10 / 17\n",
      "Segment   1/  1    1- 182\n",
      "done 002-001-1  identical.   \n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 14\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 19600.000000\n",
      "\n",
      "adding 1-2\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 17\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "   10 / 17\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP    16 / 16 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "   10 / 17\n",
      "    0 / 3\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    16 /16 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   10 / 17\n",
      "Segment   1/  1    1- 190\n",
      "done 002-001-1  identical.    identical.   \n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "   16 / 17\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 17\n",
      "njobc = 18\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "To keep the alignment length, 5 letters were DELETED.\n",
      "To know the positions of deleted letters, rerun the same command with the --mapout option.\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "   16 / 17\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 17\n",
      "njobc = 18\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "To keep the alignment length, 2 letters were DELETED.\n",
      "To know the positions of deleted letters, rerun the same command with the --mapout option.\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 18\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 32400.000000\n",
      "\n",
      "adding 4-5\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 24\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "   20 / 24\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP    23 / 23 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "   20 / 24\n",
      "    0 / 6\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP     5 /23 "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --addfull  ../testdata/HOG113/alnscratch/F7AW07_inter.fasta --keeplength ../testdata/HOG113/alnscratch/56_inter.fasta > ../testdata/HOG113/alnscratch/59_inter.fasta\n",
      "mafft --addfull  ../testdata/HOG113/alnscratch/F7AW07_inter.3di.fasta --keeplength ../testdata/HOG113/alnscratch/56_inter3di.fasta > ../testdata/HOG113/alnscratch/59_inter3di.fasta\n",
      "mafft --seed ../testdata/HOG113/alnscratch/60_inter.fasta ../testdata/HOG113/alnscratch/59_inter.fasta > ../testdata/HOG113/alnscratch/61_inter.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "STEP    23 /23 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   20 / 24\n",
      "Segment   1/  1    1- 186\n",
      "done 002-001-1  identical.   \n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 18\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 32400.000000\n",
      "\n",
      "adding 4-5\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 24\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "   20 / 24\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP    23 / 23 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "   20 / 24\n",
      "    0 / 6\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    23 /23 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   20 / 24\n",
      "Segment   1/  1    1- 195\n",
      "done 002-001-1  identical.   \n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --seed ../testdata/HOG113/alnscratch/60_inter3di.fasta ../testdata/HOG113/alnscratch/59_inter3di.fasta > ../testdata/HOG113/alnscratch/61_inter3di.fasta\n",
      "mafft --seed ../testdata/HOG113/alnscratch/62_inter.fasta ../testdata/HOG113/alnscratch/61_inter.fasta > ../testdata/HOG113/alnscratch/67_inter.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 24\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 57600.000000\n",
      "\n",
      "adding 0-1\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 26\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "   20 / 26\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP    25 / 25 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "   20 / 26\n",
      "    0 / 2\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    25 /25 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   20 / 26\n",
      "Segment   1/  1    1- 189\n",
      "STEP 002-002-1  identical.   \n",
      "Converged.\n",
      "\n",
      "done\n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 24\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 57600.000000\n",
      "\n",
      "adding 0-1\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 26\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "   20 / 26\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP    25 / 25 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "   20 / 26\n",
      "    0 / 2\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    25 /25 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   20 / 26\n",
      "Segment   1/  1    1- 201\n",
      "done 002-001-1  identical.   \n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --seed ../testdata/HOG113/alnscratch/62_inter.3di.fasta ../testdata/HOG113/alnscratch/61_inter3di.fasta > ../testdata/HOG113/alnscratch/67_inter3di.fasta\n",
      "mafft --seed ../testdata/HOG113/alnscratch/68_inter.fasta ../testdata/HOG113/alnscratch/67_inter.fasta > ../testdata/HOG113/alnscratch/71_inter.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 26\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 67600.000000\n",
      "\n",
      "adding 1-2\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 29\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "   20 / 29\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP    28 / 28 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "   20 / 29\n",
      "    0 / 3\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    28 /28 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   20 / 29\n",
      "Segment   1/  1    1- 195\n",
      "STEP 002-002-0  identical.   \n",
      "Converged.\n",
      "\n",
      "done\n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 26\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 67600.000000\n",
      "\n",
      "adding 1-2\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 29\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "   20 / 29\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP    28 / 28 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "   20 / 29\n",
      "    0 / 3\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    28 /28 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   20 / 29\n",
      "Segment   1/  1    1- 208\n",
      "STEP 002-013-1  rejected..   "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --seed ../testdata/HOG113/alnscratch/68_inter3di.fasta ../testdata/HOG113/alnscratch/67_inter3di.fasta > ../testdata/HOG113/alnscratch/71_inter3di.fasta\n",
      "mafft --seed ../testdata/HOG113/alnscratch/72_inter.fasta ../testdata/HOG113/alnscratch/71_inter.fasta > ../testdata/HOG113/alnscratch/74_inter.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "done 002-001-1  identical.   \n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 29\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 84100.000000\n",
      "\n",
      "adding 0-1\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 31\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "   20 / 31\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP    30 / 30 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "   20 / 31\n",
      "    0 / 2\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    30 /30 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   20 / 31\n",
      "Segment   1/  1    1- 194\n",
      "done 002-001-1  identical.   \n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --seed ../testdata/HOG113/alnscratch/72_inter.3di.fasta ../testdata/HOG113/alnscratch/71_inter3di.fasta > ../testdata/HOG113/alnscratch/74_inter3di.fasta\n",
      "mafft --seed ../testdata/HOG113/alnscratch/75_inter.fasta ../testdata/HOG113/alnscratch/74_inter.fasta > ../testdata/HOG113/alnscratch/76_inter.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 29\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 84100.000000\n",
      "\n",
      "adding 0-1\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 31\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "   20 / 31\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP    30 / 30 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "   20 / 31\n",
      "    0 / 2\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    30 /30 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   20 / 31\n",
      "Segment   1/  1    1- 215\n",
      "done 002-001-1  identical.   \n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 31\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 96100.000000\n",
      "\n",
      "adding 6-7\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 39\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "   30 / 39\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP    38 / 38 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "   30 / 39\n",
      "    0 / 8\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    38 /38 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   30 / 39\n",
      "Segment   1/  1    1- 198\n",
      "done 002-001-1  identical.    identical.   \n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 31\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 96100.000000\n",
      "\n",
      "adding 6-7\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "nalphabets = 256\n",
      "nused=\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 39\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "   30 / 39\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP    13 / 38 "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --textmatrix ../mafftmat/3diHEXmat.txt --seed ../testdata/HOG113/alnscratch/75_inter3di.fasta ../testdata/HOG113/alnscratch/74_inter3di.fasta > ../testdata/HOG113/alnscratch/76_inter3di.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "STEP    38 / 38  h\n",
      "done.\n",
      "\n",
      "disttbfast (text) Version 7.520\n",
      "alg=A, model=Extended, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "nalphabets = 256\n",
      "nused=\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "   30 / 39\n",
      "    0 / 8\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    38 /38 \n",
      "done.\n",
      "tbfast (text) Version 7.520\n",
      "alg=A, model=Extended, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nalphabets = 256\n",
      "nused=\n",
      "dndpre (text) Version 7.520\n",
      "alg=X, model=Extended, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum -2 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "nalphabets = 256\n",
      "nused=\n",
      "\n",
      "   30 / 39\n",
      "Segment   1/  1    1- 197\n",
      "STEP 002-013-0  rejected..    identical.    identical.    identical.    identical.    identical.   "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "final aln\n",
      "childalnsAA {<toytree.TreeNode.TreeNode object at 0x7f905d25e410>: {'fasta': '../testdata/HOG113/alnscratch/75_inter.fasta'}, <toytree.TreeNode.TreeNode object at 0x7f905b1b3cd0>: {'fasta': '../testdata/HOG113/alnscratch/74_inter.fasta'}}\n",
      "childalns3di {<toytree.TreeNode.TreeNode object at 0x7f905d25e410>: {'fasta': '../testdata/HOG113/alnscratch/75_inter3di.fasta'}, <toytree.TreeNode.TreeNode object at 0x7f905b1b3cd0>: {'fasta': '../testdata/HOG113/alnscratch/74_inter3di.fasta'}}\n",
      "finalaln ../testdata/HOG113/alnscratch/76_inter.fasta\n",
      ">_seed__seed_O67888\n",
      "----------------------------------------LDYEPVYDITADAGIRVRAK\n",
      "TL-EELFCHAILATFNEITDIDKVEP------KEEYEIQAQND-MPFLLADIINEALVLH\n",
      "ESK-HFVASECEVLELKEDF----VKVKLKGEKFDPKRHPSKLVIKAATYHRLRVEKKNE\n",
      "H--W-EA-EVIFDI\n",
      ">_seed__seed_B5YH50\n",
      "----------------------------------------MKYK-VIDVAGDVGIRAEGA\n",
      "SL-EECFINSAFGLYSLITDLAQIEP------AEEIEIIINEDNLENMLVSFLNELIFQF\n",
      "DTY-GFLGKAISI-EIKDNY----LTARIKGEKFNPEKHERKLLVKAATYHNLVLKQEDS\n",
      "F--W-IA-EIIFDI\n",
      ">_seed__seed_A9A300\n",
      "----------------------------------------MSYK-FVDHATDAIIEVTAK\n",
      "DL-QEAFSVTADAVINLTLDQDKVEE------KEQRKFVAEGKDLRYLLFSWLEEIPFLL\n",
      "ITE-GFAIKRIEF-DITQNQ----INATAYGEPLDFKKHNFKVEIKAPTFYDMEIKQNGG\n",
      "V--F-MR-FLL-DL\n",
      ">_seed__seed_Q5JCY8\n",
      "-------------------------------------------E-HYEHTADIGVRGYGS\n",
      "TL-EEAFEAVALGLFDVMVNVKKVEP------KECREVEVEEEDLEALLYSFLEELLVLH\n",
      "DME-GLVFGDVKV-RIEKTENGYKLKAKACGEVLNPEKHEPKEEVKAITYHDMKIEKLPD\n",
      "GR-W-MA-QFVPDL\n",
      ">_seed__seed_Q60334\n",
      "-------------------------------------------N-YFETTADLGVEAKGK\n",
      "SL-EEAFKEGAKGLYNIMVDIDKVDK------KEKIEFEITGEDLEELLYNFLNELLFYT\n",
      "DVE-NLVFNDFDV-KIEKNDNGYRLKCTAYGEKINKEKHNIKEEVKAVTYHKMEIKQEED\n",
      "G--W-KI-RYIVDL\n",
      ">_seed_B1L703\n",
      "-------------------------------------------R-FLELEADVGFEVWAE\n",
      "DL-NSLFEEAALSMYEIMVDVEKVER------SIERHISVSAPDIEMLLHHWLSELLFIT\n",
      "DVE-RIVFSDFEV-KISGND----LEGRAWGEFIDVERHNPKTEIKAVTYHKLSVREEGG\n",
      "V--W-RC-TVVLDI\n",
      ">_seed_Q9UX79\n",
      "-------------------------------------------E-FFEHTADVGIKSYGR\n",
      "SL-EEAFSNAALGVFEVITDTSKVRP------VEYREIYLNGYDLENLLYKWIEELLYYY\n",
      "DSE-LMIFSKFDL-MIDQDS--ITLEGKAWGERFNDKIHERRTVVKAMTYHQLSIEKTES\n",
      "G--Y-VI-TFVVDI\n",
      ">_seed_Q74CJ3\n",
      "----------------------------------------MPYRYLPDITADVAFEAWGE\n",
      "TR-EEMFCAAADALTNVMVDLASVTP------TEEIVISLANEELDLLLFSFLQELIFLK\n",
      "DAR-CFLLRVPRI-MITETEEVLHLDAIARGERIDNERHPMMVDVKAVTLHLFSVWRQEN\n",
      "D--W-WA-RVVLDI\n",
      ">_seed_B8E097\n",
      "----------------------------------------KNYE-LLDHTADIGIIVYGE\n",
      "TK-EKAFESAAEAMFDLMCPLEKIQE------KECFDIEVDGEDLESLLVTWLNELLYVF\n",
      "EVQ-KLLFRRFDVTLIGHNQ----LISHCYGERFDPKKHEITREIKAVTYNLLKIEQRED\n",
      "K--W-MI-QVVFDI\n",
      ">_seed_Q8TU10\n",
      "----------------------------------------KKYE-YLEHTADIKFLAYGN\n",
      "TL-KEVFENAALAMFNVIIDTGKVSG------ETAREVCLTSPDLESLLVDWLSELLYLF\n",
      "EVD-EIVFWKFRVEEIREEEGEYSIKAVASGEQYYPESHPFETEIKAVTYNQLELEKTAD\n",
      "G--W-KA-QVVVDI\n",
      ">_seed_B8BWG5\n",
      "-------------------------------------------E-YLDHPADIILHSWGN\n",
      "DL-SASLSNLAVCMFGYITSINLIATNPSQCLEYGSNIVAQGHDLRSLVYSFLDEWLFNF\n",
      "HDT-GFVAKEILVTEFNVEA--WRIVSTGKGEVMDVRRHPQGTEVKAITYNAMRIEEKGK\n",
      "DDRRCDV-YVVVDI\n",
      ">_seed_Q9TZN1\n",
      "-------------------------------------------E-YLDHPADIQLHSWGS\n",
      "TM-EEAFEACLVSMFGYMTDLAKVDE------MYEFYWKASGDSLDGLLFQFLDEALNSF\n",
      "HAEPCFVAKRVEILRFDKKK--FEIEFRGWGESFDTSKHETEADIKSPTYSNMQINEKPE\n",
      "RCD---I-YVIVDI\n",
      ">_seed_A2F1C4\n",
      "-------------------------------------------E-FIDHPADIQLHAWGP\n",
      "QP-DVTLEQLGKAFFEVMFETDNFKE------ETTHSISVKGKDAQNLLYNFLDEWLYVF\n",
      "DAE-DFVATRIKVTKCDFVN--LEIEATGYGELFDMKKHARRTEVKAITYASMKVDTTPG\n",
      "ASE---F-YVILDL\n",
      ">_seed_Q54BU9\n",
      "----------------------------------------------LDHTADIMFHTWGK\n",
      "DL-KEALEQMVLIMNNYMVELDSVELDDS---ATEQTISVNGHDMDSLLFALLDEFLFVF\n",
      "STE-FIIFKQVQIISFDREN--FSIKAIGKGVELDKSKHTTGTEIKAITYSCMKIEENPD\n",
      "KSD---I-HVIVDI\n",
      ">_seed_A9UTI8\n",
      "----------------------------------------------LDHTADIQVHAWGQ\n",
      "TL-KDSFEQTVVAMFGYMTELDAIEIDD----SQTSEFSVEGHDLMSLLYALLDEFLFRF\n",
      "SAEDNLVCREVEIVEFDTDN--FKITVRGKGEVFDLAKHPQGTEIKAITYSNMQIHTERA\n",
      "THD---V-YVILDI\n",
      ">_seed__seed_Q9VD92\n",
      "---------------VEFSRENFLLPE-------------MKYE-YLDHTADVQIHGWGS\n",
      "SL-KEAFEQCGVAMFGYMTELDYVSV------EQCFEIEAHGDDLESLLFHFLDELLFLF\n",
      "SAEPYLVCKKLEITKFDVEN--FEISCHCYGEPFELGKHPQGTEVKAITYSAMQIIQDVE\n",
      "ASNY-EV-FVIIDI\n",
      ">_seed__seed_D6WKY5\n",
      "---------------ETFTKEELTLPP-------------CKYE-YLDHTADVQLHAWGD\n",
      "SL-KEAYEQCGLAMFGYMTEINTVEI------KQCSEIEATGDDLDGLLFHFLDELLFLF\n",
      "SCEPFLICSKLVITEFRTEGDELKIKCKCYGEEFTLGKHPQGTEVKAITYSAMQIINDPE\n",
      "NKKF-EV-FVIIDI\n",
      ">_seed__seed_Q7PYJ8\n",
      "---------------VPLETSDHVVPE-------------VKYE-YLDHTADVQLHAWGE\n",
      "TL-QEAFEQCGMAMFGYMTELPKVEI------QKCYQIVTDTDDLENLLYRFLDELLFLF\n",
      "SAEPFLICKKLEITRFDLEQ--FTIECNCYGEEFDLAKHPQGTEVKAITYSAMQIYQKPE\n",
      "HEKN-EV-FVIIDI\n",
      ">_seed_B7PHX9\n",
      "------------------------------------------------------------\n",
      "-----------------MTEIDKVDI------RMTQDIDAEADDMVGLLFHFLDELLFIF\n",
      "SAEPFFIGRKVKILEFDKEA--FRIKARVYGEVFDLGKHPQGSEVKAITYSNMQVWDNPD\n",
      "QHE---V-FVIIDI\n",
      ">_seed_T1ELH7\n",
      "------------------------------------------------------------\n",
      "-----------------MTDISTVEQ------CESQTIEAEGDDMQSLLFHFLDEFLFLF\n",
      "SAEPFFIARRVEIKTFDREN--FKIEAVGHGEPFDIAKHPQGTEVKAITYSNMQVYDEEN\n",
      "KHE---V-FVIIDI\n",
      ">_seed_C3Y972\n",
      "------------------------------------------------------------\n",
      "-----------------MTEIETVEP------LDTYETEAQADDMENLLYQFLDEWLFAF\n",
      "SAEPFFIARDVKIVEFDREN--FRIKSVGYGETFDLSKHPQGTEVKAITYSAMQIHEEVD\n",
      "TND---V-YVVLDI\n",
      ">_seed_W5M2J6\n",
      "------------M-----NDRELDLTEAQKALKAKYPSINKKYE-YLDHTADVQIHSWGD\n",
      "TL-EEAFEQCAMGMFGYMTDTETVEP------IDTVEVESEGEDLLSLLFHFLDDWLFKF\n",
      "SADQFFVPREVKVLHIDRMR--FRMRSIGWGEEFRISKHPQGTEVKAITYSAMQIHEENK\n",
      "PE----I-FVIIDI\n",
      ">_seed_Q566V0\n",
      "------------M-----NDRELDLTEEQKALKAKYPPITKKYE-YLDHTADVQIHSWGN\n",
      "SL-EEAFEQCAMGMFGYMTDTETVEP------IDTVEVESEGDDMESLLYHFLDDWLFKF\n",
      "SADIFFIPREVKVLHIDRMR--YKIRSIGWGEEFSINKHPQGTEVKAITYSAMQIHETEQ\n",
      "PE----I-FVIIDI\n",
      ">_seed_H2MRT4\n",
      "------------M-----DDRELDLTNEQKAVKSKYPPINKKYE-YLDHTADVQIHSWGD\n",
      "SL-EEAFEQCAMGMFGYMTDTETVEP------LDTVEVESEGEDMESLLFHFLDDWLYKF\n",
      "SADLFFVPREIKVLHIDRMH--YKIRSIGWGEEFSISKHPQGTEVKAITYSAMQIHDKEK\n",
      "PE----I-FTIIDI\n",
      ">_seed_A0A5F8GYJ2\n",
      "------------MAQDGKDKRDYNLNEEQKIIK------------YLDHTADIQLHAWGD\n",
      "PL-EEAFEQCALAMFAYMTDTGTVEL------LQTVEVEAEGDDLQSLPFHFLDEWLYKL\n",
      "IANEFFIPQEVKVLSIDQRN--FKLRSIGN------------------------------\n",
      "--------------\n",
      ">_seed_F7AIV7\n",
      "------------MAQDGEDERDYNLNEEQKAIKAKYPPAIRKYE-YLDHTADVQLHAWGD\n",
      "TL-EEAFEQCAMAMFGYMTDTGTVEP------LQTVEVQAEGDDLQSLLFHFLDEWLYKF\n",
      "SANEFFIPREVKVLSIDQRN--FKLRSIGW------------------------------\n",
      "--------------\n",
      ">_seed_A0A2I3S880\n",
      "--TGSR-TEGAVMAQKEGDVRDYNLTEEQKAIKANYPPVNRNYE-YLDHTADVQLHTWGD\n",
      "TL-EEAFEQCAMAISGYMIDTRTVEP------LQTIEVETQ--EVSTFLFHFLDEWLYKS\n",
      "NADEFFIPWEVKVICIDQRH--FQLQSIGWTEEFSLSNHPQGTEVKVITYSAMQVYNKEN\n",
      "PE----V-FVIIDI\n",
      ">_seed_A0A2I2YKY3\n",
      "---GSGCSDKAVMAQKEGDVRDYNLTEEQKAIKANYPPVNRNYE-YLDHTADVQLHTWGD\n",
      "TL-EEAVEQCAMAIFGYMIDNRTVEP------LQTIEVETQ--EVSTFLFHFLDEWLYKS\n",
      "DADEFFIPWEVKVLCIDQRH--FQLQSIGWREEFSLSKHPQGTEVKVITYSAMQVYNKEN\n",
      "PE----V-FVIIDI\n",
      ">_seed_G3QX27\n",
      "IKGGSRVSNPAVMAQEEEDVRDYNLTEEQKAIKAKYPPVNRKYE-YLDHTADVQLHAWGD\n",
      "TL-EEAFEQCAMAMFGYMTDTGTVEP------LQTVEVETQGDDLQSLLFHFLDEWLYKF\n",
      "SADEFFIPREVKVLSIDQRN--FKLRSIGWGEEFSLSKHPQGTEVKAITYSAMQVYNEEK\n",
      "PE----V-FVIIDI\n",
      ">_seed_K7BTY5\n",
      "MKGGSRVSNPAVMAQEEEDVRDYNLTEEQKAIKAKYPPVNRKYE-YLDHTADVQLHAWGD\n",
      "TL-EEAFEQCAMAMFGYMTDTGTVEP------LQTVEVETQGDDLQSLLFHFLDEWLYKF\n",
      "SADEFFIPREVKVLSIDQRN--FKLRSIGWGEEFSLSKHPQGTEVKAITYSAMQVYNEEK\n",
      "PE----V-FVIIDI\n",
      ">_seed_Q8IWT0\n",
      "------------MAQEEEDVRDYNLTEEQKAIKAKYPPVNRKYE-YLDHTADVQLHAWGD\n",
      "TL-EEAFEQCAMAMFGYMTDTGTVEP------LQTVEVETQGDDLQSLLFHFLDEWLYKF\n",
      "SADEFFIPREVKVLSIDQRN--FKLRSIGWGEEFSLSKHPQGTEVKAITYSAMQVYNEEN\n",
      "PE----V-FVIIDI\n",
      ">_seed_M0R5Y9\n",
      "------------MAQEEEDVRDYNLTEEQKAIKNKYPPVNRKYE-YLDHTADVQLHAWGD\n",
      "TL-EEAFEQCAMAMFGYMTDTGTVEP------LQTVEVETQGDDLQSLLFHFLDEWLYRF\n",
      "SADEYFIPREVKVLNIDQKN--FKLRSIGWGEEFSLSKHPQGTEVKAITYSAMQVYNEEK\n",
      "PE----V-FVIIDI\n",
      ">_seed_Q505B7\n",
      "------------MAQEEEDVRDYNLTEEQKATKDKYPPVNRKYE-YLDHTADVQLHAWGD\n",
      "TL-EEAFEQCAMAMFGYMTDTGTVEP------LRTVEVETQGDDLQSLLFHFLDEWLYKF\n",
      "SADEYFIPREVKVLNIDQKN--FKLRSIGWGEEFSLSKHPQGTEVKAITYSAMQVYNEEK\n",
      "PE----VFFVIIDI\n",
      ">A0A2I3RJJ2\n",
      "--CMSRHTQPKIMAQAEEDVRDYNLTEEQKVIKAKYPPVNRKYE-YLDHTADVQLHAWGD\n",
      "TL-EEAFEQCAMAMFGYMTDTGTVEP------LPTVEVETQGDDLQSLLFHFLDEWLYKF\n",
      "SADEFFIPQEVKVL-----------------KESSLSKYPQGTEVKAITYSAMQVYNEEK\n",
      "PE----V-FVIIDI\n",
      ">A0A2I2ZWB6\n",
      "--IFGRDRVSPFMAQAEEDVRDYNLTEEQKAIKAKYPPVNRKYE-YLDHTADVQLHAWGD\n",
      "TLEEEAFEQCAMAMFGYMTDTGTVEP------LPTVEVETQGDDLQSLLFHFLDEWLYKL\n",
      "SADEFFIPREVKVL-----------------KEFSLSKHPQGTEVKAITYSAMQVYNEEK\n",
      "PE----V-FVIIGI\n",
      ">A0A2I2Y839\n",
      "------------MAQAEEDVRDYNLTEEQKAIKAKYPPVNRKYE-YLDHTADVQLHAWGD\n",
      "TL-EEAFEQCAMAMFGYMTDTGTVEP------LPTVEVETQGDDLQSLLFHFLDEWLYKL\n",
      "SADEFFIPREVKVL-------------------IFIVQAPSG--------------NRSQ\n",
      "GNN---I-FSNAGL\n",
      ">Q2YDE7\n",
      "------------MALEGEDVRDYNLTEEQKAIKAKYPPVSRKYE-YLDHTADVQLHAWGD\n",
      "TL-EEAFEQCAMAMFGYMTDTGTVEP------LQTIEVETQGDDLQSLLFHFLDEWLYKF\n",
      "SADEFFIPREVKVLNIDQRN--FKIRSIGWGEEFSLSKHPQGTEVKAITYSAMQVYNEEK\n",
      "PE----V-FVIIDI\n",
      ">F6ZCC9\n",
      "------------M-----DERYYTISEEQVAVKAKYPPVNKNYE-YLDHTADVQLHAWGE\n",
      "TL-EEAFEQCAMAMFGYMTDIETVEP------LDTVEVVTEGEDLISLLFNFLDEWLYKF\n",
      "SADQYFVPREVKVLNIDRMN--FKIRSIGWGEEFSLTKHPQGTEVKAITYSAMQIHEEEK\n",
      "AE----V-FVIIDI\n",
      ">F7AW07\n",
      "------------------------MEEKLPAYESGYHNVIKKFE-YLDHTADVQIHSWGD\n",
      "NI-EETFEQAAMAMFNYMTDIDGVRP------VKTVEITAEGTDYMSLLFHFMDEFLFLS\n",
      "SCDDFFMAREIKITNFDREN--WRMEALGYGEDFNQERHGKGTEVKAITYSAMQVIESEK\n",
      "SE----I-FVIVDI\n",
      "\n",
      "finalaln3di ../testdata/HOG113/alnscratch/76_inter3di.fasta\n",
      ">_seed__seed_O67888\n",
      "---------------------------------------DDDDDDDDPDDKDKDKDKDLD\n",
      "PQ-VRLFVLLLVVLVQFWPLVLWDFP------DKDKDKFFAA-PPVRSQVRSQVSVVCCV\n",
      "VQV-WGFN--DKDWPDDDRG----MTMMMTGTDHDDVVRIPGHDDFNGWDPVPWDWDDDP\n",
      "---SMIMTMTMTD-D\n",
      ">_seed__seed_B5YH50\n",
      "---------------------------------------DDDD-DDDDDAKDKDKAKDLD\n",
      "LQ-NRLVRVLVVLVVQFWDLVLKDQP------DKDKDKDFDQDSLVVSQVVNQVQLVCCV\n",
      "PVQ-WGFN--DWDW-DDDRR----MIIIMGGTDHDDVVRIDTHDNWNGWDPVPWDWDDDD\n",
      "---SMIMTMIMTG-D\n",
      ">_seed__seed_A9A300\n",
      "---------------------------------------DDDD-DDDDFPKDKDKDWFQD\n",
      "PQ-VRLFVVLVVLLVQFWALVLWDFP------DKDKDKAWDQDPLSSVQSSSVVSLCCCP\n",
      "VVQ-WRFN--DWDW-DDDSI----MIIIMTGTGGHDCVSGVGDDRFDGWDPVPWDWDDD-\n",
      "---STIMTITMTT-D\n",
      ">_seed__seed_Q5JCY8\n",
      "---------------------------------------DD---DDDPPW-DKQKAKDLD\n",
      "PQ-RRLFRSLLSVLVQFWQLVLWDQP------DKDKDKDDDQEDLVVSQRVSQVVLVCCV\n",
      "VVQ-WDFNGKGKDW-DADPRHI--MIIMMTGGDHDDCVGTVTHDDFDGWDPAPFDWDQDP\n",
      "V--SMIMTMGITH-D\n",
      ">_seed__seed_Q60334\n",
      "---------------------------------------DW---DDDPAF-IKIKAFDQE\n",
      "PQ-RRLFRVLQSVLVQFWQLVLWAQP------DKDKDKDFDQEPLRRSQRVNVVQVVCCV\n",
      "PVQ-WDFNGKDWDW-DADPRHI--MIIIITGGDHDDCVGTVGHDDFPGWDSVPWDWDADP\n",
      "---NHIMTMTTTH-D\n",
      ">_seed_B1L703\n",
      "---------------------------------------D----DDDPPWKDKDKDKDLD\n",
      "DF-VQLFVVLQVVLVQFWQLVQWDQP------DKDKDKDFDPDDLVVSQVSNQVQLVCCV\n",
      "VVQ-WDFN--GKGW-DDDPG----MIIMMTGGGHDDCVRTVTHDDFPGWDPVQFDWDDDP\n",
      "---SMIMTMTMTG-D\n",
      ">_seed_Q9UX79\n",
      "---------------------------------------DW---DDDPQW-IKIKGFDLE\n",
      "PQ-SRLFSSLQRLLVQFWDLVLWDQP------DKDKDKDFDPALLRLSQVVNQVQVVCCV\n",
      "PVV-WGFN--DKDW-DADRVRR--MIIMITGTDHDDVVRIPGHDDFPGFDPAPFDWDQDP\n",
      "---RYIMTIGITD-D\n",
      ">_seed_Q74CJ3\n",
      "---------------------------------------DPDD-DDVPPFKDKDKFKDLD\n",
      "LW-RRLLRVLLVLLVLFFDPVLFDQP------DKDKDKDKDLDSLVVSQVVSVVQLCCCP\n",
      "VVQ-WDWHWPTWDW-DDDPSMI--MIITIIGGDHDDCVSTVTPDRFDHKDPPQWDWDDDD\n",
      "---SMIMTMTMTG-D\n",
      ">_seed_B8E097\n",
      "---------------------------------------DPDD-DDDDPQKDKDKAKDLD\n",
      "PQ-RRLFVVLQVVLCLFWNLVLWDQP------DKDKDKFWAQAPLRQSQVVSVVVLVCCV\n",
      "PVQ-WRFN--GKGWPDDDGT----MTIIITGTDHDDVVRIDGPDDFDGKDPVPWDWDDDD\n",
      "---SMIMTMTMTR-D\n",
      ">_seed_Q8TU10\n",
      "----------------------------------------DQKDWDDDDFWIKMKGDDQE\n",
      "PQ-SRLFSLLQVVLVLFWDLVLFDFF------PKDKDKDKDLDSLRVSQVSNQVVLVCCV\n",
      "VVQ-WRFN--GKGWPDWDQDPNMIMTIIMTGTDHHDCVSIPTNDDFPGWDPVPWDWDADP\n",
      "---NHIMTMTITG-D\n",
      ">_seed_B8BWG5\n",
      "-------------------------------------------DDDDDPLKDKQKAKDLA\n",
      "PQ-RRLQVSLLSVLVQFWPLVLWDFDPVCQVVLFFFQKFADDADLNLSQRSNVSSLVSCV\n",
      "VQQ-WDFS--TKGFPDDDGVRR--MTTMGGTTGRDDCVRIPGGDDFDGWDPPPWDWDQDD\n",
      "PVRSMIMTITMTG-D\n",
      ">_seed_Q9TZN1\n",
      "-------------------------------------------DWDDDPQWIKIKGKDQD\n",
      "PQ-RRLFSSLLVVLVQFWPLVLWDQP------DKDKDKAAAPALLRLSQRSNVRVLVCLV\n",
      "DPPNWRFS--GKGWPDQDRVRR--MTIIMGGTGHDDVVRIDGNDDFPGKDPVPWDWDDDN\n",
      "---RMTMTIIITH-D\n",
      ">_seed_A2F1C4\n",
      "-------------------------------------------DDDDDPQKDKDKFKFQD\n",
      "PQ-VRLFVRLLVLLLQFFPQVLWFQP------DKDKDKDFAPAPLRLSQVSNVVVLCCCV\n",
      "VPQ-WGAN--TKGWPDDDRPRR--MTMIMGGTDHDDCVRVGTHDDFDGKDPVPWDWDDDN\n",
      "---RMTMTMTMTG-D\n",
      ">_seed_Q54BU9\n",
      "---------------------------------------------DDDPQWIKIKGKDQF\n",
      "PQ-VRLQNVLRVLLCQFWNLVLFDLDPQK---DKDKDKDAAPAQLRLSQRSNVVSLVCCV\n",
      "PVV-WRFR--TKGFPDADRVRR--MTIMMTITHHDDPVRIDGHDDFPGWDPVPWDWDDDN\n",
      "---RMIMTMTITD-D\n",
      ">_seed_A9UTI8\n",
      "---------------------------------------------DDPWDKDKDKAWAQA\n",
      "LQ-RRLFVSLLVVLVQFWPLVLWDFDPV----AKDKDKDADDAPQRLSQRSNVVVLCCCP\n",
      "PNQVDDAS--GKDFPDQDRPRR--MTIIIGTTHHDDCVRIPGGDDWDGWDPVPWDFDRPD\n",
      "---STRMTMTMTT-D\n",
      ">_seed__seed_Q9VD92\n",
      "----------------------DDDDPVP------DDQDDQAWDWDDDPQWIKIKAKDLA\n",
      "QQ-RRLFNSLLVVLCQFWPLVLWDFD------TKDKDKAAAPAPLRLSQRSSVVSLCPLV\n",
      "DPPNWRAS--GKDFPDQDRVRS--MTMIIGTTGHDDPPRIDGGDDFDGWDPVPWDWDDPP\n",
      "V-VSMIMTITITR-D\n",
      ">_seed__seed_D6WKY5\n",
      "----------------------DDDDPVN------PDDDDDDDDDDDDDQKDKDKAWDLA\n",
      "LQ-RRLFRSLVSLLVQFWPLVLWDFP------DKDKDKDKDQAPLRRSQRSSVVSLCPLV\n",
      "DPPNWRAS--GKDWPDFAHDDRMTMTMIIGTTGRDDPPSTDGGDRWDGWDPVPWDFDGDN\n",
      "P-VSTTMTMTMTG-D\n",
      ">_seed__seed_Q7PYJ8\n",
      "----------------------DDDDPDD------PDQDDQAWDWDDDPQFIKIKGFDQD\n",
      "LQ-RQLFNSLVSVLVQFWPLVLWDFD------TKDKDKAFDDADLVLSQRSSVVSLCPLV\n",
      "DPPNWRAS--GKDFPDQDGVRS--MTIIIGTTDHDDCVRIPGGDDWDGWDPVPWDFDDPP\n",
      "V-DPTGMIMTTTD-D\n",
      ">_seed_B7PHX9\n",
      "------------------------------------------------------------\n",
      "----------------FWPLVLWDQP------DKDKDKDFDPAPLRLVQRVVVVVLCCLV\n",
      "DPPNWRFS--DKDWPDQDRVRS--MTMIMGGTGRDDPVRTPGGDRFDGWDPVPWDWDDDP\n",
      "---VDTDTDTDTH-D\n",
      ">_seed_T1ELH7\n",
      "------------------------------------------------------------\n",
      "----------------DWPLVLWAQP------DKDKDKAADPAPQRLVQRVVVVQLCPLV\n",
      "DPPNWDAS--DKDWPDADRVRR--MTMIMGGTGRDDPVRTPDDDRWDGWDPVPWDWDDDP\n",
      "---PDTDTDTDTH-D\n",
      ">_seed_C3Y972\n",
      "------------------------------------------------------------\n",
      "----------------FWPLVLWAQP------DKDKDKDFDDADLVLSQRSNVRSLCPLV\n",
      "DPPNWDAS--TKGWPDQDRVRS--MTMMMGGTGHDDCVRTPGGDRFDGWDPVPWDWDDDP\n",
      "---RMTMTITMTR-D\n",
      ">_seed_W5M2J6\n",
      "-----------------DPDPDDDDDPVRVVVCPVDDDDDDDDDDDDDFDKDKDKAWDQD\n",
      "QQ-RRLQVSLVSQLVQFWPLVLWDQP------DKDKDKAAAPAQQRLSQRSSVRSLCCLP\n",
      "PNQVWRFS--GKGWPDADRVRR--MTMIMGGGHHDDVVRTPGGDDWDGWDPVPWDWDDP-\n",
      "---GIIMTMTMTT-D\n",
      ">_seed_Q566V0\n",
      "-----------------DDQPDFDDDPLLVVQLVVDDHDDAQWDWDDDPQWIKIKGKDQD\n",
      "PQ-RRLQNSLVSQLSQFFQLSLFDQP------DKDKDKDADPAQLRLSQRSNVRVLCCLP\n",
      "PNQVWRFS--GKGWPDADRVRR--MTMIMGGGHHDDCVRTGGGDDWDGWDPVPWDWDPD-\n",
      "---SMIMTITITR-D\n",
      ">_seed_H2MRT4\n",
      "-----------------DDDPDDDDDPVRVVVCPVDDDDDDDDDDDDDFDKDKDKAWDQA\n",
      "PQ-RQLQVSLVSVLVQFWPLVLFDQP------DKDKDKDADPAQLRLSQRSNVRSLCCCP\n",
      "PNFVWRFS--GKGWPDADRVRR--MTMIMGGGHHDDVVRTPGGDDWDGWDPVPWDWDGD-\n",
      "---SMTMTMTMTT-D\n",
      ">_seed_A0A5F8GYJ2\n",
      "------------DDDPDPPPPDDPDDPV-----------PVPDDDDDDDDDDDDDADDPP\n",
      "PQ-RSLQRRLQVVLVVVDDLVPQAFD------DKDKDKFFDPAPVCLSVRRSVRSVCCCP\n",
      "VHPNDDAS--GKGWLDADRVRR--MTIIMGG-----------------------------\n",
      "---------------\n",
      ">_seed_F7AIV7\n",
      "------------PADQDPADPDDDDDPVRVVVCPVDDDDRDRDHDDDDFDKDKDKAWDQA\n",
      "PQ-RQLQVSLVSQLVQFWPLVLFDQP------DKDKDKAADPAPQRLSQRRSVVSLCVCP\n",
      "PPQVWRFS--GKGWPDAPRVRR--MTMIMGG-----------------------------\n",
      "---------------\n",
      ">_seed_A0A2I3S880\n",
      "---DDDDDDDPPDDDDPPDDDDDPDDPVRVVVCVVDDDDDDDDDDDDDFDKDKDKAWAQA\n",
      "PQ-RQLQVVLVVLLVQFWPLVLWDQP------DKDKDKDADQPQ--CSQVSNVVSLCCCP\n",
      "PDFVWRFS--GKGWPDDDNVRR--MTMIMGGTDGDDVVRIPGHDRWDHKDPVPWDWDGD-\n",
      "---SMTMTMTMTT-D\n",
      ">_seed_A0A2I2YKY3\n",
      "---DDDDDDDDDDDDDDDDDDDDPDDPVRVVVCVVDDDDDDPDDDDDDFDKDKDKAWDQA\n",
      "PQ-RQLQVVLVVQLVQFWVLVLWDQP------DKDKDKDFDQDD--LSQVVSVVSLCCCV\n",
      "VPQNKGFS--GKGWPDQDNVRR--ITIIMTGIDHDDVVRIPTHDDWDHKDPVPWDWDGD-\n",
      "---SMTMTMTMTT-D\n",
      ">_seed_G3QX27\n",
      "DDYDDDDDDDDDDDDDDAADDDDDDDPVRVVVCVVDDDDDDPDHDDDDWDKDKDKAWDQA\n",
      "LQ-RQLQVSLVSQLVQFFQLCLFDQP------DKDKDKDFDPAPQRLSQVSSVVSLCCCP\n",
      "PPQVWRFS--GKGWPDAPRVRG--ITIIMGGGHHDDPVRTPGGDDWDHWDPVPWDWDGD-\n",
      "---SMTMIMTMTT-D\n",
      ">_seed_K7BTY5\n",
      "DDDDDDDDDDDDDDDDDAADDDDPDDPVRVVVCVVDDDDDHPDHDDDDWDKDKDKAWDQA\n",
      "PQ-RQLQVSLVSVLVQFFQLCLFDQP------DKDKDKDFDPAPQRLSQVSSVVSLCCCP\n",
      "PPQVWRFS--GKGWPDAPRVRG--ITIIMGGGHHDDPVGTPGGDDWDHWDPVPWDWDGP-\n",
      "---SMTMTMTMTT-D\n",
      ">_seed_Q8IWT0\n",
      "------------DPPPPPPQDPLVDDPLLVVQLVVDDDDDAQWDWDDDDQWIKIKGWDQA\n",
      "LQ-RQLQNSLVSQLSQFWPLSLWDQP------DKDKDKDFAPAQQRLSQRSNVVSLCCCP\n",
      "PNQVWDFS--TKGWPDAPRVRR--MTMIMGGGHHDDVVGIRGGDDFDGWDPVPWDWDGP-\n",
      "---SMTMIMTITN-D\n",
      ">_seed_M0R5Y9\n",
      "------------DDPDDAPDDDDPDDPVVVVVCVVDDHDPDDDGDDDDWDKDKDKAWDQA\n",
      "PQ-RQLQVSLVSVLVQFFQLCLFDQP------DKDKDKDFDPAPQRLSQVRSVVVLCVLP\n",
      "PPQVWDFS--GKGWPDAPRVRR--ITIIMGGGHRDDVVGTPGGDDWDHWDPVPWDWDGP-\n",
      "---SMTMIMTMTI-D\n",
      ">_seed_Q505B7\n",
      "------------DPPPPPPPPPLVDDPLLVVQLVVDDHQDAQKDWDDDDQKIKIKGWDQA\n",
      "PQ-RRLQNVLVVVLVQFWPLSLWDQP------DKDKDKDFAPAPLRLSQRSNVVSLCCCP\n",
      "PNQNWDFS--TKGWPDAPRVRR--MTMIMGGTHHDDVVRIRGGFDFDHWDPVPWDWDGD-\n",
      "---SMIMTMTMTGTD\n",
      ">A0A2I3RJJ2\n",
      "--DDDPPPPDPPPDDPPPPPDDPVDDPLLVVQLVVDDDDDAQWDWDDDPAAIKIKGWDQA\n",
      "LQ-RSLQNSLVSVLVVVHDLSSDDPD------DDDDDDDADPDPQRRSFVSNVVSVCCCP\n",
      "PNDVDDHS--TDGDDDDDDRDD-----------------DDDDDDDPDWDPVPWDWDGD-\n",
      "---SMTMTMTDDD-D\n",
      ">A0A2I2ZWB6\n",
      "--DDDDPPPPPPPDDPPPPPPDPVDDPLLVVLLVVDDDDDAQWDWDDDPQWIKIKGWDQA\n",
      "AQWGRLQNSLVSVLVVVDQLSSDDPD------DDDDDDAADPDPQRRSVVSSVRSQCCCP\n",
      "PPDVDRHS--GDGDDPPYDNDD-----------------DDDDDDDPDWDPVPWDWDGD-\n",
      "---SMTMTITIDR-D\n",
      ">A0A2I2Y839\n",
      "------------DDPPDQQDPDDPDDPVVVVVCVVDDDDDDDDGDDDDDDDDDDDADDPD\n",
      "PQ-RRLQRSLVSVLVVVDQLCPDDPD------DDDDDDDDDPDPVRRSVVSNVVSVCCLP\n",
      "PPDVDDHS--GDGDDDDDDDDD-----------------DDDDDDDDDDD----------\n",
      "--------------D\n",
      ">Q2YDE7\n",
      "------------DDPDDAPDDDDPDDPVVVVVCVVDDDDDDDDGDDDDWDKDKDKAWDQA\n",
      "LQ-RQLQVSLVSQLVQFFQLCLFDQP------DKDKDKDFDPAPQRLSQVSSVVSLCCLP\n",
      "PPQVWRFS--GKGWPDAPRVRG--ITMIMGGGHHDDVVGTPGGDDWDHWDPVPWDWDGP-\n",
      "---SMTMIMTMTI-D\n",
      ">F6ZCC9\n",
      "-----------------DDDDDDDDDPVVVVVCVVDDDDDDDDDDDDDFDKDKDKAWDQD\n",
      "QQ-RQLQVSLVSQLVQFWPLVLWDQP------DKDKDKDFDPDDDDLSQVSNVVSLCCLP\n",
      "PNQVWDFS--GKGWPDQDRVGR--MTIIMGGGHHDDVVRIPGGDDWDGWDPVPWDWDDD-\n",
      "---SITMTMTMTT-D\n",
      ">F7AW07\n",
      "----------------DDDDDDDDPDPPD-----PDDDDDDDDDDDDDDQKDKDKFKDQF\n",
      "PQ-VRLQVSLQVQLCQQFNLVLWAQP------DKDKDKAADDAPLRLSQRSSVVVLVCCV\n",
      "DPQNWRFS--GKGWPDADRVRR--MTMMMGGTGGDDCVRTPRHDRWDHWDPVPWDWDDD-\n",
      "---NMTMTITMTG-D\n",
      "\n",
      "nsequences 39\n",
      "39\n",
      "39\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "STEP 002-004-1  identical.   \n",
      "Converged.\n",
      "\n",
      "done\n",
      "dvtditr (text) Version 7.520\n",
      "alg=A, model=Extended, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " NW-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#prepare tree attributes\n",
    "tre = toytree.tree('../testdata/HOG113/fident_1_raw_struct_tree.PP.nwk.rooted'  )\n",
    "for i,n in enumerate(tre.treenode.traverse()):\n",
    "    n.aln = None\n",
    "    n.aln3di = None\n",
    "    n.leafset = None\n",
    "    if len(n.name) == 0:\n",
    "        n.name = 'internal_'+str(i)\n",
    "\n",
    "alnfolder = infolder+'alnscratch/'\n",
    "if not os.path.exists(alnfolder):\n",
    "    os.mkdir(infolder+'alnscratch/')\n",
    "\n",
    "#clear all files in aln scratch\n",
    "cleanup(alnfolder)\n",
    "submat = '../mafftmat/3diHEXmat.txt'\n",
    "finalaln, finalaln3di = traverse_tree_merge_mafft( tre.treenode.get_tree_root(), get_leafset(tre.treenode.get_tree_root()) , alndf , infolder+'alnscratch/' , submat = submat , verbose = True ) \n",
    "print('finalaln',finalaln)\n",
    "#print the final alignments\n",
    "print('nsequences' , len(tre.get_tip_labels()))\n",
    "finalaln = remove_redundant(finalaln)\n",
    "finalaln3di = remove_redundant(finalaln3di)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "39\n",
      "39\n",
      ">O67888\n",
      "----------------------------------------LDYEPVYDITADAGIRVRAKTL-EELFCHAILATFNEITDIDKVEP------KEEYEIQAQND-MPFLLADIINEALVLHESK-HFVASECEVLELKEDF----VKVKLKGEKFDPKRHPSKLVIKAATYHRLRVEKKNEH--W-EA-EVIFDI\n",
      ">B5YH50\n",
      "----------------------------------------MKYK-VIDVAGDVGIRAEGASL-EECFINSAFGLYSLITDLAQIEP------AEEIEIIINEDNLENMLVSFLNELIFQFDTY-GFLGKAISI-EIKDNY----LTARIKGEKFNPEKHERKLLVKAATYHNLVLKQEDSF--W-IA-EIIFDI\n",
      ">A9A300\n",
      "----------------------------------------MSYK-FVDHATDAIIEVTAKDL-QEAFSVTADAVINLTLDQDKVEE------KEQRKFVAEGKDLRYLLFSWLEEIPFLLITE-GFAIKRIEF-DITQNQ----INATAYGEPLDFKKHNFKVEIKAPTFYDMEIKQNGGV--F-MR-FLL-DL\n",
      ">Q5JCY8\n",
      "-------------------------------------------E-HYEHTADIGVRGYGSTL-EEAFEAVALGLFDVMVNVKKVEP------KECREVEVEEEDLEALLYSFLEELLVLHDME-GLVFGDVKV-RIEKTENGYKLKAKACGEVLNPEKHEPKEEVKAITYHDMKIEKLPDGR-W-MA-QFVPDL\n",
      ">Q60334\n",
      "-------------------------------------------N-YFETTADLGVEAKGKSL-EEAFKEGAKGLYNIMVDIDKVDK------KEKIEFEITGEDLEELLYNFLNELLFYTDVE-NLVFNDFDV-KIEKNDNGYRLKCTAYGEKINKEKHNIKEEVKAVTYHKMEIKQEEDG--W-KI-RYIVDL\n",
      ">B1L703\n",
      "-------------------------------------------R-FLELEADVGFEVWAEDL-NSLFEEAALSMYEIMVDVEKVER------SIERHISVSAPDIEMLLHHWLSELLFITDVE-RIVFSDFEV-KISGND----LEGRAWGEFIDVERHNPKTEIKAVTYHKLSVREEGGV--W-RC-TVVLDI\n",
      ">Q9UX79\n",
      "-------------------------------------------E-FFEHTADVGIKSYGRSL-EEAFSNAALGVFEVITDTSKVRP------VEYREIYLNGYDLENLLYKWIEELLYYYDSE-LMIFSKFDL-MIDQDS--ITLEGKAWGERFNDKIHERRTVVKAMTYHQLSIEKTESG--Y-VI-TFVVDI\n",
      ">Q74CJ3\n",
      "----------------------------------------MPYRYLPDITADVAFEAWGETR-EEMFCAAADALTNVMVDLASVTP------TEEIVISLANEELDLLLFSFLQELIFLKDAR-CFLLRVPRI-MITETEEVLHLDAIARGERIDNERHPMMVDVKAVTLHLFSVWRQEND--W-WA-RVVLDI\n",
      ">B8E097\n",
      "----------------------------------------KNYE-LLDHTADIGIIVYGETK-EKAFESAAEAMFDLMCPLEKIQE------KECFDIEVDGEDLESLLVTWLNELLYVFEVQ-KLLFRRFDVTLIGHNQ----LISHCYGERFDPKKHEITREIKAVTYNLLKIEQREDK--W-MI-QVVFDI\n",
      ">Q8TU10\n",
      "----------------------------------------KKYE-YLEHTADIKFLAYGNTL-KEVFENAALAMFNVIIDTGKVSG------ETAREVCLTSPDLESLLVDWLSELLYLFEVD-EIVFWKFRVEEIREEEGEYSIKAVASGEQYYPESHPFETEIKAVTYNQLELEKTADG--W-KA-QVVVDI\n",
      ">B8BWG5\n",
      "-------------------------------------------E-YLDHPADIILHSWGNDL-SASLSNLAVCMFGYITSINLIATNPSQCLEYGSNIVAQGHDLRSLVYSFLDEWLFNFHDT-GFVAKEILVTEFNVEA--WRIVSTGKGEVMDVRRHPQGTEVKAITYNAMRIEEKGKDDRRCDV-YVVVDI\n",
      ">Q9TZN1\n",
      "-------------------------------------------E-YLDHPADIQLHSWGSTM-EEAFEACLVSMFGYMTDLAKVDE------MYEFYWKASGDSLDGLLFQFLDEALNSFHAEPCFVAKRVEILRFDKKK--FEIEFRGWGESFDTSKHETEADIKSPTYSNMQINEKPERCD---I-YVIVDI\n",
      ">A2F1C4\n",
      "-------------------------------------------E-FIDHPADIQLHAWGPQP-DVTLEQLGKAFFEVMFETDNFKE------ETTHSISVKGKDAQNLLYNFLDEWLYVFDAE-DFVATRIKVTKCDFVN--LEIEATGYGELFDMKKHARRTEVKAITYASMKVDTTPGASE---F-YVILDL\n",
      ">Q54BU9\n",
      "----------------------------------------------LDHTADIMFHTWGKDL-KEALEQMVLIMNNYMVELDSVELDDS---ATEQTISVNGHDMDSLLFALLDEFLFVFSTE-FIIFKQVQIISFDREN--FSIKAIGKGVELDKSKHTTGTEIKAITYSCMKIEENPDKSD---I-HVIVDI\n",
      ">A9UTI8\n",
      "----------------------------------------------LDHTADIQVHAWGQTL-KDSFEQTVVAMFGYMTELDAIEIDD----SQTSEFSVEGHDLMSLLYALLDEFLFRFSAEDNLVCREVEIVEFDTDN--FKITVRGKGEVFDLAKHPQGTEIKAITYSNMQIHTERATHD---V-YVILDI\n",
      ">Q9VD92\n",
      "---------------VEFSRENFLLPE-------------MKYE-YLDHTADVQIHGWGSSL-KEAFEQCGVAMFGYMTELDYVSV------EQCFEIEAHGDDLESLLFHFLDELLFLFSAEPYLVCKKLEITKFDVEN--FEISCHCYGEPFELGKHPQGTEVKAITYSAMQIIQDVEASNY-EV-FVIIDI\n",
      ">D6WKY5\n",
      "---------------ETFTKEELTLPP-------------CKYE-YLDHTADVQLHAWGDSL-KEAYEQCGLAMFGYMTEINTVEI------KQCSEIEATGDDLDGLLFHFLDELLFLFSCEPFLICSKLVITEFRTEGDELKIKCKCYGEEFTLGKHPQGTEVKAITYSAMQIINDPENKKF-EV-FVIIDI\n",
      ">Q7PYJ8\n",
      "---------------VPLETSDHVVPE-------------VKYE-YLDHTADVQLHAWGETL-QEAFEQCGMAMFGYMTELPKVEI------QKCYQIVTDTDDLENLLYRFLDELLFLFSAEPFLICKKLEITRFDLEQ--FTIECNCYGEEFDLAKHPQGTEVKAITYSAMQIYQKPEHEKN-EV-FVIIDI\n",
      ">B7PHX9\n",
      "-----------------------------------------------------------------------------MTEIDKVDI------RMTQDIDAEADDMVGLLFHFLDELLFIFSAEPFFIGRKVKILEFDKEA--FRIKARVYGEVFDLGKHPQGSEVKAITYSNMQVWDNPDQHE---V-FVIIDI\n",
      ">T1ELH7\n",
      "-----------------------------------------------------------------------------MTDISTVEQ------CESQTIEAEGDDMQSLLFHFLDEFLFLFSAEPFFIARRVEIKTFDREN--FKIEAVGHGEPFDIAKHPQGTEVKAITYSNMQVYDEENKHE---V-FVIIDI\n",
      ">C3Y972\n",
      "-----------------------------------------------------------------------------MTEIETVEP------LDTYETEAQADDMENLLYQFLDEWLFAFSAEPFFIARDVKIVEFDREN--FRIKSVGYGETFDLSKHPQGTEVKAITYSAMQIHEEVDTND---V-YVVLDI\n",
      ">W5M2J6\n",
      "------------M-----NDRELDLTEAQKALKAKYPSINKKYE-YLDHTADVQIHSWGDTL-EEAFEQCAMGMFGYMTDTETVEP------IDTVEVESEGEDLLSLLFHFLDDWLFKFSADQFFVPREVKVLHIDRMR--FRMRSIGWGEEFRISKHPQGTEVKAITYSAMQIHEENKPE----I-FVIIDI\n",
      ">Q566V0\n",
      "------------M-----NDRELDLTEEQKALKAKYPPITKKYE-YLDHTADVQIHSWGNSL-EEAFEQCAMGMFGYMTDTETVEP------IDTVEVESEGDDMESLLYHFLDDWLFKFSADIFFIPREVKVLHIDRMR--YKIRSIGWGEEFSINKHPQGTEVKAITYSAMQIHETEQPE----I-FVIIDI\n",
      ">H2MRT4\n",
      "------------M-----DDRELDLTNEQKAVKSKYPPINKKYE-YLDHTADVQIHSWGDSL-EEAFEQCAMGMFGYMTDTETVEP------LDTVEVESEGEDMESLLFHFLDDWLYKFSADLFFVPREIKVLHIDRMH--YKIRSIGWGEEFSISKHPQGTEVKAITYSAMQIHDKEKPE----I-FTIIDI\n",
      ">A0A5F8GYJ2\n",
      "------------MAQDGKDKRDYNLNEEQKIIK------------YLDHTADIQLHAWGDPL-EEAFEQCALAMFAYMTDTGTVEL------LQTVEVEAEGDDLQSLPFHFLDEWLYKLIANEFFIPQEVKVLSIDQRN--FKLRSIGN--------------------------------------------\n",
      ">F7AIV7\n",
      "------------MAQDGEDERDYNLNEEQKAIKAKYPPAIRKYE-YLDHTADVQLHAWGDTL-EEAFEQCAMAMFGYMTDTGTVEP------LQTVEVQAEGDDLQSLLFHFLDEWLYKFSANEFFIPREVKVLSIDQRN--FKLRSIGW--------------------------------------------\n",
      ">A0A2I3S880\n",
      "--TGSR-TEGAVMAQKEGDVRDYNLTEEQKAIKANYPPVNRNYE-YLDHTADVQLHTWGDTL-EEAFEQCAMAISGYMIDTRTVEP------LQTIEVETQ--EVSTFLFHFLDEWLYKSNADEFFIPWEVKVICIDQRH--FQLQSIGWTEEFSLSNHPQGTEVKVITYSAMQVYNKENPE----V-FVIIDI\n",
      ">A0A2I2YKY3\n",
      "---GSGCSDKAVMAQKEGDVRDYNLTEEQKAIKANYPPVNRNYE-YLDHTADVQLHTWGDTL-EEAVEQCAMAIFGYMIDNRTVEP------LQTIEVETQ--EVSTFLFHFLDEWLYKSDADEFFIPWEVKVLCIDQRH--FQLQSIGWREEFSLSKHPQGTEVKVITYSAMQVYNKENPE----V-FVIIDI\n",
      ">G3QX27\n",
      "IKGGSRVSNPAVMAQEEEDVRDYNLTEEQKAIKAKYPPVNRKYE-YLDHTADVQLHAWGDTL-EEAFEQCAMAMFGYMTDTGTVEP------LQTVEVETQGDDLQSLLFHFLDEWLYKFSADEFFIPREVKVLSIDQRN--FKLRSIGWGEEFSLSKHPQGTEVKAITYSAMQVYNEEKPE----V-FVIIDI\n",
      ">K7BTY5\n",
      "MKGGSRVSNPAVMAQEEEDVRDYNLTEEQKAIKAKYPPVNRKYE-YLDHTADVQLHAWGDTL-EEAFEQCAMAMFGYMTDTGTVEP------LQTVEVETQGDDLQSLLFHFLDEWLYKFSADEFFIPREVKVLSIDQRN--FKLRSIGWGEEFSLSKHPQGTEVKAITYSAMQVYNEEKPE----V-FVIIDI\n",
      ">Q8IWT0\n",
      "------------MAQEEEDVRDYNLTEEQKAIKAKYPPVNRKYE-YLDHTADVQLHAWGDTL-EEAFEQCAMAMFGYMTDTGTVEP------LQTVEVETQGDDLQSLLFHFLDEWLYKFSADEFFIPREVKVLSIDQRN--FKLRSIGWGEEFSLSKHPQGTEVKAITYSAMQVYNEENPE----V-FVIIDI\n",
      ">M0R5Y9\n",
      "------------MAQEEEDVRDYNLTEEQKAIKNKYPPVNRKYE-YLDHTADVQLHAWGDTL-EEAFEQCAMAMFGYMTDTGTVEP------LQTVEVETQGDDLQSLLFHFLDEWLYRFSADEYFIPREVKVLNIDQKN--FKLRSIGWGEEFSLSKHPQGTEVKAITYSAMQVYNEEKPE----V-FVIIDI\n",
      ">Q505B7\n",
      "------------MAQEEEDVRDYNLTEEQKATKDKYPPVNRKYE-YLDHTADVQLHAWGDTL-EEAFEQCAMAMFGYMTDTGTVEP------LRTVEVETQGDDLQSLLFHFLDEWLYKFSADEYFIPREVKVLNIDQKN--FKLRSIGWGEEFSLSKHPQGTEVKAITYSAMQVYNEEKPE----VFFVIIDI\n",
      ">A0A2I3RJJ2\n",
      "--CMSRHTQPKIMAQAEEDVRDYNLTEEQKVIKAKYPPVNRKYE-YLDHTADVQLHAWGDTL-EEAFEQCAMAMFGYMTDTGTVEP------LPTVEVETQGDDLQSLLFHFLDEWLYKFSADEFFIPQEVKVL-----------------KESSLSKYPQGTEVKAITYSAMQVYNEEKPE----V-FVIIDI\n",
      ">A0A2I2ZWB6\n",
      "--IFGRDRVSPFMAQAEEDVRDYNLTEEQKAIKAKYPPVNRKYE-YLDHTADVQLHAWGDTLEEEAFEQCAMAMFGYMTDTGTVEP------LPTVEVETQGDDLQSLLFHFLDEWLYKLSADEFFIPREVKVL-----------------KEFSLSKHPQGTEVKAITYSAMQVYNEEKPE----V-FVIIGI\n",
      ">A0A2I2Y839\n",
      "------------MAQAEEDVRDYNLTEEQKAIKAKYPPVNRKYE-YLDHTADVQLHAWGDTL-EEAFEQCAMAMFGYMTDTGTVEP------LPTVEVETQGDDLQSLLFHFLDEWLYKLSADEFFIPREVKVL-------------------IFIVQAPSG--------------NRSQGNN---I-FSNAGL\n",
      ">Q2YDE7\n",
      "------------MALEGEDVRDYNLTEEQKAIKAKYPPVSRKYE-YLDHTADVQLHAWGDTL-EEAFEQCAMAMFGYMTDTGTVEP------LQTIEVETQGDDLQSLLFHFLDEWLYKFSADEFFIPREVKVLNIDQRN--FKIRSIGWGEEFSLSKHPQGTEVKAITYSAMQVYNEEKPE----V-FVIIDI\n",
      ">F6ZCC9\n",
      "------------M-----DERYYTISEEQVAVKAKYPPVNKNYE-YLDHTADVQLHAWGETL-EEAFEQCAMAMFGYMTDIETVEP------LDTVEVVTEGEDLISLLFNFLDEWLYKFSADQYFVPREVKVLNIDRMN--FKIRSIGWGEEFSLTKHPQGTEVKAITYSAMQIHEEEKAE----V-FVIIDI\n",
      ">F7AW07\n",
      "------------------------MEEKLPAYESGYHNVIKKFE-YLDHTADVQIHSWGDNI-EETFEQAAMAMFNYMTDIDGVRP------VKTVEITAEGTDYMSLLFHFMDEFLFLSSCDDFFMAREIKITNFDREN--WRMEALGYGEDFNQERHGKGTEVKAITYSAMQVIESEKSE----I-FVIVDI\n",
      "\n",
      "finalaln3di ../testdata/HOG113/alnscratch/76_inter3di.fasta\n",
      ">O67888\n",
      "---------------------------------------DDDDDDDDPDDKDKDKDKDLDPQ-VRLFVLLLVVLVQFWPLVLWDFP------DKDKDKFFAA-PPVRSQVRSQVSVVCCVVQV-WGFN--DKDWPDDDRG----MTMMMTGTDHDDVVRIPGHDDFNGWDPVPWDWDDDP---SMIMTMTMTD-D\n",
      ">B5YH50\n",
      "---------------------------------------DDDD-DDDDDAKDKDKAKDLDLQ-NRLVRVLVVLVVQFWDLVLKDQP------DKDKDKDFDQDSLVVSQVVNQVQLVCCVPVQ-WGFN--DWDW-DDDRR----MIIIMGGTDHDDVVRIDTHDNWNGWDPVPWDWDDDD---SMIMTMIMTG-D\n",
      ">A9A300\n",
      "---------------------------------------DDDD-DDDDFPKDKDKDWFQDPQ-VRLFVVLVVLLVQFWALVLWDFP------DKDKDKAWDQDPLSSVQSSSVVSLCCCPVVQ-WRFN--DWDW-DDDSI----MIIIMTGTGGHDCVSGVGDDRFDGWDPVPWDWDDD----STIMTITMTT-D\n",
      ">Q5JCY8\n",
      "---------------------------------------DD---DDDPPW-DKQKAKDLDPQ-RRLFRSLLSVLVQFWQLVLWDQP------DKDKDKDDDQEDLVVSQRVSQVVLVCCVVVQ-WDFNGKGKDW-DADPRHI--MIIMMTGGDHDDCVGTVTHDDFDGWDPAPFDWDQDPV--SMIMTMGITH-D\n",
      ">Q60334\n",
      "---------------------------------------DW---DDDPAF-IKIKAFDQEPQ-RRLFRVLQSVLVQFWQLVLWAQP------DKDKDKDFDQEPLRRSQRVNVVQVVCCVPVQ-WDFNGKDWDW-DADPRHI--MIIIITGGDHDDCVGTVGHDDFPGWDSVPWDWDADP---NHIMTMTTTH-D\n",
      ">B1L703\n",
      "---------------------------------------D----DDDPPWKDKDKDKDLDDF-VQLFVVLQVVLVQFWQLVQWDQP------DKDKDKDFDPDDLVVSQVSNQVQLVCCVVVQ-WDFN--GKGW-DDDPG----MIIMMTGGGHDDCVRTVTHDDFPGWDPVQFDWDDDP---SMIMTMTMTG-D\n",
      ">Q9UX79\n",
      "---------------------------------------DW---DDDPQW-IKIKGFDLEPQ-SRLFSSLQRLLVQFWDLVLWDQP------DKDKDKDFDPALLRLSQVVNQVQVVCCVPVV-WGFN--DKDW-DADRVRR--MIIMITGTDHDDVVRIPGHDDFPGFDPAPFDWDQDP---RYIMTIGITD-D\n",
      ">Q74CJ3\n",
      "---------------------------------------DPDD-DDVPPFKDKDKFKDLDLW-RRLLRVLLVLLVLFFDPVLFDQP------DKDKDKDKDLDSLVVSQVVSVVQLCCCPVVQ-WDWHWPTWDW-DDDPSMI--MIITIIGGDHDDCVSTVTPDRFDHKDPPQWDWDDDD---SMIMTMTMTG-D\n",
      ">B8E097\n",
      "---------------------------------------DPDD-DDDDPQKDKDKAKDLDPQ-RRLFVVLQVVLCLFWNLVLWDQP------DKDKDKFWAQAPLRQSQVVSVVVLVCCVPVQ-WRFN--GKGWPDDDGT----MTIIITGTDHDDVVRIDGPDDFDGKDPVPWDWDDDD---SMIMTMTMTR-D\n",
      ">Q8TU10\n",
      "----------------------------------------DQKDWDDDDFWIKMKGDDQEPQ-SRLFSLLQVVLVLFWDLVLFDFF------PKDKDKDKDLDSLRVSQVSNQVVLVCCVVVQ-WRFN--GKGWPDWDQDPNMIMTIIMTGTDHHDCVSIPTNDDFPGWDPVPWDWDADP---NHIMTMTITG-D\n",
      ">B8BWG5\n",
      "-------------------------------------------DDDDDPLKDKQKAKDLAPQ-RRLQVSLLSVLVQFWPLVLWDFDPVCQVVLFFFQKFADDADLNLSQRSNVSSLVSCVVQQ-WDFS--TKGFPDDDGVRR--MTTMGGTTGRDDCVRIPGGDDFDGWDPPPWDWDQDDPVRSMIMTITMTG-D\n",
      ">Q9TZN1\n",
      "-------------------------------------------DWDDDPQWIKIKGKDQDPQ-RRLFSSLLVVLVQFWPLVLWDQP------DKDKDKAAAPALLRLSQRSNVRVLVCLVDPPNWRFS--GKGWPDQDRVRR--MTIIMGGTGHDDVVRIDGNDDFPGKDPVPWDWDDDN---RMTMTIIITH-D\n",
      ">A2F1C4\n",
      "-------------------------------------------DDDDDPQKDKDKFKFQDPQ-VRLFVRLLVLLLQFFPQVLWFQP------DKDKDKDFAPAPLRLSQVSNVVVLCCCVVPQ-WGAN--TKGWPDDDRPRR--MTMIMGGTDHDDCVRVGTHDDFDGKDPVPWDWDDDN---RMTMTMTMTG-D\n",
      ">Q54BU9\n",
      "---------------------------------------------DDDPQWIKIKGKDQFPQ-VRLQNVLRVLLCQFWNLVLFDLDPQK---DKDKDKDAAPAQLRLSQRSNVVSLVCCVPVV-WRFR--TKGFPDADRVRR--MTIMMTITHHDDPVRIDGHDDFPGWDPVPWDWDDDN---RMIMTMTITD-D\n",
      ">A9UTI8\n",
      "---------------------------------------------DDPWDKDKDKAWAQALQ-RRLFVSLLVVLVQFWPLVLWDFDPV----AKDKDKDADDAPQRLSQRSNVVVLCCCPPNQVDDAS--GKDFPDQDRPRR--MTIIIGTTHHDDCVRIPGGDDWDGWDPVPWDFDRPD---STRMTMTMTT-D\n",
      ">Q9VD92\n",
      "----------------------DDDDPVP------DDQDDQAWDWDDDPQWIKIKAKDLAQQ-RRLFNSLLVVLCQFWPLVLWDFD------TKDKDKAAAPAPLRLSQRSSVVSLCPLVDPPNWRAS--GKDFPDQDRVRS--MTMIIGTTGHDDPPRIDGGDDFDGWDPVPWDWDDPPV-VSMIMTITITR-D\n",
      ">D6WKY5\n",
      "----------------------DDDDPVN------PDDDDDDDDDDDDDQKDKDKAWDLALQ-RRLFRSLVSLLVQFWPLVLWDFP------DKDKDKDKDQAPLRRSQRSSVVSLCPLVDPPNWRAS--GKDWPDFAHDDRMTMTMIIGTTGRDDPPSTDGGDRWDGWDPVPWDFDGDNP-VSTTMTMTMTG-D\n",
      ">Q7PYJ8\n",
      "----------------------DDDDPDD------PDQDDQAWDWDDDPQFIKIKGFDQDLQ-RQLFNSLVSVLVQFWPLVLWDFD------TKDKDKAFDDADLVLSQRSSVVSLCPLVDPPNWRAS--GKDFPDQDGVRS--MTIIIGTTDHDDCVRIPGGDDWDGWDPVPWDFDDPPV-DPTGMIMTTTD-D\n",
      ">B7PHX9\n",
      "----------------------------------------------------------------------------FWPLVLWDQP------DKDKDKDFDPAPLRLVQRVVVVVLCCLVDPPNWRFS--DKDWPDQDRVRS--MTMIMGGTGRDDPVRTPGGDRFDGWDPVPWDWDDDP---VDTDTDTDTH-D\n",
      ">T1ELH7\n",
      "----------------------------------------------------------------------------DWPLVLWAQP------DKDKDKAADPAPQRLVQRVVVVQLCPLVDPPNWDAS--DKDWPDADRVRR--MTMIMGGTGRDDPVRTPDDDRWDGWDPVPWDWDDDP---PDTDTDTDTH-D\n",
      ">C3Y972\n",
      "----------------------------------------------------------------------------FWPLVLWAQP------DKDKDKDFDDADLVLSQRSNVRSLCPLVDPPNWDAS--TKGWPDQDRVRS--MTMMMGGTGHDDCVRTPGGDRFDGWDPVPWDWDDDP---RMTMTITMTR-D\n",
      ">W5M2J6\n",
      "-----------------DPDPDDDDDPVRVVVCPVDDDDDDDDDDDDDFDKDKDKAWDQDQQ-RRLQVSLVSQLVQFWPLVLWDQP------DKDKDKAAAPAQQRLSQRSSVRSLCCLPPNQVWRFS--GKGWPDADRVRR--MTMIMGGGHHDDVVRTPGGDDWDGWDPVPWDWDDP----GIIMTMTMTT-D\n",
      ">Q566V0\n",
      "-----------------DDQPDFDDDPLLVVQLVVDDHDDAQWDWDDDPQWIKIKGKDQDPQ-RRLQNSLVSQLSQFFQLSLFDQP------DKDKDKDADPAQLRLSQRSNVRVLCCLPPNQVWRFS--GKGWPDADRVRR--MTMIMGGGHHDDCVRTGGGDDWDGWDPVPWDWDPD----SMIMTITITR-D\n",
      ">H2MRT4\n",
      "-----------------DDDPDDDDDPVRVVVCPVDDDDDDDDDDDDDFDKDKDKAWDQAPQ-RQLQVSLVSVLVQFWPLVLFDQP------DKDKDKDADPAQLRLSQRSNVRSLCCCPPNFVWRFS--GKGWPDADRVRR--MTMIMGGGHHDDVVRTPGGDDWDGWDPVPWDWDGD----SMTMTMTMTT-D\n",
      ">A0A5F8GYJ2\n",
      "------------DDDPDPPPPDDPDDPV-----------PVPDDDDDDDDDDDDDADDPPPQ-RSLQRRLQVVLVVVDDLVPQAFD------DKDKDKFFDPAPVCLSVRRSVRSVCCCPVHPNDDAS--GKGWLDADRVRR--MTIIMGG--------------------------------------------\n",
      ">F7AIV7\n",
      "------------PADQDPADPDDDDDPVRVVVCPVDDDDRDRDHDDDDFDKDKDKAWDQAPQ-RQLQVSLVSQLVQFWPLVLFDQP------DKDKDKAADPAPQRLSQRRSVVSLCVCPPPQVWRFS--GKGWPDAPRVRR--MTMIMGG--------------------------------------------\n",
      ">A0A2I3S880\n",
      "---DDDDDDDPPDDDDPPDDDDDPDDPVRVVVCVVDDDDDDDDDDDDDFDKDKDKAWAQAPQ-RQLQVVLVVLLVQFWPLVLWDQP------DKDKDKDADQPQ--CSQVSNVVSLCCCPPDFVWRFS--GKGWPDDDNVRR--MTMIMGGTDGDDVVRIPGHDRWDHKDPVPWDWDGD----SMTMTMTMTT-D\n",
      ">A0A2I2YKY3\n",
      "---DDDDDDDDDDDDDDDDDDDDPDDPVRVVVCVVDDDDDDPDDDDDDFDKDKDKAWDQAPQ-RQLQVVLVVQLVQFWVLVLWDQP------DKDKDKDFDQDD--LSQVVSVVSLCCCVVPQNKGFS--GKGWPDQDNVRR--ITIIMTGIDHDDVVRIPTHDDWDHKDPVPWDWDGD----SMTMTMTMTT-D\n",
      ">G3QX27\n",
      "DDYDDDDDDDDDDDDDDAADDDDDDDPVRVVVCVVDDDDDDPDHDDDDWDKDKDKAWDQALQ-RQLQVSLVSQLVQFFQLCLFDQP------DKDKDKDFDPAPQRLSQVSSVVSLCCCPPPQVWRFS--GKGWPDAPRVRG--ITIIMGGGHHDDPVRTPGGDDWDHWDPVPWDWDGD----SMTMIMTMTT-D\n",
      ">K7BTY5\n",
      "DDDDDDDDDDDDDDDDDAADDDDPDDPVRVVVCVVDDDDDHPDHDDDDWDKDKDKAWDQAPQ-RQLQVSLVSVLVQFFQLCLFDQP------DKDKDKDFDPAPQRLSQVSSVVSLCCCPPPQVWRFS--GKGWPDAPRVRG--ITIIMGGGHHDDPVGTPGGDDWDHWDPVPWDWDGP----SMTMTMTMTT-D\n",
      ">Q8IWT0\n",
      "------------DPPPPPPQDPLVDDPLLVVQLVVDDDDDAQWDWDDDDQWIKIKGWDQALQ-RQLQNSLVSQLSQFWPLSLWDQP------DKDKDKDFAPAQQRLSQRSNVVSLCCCPPNQVWDFS--TKGWPDAPRVRR--MTMIMGGGHHDDVVGIRGGDDFDGWDPVPWDWDGP----SMTMIMTITN-D\n",
      ">M0R5Y9\n",
      "------------DDPDDAPDDDDPDDPVVVVVCVVDDHDPDDDGDDDDWDKDKDKAWDQAPQ-RQLQVSLVSVLVQFFQLCLFDQP------DKDKDKDFDPAPQRLSQVRSVVVLCVLPPPQVWDFS--GKGWPDAPRVRR--ITIIMGGGHRDDVVGTPGGDDWDHWDPVPWDWDGP----SMTMIMTMTI-D\n",
      ">Q505B7\n",
      "------------DPPPPPPPPPLVDDPLLVVQLVVDDHQDAQKDWDDDDQKIKIKGWDQAPQ-RRLQNVLVVVLVQFWPLSLWDQP------DKDKDKDFAPAPLRLSQRSNVVSLCCCPPNQNWDFS--TKGWPDAPRVRR--MTMIMGGTHHDDVVRIRGGFDFDHWDPVPWDWDGD----SMIMTMTMTGTD\n",
      ">A0A2I3RJJ2\n",
      "--DDDPPPPDPPPDDPPPPPDDPVDDPLLVVQLVVDDDDDAQWDWDDDPAAIKIKGWDQALQ-RSLQNSLVSVLVVVHDLSSDDPD------DDDDDDDADPDPQRRSFVSNVVSVCCCPPNDVDDHS--TDGDDDDDDRDD-----------------DDDDDDDPDWDPVPWDWDGD----SMTMTMTDDD-D\n",
      ">A0A2I2ZWB6\n",
      "--DDDDPPPPPPPDDPPPPPPDPVDDPLLVVLLVVDDDDDAQWDWDDDPQWIKIKGWDQAAQWGRLQNSLVSVLVVVDQLSSDDPD------DDDDDDAADPDPQRRSVVSSVRSQCCCPPPDVDRHS--GDGDDPPYDNDD-----------------DDDDDDDPDWDPVPWDWDGD----SMTMTITIDR-D\n",
      ">A0A2I2Y839\n",
      "------------DDPPDQQDPDDPDDPVVVVVCVVDDDDDDDDGDDDDDDDDDDDADDPDPQ-RRLQRSLVSVLVVVDQLCPDDPD------DDDDDDDDDPDPVRRSVVSNVVSVCCLPPPDVDDHS--GDGDDDDDDDDD-----------------DDDDDDDDDDD------------------------D\n",
      ">Q2YDE7\n",
      "------------DDPDDAPDDDDPDDPVVVVVCVVDDDDDDDDGDDDDWDKDKDKAWDQALQ-RQLQVSLVSQLVQFFQLCLFDQP------DKDKDKDFDPAPQRLSQVSSVVSLCCLPPPQVWRFS--GKGWPDAPRVRG--ITMIMGGGHHDDVVGTPGGDDWDHWDPVPWDWDGP----SMTMIMTMTI-D\n",
      ">F6ZCC9\n",
      "-----------------DDDDDDDDDPVVVVVCVVDDDDDDDDDDDDDFDKDKDKAWDQDQQ-RQLQVSLVSQLVQFWPLVLWDQP------DKDKDKDFDPDDDDLSQVSNVVSLCCLPPNQVWDFS--GKGWPDQDRVGR--MTIIMGGGHHDDVVRIPGGDDWDGWDPVPWDWDDD----SITMTMTMTT-D\n",
      ">F7AW07\n",
      "----------------DDDDDDDDPDPPD-----PDDDDDDDDDDDDDDQKDKDKFKDQFPQ-VRLQVSLQVQLCQQFNLVLWAQP------DKDKDKAADDAPLRLSQRSSVVVLVCCVDPQNWRFS--GKGWPDADRVRR--MTMMMGGTGGDDCVRTPRHDRWDHWDPVPWDWDDD----NMTMTITMTG-D\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "finalaln = remove_seeds(finalaln)\n",
    "finalaln3di = remove_seeds(finalaln3di)\n",
    "#print number of sequences in the final alignment   \n",
    "with open(finalaln) as f:\n",
    "    print( f.read().count('>'))\n",
    "with open(finalaln3di) as f:\n",
    "    print( f.read().count('>'))\n",
    "\n",
    "with open(finalaln) as out:\n",
    "    print(out.read())\n",
    "\n",
    "print('finalaln3di',finalaln3di)\n",
    "with open(finalaln3di) as out:\n",
    "    print(out.read())\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "foldtree",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
