{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import os\n",
    "import glob\n",
    "import toytree\n",
    "import tqdm\n",
    "import pandas as pd\n",
    "from subprocess import PIPE, Popen\n",
    "import shlex\n",
    "from Bio import SeqIO\n",
    "import random\n",
    "import itertools\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "def remove_seeds( alnfile):\n",
    "    \"\"\"\n",
    "    this function removes seeds from an alignment\n",
    "    \"\"\"\n",
    "    aln = SeqIO.parse(alnfile, 'fasta')\n",
    "    sequences = []\n",
    "    for s in aln:\n",
    "        sequences.append(s)\n",
    "    #write new aln\n",
    "    \n",
    "    with open(alnfile, 'w') as f:\n",
    "        for s in sequences:\n",
    "            f.write('>' + str(s.id).replace('seed','').replace('_' , '') + '\\n')\n",
    "            f.write(str(s.seq) + '\\n')\n",
    "    \n",
    "    return alnfile\n",
    "\n",
    "def Fident(str1,str2 , verbose = False):\n",
    "    #minlen= min( (len(str1),len(str2))  )\n",
    "    #str1 = str1[:minlen]\n",
    "    #str2 = str2[:minlen]\n",
    "    str1 = np.array(list(str1))\n",
    "    str2 = np.array(list(str2))            \n",
    "    return len(np.where( (str1 == str2 ) & (str1 != '-' ) & (str2 != '-')  )[0]) / len(str1)\n",
    "\n",
    "def copyaln( aln, seq):\n",
    "    seqiter = iter(seq)\n",
    "    newaln = ''\n",
    "    for i,char in enumerate(aln):\n",
    "        if char == '-':\n",
    "            newaln += '-'\n",
    "        else:\n",
    "            newaln+=next(seqiter)\n",
    "    return newaln\n",
    "\n",
    "def read_dbfiles3di(  AADB , threeDidb):\n",
    "    #find positions \n",
    "    threeDiseq = [ l.strip().replace('\\x00','') for l in open(threeDidb)]\n",
    "    lookup = AADB+'.lookup'\n",
    "    ids = [ l.split()[1].strip() for l in open(lookup)]\n",
    "    AAs = [ l.strip().replace('\\x00','') for l in open(AADB)]\n",
    "\n",
    "    mapper3di = dict(zip(ids,threeDiseq))\n",
    "    mapperAA = dict(zip(ids,AAs))\n",
    "    \n",
    "    return mapper3di, mapperAA\n",
    "\n",
    "def calc_fident_crossaln(row , verbose = False):\n",
    "    #amino acid representations of alns using AAand3di or just 3di\n",
    "    qaln_2, taln_2 = row.qaln , row.taln\n",
    "    #start and stop of aln\n",
    "    \n",
    "    qstart_2, qend_2, tstart_2 , tend_2 = row.qstart, row.qend , row.tstart , row.tend\n",
    "    #indexing starts at 1...\n",
    "    \n",
    "    #3di of the query and target\n",
    "    structQ, structT = row['3diq'], row['3dit']\n",
    "    AAq, AAt = row['AAq'], row['AAt']\n",
    "\n",
    "    #add gaps\n",
    "    t3diAA_newgaps = copyaln(taln_2, structT[tstart_2-1:tend_2]) \n",
    "    q3diAA_newgaps = copyaln(qaln_2, structQ[qstart_2-1:qend_2])\n",
    "    row = pd.Series( { '3di_qaln_mode2':q3diAA_newgaps , '3di_taln_mode2':t3diAA_newgaps })\n",
    "    #return columns\n",
    "    return row\n",
    "\n",
    "def get_leafset( treenode ):\n",
    "    \"\"\"\n",
    "    this function returns the leafset of a node\n",
    "    \"\"\"\n",
    "    if treenode.is_leaf():\n",
    "        return [treenode.name]\n",
    "    else:\n",
    "        return treenode.get_leaf_names()\n",
    "\n",
    "\n",
    "def mafft_profile(aln1,aln2, outprofile , submat = None):\n",
    "    \"\"\"\n",
    "    this function aligns two alignments using MAFFT\n",
    "    \"\"\"\n",
    "    #make profile\n",
    "    if submat:\n",
    "        cmd = 'mafft --textmatrix {} --seed {} {} > {}'.format(submat, aln1,aln2, outprofile)\n",
    "    else:\n",
    "        cmd = 'mafft --seed {} {} > {}'.format(aln1,aln2, outprofile)\n",
    "\n",
    "    print(cmd)\n",
    "    subprocess.run(cmd , shell=True)\n",
    "    return outprofile\n",
    "\n",
    "def mafft_addfull(aln1,aln2, outprofile , submat = None):\n",
    "    \"\"\"\n",
    "    this function aligns two alignments using MAFFT\n",
    "    \"\"\"\n",
    "    #make profile\n",
    "    profile = aln1 + '.profile'\n",
    "    if submat:\n",
    "        cmd = 'mafft --textmatrix {} --addfull {} --keeplength {} > {}'.format(submat, aln1,aln2, outprofile)\n",
    "    else:\n",
    "        cmd = 'mafft --addfull  {} --keeplength {} > {}'.format(aln1,aln2, outprofile)\n",
    "    print(cmd)\n",
    "\n",
    "    subprocess.run(cmd , shell = True)\n",
    "    return outprofile\n",
    "\n",
    "\n",
    "def sub2fasta( sub, outfile , fastacol1='qaln' , fastacol2='taln' ):\n",
    "    with open(outfile, 'w') as f:\n",
    "        f.write('>' + sub['query'] + '\\n')\n",
    "        f.write(sub[fastacol1] + '\\n')\n",
    "        f.write('>' + sub['target'] + '\\n')\n",
    "        f.write(sub[fastacol2] + '\\n')    \n",
    "    return outfile\n",
    "\n",
    "def retalns(allvall, leafname1,leafname2):\n",
    "    sub = allvall[allvall['query'].isin( leafname1)]\n",
    "    sub = sub[sub['target'].isin(leafname2)]\n",
    "    sub = sub[sub['query'] != sub['target']]\n",
    "    #get max prot lenght aligned\n",
    "    sub['alnlen'] = sub.apply(lambda x: max(x['qend'] - x['qstart'] , x['tend'] - x['tstart']) , axis = 1)\n",
    "    sub = sub[sub['alnlen'] == sub['alnlen'].max()]\n",
    "    if len(sub)==0:\n",
    "        print(leafname1, leafname2)\n",
    "        raise Exception('no sub')\n",
    "    return sub.iloc[0]\n",
    "\n",
    "def get_fasta_leafset(fasta):\n",
    "    \"\"\"\n",
    "    this function returns the leafset of a fasta file\n",
    "    \"\"\"\n",
    "    aln = SeqIO.parse(fasta, 'fasta')\n",
    "    leafset = []\n",
    "    for s in aln:\n",
    "        leafset.append(s.id)\n",
    "    return leafset\n",
    "\n",
    "#traverse tree from root to leaves recursively\n",
    "def traverse_tree_merge_mafft( treenode, topleafset, allvall , alnfolder , submat = None , verbose = False):\n",
    "    \"\"\"\n",
    "    this function traverses a tree from root to leaves recursively\n",
    "    it returns a dictionary with the iteratively built alignment\n",
    "    \"\"\"\n",
    "    if verbose == True:\n",
    "        print('traverse', treenode.name , treenode.is_leaf() , treenode.leafset)\n",
    "    \n",
    "    if treenode.is_leaf():\n",
    "        print(treenode, treenode.name)\n",
    "        topleafset.remove(treenode.name)\n",
    "        #if the node is a leaf, then we need to add it to the alignment with one of the pivots in the current leafset\n",
    "        #select the alignment of the leaf with itself\n",
    "        sub = allvall[allvall['query'].isin( [treenode.name] )]\n",
    "        sub = sub.iloc[0]\n",
    "        \n",
    "        assert len(sub) > 0\n",
    "\n",
    "\n",
    "        with open(alnfolder + treenode.name + '_inter.fasta', 'w') as f:\n",
    "            f.write('>' + sub['query'] + '\\n')\n",
    "            f.write(sub['AAq'] + '\\n')\n",
    "        with open(alnfolder + treenode.name + '_inter.3di.fasta', 'w') as f:\n",
    "            f.write('>' + sub['query'] + '\\n')\n",
    "            f.write(sub['3diq'] + '\\n')\n",
    "        treenode.aln = alnfolder + treenode.name + '_inter.fasta'\n",
    "        treenode.aln3di = alnfolder + treenode.name + '_inter.3di.fasta'\n",
    "        return treenode.aln, treenode.aln3di\n",
    "    \n",
    "    else:\n",
    "        childalns3di = {}\n",
    "        childalnsAA = {}\n",
    "       \n",
    "        #treenode.leafset = get_leafset(treenode)\n",
    "        #get the intersection of the child leafsets\n",
    "        treenode.leafset = get_leafset(treenode)\n",
    "        children = treenode.get_children()\n",
    "        \n",
    "        if len(children) == 2 and children[0].is_leaf() and children[1].is_leaf():\n",
    "            #treat the case of a cherry\n",
    "            print('cherry', children[0].name , children[1].name)\n",
    "            treenode.aln = sub2fasta( retalns(allvall, [children[0].name] , [children[1].name]) , alnfolder + treenode.name + '_inter.fasta')\n",
    "            treenode.aln3di = sub2fasta( retalns(allvall, [children[0].name] , [children[1].name]) , alnfolder + treenode.name + '_inter.3di.fasta' , fastacol1='3di_qaln_mode2' , fastacol2='3di_taln_mode2')\n",
    "            return treenode.aln, treenode.aln3di\n",
    "        \n",
    "        else:\n",
    "            #not a cherry. one or both sides is a subtree\n",
    "            print('not cherry', treenode.name  )\n",
    "            print( 'children', [c.name for c in children])\n",
    "            for c in treenode.get_children():\n",
    "                #make sub aln for each child\n",
    "                if verbose == True:\n",
    "                    print('traverse', c.name , c.is_leaf() , c.leafset)\n",
    "                if not c.aln:\n",
    "                    c.aln,c.aln3di = traverse_tree_merge_mafft(c , treenode.leafset , allvall, alnfolder , verbose = verbose)\n",
    "                childalnsAA[c] = { 'fasta': c.aln  }\n",
    "                childalns3di[c] = { 'fasta': c.aln3di  }\n",
    "            \n",
    "                \n",
    "\n",
    "            if len(children) == 2:\n",
    "                c1,c2 = children\n",
    "                \n",
    "                if c1.is_leaf():\n",
    "                    treenode.aln = mafft_addfull(childalnsAA[c1]['fasta'], childalnsAA[c2]['fasta'], alnfolder + treenode.name + '_inter.fasta' )\n",
    "                    treenode.aln3di = mafft_addfull(childalns3di[c1]['fasta'], childalns3di[c2]['fasta'], alnfolder + treenode.name + '_inter3di.fasta' , submat =submat )\n",
    "\n",
    "                elif c2.is_leaf():\n",
    "                    treenode.aln = mafft_addfull(childalnsAA[c2]['fasta'], childalnsAA[c1]['fasta'], alnfolder + treenode.name + '_inter.fasta' )\n",
    "                    treenode.aln3di = mafft_addfull(childalns3di[c1]['fasta'], childalns3di[c2]['fasta'], alnfolder + treenode.name + '_inter3di.fasta' , submat = submat)\n",
    "\n",
    "                else:\n",
    "                    with open(childalnsAA[c2]['fasta'], 'r') as f:\n",
    "                        c2seqs = f.read().count('>')\n",
    "                    with open(childalnsAA[c1]['fasta'], 'r') as f:\n",
    "                        c1seqs = f.read().count('>')\n",
    "                    \n",
    "                    if c1seqs > c2seqs:\n",
    "                        treenode.aln = mafft_profile(childalnsAA[c1]['fasta'], childalnsAA[c2]['fasta'], alnfolder + treenode.name + '_inter.fasta' )\n",
    "                        treenode.aln3di = mafft_profile(childalns3di[c1]['fasta'], childalns3di[c2]['fasta'], alnfolder + treenode.name + '_inter3di.fasta' , submat =submat )\n",
    "                    else:\n",
    "                        treenode.aln = mafft_profile(childalnsAA[c2]['fasta'], childalnsAA[c1]['fasta'], alnfolder + treenode.name + '_inter.fasta' )\n",
    "                        treenode.aln3di = mafft_profile(childalns3di[c2]['fasta'], childalns3di[c1]['fasta'], alnfolder + treenode.name + '_inter3di.fasta' , submat =submat )\n",
    "            \n",
    "            elif len(children) > 2 and treenode.up == None:\n",
    "                print('final aln')\n",
    "                print('childalnsAA', childalnsAA)\n",
    "                print('childalns3di', childalns3di)\n",
    "                \n",
    "\n",
    "                print([(c.aln,c.aln3di,type(c.aln), type(c.aln3di)) for c in children])\n",
    "                \n",
    "                children = [c for c in treenode.get_children() if c.aln and c.aln3di]\n",
    "                \n",
    "                for c in children:\n",
    "                    #print alns\n",
    "                    print('aln' + c.name)\n",
    "                    with open(childalnsAA[c]['fasta'], 'r') as f:\n",
    "                        print(f.read())\n",
    "                    print('aln3di' + c.name )\n",
    "                    with open(childalns3di[c]['fasta'], 'r') as f:\n",
    "                        print(f.read())\n",
    "\n",
    "                c1,c2 = children[0],children[1]\n",
    "                if c1.is_leaf():\n",
    "                    rootfasta = mafft_addfull(childalnsAA[c1]['fasta'], childalnsAA[c2]['fasta'], alnfolder + treenode.name + '_root.fasta' )\n",
    "                    rootfasta3di = mafft_addfull(childalns3di[c1]['fasta'], childalns3di[c2]['fasta'], alnfolder + treenode.name + '_inter3di.fasta' , submat =submat )\n",
    "                elif c2.is_leaf():\n",
    "                    rootfasta = mafft_addfull(childalnsAA[c2]['fasta'], childalnsAA[c1]['fasta'], alnfolder + treenode.name + '_root.fasta' )\n",
    "                    rootfasta3di = mafft_addfull(childalns3di[c1]['fasta'], childalns3di[c2]['fasta'], alnfolder + treenode.name + '_inter3di.fasta' , submat = submat)\n",
    "                else:                \n",
    "                    rootfasta = mafft_profile(childalnsAA[c1]['fasta'], childalnsAA[c2]['fasta'], alnfolder + treenode.name + '_root.fasta' )\n",
    "                    rootfasta3di = mafft_profile(childalns3di[c1]['fasta'], childalns3di[c2]['fasta'], alnfolder + treenode.name + '_inter3di.fasta' , submat =  submat)\n",
    "                \n",
    "                print('aln1')\n",
    "                with open(rootfasta , 'r') as f:\n",
    "                    print(f.read())\n",
    "                with open(rootfasta3di , 'r') as f:\n",
    "                    print(f.read())\n",
    "                \n",
    "                for i,c in enumerate(children[1:]):\n",
    "                    if c.is_leaf():\n",
    "                        rootfasta = mafft_addfull( childalnsAA[c]['fasta'] , rootfasta, rootfasta +'.iter' )\n",
    "                        rootfasta3di = mafft_addfull( childalns3di[c]['fasta'], rootfasta3di , rootfasta3di+'.iter'  , submat =submat )\n",
    "                    else:\n",
    "                        rootfasta = mafft_profile(rootfasta, childalnsAA[c]['fasta'], rootfasta+'.iter' )\n",
    "                        rootfasta3di = mafft_profile(rootfasta3di, childalns3di[c]['fasta'], rootfasta3di+'.iter'  , submat =submat )\n",
    "                    print('aln'+str(i))\n",
    "                    with open(rootfasta , 'r') as f:\n",
    "                        print(f.read())\n",
    "                    with open(rootfasta3di , 'r') as f:\n",
    "                        print(f.read())\n",
    "                \n",
    "\n",
    "                treenode.aln = rootfasta\n",
    "                treenode.aln3di = rootfasta3di\n",
    "                \n",
    "                with open(treenode.aln , 'r') as f:\n",
    "                    print(f.read())\n",
    "                with open(treenode.aln3di , 'r') as f:\n",
    "                    print(f.read())\n",
    "                \n",
    "            if verbose == True:\n",
    "                #check if node is root  \n",
    "                if treenode.up == None:\n",
    "                    print('final aln')\n",
    "                    print('childalnsAA', childalnsAA)\n",
    "                    print('childalns3di', childalns3di)\n",
    "            return treenode.aln, treenode.aln3di\n",
    "\n",
    "\n",
    "def remove_redundant( alignment ):\n",
    "    \"\"\"\n",
    "    this function removes redundant sequences from an alignment\n",
    "    \"\"\"\n",
    "    aln = SeqIO.parse(alignment, 'fasta')\n",
    "    seqs = []\n",
    "    ids = []\n",
    "    for s in aln:\n",
    "        if s.id not in ids:\n",
    "            seqs.append(s)\n",
    "            ids.append(s.id)\n",
    "    \n",
    "    with open(alignment, 'w') as f:\n",
    "        for s in seqs:\n",
    "            f.write('>' + s.id + '\\n')\n",
    "            f.write(str(s.seq) + '\\n')\n",
    "    return alignment\n",
    "\n",
    "#remove all alns except the final merged one\n",
    "def cleanup( filedir ):\n",
    "    \"\"\"\n",
    "    this function removes all alns except the final merged one\n",
    "    \"\"\"\n",
    "    for f in glob.glob(filedir + '*inter.fasta'):\n",
    "        os.remove(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#traverse tree from root to leaves recursively\n",
    "def traverse_tree_merge( treenode, topleafset, allvall , alnfolder , verbose = False):\n",
    "    \"\"\"\n",
    "    this function traverses a tree from root to leaves recursively\n",
    "    it returns a dictionary with the iteratively built alignment\n",
    "    \"\"\"\n",
    "    if verbose == True:\n",
    "        print('traverse', treenode.name , treenode.is_leaf() , treenode.leafset)\n",
    "    \n",
    "    if treenode.is_leaf():\n",
    "        topleafset.remove(treenode.name)\n",
    "        #if the node is a leaf, then we need to add it to the alignment with one of the pivots in the current leafset\n",
    "        sub = retalns(allvall, [treenode.name] , topleafset)  \n",
    "        treenode.aln = sub2fasta(sub, alnfolder + treenode.name + '_inter.fasta')\n",
    "        treenode.aln3di = sub2fasta(sub, alnfolder + treenode.name + '_inter.3di.fasta' , fastacol1='3di_qaln_mode2' , fastacol2='3di_taln_mode2')\n",
    "        return treenode.aln, treenode.aln3di\n",
    "    \n",
    "    else:\n",
    "        childalns3di = {}\n",
    "        childalnsAA = {}\n",
    "        bridges3di = {}\n",
    "        bridgesAA = {}\n",
    "        #treenode.leafset = get_leafset(treenode)\n",
    "        #get the intersection of the child leafsets\n",
    "        treenode.leafset = get_leafset(treenode)\n",
    "        children = treenode.get_children()\n",
    "        \n",
    "        if len(children) == 2 and children[0].is_leaf() and children[1].is_leaf():\n",
    "            #treat the case of a cherry\n",
    "            print('cherry', children[0].name , children[1].name)\n",
    "            treenode.aln = sub2fasta( retalns(allvall, [children[0].name] , [children[1].name]) , alnfolder + treenode.name + '_inter.fasta')\n",
    "            treenode.aln3di = sub2fasta( retalns(allvall, [children[0].name] , [children[1].name]) , alnfolder + treenode.name + '_inter.3di.fasta' , fastacol1='3di_qaln_mode2' , fastacol2='3di_taln_mode2')\n",
    "            return treenode.aln, treenode.aln3di\n",
    "        \n",
    "        else:\n",
    "            #not a cherry. one or both sides is a subtree\n",
    "            print('not cherry', treenode.name  )\n",
    "            print( 'children', [c.name for c in children])\n",
    "            for c in treenode.get_children():\n",
    "                #make sub aln for each child\n",
    "                if verbose == True:\n",
    "                    print('traverse', c.name , c.is_leaf() , c.leafset)\n",
    "                if not c.aln:\n",
    "                    c.aln,c.aln3di = traverse_tree_merge(c , treenode.leafset , allvall, alnfolder , verbose = verbose)\n",
    "                childalnsAA[c] = { 'fasta': c.aln , 'protset':set(get_fasta_leafset(c.aln) ) }\n",
    "                childalns3di[c] = { 'fasta': c.aln3di , 'protset':set(get_fasta_leafset(c.aln3di) ) }\n",
    "            \n",
    "            \"\"\"\n",
    "            for c1,c2 in itertools.combinations(treenode.get_children(),2):\n",
    "                bridge = retalns(allvall, childalnsAA[c1]['protset'] , childalnsAA[c2]['protset'] )\n",
    "                bridgesAA[(c1,c2)] = { 'fasta': sub2fasta(bridge, alnfolder + treenode.name + '_bridge.fasta') , 'protset':set([bridge.query , bridge.target]) }\n",
    "                bridges3di[(c1,c2)] = { 'fasta' : sub2fasta(bridge, alnfolder + treenode.name + '_bridge.3di.fasta' , fastacol1='3di_qaln_mode2' , fastacol2='3di_taln_mode2') , 'protset':set([bridge.query, bridge.target]) }\n",
    "            \n",
    "            #successively merge the alignments of the children\n",
    "            for i, c in enumerate(itertools.combinations(treenode.get_children(),2)):\n",
    "                c1,c2 = c\n",
    "                if verbose == True:\n",
    "                    print('merge', c1.name , c2.name)\n",
    "                if i == 0:\n",
    "                    #first merge\n",
    "                    try:\n",
    "                        print('first merge')\n",
    "                        \n",
    "                        alnAA = mergealns( childalnsAA[c1]['fasta'], bridgesAA[(c1,c2)]['fasta'] ,alnfolder + treenode.name + '_inter.fasta' , verbose=verbose)\n",
    "                        aln3di = mergealns( childalns3di[c1]['fasta'], bridges3di[(c1,c2)]['fasta'] ,alnfolder + treenode.name + '_inter3di.fasta', verbose=verbose)\n",
    "\n",
    "                        print('2 merge')\n",
    "\n",
    "                        alnAA = mergealns( childalnsAA[c2]['fasta'], alnAA , alnfolder + treenode.name + '_inter.fasta' , verbose=verbose)\n",
    "                        aln3di = mergealns( childalns3di[c2]['fasta'], aln3di ,alnfolder + treenode.name + '_inter3di.fasta', verbose=verbose)\n",
    "\n",
    "                    except:\n",
    "                        print( treenode , childalnsAA , childalns3di , bridgesAA , bridges3di)\n",
    "                        raise Exception('merge error 1')\n",
    "                else:\n",
    "                    try:\n",
    "                        print('3 merge')\n",
    "\n",
    "                        alnAA = mergealns( childalnsAA[c1]['fasta'], bridgesAA[(c1,c2)]['fasta'] , alnfolder + treenode.name + '_inter.fasta' , verbose=verbose)\n",
    "                        alnAA = mergealns( childalnsAA[c1]['fasta'], alnAA , alnfolder + treenode.name + '_inter.fasta' , verbose=verbose)\n",
    "\n",
    "                        print('4 merge')\n",
    "\n",
    "                        aln3di = mergealns( childalns3di[c2]['fasta'], bridges3di[(c1,c2)]['fasta'] , alnfolder + treenode.name + '_inter3di.fasta'   , verbose=verbose)\n",
    "                        aln3di = mergealns( childalns3di[c2]['fasta'], aln3di , alnfolder + treenode.name + '_inter3di.fasta' , verbose=verbose)\n",
    "\n",
    "                        \n",
    "                    except:\n",
    "                        print( treenode )\n",
    "                        print( childalnsAA , childalns3di , bridgesAA , bridges3di)\n",
    "                        raise Exception('merge error 2') \n",
    "            \"\"\"\n",
    "\n",
    "            treenode.aln = alnAA\n",
    "            treenode.aln3di = aln3di\n",
    "            if verbose == True:\n",
    "                #check if node is root  \n",
    "                if treenode.up == None:\n",
    "                    print('final aln')\n",
    "                    print('childalnsAA', childalnsAA)\n",
    "                    print('childalns3di', childalns3di)\n",
    "            return treenode.aln, treenode.aln3di\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def aln_mapping( s1 , s2, maxaln = 0 , coidx1= 0 , coidx2 = 0 , start1 = 0 , start2=0 ,  verbose = False):\n",
    "    #build a dictionary of the positions of the characters in the string\n",
    "    #convolve the strings\n",
    "    maxaln, maxcount = convolve_strings(s1,s2)\n",
    "\n",
    "    if start1 != 0 or start2 != 0:\n",
    "        print('start', start1, start2)\n",
    "    if verbose == True:\n",
    "        print('maxaln', maxaln, maxcount)\n",
    "    \n",
    "    #find starting points\n",
    "    if len(s1) < len(s2):\n",
    "        if maxaln < 0:\n",
    "            coidx1 = np.abs(maxaln)\n",
    "            coidx2 = 0\n",
    "        else:\n",
    "            coidx1 = 0\n",
    "            coidx2 = maxaln\n",
    "    else:\n",
    "        if maxaln < 0:\n",
    "            coidx1 = 0\n",
    "            coidx2 = np.abs(maxaln)\n",
    "        else:\n",
    "            coidx1 = maxaln\n",
    "            coidx2 = 0\n",
    "    \n",
    "    substr1 = s1[coidx1:]\n",
    "    substr2 = s2[coidx2:]\n",
    "\n",
    "    print('substr1', substr1)\n",
    "    print('substr2', substr2)\n",
    "    print('coidx1', coidx1)\n",
    "    print('coidx2', coidx2)\n",
    "    \n",
    "    #find equivalent positions in the strings\n",
    "    maps1 = {}\n",
    "    maps2 = {}\n",
    "\n",
    "    oppositemap1 = {}\n",
    "    oppositemap2 = {}\n",
    "\n",
    "    for i, char in enumerate(substr1):\n",
    "        if substr2[i] == char:\n",
    "            maps1[i+coidx1] = i + start1\n",
    "            maps2[i+coidx2] = i + start2\n",
    "            \n",
    "            oppositemap1[i+coidx1] = i + start2\n",
    "            oppositemap2[i+coidx2] = i + start1\n",
    "        else:\n",
    "            #if there is mismatch convolve the remaining strings\n",
    "            print('mismatch')\n",
    "            sub1,sub2 , om1 , om2 = aln_mapping( substr1[i:] , substr2[i:], maxaln = 0 , start1= start1+coidx1+i , start2 = start2+coidx2+i , verbose = False)\n",
    "            maps1.update( sub1 )\n",
    "            maps2.update( sub2 )\n",
    "            oppositemap1.update( om1 )\n",
    "            oppositemap2.update( om2 )\n",
    "            break\n",
    "    \n",
    "    return maps1, maps2 , oppositemap1, oppositemap2\n",
    "\n",
    "def aln_mapping_full( s1 , s2, maxaln = 0 , coidx1= 0 , coidx2 = 0 , verbose = False):\n",
    "    maps1, maps2 , om1 , om2 = aln_mapping( s1 , s2, maxaln = 0 , coidx1= 0 , coidx2 = 0 , verbose = False)\n",
    "    #add the reverse mapping\n",
    "    revmap1 = { v:k for k,v in maps1.items()}\n",
    "    revmap2 = { v:k for k,v in maps2.items()}\n",
    "\n",
    "    revmapom1 = { v:k for k,v in om1.items()}   \n",
    "    revmapom2 = { v:k for k,v in om2.items()}\n",
    "    \n",
    "    return maps1, maps2, om1, om2 , revmap1, revmap2 , revmapom1, revmapom2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def convolve_strings(str1, str2):\n",
    "    # Determine the lengths of the strings\n",
    "    len1, len2 = len(str1), len(str2)\n",
    "\n",
    "    if len(str1) < len(str2):\n",
    "        str1, str2 = str2, str1\n",
    "        len1, len2 = len2, len1\n",
    "    max_alignment = 0\n",
    "    max_count = 0\n",
    "    # Slide str2 over str1, starting with one character overlap\n",
    "    # and continue until str2 is again overlapping by just one character\n",
    "    for i in range(-len2 + 1, len1):\n",
    "        count = 0\n",
    "        for j in range(len2):\n",
    "            if 0 <= i + j < len1 and str1[i + j] == str2[j]:\n",
    "                count += 1\n",
    "        if count > max_count:\n",
    "            max_count = count\n",
    "            max_alignment = i\n",
    "    return max_alignment, max_count\n",
    "\n",
    "\n",
    "def alnchop(s1,s2,rawaln1,rawaln2,aln1,aln2,maxaln = 0):\n",
    "\n",
    "    #align the two sequences\n",
    "    coidx1 = 0\n",
    "    coidx2 = 0\n",
    "\n",
    "    if len(s1) < len(s2):\n",
    "        if maxaln < 0:\n",
    "            coidx1 = np.abs(maxaln)\n",
    "            coidx2 = 0\n",
    "        else:\n",
    "            coidx1 = 0\n",
    "            coidx2 = maxaln\n",
    "    else:\n",
    "        if maxaln < 0:\n",
    "            coidx1 = 0\n",
    "            coidx2 = np.abs(maxaln)\n",
    "        else:\n",
    "            coidx1 = maxaln\n",
    "            coidx2 = 0\n",
    "    discardcount = 0 \n",
    "    rawaln1 = iter(rawaln1)\n",
    "    while coidx1 > 0:\n",
    "        rchar1 = next(rawaln1)\n",
    "        discardcount += 1\n",
    "        if rchar1 != '-':\n",
    "            coidx1 -= 1\n",
    "    aln1 = aln1[discardcount:]\n",
    "\n",
    "\n",
    "    discardcount = 0 \n",
    "    rawaln2 = iter(rawaln2)\n",
    "    while coidx1 > 0:\n",
    "        rchar2 = next(rawaln2)\n",
    "        discardcount += 1\n",
    "        if rchar2 != '-':\n",
    "            coidx2 -= 1\n",
    "    aln2 = aln2[discardcount:]\n",
    "\n",
    "\n",
    "    rawaln1 = ''.join([ s for s in iter(rawaln1)])\n",
    "    rawaln2 = ''.join([ s for s in iter(rawaln2)])\n",
    "\n",
    "    return aln1, aln2, rawaln1, rawaln2\n",
    "\n",
    "\n",
    "def mergealns( aln1f, aln2f, outfile , verbose = False):\n",
    "    if set(get_fasta_leafset(aln1f)) == set(get_fasta_leafset(aln2f)):\n",
    "        print('identical')\n",
    "        return aln1f\n",
    "\n",
    "    #find sequences in common between the two alignments\n",
    "    aln1 = SeqIO.parse(aln1f, 'fasta')\n",
    "    aln2 = SeqIO.parse(aln2f, 'fasta')\n",
    "    ids1 = {s.id:str(s.seq) for s in aln1}\n",
    "    ids2 = {s.id:str(s.seq) for s in aln2}\n",
    "    aln1 = SeqIO.parse(aln1f, 'fasta')\n",
    "    aln2 = SeqIO.parse(aln2f, 'fasta')\n",
    "    idlist = [ s.id for s in aln1] + [ s.id for s in aln2]\n",
    "    commonids = set(ids1.keys()).intersection(set(ids2.keys()))\n",
    "    try:\n",
    "        assert len(commonids) > 0\n",
    "    except:\n",
    "        print('no common ids')\n",
    "        print('ids1', ids1)\n",
    "        print('ids2', ids2)\n",
    "        raise Exception('no common ids')\n",
    "    #transform both alignments into numpy matrices\n",
    "    aln1 = SeqIO.parse(aln1f, 'fasta')\n",
    "    aln2 = SeqIO.parse(aln2f, 'fasta')\n",
    "    aln1 = np.array([ list(str(s.seq)) for s in aln1])\n",
    "    aln2 = np.array([ list(str(s.seq)) for s in aln2])\n",
    "    nrows1 = aln1.shape[0]\n",
    "    nrows2 = aln2.shape[0]\n",
    "\n",
    "    #generate a list of column arrays\n",
    "    aln1 = [ aln1[:,i] for i in range(aln1.shape[1])]\n",
    "    aln2 = [ aln2[:,i] for i in range(aln2.shape[1])]\n",
    "    #find the best common sequence\n",
    "    maxconv = 0\n",
    "    maxaln = 0\n",
    "\n",
    "    print(ids1)\n",
    "    print(ids2)\n",
    "    \n",
    "    for commonid in commonids:\n",
    "        s1t = ids1[commonid]\n",
    "        s2t = ids2[commonid]\n",
    "        s1t = s1t.replace('-','')\n",
    "        s2t = s2t.replace('-','')\n",
    "        #if the common subsequence is not found start by removing the first character of the common sequence\n",
    "        #convolution of the two sequences\n",
    "        aln, count = convolve_strings(s1t,s2t)\n",
    "        if count > maxconv:\n",
    "            maxconv = count\n",
    "            maxaln = aln\n",
    "            ID = commonid\n",
    "            s1 = s1t\n",
    "            s2 = s2t\n",
    "\n",
    "    rawaln1 = ids1[ID]\n",
    "    rawaln2 = ids2[ID]\n",
    "    print('pivot' , ID)\n",
    "    print('s1',s1)\n",
    "    print('s2',s2)\n",
    "\n",
    "    print('rawaln1', rawaln1)\n",
    "    print('rawaln2', rawaln2)\n",
    "    \n",
    "    print('maxaln', maxaln)\n",
    "    #use the sequence convolution to align the two alignment arrays\n",
    "    if len(s1) < len(s2):\n",
    "        if maxaln < 0:\n",
    "            coidx1 = np.abs(maxaln)\n",
    "            coidx2 = 0\n",
    "        else:\n",
    "            coidx1 = 0\n",
    "            coidx2 = maxaln\n",
    "    else:\n",
    "        if maxaln < 0:\n",
    "            coidx1 = 0\n",
    "            coidx2 = np.abs(maxaln)\n",
    "        else:\n",
    "            coidx1 = maxaln\n",
    "            coidx2 = 0\n",
    "    print('coidx1', coidx1)\n",
    "    print('coidx2', coidx2)\n",
    "\n",
    "\n",
    "    #remove the leading gaps\n",
    "    for i in range(len(rawaln1)):\n",
    "        if rawaln1 != '-':\n",
    "            break\n",
    "    aln1 = aln1[i:]\n",
    "    rawaln1 = rawaln1[i:]\n",
    "    for i in range(len(rawaln1)):\n",
    "        if rawaln1 != '-':\n",
    "            break\n",
    "    aln2 = aln2[i:]\n",
    "    rawaln2 = rawaln2[i:]\n",
    "\n",
    "    #remove the trailing gaps\n",
    "    for i in range(len(rawaln1)):\n",
    "        if rawaln1[-i] != '-':\n",
    "            break\n",
    "    aln1 = aln1[:len(rawaln1)-i]\n",
    "    rawaln1 = rawaln1[:len(rawaln1)-i]\n",
    "    for i in range(len(rawaln2)):\n",
    "        if rawaln2[-i] != '-':\n",
    "            break\n",
    "    aln2 = aln2[:len(rawaln2)-i]\n",
    "    rawaln2 = rawaln2[:len(rawaln2)-i]\n",
    "    \n",
    "\n",
    "    #construct alignment with common sequence\n",
    "\n",
    "    rawaln1 = iter(rawaln1)\n",
    "    rawaln2 = iter(rawaln2)    \n",
    "    char1 = next(rawaln1)\n",
    "    char2 = next(rawaln2)\n",
    "    i = 1\n",
    "    j = 1 \n",
    "    pchar1 = char1\n",
    "    pchar2 = char2\n",
    "    newaln1 = []\n",
    "    newaln2 = []\n",
    "    convolved = False\n",
    "    while True:\n",
    "        try:\n",
    "            if pchar1 == '-' and char1 != '-':\n",
    "                print('end insertion1')\n",
    "\n",
    "            if pchar2 == '-' and char2 != '-':\n",
    "                print('end insertion2')\n",
    "\n",
    "            if char1 == '-' and char2 != '-':\n",
    "                print('insertion1')\n",
    "                newaln2.append(['-']*nrows2)\n",
    "                newaln1.append(aln1[i])\n",
    "                pchar1 = char1\n",
    "                char1 = next(rawaln1)\n",
    "                i +=1\n",
    "\n",
    "            elif char2 == '-' and char1 != '-':\n",
    "                print('insertion2')\n",
    "                newaln2.append(aln2[j])\n",
    "                newaln1.append(['-']*nrows1)\n",
    "                pchar2 = char2\n",
    "                char2 = next(rawaln2)\n",
    "                j +=1\n",
    "\n",
    "            elif char1 == char2 and char1 != '-' and char2 != '-':\n",
    "                char1 = next(rawaln1)\n",
    "                char2 = next(rawaln2)\n",
    "                newaln2.append(aln2[j])\n",
    "                newaln1.append(aln1[i])\n",
    "                pchar1 = char1\n",
    "                pchar2 = char2\n",
    "                j+= 1\n",
    "                i+= 1\n",
    "\n",
    "            elif char1 != '-' and char2 != '-' and char1 != char2:\n",
    "                convolved = True\n",
    "                print('mismatch')\n",
    "                #mismatch reconvolve remaining strings\n",
    "                rawaln1 = ''.join([ s for s in iter(rawaln1)])\n",
    "                rawaln2 = ''.join([ s for s in iter(rawaln2)])\n",
    "                s1 = rawaln1.replace('-','')\n",
    "                s2 = rawaln2.replace('-','')\n",
    "\n",
    "                maxaln, count = convolve_strings(s1,s2)\n",
    "                if len(s1) < len(s2):\n",
    "                    if maxaln < 0:\n",
    "                        coidx1 = np.abs(maxaln)\n",
    "                        coidx2 = 0\n",
    "                    else:\n",
    "                        coidx1 = 0\n",
    "                        coidx2 = maxaln\n",
    "                else:\n",
    "                    if maxaln < 0:\n",
    "                        coidx1 = 0\n",
    "                        coidx2 = np.abs(maxaln)\n",
    "                    else:\n",
    "                        coidx1 = maxaln\n",
    "                        coidx2 = 0\n",
    "                \n",
    "                rawaln1 = iter(rawaln1)\n",
    "                discardcount1 = 0\n",
    "                count1 = 0\n",
    "                while count1 < coidx1:\n",
    "                    discardcount1 += 1\n",
    "                    char1 = next(rawaln1)\n",
    "                    if char1 != '-':\n",
    "                        count1 += 1\n",
    "                aln1 = aln1[discardcount1:]\n",
    "                rawaln2 = iter(rawaln2) \n",
    "                discardcount2 = 0\n",
    "                count2 = 0\n",
    "                while count2 < coidx2:\n",
    "                    discardcount2 += 1\n",
    "                    char2 = next(rawaln2)\n",
    "                    if char2 != '-':\n",
    "                        count2 += 1\n",
    "                aln2 = aln2[discardcount2:]\n",
    "\n",
    "                char1 = next(rawaln1)\n",
    "                char2 = next(rawaln2)\n",
    "                \n",
    "\n",
    "                i = 0 \n",
    "                j = 0\n",
    "                print('char1', char1)\n",
    "                print('char2', char2)\n",
    "\n",
    "                print( 'newaln1' , newaln1)\n",
    "\n",
    "                print( 'newaln2' , newaln2)\n",
    "                \n",
    "            else:\n",
    "                print('end')\n",
    "                break\n",
    "        \n",
    "        except StopIteration:\n",
    "            break\n",
    "    try:\n",
    "        newaln1 = np.vstack(newaln1).T\n",
    "        newaln2 = np.vstack(newaln2).T\n",
    "        newaln = np.concatenate((newaln1, newaln2), axis = 0)\n",
    "    except:\n",
    "        print( 'aln err')\n",
    "        print( 'char1', char1 )\n",
    "        print( 'char2' , char2 )\n",
    "        print( 'aln1', aln1 ) \n",
    "        print( 'aln2' , aln2 )\n",
    "        print( 'maxaln'  , maxaln)\n",
    "        raise Exception('aln err: newaln not created')\n",
    "    #write out the new alignment\n",
    "    with open(outfile, 'w') as f:\n",
    "        for i in range(newaln.shape[0]):\n",
    "            #print('>' + idlist[i] + '\\n' + ''.join(list(newaln[i,:])) + '\\n')\n",
    "            f.write('>' + idlist[i] + '\\n')\n",
    "            f.write(''.join(list(newaln[i,:])) + '\\n')\n",
    "    return outfile  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import toytree\n",
    "import os\n",
    "import pandas as pd\n",
    "import glob\n",
    "\n",
    "infolder = '../../testcat/4.10.530/'\n",
    "alndf = pd.read_table(infolder + 'allvall_1.csv', header = None)\n",
    "\n",
    "mapper3di, mapperAA = read_dbfiles3di( infolder+\"outdb\" , infolder+\"outdb_ss\")\n",
    "#add the 3di alignment to the dataframe\n",
    "columns = 'query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,lddt,qaln,taln,cigar,lntmscore'.split(',')\n",
    "alndf.columns = columns\n",
    "\n",
    "alndf['3diq']= alndf['query'].map(mapper3di)\n",
    "alndf['3dit']= alndf['target'].map(mapper3di)\n",
    "alndf['AAq']= alndf['query'].map(mapperAA)\n",
    "alndf['AAt']= alndf['target'].map(mapperAA)\n",
    "\n",
    "#output a fasta with the 3di sequences\n",
    "res = alndf.apply(calc_fident_crossaln , axis = 1)\n",
    "alndf = pd.concat([alndf,res] , axis = 1)\n",
    "\n",
    "with open(infolder + '3diseqs.fasta' , 'w') as out:\n",
    "    for seq in alndf['query'].unique():\n",
    "        out.write('>'+seq.replace('.pdb', '' )+'\\n')\n",
    "        out.write(mapper3di[seq]+'\\n')\n",
    "alndf['query'] = alndf['query'].map(lambda x :x.replace('.pdb', ''))\n",
    "alndf['target'] = alndf['target'].map(lambda x :x.replace('.pdb', ''))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    query  target  fident  alnlen  mismatch  gapopen  qstart  qend  tstart  \\\n",
      "0  O00602  O00602   1.000     217         0        0       1   217       1   \n",
      "1  O00602  Q15485   0.850     214        32        0       4   217       1   \n",
      "2  O00602  Q8N539   0.495     220       104        3       4   217       1   \n",
      "3  O00602  O75636   0.523     214        96        3       5   215       1   \n",
      "4  O00602  Q9U8W8   0.472     218       108        3       6   217       2   \n",
      "\n",
      "   tend  ...                                               qaln  \\\n",
      "0   217  ...  SCATGPRNCKDLLDRGYFLSGWHTIYLPDCRPLTVLCDMDTDGGGW...   \n",
      "1   214  ...  TGPRNCKDLLDRGYFLSGWHTIYLPDCRPLTVLCDMDTDGGGWTVF...   \n",
      "2   219  ...  TGPRNCKDLLDRGYFLSGWHTIYLPDCR-PLTVLCDMDTDGGGWTV...   \n",
      "3   211  ...  GPRNCKDLLDRGYFLSGWHTIYLPDCRPLTVLCDMDTDGGGWTVFQ...   \n",
      "4   218  ...  PRNCKDLLDRGYFLSGWHTIYLPD---CRPLTVLCDMDTDGGGWTV...   \n",
      "\n",
      "                                                taln               cigar  \\\n",
      "0  SCATGPRNCKDLLDRGYFLSGWHTIYLPDCRPLTVLCDMDTDGGGW...                217M   \n",
      "1  TGPRTCKDLLDRGHFLSGWHTIYLPDCRPLTVLCDMDTDGGGWTVF...                214M   \n",
      "2  SRPRDCLDVLLSGQQDDGVYSVFPTHYPAGFQVYCDMRTDGGGWTV...  28M1D85M5D14M1I86M   \n",
      "3  GPRNCRELLSQGATLSGWYHLCLPEGRALPVFCDMDTEGGGWLVFQ...   179M1D2M2D3M3I24M   \n",
      "4  PTDCADILLNGYRSSGGYRIWPKSWMTVGTLNVYCDMETDGGGWTV...  24M3D29M3D72M1I86M   \n",
      "\n",
      "  lntmscore                                               3diq  \\\n",
      "0    1.0000  DLVPFAAFQLSVVVVPPQCWAWDWHAFPVRDIDIFTWDNPPPNTRW...   \n",
      "1    0.9945  DLVPFAAFQLSVVVVPPQCWAWDWHAFPVRDIDIFTWDNPPPNTRW...   \n",
      "2    0.9651  DLVPFAAFQLSVVVVPPQCWAWDWHAFPVRDIDIFTWDNPPPNTRW...   \n",
      "3    0.9638  DLVPFAAFQLSVVVVPPQCWAWDWHAFPVRDIDIFTWDNPPPNTRW...   \n",
      "4    0.9517  DLVPFAAFQLSVVVVPPQCWAWDWHAFPVRDIDIFTWDNPPPNTRW...   \n",
      "\n",
      "                                                3dit  \\\n",
      "0  DLVPFAAFQLSVVVVPPQCWAWDWHAFPVRDIDIFTWDNPPPNTRW...   \n",
      "1  DFAAFLQVVVVVPPQDWAWDWHAFPVRDIDIFTWANPPPNTGWTWQ...   \n",
      "2  DAAQFQQSVVVVPDQDWAKDWHAFPVRRVTDIFTWFCPPPRTRWTW...   \n",
      "3  DALFQVVVVVVPPQDWAWDWHQFPVRDIAIFTWDCDPPNTGKTWQW...   \n",
      "4  DAQALLSVVVVPDQDWAWDWHDQVVCVPVNIDIFTWDCPPPRTSWT...   \n",
      "\n",
      "                                                 AAq  \\\n",
      "0  SCATGPRNCKDLLDRGYFLSGWHTIYLPDCRPLTVLCDMDTDGGGW...   \n",
      "1  SCATGPRNCKDLLDRGYFLSGWHTIYLPDCRPLTVLCDMDTDGGGW...   \n",
      "2  SCATGPRNCKDLLDRGYFLSGWHTIYLPDCRPLTVLCDMDTDGGGW...   \n",
      "3  SCATGPRNCKDLLDRGYFLSGWHTIYLPDCRPLTVLCDMDTDGGGW...   \n",
      "4  SCATGPRNCKDLLDRGYFLSGWHTIYLPDCRPLTVLCDMDTDGGGW...   \n",
      "\n",
      "                                                 AAt  \\\n",
      "0  SCATGPRNCKDLLDRGYFLSGWHTIYLPDCRPLTVLCDMDTDGGGW...   \n",
      "1  TGPRTCKDLLDRGHFLSGWHTIYLPDCRPLTVLCDMDTDGGGWTVF...   \n",
      "2  SRPRDCLDVLLSGQQDDGVYSVFPTHYPAGFQVYCDMRTDGGGWTV...   \n",
      "3  GPRNCRELLSQGATLSGWYHLCLPEGRALPVFCDMDTEGGGWLVFQ...   \n",
      "4  DPTDCADILLNGYRSSGGYRIWPKSWMTVGTLNVYCDMETDGGGWT...   \n",
      "\n",
      "                                      3di_qaln_mode2  \\\n",
      "0  DLVPFAAFQLSVVVVPPQCWAWDWHAFPVRDIDIFTWDNPPPNTRW...   \n",
      "1  PFAAFQLSVVVVPPQCWAWDWHAFPVRDIDIFTWDNPPPNTRWTWQ...   \n",
      "2  PFAAFQLSVVVVPPQCWAWDWHAFPVRD-IDIFTWDNPPPNTRWTW...   \n",
      "3  FAAFQLSVVVVPPQCWAWDWHAFPVRDIDIFTWDNPPPNTRWTWQW...   \n",
      "4  AAFQLSVVVVPPQCWAWDWHAFPV---RDIDIFTWDNPPPNTRWTW...   \n",
      "\n",
      "                                      3di_taln_mode2  \n",
      "0  DLVPFAAFQLSVVVVPPQCWAWDWHAFPVRDIDIFTWDNPPPNTRW...  \n",
      "1  DFAAFLQVVVVVPPQDWAWDWHAFPVRDIDIFTWANPPPNTGWTWQ...  \n",
      "2  DAAQFQQSVVVVPDQDWAKDWHAFPVRRVTDIFTWFCPPPRTRWTW...  \n",
      "3  DALFQVVVVVVPPQDWAWDWHQFPVRDIAIFTWDCDPPNTGKTWQW...  \n",
      "4  AQALLSVVVVPDQDWAWDWHDQVVCVPVNIDIFTWDCPPPRTSWTF...  \n",
      "\n",
      "[5 rows x 23 columns]\n"
     ]
    }
   ],
   "source": [
    "print(alndf.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "      /-Q8N539\n",
      "   /-|\n",
      "  |   \\-Q9U8W8\n",
      "  |\n",
      "  |   /-P24821\n",
      "  |--|\n",
      "  |  |   /-O75636\n",
      "  |   \\-|\n",
      "  |     |   /-Q15485\n",
      "--|      \\-|\n",
      "  |         \\-O00602\n",
      "  |\n",
      "  |      /-Q9BY76\n",
      "  |   /-|\n",
      "  |  |  |   /-Q15389\n",
      "  |  |   \\-|\n",
      "  |  |      \\-O15123\n",
      "   \\-|\n",
      "     |      /-P04115\n",
      "     |   /-|\n",
      "     |  |  |   /-P02679\n",
      "     |  |   \\-|\n",
      "      \\-|      \\-O93568\n",
      "        |\n",
      "        |   /-P02671\n",
      "         \\-|\n",
      "           |   /-P02678\n",
      "            \\-|\n",
      "              |   /-P02675\n",
      "               \\-|\n",
      "                  \\-Q02020\n",
      "traverse 29 False None\n",
      "not cherry 29\n",
      "children ['28', '27', '26']\n",
      "traverse 28 False None\n",
      "traverse 28 False None\n",
      "cherry Q8N539 Q9U8W8\n",
      "traverse 27 False None\n",
      "traverse 27 False None\n",
      "not cherry 27\n",
      "children ['P24821', '25']\n",
      "traverse P24821 True None\n",
      "traverse P24821 True None\n",
      "\n",
      "--P24821 P24821\n",
      "traverse 25 False None\n",
      "traverse 25 False None\n",
      "not cherry 25\n",
      "children ['O75636', '22']\n",
      "traverse O75636 True None\n",
      "traverse O75636 True None\n",
      "\n",
      "--O75636 O75636\n",
      "traverse 22 False None\n",
      "traverse 22 False None\n",
      "cherry Q15485 O00602\n",
      "mafft --addfull  ../../testcat/4.10.530/alnscratch/O75636_inter.fasta --keeplength ../../testcat/4.10.530/alnscratch/22_inter.fasta > ../../testcat/4.10.530/alnscratch/25_inter.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    1 / 2\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 2\n",
      "njobc = 3\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --addfull  ../../testcat/4.10.530/alnscratch/O75636_inter.3di.fasta --keeplength ../../testcat/4.10.530/alnscratch/22_inter.3di.fasta > ../../testcat/4.10.530/alnscratch/25_inter3di.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    1 / 2\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 2\n",
      "njobc = 3\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --addfull  ../../testcat/4.10.530/alnscratch/P24821_inter.fasta --keeplength ../../testcat/4.10.530/alnscratch/25_inter.fasta > ../../testcat/4.10.530/alnscratch/27_inter.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    2 / 3\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 3\n",
      "njobc = 4\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "To keep the alignment length, 9 letters were DELETED.\n",
      "To know the positions of deleted letters, rerun the same command with the --mapout option.\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --addfull  ../../testcat/4.10.530/alnscratch/P24821_inter.3di.fasta --keeplength ../../testcat/4.10.530/alnscratch/25_inter3di.fasta > ../../testcat/4.10.530/alnscratch/27_inter3di.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    2 / 3\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 3\n",
      "njobc = 4\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "To keep the alignment length, 10 letters were DELETED.\n",
      "To know the positions of deleted letters, rerun the same command with the --mapout option.\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "traverse 26 False None\n",
      "traverse 26 False None\n",
      "not cherry 26\n",
      "children ['24', '23']\n",
      "traverse 24 False None\n",
      "traverse 24 False None\n",
      "not cherry 24\n",
      "children ['Q9BY76', '21']\n",
      "traverse Q9BY76 True None\n",
      "traverse Q9BY76 True None\n",
      "\n",
      "--Q9BY76 Q9BY76\n",
      "traverse 21 False None\n",
      "traverse 21 False None\n",
      "cherry Q15389 O15123\n",
      "mafft --addfull  ../../testcat/4.10.530/alnscratch/Q9BY76_inter.fasta --keeplength ../../testcat/4.10.530/alnscratch/21_inter.fasta > ../../testcat/4.10.530/alnscratch/24_inter.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    1 / 2\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 2\n",
      "njobc = 3\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "To keep the alignment length, 9 letters were DELETED.\n",
      "To know the positions of deleted letters, rerun the same command with the --mapout option.\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --addfull  ../../testcat/4.10.530/alnscratch/Q9BY76_inter.3di.fasta --keeplength ../../testcat/4.10.530/alnscratch/21_inter.3di.fasta > ../../testcat/4.10.530/alnscratch/24_inter3di.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    1 / 2\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 2\n",
      "njobc = 3\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "To keep the alignment length, 5 letters were DELETED.\n",
      "To know the positions of deleted letters, rerun the same command with the --mapout option.\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "traverse 23 False None\n",
      "traverse 23 False None\n",
      "not cherry 23\n",
      "children ['20', '19']\n",
      "traverse 20 False None\n",
      "traverse 20 False None\n",
      "not cherry 20\n",
      "children ['P04115', '18']\n",
      "traverse P04115 True None\n",
      "traverse P04115 True None\n",
      "\n",
      "--P04115 P04115\n",
      "traverse 18 False None\n",
      "traverse 18 False None\n",
      "cherry P02679 O93568\n",
      "mafft --addfull  ../../testcat/4.10.530/alnscratch/P04115_inter.fasta --keeplength ../../testcat/4.10.530/alnscratch/18_inter.fasta > ../../testcat/4.10.530/alnscratch/20_inter.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    1 / 2\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 2\n",
      "njobc = 3\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "To keep the alignment length, 34 letters were DELETED.\n",
      "To know the positions of deleted letters, rerun the same command with the --mapout option.\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --addfull  ../../testcat/4.10.530/alnscratch/P04115_inter.3di.fasta --keeplength ../../testcat/4.10.530/alnscratch/18_inter.3di.fasta > ../../testcat/4.10.530/alnscratch/20_inter3di.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    1 / 2\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 2\n",
      "njobc = 3\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "To keep the alignment length, 34 letters were DELETED.\n",
      "To know the positions of deleted letters, rerun the same command with the --mapout option.\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "traverse 19 False None\n",
      "traverse 19 False None\n",
      "not cherry 19\n",
      "children ['P02671', '17']\n",
      "traverse P02671 True None\n",
      "traverse P02671 True None\n",
      "\n",
      "--P02671 P02671\n",
      "traverse 17 False None\n",
      "traverse 17 False None\n",
      "not cherry 17\n",
      "children ['P02678', '16']\n",
      "traverse P02678 True None\n",
      "traverse P02678 True None\n",
      "\n",
      "--P02678 P02678\n",
      "traverse 16 False None\n",
      "traverse 16 False None\n",
      "cherry P02675 Q02020\n",
      "mafft --addfull  ../../testcat/4.10.530/alnscratch/P02678_inter.fasta --keeplength ../../testcat/4.10.530/alnscratch/16_inter.fasta > ../../testcat/4.10.530/alnscratch/17_inter.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    1 / 2\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 2\n",
      "njobc = 3\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "To keep the alignment length, 16 letters were DELETED.\n",
      "To know the positions of deleted letters, rerun the same command with the --mapout option.\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --addfull  ../../testcat/4.10.530/alnscratch/P02678_inter.3di.fasta --keeplength ../../testcat/4.10.530/alnscratch/16_inter.3di.fasta > ../../testcat/4.10.530/alnscratch/17_inter3di.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    1 / 2\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 2\n",
      "njobc = 3\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "To keep the alignment length, 16 letters were DELETED.\n",
      "To know the positions of deleted letters, rerun the same command with the --mapout option.\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --addfull  ../../testcat/4.10.530/alnscratch/P02671_inter.fasta --keeplength ../../testcat/4.10.530/alnscratch/17_inter.fasta > ../../testcat/4.10.530/alnscratch/19_inter.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    2 / 3\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 3\n",
      "njobc = 4\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "To keep the alignment length, 3 letters were DELETED.\n",
      "To know the positions of deleted letters, rerun the same command with the --mapout option.\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --addfull  ../../testcat/4.10.530/alnscratch/P02671_inter.3di.fasta --keeplength ../../testcat/4.10.530/alnscratch/17_inter3di.fasta > ../../testcat/4.10.530/alnscratch/19_inter3di.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "    2 / 3\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.520\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 3\n",
      "njobc = 4\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 0 / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "To keep the alignment length, 3 letters were DELETED.\n",
      "To know the positions of deleted letters, rerun the same command with the --mapout option.\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --seed ../../testcat/4.10.530/alnscratch/20_inter.fasta ../../testcat/4.10.530/alnscratch/19_inter.fasta > ../../testcat/4.10.530/alnscratch/23_inter.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 4\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 1600.000000\n",
      "\n",
      "adding 1-2\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 7\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "    0 / 7\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP     6 / 6 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "    0 / 7\n",
      "    0 / 3\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP     6 /6 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "    0 / 7\n",
      "Segment   1/  1    1- 311\n",
      "done 002-001-1  identical.   \n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --seed ../../testcat/4.10.530/alnscratch/20_inter3di.fasta ../../testcat/4.10.530/alnscratch/19_inter3di.fasta > ../../testcat/4.10.530/alnscratch/23_inter3di.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 4\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 1600.000000\n",
      "\n",
      "adding 1-2\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 7\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "    0 / 7\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP     6 / 6 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "    0 / 7\n",
      "    0 / 3\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP     6 /6 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "    0 / 7\n",
      "Segment   1/  1    1- 312\n",
      "done 002-001-1  identical.    identical.   \n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --seed ../../testcat/4.10.530/alnscratch/24_inter.fasta ../../testcat/4.10.530/alnscratch/23_inter.fasta > ../../testcat/4.10.530/alnscratch/26_inter.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 7\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 4900.000000\n",
      "\n",
      "adding 1-2\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 10\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "    0 / 10\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP     9 / 9 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "    0 / 10\n",
      "    0 / 3\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP     9 /9 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "    0 / 10\n",
      "Segment   1/  1    1- 311\n",
      "done 002-001-1  identical.    identical.   \n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --seed ../../testcat/4.10.530/alnscratch/24_inter3di.fasta ../../testcat/4.10.530/alnscratch/23_inter3di.fasta > ../../testcat/4.10.530/alnscratch/26_inter3di.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 7\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 4900.000000\n",
      "\n",
      "adding 1-2\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 10\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "    0 / 10\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP     9 / 9 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "    0 / 10\n",
      "    0 / 3\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP     9 /9 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "    0 / 10\n",
      "Segment   1/  1    1- 317\n",
      "done 002-001-1  identical.    accepted.\n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "final aln\n",
      "childalnsAA {<toytree.TreeNode.TreeNode object at 0x7f18d57b3d60>: {'fasta': '../../testcat/4.10.530/alnscratch/28_inter.fasta'}, <toytree.TreeNode.TreeNode object at 0x7f18d57c4790>: {'fasta': '../../testcat/4.10.530/alnscratch/27_inter.fasta'}, <toytree.TreeNode.TreeNode object at 0x7f18d57c5c30>: {'fasta': '../../testcat/4.10.530/alnscratch/26_inter.fasta'}}\n",
      "childalns3di {<toytree.TreeNode.TreeNode object at 0x7f18d57b3d60>: {'fasta': '../../testcat/4.10.530/alnscratch/28_inter.3di.fasta'}, <toytree.TreeNode.TreeNode object at 0x7f18d57c4790>: {'fasta': '../../testcat/4.10.530/alnscratch/27_inter3di.fasta'}, <toytree.TreeNode.TreeNode object at 0x7f18d57c5c30>: {'fasta': '../../testcat/4.10.530/alnscratch/26_inter3di.fasta'}}\n",
      "[('../../testcat/4.10.530/alnscratch/28_inter.fasta', '../../testcat/4.10.530/alnscratch/28_inter.3di.fasta', <class 'str'>, <class 'str'>), ('../../testcat/4.10.530/alnscratch/27_inter.fasta', '../../testcat/4.10.530/alnscratch/27_inter3di.fasta', <class 'str'>, <class 'str'>), ('../../testcat/4.10.530/alnscratch/26_inter.fasta', '../../testcat/4.10.530/alnscratch/26_inter3di.fasta', <class 'str'>, <class 'str'>)]\n",
      "aln28\n",
      ">Q8N539\n",
      "PRDCLDVLLSGQQDDGVYSVFPTHY--PAGFQVYCDMRTDGGGWTVFQRREDGSVN---FFRGWDAYRDGFGRLTGEHWLGLKRIHALTTQAAYELHVDLEDFENGTAYARYGSFGVGLFSVDPEEDGYPLTVADYSGTAGDSLLKHSGMRFTTKDRDSDHSENNCAAFYRGAWWYRNCHTSNLNGQYLRGAHASYADGVEWSSWTGWQYSLKFSEMKIRPV\n",
      ">Q9U8W8\n",
      "PTDCADILLNGYRSSGGYRIWPKSWMTVGTLNVYCDMETDGGGWTVIQRRGNYGNPSDYFYKPWKNYKLGFGNIEKDFWLGNDRIFALTNQRNYMIRFDLKDKENDTRYAIYQDFWI-----ENEDYLYCLHIGNYSGDAGNSFGRHNGHNFSTIDKDHDTHETHCAQTYKGGWWYDRCHESNLNGLYLNGEHNSYADGIEWRAWKGYHYSLPQVEMKIRPV\n",
      "\n",
      "aln3di28\n",
      ">Q8N539\n",
      "AQFQQSVVVVPDQDWAKDWHAFPVR--RVTDIFTWFCPPPRTRWTWFFKFAAQPDQ---FQDDDVCQQCKDDDRVHGIGNHLVVQQSNCQRAFKKKKKWFAAPPRDIKIKIWRGWHFCNPPPDRVVQQRAIATHDIDMDQDDWPVVLHRAHKFHPPRARECDPDGVCVQLVHIGSAHNGGQAGQSFHQPHFFDDDFQNAGDNCSHRHGGTHTNMMIMIMDHD\n",
      ">Q9U8W8\n",
      "AQALLSVVVVPDQDWAWDWHDQVVCVPVNIDIFTWDCPPPRTSWTFFFKFAPDPDALLVQLDDPVCQQCKDDDNVHGIGNHLVVQQRSCPSAWKKKKKWFAAPVRDIWMKIWGRWHF-----HDVVQQRATATHDMDIPQDPFCVVLHGAHWAHPVGAREDAPDRQCVQQRHTGRAYVGGQARQSFGDPRFFDDRFQNAGHNPSHPHGGTHTNIMIMIMHHP\n",
      "\n",
      "aln27\n",
      ">Q15485\n",
      "TGPRTCKDLLDRGHFLSGWHTIYLPDCRPLTVLCDMDTDGGGWTVFQRRVDGSVDFYRDW\n",
      "ATYKQGFGSRLGEFWLGNDNIHALTAQGTSELRVDLVDFEDNYQFAKYRSFKVADEAEKY\n",
      "NLVLGAFVEGSAGDSLTFHNNQSFSTKDQDNDLNTGNCAVMFQGAWWYKNCHVSNLNGRY\n",
      "LRGTHGSFANGINWKSGKGYNYSYKVSEMKVRPA\n",
      ">O00602\n",
      "TGPRNCKDLLDRGYFLSGWHTIYLPDCRPLTVLCDMDTDGGGWTVFQRRMDGSVDFYRDW\n",
      "AAYKQGFGSQLGEFWLGNDNIHALTAQGSSELRVDLVDFEGNHQFAKYKSFKVADEAEKY\n",
      "KLVLGAFVGGSAGNSLTGHNNNFFSTKDQDNDVSSSNCAEKFQGAWWYADCHASNLNGLY\n",
      "LMGPHESFANGINWSAAKGYKYSYKVSEMKVRPA\n",
      ">O75636\n",
      "-GPRNCRELLSQGATLSGWYHLCLPEGRALPVFCDMDTEGGGWLVFQRRQDGSVDFFRSW\n",
      "SSYRAGFGNQESEFWLGNENLHQLTLQGNWELRVELEDFNGNRTFAHYATFRLLGEVDHY\n",
      "QLALGKFSEGTAGDSLSLHSGRPFTTYDADHDSSNSNCAVIVHGAWWYASCYRSNLNGRY\n",
      "AVSEAAAHKYGIDWASGRGVGHPYRRVRMMLR--\n",
      ">P24821\n",
      "PFPKDCSQAMLNGDTTSGLYTIYLDKAQALEVFCDMTSDGGGWIVFLRRKNGRENFYQNW\n",
      "KAYAAGFGDRREEFWLGLDNLNKITAQGQYELRVDLRD-HGETAFAVYDKFSVGDAKTRY\n",
      "KLKVEGYS-GTAGDSMAYHNGRSFSTFDK-------NCALSYKGAFWYRNCHRVNLMGRY\n",
      "GDNNH---SQGVNWFHWKGHEHSIQFAEMKLRPS\n",
      "\n",
      "aln3di27\n",
      ">Q15485\n",
      "DFAAFLQVVVVVPPQDWAWDWHAFPVRDIDIFTWANPPPNTGWTWQWKFAAQPDQQQDAP\n",
      "VDQQCKDDDSVHTIGNHLVVQQRSQVPAFKKKKKWFAAPVGDIKIWIFRGWHFHDVVLQR\n",
      "AIATHDTDDMQQDDFCVVLHRAHKDHQVRAREQDPDRQCVLQVHIGSAHNGGQAGQSFGC\n",
      "PHFFAPRFQNARHHPSHPHRGTHTNMIIMIIHGD\n",
      ">O00602\n",
      "PFAAFQLSVVVVPPQCWAWDWHAFPVRDIDIFTWDNPPPNTRWTWQWKFAAQPDQQQDAP\n",
      "VCQQCKDDDSVHTIGNHLVVQQRSCVPAFKKKKKWFAAPVGDIKIWIFRGWHFDDVVLQR\n",
      "AIATHDTDDMLLDDWCVVLHRAHKHHQVRAREQDPDRQCVQLVHIGSAHNGGQAGQRFHP\n",
      "PHFFADRFQRARDNDSHVHGGTHTNMIIMIIHHD\n",
      ">O75636\n",
      "-DALFQVVVVVVPPQDWAWDWHQFPVRDIAIFTWDCDPPNTGKTWQWKFADQPDQFQDDL\n",
      "VCQQCKDDDRVHTIGRHLVVQQRSPVPHFKKKKKWFAADVGDIKIWIFRTWHFHDVVQQR\n",
      "AIATHDTDDMPLDDWCVVLHRAHKDHQVRAREQDPDRQCVVQVHIGSAHNHTQAGQSGHD\n",
      "DPHVVRWDQHGRHHCSHCHRRGHTRIMIMIID--\n",
      ">P24821\n",
      "DAAAFQLSVVVVVDQDWAWDWHAQPVR-IDIFTWQNPDPRTRKTWFFKFAAQPDAQQDDP\n",
      "VCQQCKDDDRVYGIGRHLVVQQRSCVNAFKKKKKWFDAP-HDIKMKIFLGWGFHDVVLLR\n",
      "FIAGHDI-DMDQDPFCVVLHGAHKDHPV----NDP---CVVLSHIGSADVGGQAGQRFHD\n",
      "P---AQDFQNAGHNDSRPRRRDHTNIMIMIMDHS\n",
      "\n",
      "aln26\n",
      ">_seed_Q15389\n",
      "--------------------------------------------------FRDCADVYQA\n",
      "GFNKSGIYTIYINNMPEPKKVFCNMDVNGGGWTVIQHREDGSLDFQRGWKEYKMGFGNPS\n",
      "---------------GEYWLGNEFIFAITSQRQ--YMLRIELMDWEGNRAYSQYDRFHIG\n",
      "NEKQNYRLYLKGH-TGTAG----------------KQSSLILHGADFSTKDADNDNC---\n",
      "--MCKCALMLTGGWWFDACGPSNLNGMFYTAGQNHGKL------NGIKWHYFKGPSYSLR\n",
      "STTMMIRPLD--F\n",
      ">_seed_O15123\n",
      "--------------------------------------------------FRDCAEVFKS\n",
      "GHTTNGIYTLTFPNSTEEIKAYCDMEAGGGGWTIIQRREDGSVDFQRTWKEYKVGFGNPS\n",
      "---------------GEYWLGNEFVSQLTNQQR--YVLKIHLKDWEGNEAYSLYEHFYLS\n",
      "SEELNYRIHLKGL-TGTAG----------------KISSISQPGNDFSTKDGDNDKC---\n",
      "--ICKCSQMLTGGWWFDACGPSNLNGMYYPQRQNTNKF------NGIKWYYWKGSGYSLK\n",
      "ATTMMIRPAD--F\n",
      ">_seed_Q9BY76\n",
      "--------------------------------------------------PRDCQELFQV\n",
      "GERQSGLFEIQ-PQGSPPFLVNCKM-TSDGGWTVIQRRHDGSVDFNRPWEAYKAGFGDPH\n",
      "---------------GEFWLGLEKVHSITGDRN--SRLAVQLRDWDGNAELLQFS-VHLG\n",
      "GEDTAYSLQLTA---PVAG----------------QLGATTVPPSGFSTWDQDHDLR---\n",
      "--RDKCAKSLSGGWWFGTCSHSNLNGQYFPQRQKLKK--------GIFWKTWRGRYYPLQ\n",
      "ATTMLIQPM----\n",
      ">_seed_P02679\n",
      "HDS---SIRYLQEIYNSNNQKIVNLKEKVAQLEAQCQEPCKDTVQIHDITGKDCQDIANK\n",
      "GAKQSGLYFIKPLKANQQFLVYCEIDGSGNGWTVFQKRLDGSVDFKKNWIQYKEGFGHLS\n",
      "-----------PTGTTEFWLGNEKIHLISTQSAIPYALRVELEDWNGRTSTADYAMFKVG\n",
      "PEADKYRLTYAYFAGGDAGDAF-DGFD--FGDDPSDKFFTSHKGMQFSTWDNDNDKF---\n",
      "--EGNCAEQDGSGWWMNKCHAGHLNGVYYQGGTYSKASTPNGYDNGIIWATWKTRWYSMK\n",
      "KTTMKIIPFNRLT\n",
      ">_seed_O93568\n",
      "HEN---TIQQLTDMHIMNSNKITQLKQKIAQLESHCQEPCKDTAEIQETTGRDCQDIANK\n",
      "GARKSGLYFIKPQKAKQSFLVYCEIDTYGNGWTVLQRRLDGSEDFRRNWVQYKEGFGHLS\n",
      "-----------PDDTTEFWLGNEKIHLITTQSTLPYALRIELEDWSGKKGTADYAVFKVG\n",
      "TEEDKYRLTYAYFIGGEAGDAF-DGFN--FGDDPSDKSYTYHNGMRFSTFDNDNDNF---\n",
      "--EGNCAEQDGSGWWMNRCHAGHLNGPYYIGGVYSRDTGTNSYDNGIIWATWRDRWYSMK\n",
      "KTTMKIIPFNRLS\n",
      ">_seed_P04115\n",
      "HDA---QIQELSEMWRVNQQFVTRLQQQLVDIRQTCSRSCQDTTKISPITGKDCQQVVDN\n",
      "GGKDSGLYYIKPLKAKQPFLVFCEIE-NGNGWTVIQHRHDGSVNFTRDWVSYREGFGYLA\n",
      "-----------PTLTTEFWLGNEKIHLLTGQQA--YRLRIDLTDWENTHRYADYGHFKLT\n",
      "PESDEYRLFYSMYLDGDAGNAF-DGFD--FGDDPQDKFYTTHLGMLFSTPERDNDKY---\n",
      "--EGSCAEQDGSGWWMNRCHAGHLNGKYYFGGNYRKTDVEFPYDDGIIWATWHDRWYSLK\n",
      "MTTMKLLPMGRLS\n",
      ">P02675\n",
      "NSNIPTNLRVLRSILENLRSKIQKLESDVSAQMEYCRTPCTVSCNIPVVSGKECEEIIRK\n",
      "GGETSEMYLIQPDSSVKPYRVYCDMNTENGGWTVIQNRQDGSVDFGRKWDPYKQGFGNVA\n",
      "TNTDGKNYCGLP---GEYWLGNDKISQLTRMGP--TELLIEMEDWKGDKVKAHYGGFTVQ\n",
      "NEANKYQISVNKY-RGTAGNALMDGASQLMG---ENRTMTIHNGMFFSTYDRDNDGWLTS\n",
      "DPRKQCSKEDGGGWWYNRCHAANPNGRYYWGGQYTWDMAKHGTDDGVVWMNWKGSWYSMR\n",
      "KMSMKIRPF---F\n",
      ">Q02020\n",
      "DNNIPSSLRVLRAVIDSLHKKIQKLENAIATQTDYCRSPCVASCNIPVVSGRECEDIYRK\n",
      "GGETSEMYIIQPDPFTTPYRVYCDMETDNGGWTLIQNRQDGSVNFGRAWDEYKRGFGNIA\n",
      "-KSGGKKYCDTP---GEYWLGNDKISQLTKIGP--TKVLIEMEDWNGDKVSALYGGFTIH\n",
      "NEGNKYQLSVSNY-KGNAGNALMEGASQLYG---ENRTMTIHNGMYFSTYDRDNDGWLTT\n",
      "DPRKQCSKEDGGGWWYNRCHAANPNGRYYWGGTYSWDMAKHGTDDGIVWMNWKGSWYSMK\n",
      "KMSMKIKPY---F\n",
      ">P02678\n",
      "ESTVAGSLRSMKSVLEHLRAKMQRMEEAIKTQKELCSAPCTVNCRVPVVSGMHCEDIYRN\n",
      "GGRTSEAYYIQPDLFSEPYKVFCDMESHGGGWTVVQNRVDGSSNFARDWNTYKAEFGNIA\n",
      "-FGNGKSICNIP---GEYWLGTKTVHQLTKQHT--QQVLFDMSDWEGSSVYAQYASFRPE\n",
      "NEAQGYRLWVEDY-SGNAGNALLEGATQLMG---DNRTMTIHNGMQFSTFDRDNDNWNPG\n",
      "DPTKHCSREDAGGWWYNRCHAANPNGRYYWGGIYTKEQADYGTDDGVVWMNWKGSWYSMR\n",
      "QMAMKLRPK----\n",
      ">P02671\n",
      "------------------------------------------------------------\n",
      "-----------------------------GGWLLIQQRMDGSLNFNRTWQDYKRGFGSLN\n",
      "--DEGE---------GEFWLGNDYLHLLTQRGS---VLRVELEDWAGNEAYAEYH-FRVG\n",
      "SEAEGYALQVSSY-EGTAGDALIEGSVE------EGAEYTSHNNMQFSTFDRDADQW---\n",
      "--EENCAEVYGGGWWYNNCQAANLNGIYYPGGSYDRNNSPYEIENGVVWVSFRGADYSLR\n",
      "AVRMKIRPL---V\n",
      "\n",
      "aln3di26\n",
      ">_seed_Q15389\n",
      "--------------------------------------------------FQFLLVVVVV\n",
      "PPQDWDWDWGDFPPDPDIAIFTWDCPPPNTRKTWQFKAAPQPDAQQDDPVCQACWDDDR-\n",
      "-----------VHTIGNHLVSQQRSCVV--DWKKKKKWFAAQPGDIKMKIFLTWAFDHVV\n",
      "SLRKIATHDI-DIDQ---------ADDF----LADRPIAHKDHPVRQGEQAPD-----RV\n",
      "CVVQNGIGRAHVHTRFDQSFGADDD------PRQAQQQNAGGHDRRPHRRGHTNIMIMIM\n",
      "DDS--VD\n",
      ">_seed_O15123\n",
      "--------------------------------------------------DLFLLRVVVV\n",
      "PNLDWAWDWHAQPPGRDIFIFTWDNPPPRTRKTWQFKDAAQPDQQQDDLVCQQCKDDDN-\n",
      "-----------VHGIGRHLVSQQRVLVV--AFKKKKKWFAALVGDIKIKIFRGWHFHHVV\n",
      "QQRWIPTHDI-DMDL---------APDF----QQDPPTAHKAHPHRQRECAPD-----RC\n",
      "CVVQNHIDRAHVHTQADQSFGDDDN------PRQAQASNAGGHDRRPHPRGHTNMMIMIM\n",
      "HGS--PR\n",
      ">_seed_Q9BY76\n",
      "--------------------------------------------------AQAQQSVQVV\n",
      "PDADFAWDWHHFPPG-DIDIFTWDQDPVGI-KTWQFKAAAQPDAQQDALVCQACWDDDR-\n",
      "-----------VHTIGHHLVVVCSNQPQ--FWKKKWKWFAAPVGDIDIDI-DTWGFDDVV\n",
      "QQRWTWAAPV-RVVV---------LPDP----DADRDIFHKDHQVRADPPDPD-----GL\n",
      "NVQAGHIDRCPDSQAFDQSFADDDD------DPVSNLSRAGDGDVPPHRSGHTRMMIMIM\n",
      "DGD----\n",
      ">_seed_P02679\n",
      "C---VVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVPPDDDDDPADEDDQEEAAQLRSLVV\n",
      "PDAAWFKYWYDFPPDPDIDIFTWDHDPVSFTKTWFFKFAAQPDFFLDAPVCQACWDWDGG\n",
      "RVR--------PTGIGRHLVVQQRSQPSPDFGKKKKKWFAAPVGDIKIWIFARWHFDDVV\n",
      "LQTATAGDDTPDMPLDPFLQADDPPPDPCRSVQAHRHRAHKDAPVRAGEPDPH-----HP\n",
      "RPLLRHIDSAHNHHQAAQSFHDDPNFADDLVRGGPRAQTARHRVSHDPSRGHTNMIIIIM\n",
      "DGVVPPD\n",
      ">_seed_O93568\n",
      "V---VVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVPDDDDDDQQDADEFEELAFLRSLVV\n",
      "PDAQWFKYWYAHAVRPDIDIFTWDHHPVSFIWTWFWKFAALPDAFLDDDVCQACKDWDGG\n",
      "GVP--------PITIGRHLVVQQRSLVSDPFKKKKKKKFAAPVGDIKIWIFIRWHFDDVV\n",
      "QQRATATDDTDDMPLDQFLQFDPPVPDPCRRVQAHQHGAHKDHQVRARECDPD-----RQ\n",
      "NVLLSHIGRGHVHHQAAQSGGADPPFADDWDPDPSTDLNAGDHPVHDDSRGHTNIIIMIM\n",
      "GGVVVVD\n",
      ">_seed_P04115\n",
      "V---VVVVVVVVVVVVVVVVVVVVVVVVVVVVVPVPPDDDDDVVPADPAEEQAQLVSLVS\n",
      "PDDQWFWYFYAHDVRPDTGIFTWHDDP-SFTWTFFWKDAAQPDQQLDDDQCQACKDWDGG\n",
      "RDP--------RTGIGNHLVVQQRSLVR--AKKKKKKWFAAVVRDIKIWIFTSWHFHDCV\n",
      "LPTQTATPGTDDMDAADFLQFDPPPPDPCRSVQAHQHRAHWAHLVGARECDPD-----RL\n",
      "RVLQSHIGRHHVHHQAAQSFHDDPPFADADPDDPNRFQRAGDHNVHPDSRGHTNMMTIII\n",
      "DDDDPPD\n",
      ">P02675\n",
      "DDPVPVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVDDDDAQADEDPQEEQALLVSVVV\n",
      "PPQAFAKHWYDNDPVDDIDIWTWFNPPPNTSWTWQFKFAAQPDALQDDPVCQAQKDDFRF\n",
      "HDPVVDNHGRGGTGIHNHLPVQQVVLVS--AFKKKKKKFAAPVGDIWIKIFRGWHFHDVV\n",
      "QLGWIATHDI-DIDLADPQAQHADPDDDPVRVQAHLHGAHKDFLVGAGEQDPDDDNCDDQ\n",
      "HNQAGGIDRAHNGHQFAQRFHDDPPFFDAQVVGPRSFLHEGDHNSHPHSGTHTNIIIMIM\n",
      "DHN---D\n",
      ">Q02020\n",
      "PPNVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVDDDDAQADEDPQEELAQLVSVVV\n",
      "PPQAWAKHWYPVDPPDDIGIFTWDCPPPNTSWTWAFKFAAQPDAFQDAPQCQAQKDWFQF\n",
      "HPPD-DSHTRGGTGIGRHLVVVQSNCVS--AQKKKKKKFAAPVGDIKIKIFSGWHFHDVV\n",
      "QLRQTATHPI-DIDLDPFQAQAAPPDDDPVRVQAHQHRAGKDHLVGAREQCDDPPNCPGQ\n",
      "RNQLRHIGRHHNHHQAAQSFGDDPPFFDALVVGPNSFLNAGDNCNHPHSGGHTNIITMIM\n",
      "HHN---D\n",
      ">P02678\n",
      "CPVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVPVDPDDAQADADPQEEAALLRSVVV\n",
      "PPQDWFKHWHNVDPPDDTDIFTWHNPPPNTRWTWQFKDAAQPDAFQDAPVCQQQKDWDQF\n",
      "HQPP-VRTTRHGTGIGHHLQVVQVNCVV--FWKKKKKKFAAPVGDIWIKIFRTWHFHHVV\n",
      "VLRATATHDI-DIDQDQCQQFNLVVDDDPVGVQAHLHRAHKHHLPGARAQQDVPDNPQRL\n",
      "NVQQRHIGRHHSHHRFAASGRDDPDFFDALVVGPVSFLHAGDNCSHRHSGTHTNIMIMII\n",
      "HR----D\n",
      ">P02671\n",
      "-------------------------------------DDKD-------------------\n",
      "---------------------------------FFFKDAALPDAAQDDPVCLQCKDWDA-\n",
      "-DP------VRDTGIGRHLVVQQVVDVQ--KWKKK-KWFAAPVGDIWIKIW-TWHFDDVV\n",
      "LLSFIATDDI-DIDQDPCAQQN---DDVVGRQFAHQHRAHKAHQVGDDECDPD-----GC\n",
      "NVVLRHIGRAHNGHQFDQNFGDDPDFFDALCPDPPSALRAHDRDSHPHSGTHTNIMTMII\n",
      "DHV---D\n",
      "\n",
      "mafft --seed ../../testcat/4.10.530/alnscratch/28_inter.fasta ../../testcat/4.10.530/alnscratch/27_inter.fasta > ../../testcat/4.10.530/alnscratch/29_root.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 4\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 1600.000000\n",
      "\n",
      "adding 0-1\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 6\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "    0 / 6\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP     5 / 5 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "    0 / 6\n",
      "    0 / 2\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP     5 /5 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "    0 / 6\n",
      "Segment   1/  1    1- 227\n",
      "done 002-001-1  identical.   \n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --textmatrix ../mafftmat/3diHEXmat.txt --seed ../../testcat/4.10.530/alnscratch/28_inter.3di.fasta ../../testcat/4.10.530/alnscratch/27_inter3di.fasta > ../../testcat/4.10.530/alnscratch/29_inter3di.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 4\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 1600.000000\n",
      "\n",
      "adding 0-1\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "nalphabets = 256\n",
      "nused=\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 6\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "    0 / 6\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP     5 / 5 \n",
      "done.\n",
      "\n",
      "disttbfast (text) Version 7.520\n",
      "alg=A, model=Extended, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "nalphabets = 256\n",
      "nused=\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "    0 / 6\n",
      "    0 / 2\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP     5 /5 \n",
      "done.\n",
      "tbfast (text) Version 7.520\n",
      "alg=A, model=Extended, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nalphabets = 256\n",
      "nused=\n",
      "dndpre (text) Version 7.520\n",
      "alg=X, model=Extended, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum -2 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "nalphabets = 256\n",
      "nused=\n",
      "\n",
      "    0 / 6\n",
      "Segment   1/  1    1- 226\n",
      "STEP 002-001-1  identical.    accepted. identical.    identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "done\n",
      "dvtditr (text) Version 7.520\n",
      "alg=A, model=Extended, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " NW-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "aln1\n",
      ">_seed_Q8N539\n",
      "--PRDCLDVLLSGQQDDGVYSVF-PTHY--PAGFQVYCDMRTDGGGWTVFQRREDGSVN-\n",
      "--FFRGWDAYRDGFGRLTGEHWLGLKRIHALTTQAAYELHVDLEDFENGTAYARYGSFGV\n",
      "GLFSVDPEEDGYPLTVADYS-GTAGDSLLKHSGMRFTTKDRDSDHSENNCAAFYRGAWWY\n",
      "RNCHTSNLNGQYLRGAHASYADGVEWSSWTGWQYSLKFSEMKIRPV\n",
      ">_seed_Q9U8W8\n",
      "--PTDCADILLNGYRSSGGYRIW-PKSWMTVGTLNVYCDMETDGGGWTVIQRRGNYGNPS\n",
      "DYFYKPWKNYKLGFGNIEKDFWLGNDRIFALTNQRNYMIRFDLKDKENDTRYAIYQDFWI\n",
      "-----ENEDYLYCLHIGNYS-GDAGNSFGRHNGHNFSTIDKDHDTHETHCAQTYKGGWWY\n",
      "DRCHESNLNGLYLNGEHNSYADGIEWRAWKGYHYSLPQVEMKIRPV\n",
      ">Q15485\n",
      "TGPRTCKDLLDRGHFLSGWHTIYLPDCR--P--LTVLCDMDTDGGGWTVFQRRVDGSVD-\n",
      "--FYRDWATYKQGFGSRLGEFWLGNDNIHALTAQGTSELRVDLVDFEDNYQFAKYRS---\n",
      "--FKVADEAEKYNLVLGAFVEGSAGDSLTFHNNQSFSTKDQDNDLNTGNCAVMFQGAWWY\n",
      "KNCHVSNLNGRYLRGTHGSFANGINWKSGKGYNYSYKVSEMKVRPA\n",
      ">O00602\n",
      "TGPRNCKDLLDRGYFLSGWHTIYLPDCR--P--LTVLCDMDTDGGGWTVFQRRMDGSVD-\n",
      "--FYRDWAAYKQGFGSQLGEFWLGNDNIHALTAQGSSELRVDLVDFEGNHQFAKYKS---\n",
      "--FKVADEAEKYKLVLGAFVGGSAGNSLTGHNNNFFSTKDQDNDVSSSNCAEKFQGAWWY\n",
      "ADCHASNLNGLYLMGPHESFANGINWSAAKGYKYSYKVSEMKVRPA\n",
      ">O75636\n",
      "-GPRNCRELLSQGATLSGWYHLCLPEGR--A--LPVFCDMDTEGGGWLVFQRRQDGSVD-\n",
      "--FFRSWSSYRAGFGNQESEFWLGNENLHQLTLQGNWELRVELEDFNGNRTFAHYAT---\n",
      "--FRLLGEVDHYQLALGKFSEGTAGDSLSLHSGRPFTTYDADHDSSNSNCAVIVHGAWWY\n",
      "ASCYRSNLNGRYAVSEAAAHKYGIDWASGRGVGHPYRRVRMMLR--\n",
      ">P24821\n",
      "PFPKDCSQAMLNGDTTSGLYTIYLDKAQ--A--LEVFCDMTSDGGGWIVFLRRKNGREN-\n",
      "--FYQNWKAYAAGFGDRREEFWLGLDNLNKITAQGQYELRVDLRD-HGETAFAVYDK---\n",
      "--FSVGDAKTRYKLKVEGYS-GTAGDSMAYHNGRSFSTFDK-------NCALSYKGAFWY\n",
      "RNCHRVNLMGRYGDNNHSQ---GVNWFHWKGHEHSIQFAEMKLRPS\n",
      "\n",
      ">_seed_Q8N539\n",
      "--AQFQQSVVVVPDQDWAKDWHAFPVR--RVTDIFTWFCPPPRTRWTWFFKFAAQPDQ--\n",
      "-FQDDDVCQQCKDDDRVHGIGNHLVVQQSNCQRAFKKKKKWFAAPPRDIKIKIWRGWHFC\n",
      "NPPPDRVVQQRAIATHDI-DMDQDDWPVVLHRAHKFHPPRARECDPDGVCVQLVHIGSAH\n",
      "NGGQAGQSFHQPHFFDDDFQNAGDNCSHRHGGTHTNMMIMIMDHD\n",
      ">_seed_Q9U8W8\n",
      "--AQALLSVVVVPDQDWAWDWHDQVVCVPVNIDIFTWDCPPPRTSWTFFFKFAPDPDALL\n",
      "VQLDDPVCQQCKDDDNVHGIGNHLVVQQRSCPSAWKKKKKWFAAPVRDIWMKIWGRWHF-\n",
      "----HDVVQQRATATHDM-DIPQDPFCVVLHGAHWAHPVGAREDAPDRQCVQQRHTGRAY\n",
      "VGGQARQSFGDPRFFDDRFQNAGHNPSHPHGGTHTNIMIMIMHHP\n",
      ">Q15485\n",
      "DFAAFLQVVVVVPPQDWAWDWHAFPVR---DIDIFTWANPPPNTGWTWQWKFAAQPDQ--\n",
      "-QQDAPVDQQCKDDDSVHTIGNHLVVQQRSQVPAFKKKKKWFAAPVGDIKIWIFRGWHF-\n",
      "----HDVVLQRAIATHDTDDMQQDDFCVVLHRAHKDHQVRAREQDPDRQCVLQVHIGSAH\n",
      "NGGQAGQSFGCPHFFAPRFQNARHHPSHPHRGTHTNMIIMIIHGD\n",
      ">O00602\n",
      "PFAAFQLSVVVVPPQCWAWDWHAFPVR---DIDIFTWDNPPPNTRWTWQWKFAAQPDQ--\n",
      "-QQDAPVCQQCKDDDSVHTIGNHLVVQQRSCVPAFKKKKKWFAAPVGDIKIWIFRGWHF-\n",
      "----DDVVLQRAIATHDTDDMLLDDWCVVLHRAHKHHQVRAREQDPDRQCVQLVHIGSAH\n",
      "NGGQAGQRFHPPHFFADRFQRARDNDSHVHGGTHTNMIIMIIHHD\n",
      ">O75636\n",
      "D-ALFQVVVVVVPPQDWAWDWHQFPVR---DIAIFTWDCDPPNTGKTWQWKFADQPDQ--\n",
      "-FQDDLVCQQCKDDDRVHTIGRHLVVQQRSPVPHFKKKKKWFAADVGDIKIWIFRTWHF-\n",
      "----HDVVQQRAIATHDTDDMPLDDWCVVLHRAHKDHQVRAREQDPDRQCVVQVHIGSAH\n",
      "NHTQAGQSGHDDPHVVRWDQHGRHHCSHCHRRGHTRIMIMIID--\n",
      ">P24821\n",
      "DAAAFQLSVVVVVDQDWAWDWHAQPVR----IDIFTWQNPDPRTRKTWFFKFAAQPDA--\n",
      "-QQDDPVCQQCKDDDRVYGIGRHLVVQQRSCVNAFKKKKKWFDAP-HDIKMKIFLGWGF-\n",
      "----HDVVLLRFIAGHDI-DMDQDPFCVVLHGAHKDHPVND-------PCVVLSHIGSAD\n",
      "VGGQAGQRFHDP---AQDFQNAGHNDSRPRRRDHTNIMIMIMDHS\n",
      "\n",
      "mafft --seed ../../testcat/4.10.530/alnscratch/29_root.fasta ../../testcat/4.10.530/alnscratch/27_inter.fasta > ../../testcat/4.10.530/alnscratch/29_root.fasta.iter\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 4\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 1600.000000\n",
      "\n",
      "adding 4-5\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 10\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "    0 / 10\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP     9 / 9 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "    0 / 10\n",
      "    0 / 6\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP     9 /9 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "    0 / 10\n",
      "Segment   1/  1    1- 227\n",
      "STEP 002-008-1  identical.   \n",
      "Converged.\n",
      "\n",
      "done\n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --textmatrix ../mafftmat/3diHEXmat.txt --seed ../../testcat/4.10.530/alnscratch/29_inter3di.fasta ../../testcat/4.10.530/alnscratch/27_inter3di.fasta > ../../testcat/4.10.530/alnscratch/29_inter3di.fasta.iter\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 4\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 1600.000000\n",
      "\n",
      "adding 4-5\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "nalphabets = 256\n",
      "nused=\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 10\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "    0 / 10\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP     9 / 9 \n",
      "done.\n",
      "\n",
      "disttbfast (text) Version 7.520\n",
      "alg=A, model=Extended, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "nalphabets = 256\n",
      "nused=\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "    0 / 10\n",
      "    0 / 6\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP     9 /9 \n",
      "done.\n",
      "tbfast (text) Version 7.520\n",
      "alg=A, model=Extended, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nalphabets = 256\n",
      "nused=\n",
      "dndpre (text) Version 7.520\n",
      "alg=X, model=Extended, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum -2 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "nalphabets = 256\n",
      "nused=\n",
      "\n",
      "    0 / 10\n",
      "Segment   1/  1    1- 226\n",
      "STEP 002-008-1  identical.    identical.    identical.    identical.    identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "done\n",
      "dvtditr (text) Version 7.520\n",
      "alg=A, model=Extended, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " NW-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "aln0\n",
      ">_seed__seed_Q8N539\n",
      "--PRDCLDVLLSGQQDDGVYSVF-PTHY--PAGFQVYCDMRTDGGGWTVFQRREDGSVN-\n",
      "--FFRGWDAYRDGFGRLTGEHWLGLKRIHALTTQAAYELHVDLEDFENGTAYARYGSFGV\n",
      "GLFSVDPEEDGYPLTVADYS-GTAGDSLLKHSGMRFTTKDRDSDHSENNCAAFYRGAWWY\n",
      "RNCHTSNLNGQYLRGAHASYADGVEWSSWTGWQYSLKFSEMKIRPV\n",
      ">_seed__seed_Q9U8W8\n",
      "--PTDCADILLNGYRSSGGYRIW-PKSWMTVGTLNVYCDMETDGGGWTVIQRRGNYGNPS\n",
      "DYFYKPWKNYKLGFGNIEKDFWLGNDRIFALTNQRNYMIRFDLKDKENDTRYAIYQDFWI\n",
      "-----ENEDYLYCLHIGNYS-GDAGNSFGRHNGHNFSTIDKDHDTHETHCAQTYKGGWWY\n",
      "DRCHESNLNGLYLNGEHNSYADGIEWRAWKGYHYSLPQVEMKIRPV\n",
      ">_seed_Q15485\n",
      "TGPRTCKDLLDRGHFLSGWHTIYLPDCR--P--LTVLCDMDTDGGGWTVFQRRVDGSVD-\n",
      "--FYRDWATYKQGFGSRLGEFWLGNDNIHALTAQGTSELRVDLVDFEDNYQFAKYRS---\n",
      "--FKVADEAEKYNLVLGAFVEGSAGDSLTFHNNQSFSTKDQDNDLNTGNCAVMFQGAWWY\n",
      "KNCHVSNLNGRYLRGTHGSFANGINWKSGKGYNYSYKVSEMKVRPA\n",
      ">_seed_O00602\n",
      "TGPRNCKDLLDRGYFLSGWHTIYLPDCR--P--LTVLCDMDTDGGGWTVFQRRMDGSVD-\n",
      "--FYRDWAAYKQGFGSQLGEFWLGNDNIHALTAQGSSELRVDLVDFEGNHQFAKYKS---\n",
      "--FKVADEAEKYKLVLGAFVGGSAGNSLTGHNNNFFSTKDQDNDVSSSNCAEKFQGAWWY\n",
      "ADCHASNLNGLYLMGPHESFANGINWSAAKGYKYSYKVSEMKVRPA\n",
      ">_seed_O75636\n",
      "-GPRNCRELLSQGATLSGWYHLCLPEGR--A--LPVFCDMDTEGGGWLVFQRRQDGSVD-\n",
      "--FFRSWSSYRAGFGNQESEFWLGNENLHQLTLQGNWELRVELEDFNGNRTFAHYAT---\n",
      "--FRLLGEVDHYQLALGKFSEGTAGDSLSLHSGRPFTTYDADHDSSNSNCAVIVHGAWWY\n",
      "ASCYRSNLNGRYAVSEAAAHKYGIDWASGRGVGHPYRRVRMMLR--\n",
      ">_seed_P24821\n",
      "PFPKDCSQAMLNGDTTSGLYTIYLDKAQ--A--LEVFCDMTSDGGGWIVFLRRKNGREN-\n",
      "--FYQNWKAYAAGFGDRREEFWLGLDNLNKITAQGQYELRVDLRD-HGETAFAVYDK---\n",
      "--FSVGDAKTRYKLKVEGYS-GTAGDSMAYHNGRSFSTFDK-------NCALSYKGAFWY\n",
      "RNCHRVNLMGRYGDNNHSQ---GVNWFHWKGHEHSIQFAEMKLRPS\n",
      ">Q15485\n",
      "TGPRTCKDLLDRGHFLSGWHTIYLPDCR--P--LTVLCDMDTDGGGWTVFQRRVDGSVD-\n",
      "--FYRDWATYKQGFGSRLGEFWLGNDNIHALTAQGTSELRVDLVDFEDNYQFAKYRS---\n",
      "--FKVADEAEKYNLVLGAFVEGSAGDSLTFHNNQSFSTKDQDNDLNTGNCAVMFQGAWWY\n",
      "KNCHVSNLNGRYLRGTHGSFANGINWKSGKGYNYSYKVSEMKVRPA\n",
      ">O00602\n",
      "TGPRNCKDLLDRGYFLSGWHTIYLPDCR--P--LTVLCDMDTDGGGWTVFQRRMDGSVD-\n",
      "--FYRDWAAYKQGFGSQLGEFWLGNDNIHALTAQGSSELRVDLVDFEGNHQFAKYKS---\n",
      "--FKVADEAEKYKLVLGAFVGGSAGNSLTGHNNNFFSTKDQDNDVSSSNCAEKFQGAWWY\n",
      "ADCHASNLNGLYLMGPHESFANGINWSAAKGYKYSYKVSEMKVRPA\n",
      ">O75636\n",
      "-GPRNCRELLSQGATLSGWYHLCLPEGR--A--LPVFCDMDTEGGGWLVFQRRQDGSVD-\n",
      "--FFRSWSSYRAGFGNQESEFWLGNENLHQLTLQGNWELRVELEDFNGNRTFAHYAT---\n",
      "--FRLLGEVDHYQLALGKFSEGTAGDSLSLHSGRPFTTYDADHDSSNSNCAVIVHGAWWY\n",
      "ASCYRSNLNGRYAVSEAAAHKYGIDWASGRGVGHPYRRVRMMLR--\n",
      ">P24821\n",
      "PFPKDCSQAMLNGDTTSGLYTIYLDKAQ--A--LEVFCDMTSDGGGWIVFLRRKNGREN-\n",
      "--FYQNWKAYAAGFGDRREEFWLGLDNLNKITAQGQYELRVDLRD-HGETAFAVYDK---\n",
      "--FSVGDAKTRYKLKVEGYS-GTAGDSMAYHNGRSFSTFDK-------NCALSYKGAFWY\n",
      "RNCHRVNLMGRYGDNNHSQ---GVNWFHWKGHEHSIQFAEMKLRPS\n",
      "\n",
      ">_seed__seed_Q8N539\n",
      "--AQFQQSVVVVPDQDWAKDWHAFPVR--RVTDIFTWFCPPPRTRWTWFFKFAAQPDQ--\n",
      "-FQDDDVCQQCKDDDRVHGIGNHLVVQQSNCQRAFKKKKKWFAAPPRDIKIKIWRGWHFC\n",
      "NPPPDRVVQQRAIATHDI-DMDQDDWPVVLHRAHKFHPPRARECDPDGVCVQLVHIGSAH\n",
      "NGGQAGQSFHQPHFFDDDFQNAGDNCSHRHGGTHTNMMIMIMDHD\n",
      ">_seed__seed_Q9U8W8\n",
      "--AQALLSVVVVPDQDWAWDWHDQVVCVPVNIDIFTWDCPPPRTSWTFFFKFAPDPDALL\n",
      "VQLDDPVCQQCKDDDNVHGIGNHLVVQQRSCPSAWKKKKKWFAAPVRDIWMKIWGRWHF-\n",
      "----HDVVQQRATATHDM-DIPQDPFCVVLHGAHWAHPVGAREDAPDRQCVQQRHTGRAY\n",
      "VGGQARQSFGDPRFFDDRFQNAGHNPSHPHGGTHTNIMIMIMHHP\n",
      ">_seed_Q15485\n",
      "DFAAFLQVVVVVPPQDWAWDWHAFPVR---DIDIFTWANPPPNTGWTWQWKFAAQPDQ--\n",
      "-QQDAPVDQQCKDDDSVHTIGNHLVVQQRSQVPAFKKKKKWFAAPVGDIKIWIFRGWHF-\n",
      "----HDVVLQRAIATHDTDDMQQDDFCVVLHRAHKDHQVRAREQDPDRQCVLQVHIGSAH\n",
      "NGGQAGQSFGCPHFFAPRFQNARHHPSHPHRGTHTNMIIMIIHGD\n",
      ">_seed_O00602\n",
      "PFAAFQLSVVVVPPQCWAWDWHAFPVR---DIDIFTWDNPPPNTRWTWQWKFAAQPDQ--\n",
      "-QQDAPVCQQCKDDDSVHTIGNHLVVQQRSCVPAFKKKKKWFAAPVGDIKIWIFRGWHF-\n",
      "----DDVVLQRAIATHDTDDMLLDDWCVVLHRAHKHHQVRAREQDPDRQCVQLVHIGSAH\n",
      "NGGQAGQRFHPPHFFADRFQRARDNDSHVHGGTHTNMIIMIIHHD\n",
      ">_seed_O75636\n",
      "D-ALFQVVVVVVPPQDWAWDWHQFPVR---DIAIFTWDCDPPNTGKTWQWKFADQPDQ--\n",
      "-FQDDLVCQQCKDDDRVHTIGRHLVVQQRSPVPHFKKKKKWFAADVGDIKIWIFRTWHF-\n",
      "----HDVVQQRAIATHDTDDMPLDDWCVVLHRAHKDHQVRAREQDPDRQCVVQVHIGSAH\n",
      "NHTQAGQSGHDDPHVVRWDQHGRHHCSHCHRRGHTRIMIMIID--\n",
      ">_seed_P24821\n",
      "DAAAFQLSVVVVVDQDWAWDWHAQPVR----IDIFTWQNPDPRTRKTWFFKFAAQPDA--\n",
      "-QQDDPVCQQCKDDDRVYGIGRHLVVQQRSCVNAFKKKKKWFDAP-HDIKMKIFLGWGF-\n",
      "----HDVVLLRFIAGHDI-DMDQDPFCVVLHGAHKDHPVND-------PCVVLSHIGSAD\n",
      "VGGQAGQRFHDP---AQDFQNAGHNDSRPRRRDHTNIMIMIMDHS\n",
      ">Q15485\n",
      "DFAAFLQVVVVVPPQDWAWDWHAFPVR---DIDIFTWANPPPNTGWTWQWKFAAQPDQ--\n",
      "-QQDAPVDQQCKDDDSVHTIGNHLVVQQRSQVPAFKKKKKWFAAPVGDIKIWIFRGWHF-\n",
      "----HDVVLQRAIATHDTDDMQQDDFCVVLHRAHKDHQVRAREQDPDRQCVLQVHIGSAH\n",
      "NGGQAGQSFGCPHFFAPRFQNARHHPSHPHRGTHTNMIIMIIHGD\n",
      ">O00602\n",
      "PFAAFQLSVVVVPPQCWAWDWHAFPVR---DIDIFTWDNPPPNTRWTWQWKFAAQPDQ--\n",
      "-QQDAPVCQQCKDDDSVHTIGNHLVVQQRSCVPAFKKKKKWFAAPVGDIKIWIFRGWHF-\n",
      "----DDVVLQRAIATHDTDDMLLDDWCVVLHRAHKHHQVRAREQDPDRQCVQLVHIGSAH\n",
      "NGGQAGQRFHPPHFFADRFQRARDNDSHVHGGTHTNMIIMIIHHD\n",
      ">O75636\n",
      "D-ALFQVVVVVVPPQDWAWDWHQFPVR---DIAIFTWDCDPPNTGKTWQWKFADQPDQ--\n",
      "-FQDDLVCQQCKDDDRVHTIGRHLVVQQRSPVPHFKKKKKWFAADVGDIKIWIFRTWHF-\n",
      "----HDVVQQRAIATHDTDDMPLDDWCVVLHRAHKDHQVRAREQDPDRQCVVQVHIGSAH\n",
      "NHTQAGQSGHDDPHVVRWDQHGRHHCSHCHRRGHTRIMIMIID--\n",
      ">P24821\n",
      "DAAAFQLSVVVVVDQDWAWDWHAQPVR----IDIFTWQNPDPRTRKTWFFKFAAQPDA--\n",
      "-QQDDPVCQQCKDDDRVYGIGRHLVVQQRSCVNAFKKKKKWFDAP-HDIKMKIFLGWGF-\n",
      "----HDVVLLRFIAGHDI-DMDQDPFCVVLHGAHKDHPVND-------PCVVLSHIGSAD\n",
      "VGGQAGQRFHDP---AQDFQNAGHNDSRPRRRDHTNIMIMIMDHS\n",
      "\n",
      "mafft --seed ../../testcat/4.10.530/alnscratch/29_root.fasta.iter ../../testcat/4.10.530/alnscratch/26_inter.fasta > ../../testcat/4.10.530/alnscratch/29_root.fasta.iter.iter\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 10\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 10000.000000\n",
      "\n",
      "adding 8-9\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 20\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "   10 / 20\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP    19 / 19 \n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "   10 / 20\n",
      "    0 / 10\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    19 /19 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.520\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   10 / 20\n",
      "Segment   1/  1    1- 322\n",
      "done 002-001-1  identical.    accepted.\n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --textmatrix ../mafftmat/3diHEXmat.txt --seed ../../testcat/4.10.530/alnscratch/29_inter3di.fasta.iter ../../testcat/4.10.530/alnscratch/26_inter3di.fasta > ../../testcat/4.10.530/alnscratch/29_inter3di.fasta.iter.iter\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "############################################################################\n",
      "# Warning:\n",
      "#   Progressive alignment method is incompatible with the --seed option.\n",
      "#   Automatically switched to the iterative refinement method.\n",
      "#   \n",
      "# Also consider using the '--add' option, which is compatible with\n",
      "#   the progressive method and FASTER than the '--seed' option.\n",
      "#   Usage is:\n",
      "#   % mafft --add newSequences existingAlignment > output\n",
      "############################################################################\n",
      "nhomologs = 10\n",
      "seedoffset = 0\n",
      "seed = seed1\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, -0.12\n",
      "tsuyosa = 10000.000000\n",
      "\n",
      "adding 8-9\n",
      "multi2hat3s (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "nalphabets = 256\n",
      "nused=\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "    1 / 20\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "   10 / 20\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/1... \n",
      "STEP    19 / 19 \n",
      "done.\n",
      "\n",
      "disttbfast (text) Version 7.520\n",
      "alg=A, model=Extended, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "nalphabets = 256\n",
      "nused=\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 0, compacttree = 0\n",
      "Making a distance matrix from msa .. \n",
      "done.                                           \n",
      "Constructing a UPGMA tree ... \n",
      "   10 / 20\n",
      "    0 / 10\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    19 /19 \n",
      "done.\n",
      "tbfast (text) Version 7.520\n",
      "alg=A, model=Extended, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "nalphabets = 256\n",
      "nused=\n",
      "dndpre (text) Version 7.520\n",
      "alg=X, model=Extended, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum -2 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "Loading 'hat3' ... done.\n",
      "nalphabets = 256\n",
      "nused=\n",
      "\n",
      "   10 / 20\n",
      "Segment   1/  1    1- 319\n",
      "STEP 002-003-1  identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    accepted. identical.    identical.    rejected. identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.   "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "aln1\n",
      ">_seed__seed__seed_Q8N539\n",
      "--------------------------------------------------PRDCLDVLLS\n",
      "GQQDDGVYSVF-PTHY--PAGFQVYCDMRTDGGGWTVFQRREDGSVN---FFRGWDAYRD\n",
      "GFGRLT---------------GEHWLGLKRIHALTTQAA--YELHVDLEDFENGTAYARY\n",
      "GSFGVGLFSVDPEEDGYPLTVADYS-GTAGDSL-------------------LKHSGMRF\n",
      "TTKDRDSDHS-----ENNCAAFYRGAWWYRNCHTSNLNGQYLRGAHAS-------YADGV\n",
      "EWSSWTGWQYSLKFSEMKIRPV----\n",
      ">_seed__seed__seed_Q9U8W8\n",
      "--------------------------------------------------PTDCADILLN\n",
      "GYRSSGGYRIW-PKSWMTVGTLNVYCDMETDGGGWTVIQRRGNYGNPSDYFYKPWKNYKL\n",
      "GFGNIE---------------KDFWLGNDRIFALTNQRN--YMIRFDLKDKENDTRYAIY\n",
      "QDFWI-----ENEDYLYCLHIGNYS-GDAGNSF-------------------GRHNGHNF\n",
      "STIDKDHDTH-----ETHCAQTYKGGWWYDRCHESNLNGLYLNGEHNS-------YADGI\n",
      "EWRAWKGYHYSLPQVEMKIRPV----\n",
      ">_seed__seed_Q15485\n",
      "------------------------------------------------TGPRTCKDLLDR\n",
      "GHFLSGWHTIYLPDCR--P--LTVLCDMDTDGGGWTVFQRRVDGSVD---FYRDWATYKQ\n",
      "GFGSRL---------------GEFWLGNDNIHALTAQGT--SELRVDLVDFEDNYQFAKY\n",
      "RS-----FKVADEAEKYNLVLGAFVEGSAGDSL-------------------TFHNNQSF\n",
      "STKDQDNDLN-----TGNCAVMFQGAWWYKNCHVSNLNGRYLRGTHGS-------FANGI\n",
      "NWKSGKGYNYSYKVSEMKVRPA----\n",
      ">_seed__seed_O00602\n",
      "------------------------------------------------TGPRNCKDLLDR\n",
      "GYFLSGWHTIYLPDCR--P--LTVLCDMDTDGGGWTVFQRRMDGSVD---FYRDWAAYKQ\n",
      "GFGSQL---------------GEFWLGNDNIHALTAQGS--SELRVDLVDFEGNHQFAKY\n",
      "KS-----FKVADEAEKYKLVLGAFVGGSAGNSL-------------------TGHNNNFF\n",
      "STKDQDNDVS-----SSNCAEKFQGAWWYADCHASNLNGLYLMGPHES-------FANGI\n",
      "NWSAAKGYKYSYKVSEMKVRPA----\n",
      ">_seed__seed_O75636\n",
      "-------------------------------------------------GPRNCRELLSQ\n",
      "GATLSGWYHLCLPEGR--A--LPVFCDMDTEGGGWLVFQRRQDGSVD---FFRSWSSYRA\n",
      "GFGNQE---------------SEFWLGNENLHQLTLQGN--WELRVELEDFNGNRTFAHY\n",
      "AT-----FRLLGEVDHYQLALGKFSEGTAGDSL-------------------SLHSGRPF\n",
      "TTYDADHDSS-----NSNCAVIVHGAWWYASCYRSNLNGRYAVSEAAA-------HKYGI\n",
      "DWASGRGVGHPYRRVRMMLR------\n",
      ">_seed__seed_P24821\n",
      "------------------------------------------------PFPKDCSQAMLN\n",
      "GDTTSGLYTIYLDKAQ--A--LEVFCDMTSDGGGWIVFLRRKNGREN---FYQNWKAYAA\n",
      "GFGDRR---------------EEFWLGLDNLNKITAQGQ--YELRVDLRD-HGETAFAVY\n",
      "DK-----FSVGDAKTRYKLKVEGYS-GTAGDSM-------------------AYHNGRSF\n",
      "STFDK------------NCALSYKGAFWYRNCHRVNLMGRYGDNNHSQ----------GV\n",
      "NWFHWKGHEHSIQFAEMKLRPS----\n",
      ">_seed_Q15485\n",
      "------------------------------------------------TGPRTCKDLLDR\n",
      "GHFLSGWHTIYLPDCR--P--LTVLCDMDTDGGGWTVFQRRVDGSVD---FYRDWATYKQ\n",
      "GFGSRL---------------GEFWLGNDNIHALTAQGT--SELRVDLVDFEDNYQFAKY\n",
      "RS-----FKVADEAEKYNLVLGAFVEGSAGDSL-------------------TFHNNQSF\n",
      "STKDQDNDLN-----TGNCAVMFQGAWWYKNCHVSNLNGRYLRGTHGS-------FANGI\n",
      "NWKSGKGYNYSYKVSEMKVRPA----\n",
      ">_seed_O00602\n",
      "------------------------------------------------TGPRNCKDLLDR\n",
      "GYFLSGWHTIYLPDCR--P--LTVLCDMDTDGGGWTVFQRRMDGSVD---FYRDWAAYKQ\n",
      "GFGSQL---------------GEFWLGNDNIHALTAQGS--SELRVDLVDFEGNHQFAKY\n",
      "KS-----FKVADEAEKYKLVLGAFVGGSAGNSL-------------------TGHNNNFF\n",
      "STKDQDNDVS-----SSNCAEKFQGAWWYADCHASNLNGLYLMGPHES-------FANGI\n",
      "NWSAAKGYKYSYKVSEMKVRPA----\n",
      ">_seed_O75636\n",
      "-------------------------------------------------GPRNCRELLSQ\n",
      "GATLSGWYHLCLPEGR--A--LPVFCDMDTEGGGWLVFQRRQDGSVD---FFRSWSSYRA\n",
      "GFGNQE---------------SEFWLGNENLHQLTLQGN--WELRVELEDFNGNRTFAHY\n",
      "AT-----FRLLGEVDHYQLALGKFSEGTAGDSL-------------------SLHSGRPF\n",
      "TTYDADHDSS-----NSNCAVIVHGAWWYASCYRSNLNGRYAVSEAAA-------HKYGI\n",
      "DWASGRGVGHPYRRVRMMLR------\n",
      ">_seed_P24821\n",
      "------------------------------------------------PFPKDCSQAMLN\n",
      "GDTTSGLYTIYLDKAQ--A--LEVFCDMTSDGGGWIVFLRRKNGREN---FYQNWKAYAA\n",
      "GFGDRR---------------EEFWLGLDNLNKITAQGQ--YELRVDLRD-HGETAFAVY\n",
      "DK-----FSVGDAKTRYKLKVEGYS-GTAGDSM-------------------AYHNGRSF\n",
      "STFDK------------NCALSYKGAFWYRNCHRVNLMGRYGDNNHSQ----------GV\n",
      "NWFHWKGHEHSIQFAEMKLRPS----\n",
      ">_seed_Q15389\n",
      "--------------------------------------------------FRDCADVYQA\n",
      "GFNKSGIYTIYINNMP-EP--KKVFCNMDVNGGGWTVIQHREDGSLD---FQRGWKEYKM\n",
      "GFGNPS---------------GEYWLGNEFIFAITSQRQ--YMLRIELMDWEGNRAYSQY\n",
      "DR-----FHIGNEKQNYRLYLKGHT-GTAGKQS------------------SLILHGADF\n",
      "STKDADNDNC-----MCKCALMLTGGWWFDACGPSNLNGMFYTAGQNH------GKLNGI\n",
      "KWHYFKGPSYSLRSTTMMIRPLD--F\n",
      ">_seed_O15123\n",
      "--------------------------------------------------FRDCAEVFKS\n",
      "GHTTNGIYTLTFPNST-EE--IKAYCDMEAGGGGWTIIQRREDGSVD---FQRTWKEYKV\n",
      "GFGNPS---------------GEYWLGNEFVSQLTNQQR--YVLKIHLKDWEGNEAYSLY\n",
      "EH-----FYLSSEELNYRIHLKGLT-GTAGKIS------------------SISQPGNDF\n",
      "STKDGDNDKC-----ICKCSQMLTGGWWFDACGPSNLNGMYYPQRQNT------NKFNGI\n",
      "KWYYWKGSGYSLKATTMMIRPAD--F\n",
      ">_seed_Q9BY76\n",
      "--------------------------------------------------PRDCQELFQV\n",
      "GERQSGLFEIQ-PQGS-PP--FLVNCKMTSD-GGWTVIQRRHDGSVD---FNRPWEAYKA\n",
      "GFGDPH---------------GEFWLGLEKVHSITGDRN--SRLAVQLRDWDGNAELLQF\n",
      "S------VHLGGEDTAYSLQLTAPVAGQLGATTVPPSG---------------------F\n",
      "STWDQDHDLR-----RDKCAKSLSGGWWFGTCSHSNLNGQYFPQRQ--------KLKKGI\n",
      "FWKTWRGRYYPLQATTMLIQPM----\n",
      ">_seed_P02679\n",
      "HDS---SIRYLQEIYNSNNQKIVNLKEKVAQLEAQCQEPCKDTVQIHDITGKDCQDIANK\n",
      "GAKQSGLYFIK-PLKANQQ--FLVYCEIDGSGNGWTVFQKRLDGSVD---FKKNWIQYKE\n",
      "GFGHLS-----------PTGTTEFWLGNEKIHLISTQSAIPYALRVELEDWNGRTSTADY\n",
      "AM-----FKVGPEADKYRLTYAYFAGGDAGDAF---DGFD--FGDDPSDKFFTSHKGMQF\n",
      "STWDNDNDKF-----EGNCAEQDGSGWWMNKCHAGHLNGVYYQGGTYSKASTPNGYDNGI\n",
      "IWATWKTRWYSMKKTTMKIIPFNRLT\n",
      ">_seed_O93568\n",
      "HEN---TIQQLTDMHIMNSNKITQLKQKIAQLESHCQEPCKDTAEIQETTGRDCQDIANK\n",
      "GARKSGLYFIK-PQKAKQS--FLVYCEIDTYGNGWTVLQRRLDGSED---FRRNWVQYKE\n",
      "GFGHLS-----------PDDTTEFWLGNEKIHLITTQSTLPYALRIELEDWSGKKGTADY\n",
      "AV-----FKVGTEEDKYRLTYAYFIGGEAGDAF---DGFN--FGDDPSDKSYTYHNGMRF\n",
      "STFDNDNDNF-----EGNCAEQDGSGWWMNRCHAGHLNGPYYIGGVYSRDTGTNSYDNGI\n",
      "IWATWRDRWYSMKKTTMKIIPFNRLS\n",
      ">_seed_P04115\n",
      "HDA---QIQELSEMWRVNQQFVTRLQQQLVDIRQTCSRSCQDTTKISPITGKDCQQVVDN\n",
      "GGKDSGLYYIK-PLKAKQP--FLVFCEIE-NGNGWTVIQHRHDGSVN---FTRDWVSYRE\n",
      "GFGYLA-----------PTLTTEFWLGNEKIHLLTGQQA--YRLRIDLTDWENTHRYADY\n",
      "GH-----FKLTPESDEYRLFYSMYLDGDAGNAF---DGFD--FGDDPQDKFYTTHLGMLF\n",
      "STPERDNDKY-----EGSCAEQDGSGWWMNRCHAGHLNGKYYFGGNYRKTDVEFPYDDGI\n",
      "IWATWHDRWYSLKMTTMKLLPMGRLS\n",
      ">P02675\n",
      "NSNIPTNLRVLRSILENLRSKIQKLESDVSAQMEYCRTPCTVSCNIPVVSGKECEEIIRK\n",
      "GGETSEMYLIQ-PDSSVKP--YRVYCDMNTENGGWTVIQNRQDGSVD---FGRKWDPYKQ\n",
      "GFGNVATNTDGKNYCGLP---GEYWLGNDKISQLTRMGP--TELLIEMEDWKGDKVKAHY\n",
      "GG-----FTVQNEANKYQISVNKYR-GTAGNALM--DGASQLMGE---NRTMTIHNGMFF\n",
      "STYDRDNDGWLTSDPRKQCSKEDGGGWWYNRCHAANPNGRYYWGGQYTWDMAKHGTDDGV\n",
      "VWMNWKGSWYSMRKMSMKIRPF---F\n",
      ">Q02020\n",
      "DNNIPSSLRVLRAVIDSLHKKIQKLENAIATQTDYCRSPCVASCNIPVVSGRECEDIYRK\n",
      "GGETSEMYIIQ-PDPFTTP--YRVYCDMETDNGGWTLIQNRQDGSVN---FGRAWDEYKR\n",
      "GFGNIA-KSGGKKYCDTP---GEYWLGNDKISQLTKIGP--TKVLIEMEDWNGDKVSALY\n",
      "GG-----FTIHNEGNKYQLSVSNYK-GNAGNALM--EGASQLYGE---NRTMTIHNGMYF\n",
      "STYDRDNDGWLTTDPRKQCSKEDGGGWWYNRCHAANPNGRYYWGGTYSWDMAKHGTDDGI\n",
      "VWMNWKGSWYSMKKMSMKIKPY---F\n",
      ">P02678\n",
      "ESTVAGSLRSMKSVLEHLRAKMQRMEEAIKTQKELCSAPCTVNCRVPVVSGMHCEDIYRN\n",
      "GGRTSEAYYIQ-PDLFSEP--YKVFCDMESHGGGWTVVQNRVDGSSN---FARDWNTYKA\n",
      "EFGNIA-FGNGKSICNIP---GEYWLGTKTVHQLTKQHT--QQVLFDMSDWEGSSVYAQY\n",
      "AS-----FRPENEAQGYRLWVEDYS-GNAGNALL--EGATQLMGD---NRTMTIHNGMQF\n",
      "STFDRDNDNWNPGDPTKHCSREDAGGWWYNRCHAANPNGRYYWGGIYTKEQADYGTDDGV\n",
      "VWMNWKGSWYSMRQMAMKLRPK----\n",
      ">P02671\n",
      "------------------------------------------------------------\n",
      "--------------------------------GGWLLIQQRMDGSLN---FNRTWQDYKR\n",
      "GFGSLN--DEGE---------GEFWLGNDYLHLLTQRGSV---LRVELEDWAGNEAYAEY\n",
      "H------FRVGSEAEGYALQVSSYE-GTAGDALI--EGSVEEGAE------YTSHNNMQF\n",
      "STFDRDADQW-----EENCAEVYGGGWWYNNCQAANLNGIYYPGGSYDRNNSPYEIENGV\n",
      "VWVSFRGADYSLRAVRMKIRPL---V\n",
      "\n",
      ">_seed__seed__seed_Q8N539\n",
      "--------------------------------------------------AQFQQSVVVV\n",
      "PDQDWAKDWHAFPVR--RVTDIFTWFCPPPRTRWTWFFKFAAQPDQ---FQDDDVCQQCK\n",
      "DDDR------------VHGIGNHLVVQQSNCQR--AFKKKKKWFAAPPRDIKIKIWRGWH\n",
      "FCNPPPDRVVQQRAIATHDI-DMDQDDWPVV--------------LHRAHKFHPPRAREC\n",
      "DP-----DGVCVQLVHIGSAHNGGQAGQSFHQPHFFDDD-------FQNAGDNCSHRHGG\n",
      "THTNMMIMIMDHD----\n",
      ">_seed__seed__seed_Q9U8W8\n",
      "--------------------------------------------------AQALLSVVVV\n",
      "PDQDWAWDWHDQVVCVPVNIDIFTWDCPPPRTSWTFFFKFAPDPDALLVQLDDPVCQQCK\n",
      "DDDN------------VHGIGNHLVVQQRSCPS--AWKKKKKWFAAPVRDIWMKIWGRWH\n",
      "F-----HDVVQQRATATHDM-DIPQDPFCVV--------------LHGAHWAHPVGARED\n",
      "AP-----DRQCVQQRHTGRAYVGGQARQSFGDPRFFDDR-------FQNAGHNPSHPHGG\n",
      "THTNIMIMIMHHP----\n",
      ">_seed__seed_Q15485\n",
      "------------------------------------------------DFAAFLQVVVVV\n",
      "PPQDWAWDWHAFPVR---DIDIFTWANPPPNTGWTWQWKFAAQPDQ---QQDAPVDQQCK\n",
      "DDDS------------VHTIGNHLVVQQRSQVP--AFKKKKKWFAAPVGDIKIWIFRGWH\n",
      "F-----HDVVLQRAIATHDTDDMQQDDFCVV--------------LHRAHKDHQVRAREQ\n",
      "DP-----DRQCVLQVHIGSAHNGGQAGQSFGCPHFFAPR-------FQNARHHPSHPHRG\n",
      "THTNMIIMIIHGD----\n",
      ">_seed__seed_O00602\n",
      "------------------------------------------------PFAAFQLSVVVV\n",
      "PPQCWAWDWHAFPVR---DIDIFTWDNPPPNTRWTWQWKFAAQPDQ---QQDAPVCQQCK\n",
      "DDDS------------VHTIGNHLVVQQRSCVP--AFKKKKKWFAAPVGDIKIWIFRGWH\n",
      "F-----DDVVLQRAIATHDTDDMLLDDWCVV--------------LHRAHKHHQVRAREQ\n",
      "DP-----DRQCVQLVHIGSAHNGGQAGQRFHPPHFFADR-------FQRARDNDSHVHGG\n",
      "THTNMIIMIIHHD----\n",
      ">_seed__seed_O75636\n",
      "------------------------------------------------D-ALFQVVVVVV\n",
      "PPQDWAWDWHQFPVR---DIAIFTWDCDPPNTGKTWQWKFADQPDQ---FQDDLVCQQCK\n",
      "DDDR------------VHTIGRHLVVQQRSPVP--HFKKKKKWFAADVGDIKIWIFRTWH\n",
      "F-----HDVVQQRAIATHDTDDMPLDDWCVV--------------LHRAHKDHQVRAREQ\n",
      "DP-----DRQCVVQVHIGSAHNHTQAGQSGHDDPHVVRW-------DQHGRHHCSHCHRR\n",
      "GHTRIMIMIID------\n",
      ">_seed__seed_P24821\n",
      "------------------------------------------------DAAAFQLSVVVV\n",
      "VDQDWAWDWHAQPVR----IDIFTWQNPDPRTRKTWFFKFAAQPDA---QQDDPVCQQCK\n",
      "DDDR------------VYGIGRHLVVQQRSCVN--AFKKKKKWFDAP-HDIKMKIFLGWG\n",
      "F-----HDVVLLRFIAGHDI-DMDQDPFCVV--------------LHGAHKDHPVND---\n",
      "---------PCVVLSHIGSADVGGQAGQRFHDP---AQD-------FQNAGHNDSRPRRR\n",
      "DHTNIMIMIMDHS----\n",
      ">_seed_Q15485\n",
      "------------------------------------------------DFAAFLQVVVVV\n",
      "PPQDWAWDWHAFPVR---DIDIFTWANPPPNTGWTWQWKFAAQPDQ---QQDAPVDQQCK\n",
      "DDDS------------VHTIGNHLVVQQRSQVP--AFKKKKKWFAAPVGDIKIWIFRGWH\n",
      "F-----HDVVLQRAIATHDTDDMQQDDFCVV--------------LHRAHKDHQVRAREQ\n",
      "DP-----DRQCVLQVHIGSAHNGGQAGQSFGCPHFFAPR-------FQNARHHPSHPHRG\n",
      "THTNMIIMIIHGD----\n",
      ">_seed_O00602\n",
      "------------------------------------------------PFAAFQLSVVVV\n",
      "PPQCWAWDWHAFPVR---DIDIFTWDNPPPNTRWTWQWKFAAQPDQ---QQDAPVCQQCK\n",
      "DDDS------------VHTIGNHLVVQQRSCVP--AFKKKKKWFAAPVGDIKIWIFRGWH\n",
      "F-----DDVVLQRAIATHDTDDMLLDDWCVV--------------LHRAHKHHQVRAREQ\n",
      "DP-----DRQCVQLVHIGSAHNGGQAGQRFHPPHFFADR-------FQRARDNDSHVHGG\n",
      "THTNMIIMIIHHD----\n",
      ">_seed_O75636\n",
      "------------------------------------------------D-ALFQVVVVVV\n",
      "PPQDWAWDWHQFPVR---DIAIFTWDCDPPNTGKTWQWKFADQPDQ---FQDDLVCQQCK\n",
      "DDDR------------VHTIGRHLVVQQRSPVP--HFKKKKKWFAADVGDIKIWIFRTWH\n",
      "F-----HDVVQQRAIATHDTDDMPLDDWCVV--------------LHRAHKDHQVRAREQ\n",
      "DP-----DRQCVVQVHIGSAHNHTQAGQSGHDDPHVVRW-------DQHGRHHCSHCHRR\n",
      "GHTRIMIMIID------\n",
      ">_seed_P24821\n",
      "------------------------------------------------DAAAFQLSVVVV\n",
      "VDQDWAWDWHAQPVR----IDIFTWQNPDPRTRKTWFFKFAAQPDA---QQDDPVCQQCK\n",
      "DDDR------------VYGIGRHLVVQQRSCVN--AFKKKKKWFDAP-HDIKMKIFLGWG\n",
      "F-----HDVVLLRFIAGHDI-DMDQDPFCVV--------------LHGAHKDHPVND---\n",
      "---------PCVVLSHIGSADVGGQAGQRFHDP---AQD-------FQNAGHNDSRPRRR\n",
      "DHTNIMIMIMDHS----\n",
      ">_seed_Q15389\n",
      "--------------------------------------------------FQFLLVVVVV\n",
      "PPQDWDWDWGDFPPD--PDIAIFTWDCPPPNTRKTWQFKAAPQPDA---QQDDPVCQACW\n",
      "DDDR------------VHTIGNHLVSQQRSCVV--DWKKKKKWFAAQPGDIKMKIFLTWA\n",
      "F-----DHVVSLRKIATHDI-DIDQADDFLAD-------------RPIAHKDHPVRQGEQ\n",
      "AP-----DRVCVVQNGIGRAHVHTRFDQSFGADDDPRQAQ------QQNAGGHDRRPHRR\n",
      "GHTNIMIMIMDDSVD--\n",
      ">_seed_O15123\n",
      "--------------------------------------------------DLFLLRVVVV\n",
      "PNLDWAWDWHAQPPG--RDIFIFTWDNPPPRTRKTWQFKDAAQPDQ---QQDDLVCQQCK\n",
      "DDDN------------VHGIGRHLVSQQRVLVV--AFKKKKKWFAALVGDIKIKIFRGWH\n",
      "F-----HHVVQQRWIPTHDI-DMDLAPDFQQD-------------PPTAHKAHPHRQREC\n",
      "AP-----DRCCVVQNHIDRAHVHTQADQSFGDDDNPRQAQ------ASNAGGHDRRPHPR\n",
      "GHTNMMIMIMHGSPR--\n",
      ">_seed_Q9BY76\n",
      "--------------------------------------------------AQAQQSVQVV\n",
      "PDADFAWDWHHFPPG---DIDIFTWDQDPV-GIKTWQFKAAAQPDA---QQDALVCQACW\n",
      "DDDR------------VHTIGHHLVVVCSNQPQ--FWKKKWKWFAAPVGDIDIDID-TWG\n",
      "F-----DDVVQQRWT----------WAAPVRVVVLPDP----DADRDIFHKDHQVRADPP\n",
      "DP-----DGLNVQAGHIDRCPDSQAFDQSFADDDDDPVSN------LSRAGDGDVPPHRS\n",
      "GHTRMMIMIMDGD----\n",
      ">_seed_P02679\n",
      "---CVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVPPDDDDDPADEDDQEEAAQLRSLVV\n",
      "PDAAWFKYWYDFPPD--PDIDIFTWDHDPVSFTKTWFFKFAAQPDF---FLDAPVCQACW\n",
      "DWDGGRVR--------PTGIGRHLVVQQRSQPSPDFGKKKKKWFAAPVGDIKIWIFARWH\n",
      "F-----DDVVLQTATAGDDTPDMPLDPFLQADDPPPDPCRSVQAHRHRAHKDAPVRAGEP\n",
      "DP-----HHPRPLLRHIDSAHNHHQAAQSFHDDPNFADDLVRGGPRAQTARHRVSHDPSR\n",
      "GHTNMIIIIMDGVVPPD\n",
      ">_seed_O93568\n",
      "---VVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVPDDDDDDQQDADEFEELAFLRSLVV\n",
      "PDAQWFKYWYAHAVR--PDIDIFTWDHHPVSFIWTWFWKFAALPDA---FLDDDVCQACK\n",
      "DWDGGGVP--------PITIGRHLVVQQRSLVSDPFKKKKKKKFAAPVGDIKIWIFIRWH\n",
      "F-----DDVVQQRATATDDTDDMPLDQFLQFDPPVPDPCRRVQAHQHGAHKDHQVRAREC\n",
      "DP-----DRQNVLLSHIGRGHVHHQAAQSGGADPPFADDWDPDPSTDLNAGDHPVHDDSR\n",
      "GHTNIIIMIMGGVVVVD\n",
      ">_seed_P04115\n",
      "---VVVVVVVVVVVVVVVVVVVVVVVVVVVVVVPVPPDDDDDVVPADPAEEQAQLVSLVS\n",
      "PDDQWFWYFYAHDVR--PDTGIFTW-HDDPSFTWTFFWKDAAQPDQ---QLDDDQCQACK\n",
      "DWDGGRDP--------RTGIGNHLVVQQRSLVR--AKKKKKKWFAAVVRDIKIWIFTSWH\n",
      "F-----HDCVLPTQTATPGTDDMDAADFLQFDPPPPDPCRSVQAHQHRAHWAHLVGAREC\n",
      "DP-----DRLRVLQSHIGRHHVHHQAAQSFHDDPPFADADPDDPNRFQRAGDHNVHPDSR\n",
      "GHTNMMTIIIDDDDPPD\n",
      ">P02675\n",
      "DDPVPVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVDDDDAQADEDPQEEQALLVSVVV\n",
      "PPQAFAKHWYDNDPV--DDIDIWTWFNPPPNTSWTWQFKFAAQPDA---LQDDPVCQAQK\n",
      "DDFRFHDPVVDNHGRGGTGIHNHLPVQQVVLVS--AFKKKKKKFAAPVGDIWIKIFRGWH\n",
      "F-----HDVVQLGWIATHDI-DIDLADPQAQHADPDDDPVRVQAHLHGAHKDFLVGAGEQ\n",
      "DPDDDNCDDQHNQAGGIDRAHNGHQFAQRFHDDPPFFDAQVVGPRSFLHEGDHNSHPHSG\n",
      "THTNIIIMIMDHND---\n",
      ">Q02020\n",
      "PPNVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVDDDDAQADEDPQEELAQLVSVVV\n",
      "PPQAWAKHWYPVDPP--DDIGIFTWDCPPPNTSWTWAFKFAAQPDA---FQDAPQCQAQK\n",
      "DWFQFHPP-DDSHTRGGTGIGRHLVVVQSNCVS--AQKKKKKKFAAPVGDIKIKIFSGWH\n",
      "F-----HDVVQLRQTATHPI-DIDLDPFQAQAAPPDDDPVRVQAHQHRAGKDHLVGAREQ\n",
      "CDDPPNCPGQRNQLRHIGRHHNHHQAAQSFGDDPPFFDALVVGPNSFLNAGDNCNHPHSG\n",
      "GHTNIITMIMHHND---\n",
      ">P02678\n",
      "CPVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVPVDPDDAQADADPQEEAALLRSVVV\n",
      "PPQDWFKHWHNVDPP--DDTDIFTWHNPPPNTRWTWQFKDAAQPDA---FQDAPVCQQQK\n",
      "DWDQFHQP-PVRTTRHGTGIGHHLQVVQVNCVV--FWKKKKKKFAAPVGDIWIKIFRTWH\n",
      "F-----HHVVVLRATATHDI-DIDQDQCQQFNLVVDDDPVGVQAHLHRAHKHHLPGARAQ\n",
      "QDVPDNPQRLNVQQRHIGRHHSHHRFAASGRDDPDFFDALVVGPVSFLHAGDNCSHRHSG\n",
      "THTNIMIMIIHRD----\n",
      ">P02671\n",
      "------------------------------------------------------------\n",
      "-------------------------------DDKDFFFKDAALPDA---AQDDPVCLQCK\n",
      "DWDADPVR--------DTGIGRHLVVQQVVDV---QKWKKKKWFAAPVGDIWIKIW-TWH\n",
      "F-----DDVVLLSFIATDDI-DIDQDPCAQQN---DDVVGRQFAHQHRAHKAHQVGDDEC\n",
      "DP-----DGCNVVLRHIGRAHNGHQFDQNFGDDPDFFDALCPDPPSALRAHDRDSHPHSG\n",
      "THTNIMTMIIDHVD---\n",
      "\n",
      ">_seed__seed__seed_Q8N539\n",
      "--------------------------------------------------PRDCLDVLLS\n",
      "GQQDDGVYSVF-PTHY--PAGFQVYCDMRTDGGGWTVFQRREDGSVN---FFRGWDAYRD\n",
      "GFGRLT---------------GEHWLGLKRIHALTTQAA--YELHVDLEDFENGTAYARY\n",
      "GSFGVGLFSVDPEEDGYPLTVADYS-GTAGDSL-------------------LKHSGMRF\n",
      "TTKDRDSDHS-----ENNCAAFYRGAWWYRNCHTSNLNGQYLRGAHAS-------YADGV\n",
      "EWSSWTGWQYSLKFSEMKIRPV----\n",
      ">_seed__seed__seed_Q9U8W8\n",
      "--------------------------------------------------PTDCADILLN\n",
      "GYRSSGGYRIW-PKSWMTVGTLNVYCDMETDGGGWTVIQRRGNYGNPSDYFYKPWKNYKL\n",
      "GFGNIE---------------KDFWLGNDRIFALTNQRN--YMIRFDLKDKENDTRYAIY\n",
      "QDFWI-----ENEDYLYCLHIGNYS-GDAGNSF-------------------GRHNGHNF\n",
      "STIDKDHDTH-----ETHCAQTYKGGWWYDRCHESNLNGLYLNGEHNS-------YADGI\n",
      "EWRAWKGYHYSLPQVEMKIRPV----\n",
      ">_seed__seed_Q15485\n",
      "------------------------------------------------TGPRTCKDLLDR\n",
      "GHFLSGWHTIYLPDCR--P--LTVLCDMDTDGGGWTVFQRRVDGSVD---FYRDWATYKQ\n",
      "GFGSRL---------------GEFWLGNDNIHALTAQGT--SELRVDLVDFEDNYQFAKY\n",
      "RS-----FKVADEAEKYNLVLGAFVEGSAGDSL-------------------TFHNNQSF\n",
      "STKDQDNDLN-----TGNCAVMFQGAWWYKNCHVSNLNGRYLRGTHGS-------FANGI\n",
      "NWKSGKGYNYSYKVSEMKVRPA----\n",
      ">_seed__seed_O00602\n",
      "------------------------------------------------TGPRNCKDLLDR\n",
      "GYFLSGWHTIYLPDCR--P--LTVLCDMDTDGGGWTVFQRRMDGSVD---FYRDWAAYKQ\n",
      "GFGSQL---------------GEFWLGNDNIHALTAQGS--SELRVDLVDFEGNHQFAKY\n",
      "KS-----FKVADEAEKYKLVLGAFVGGSAGNSL-------------------TGHNNNFF\n",
      "STKDQDNDVS-----SSNCAEKFQGAWWYADCHASNLNGLYLMGPHES-------FANGI\n",
      "NWSAAKGYKYSYKVSEMKVRPA----\n",
      ">_seed__seed_O75636\n",
      "-------------------------------------------------GPRNCRELLSQ\n",
      "GATLSGWYHLCLPEGR--A--LPVFCDMDTEGGGWLVFQRRQDGSVD---FFRSWSSYRA\n",
      "GFGNQE---------------SEFWLGNENLHQLTLQGN--WELRVELEDFNGNRTFAHY\n",
      "AT-----FRLLGEVDHYQLALGKFSEGTAGDSL-------------------SLHSGRPF\n",
      "TTYDADHDSS-----NSNCAVIVHGAWWYASCYRSNLNGRYAVSEAAA-------HKYGI\n",
      "DWASGRGVGHPYRRVRMMLR------\n",
      ">_seed__seed_P24821\n",
      "------------------------------------------------PFPKDCSQAMLN\n",
      "GDTTSGLYTIYLDKAQ--A--LEVFCDMTSDGGGWIVFLRRKNGREN---FYQNWKAYAA\n",
      "GFGDRR---------------EEFWLGLDNLNKITAQGQ--YELRVDLRD-HGETAFAVY\n",
      "DK-----FSVGDAKTRYKLKVEGYS-GTAGDSM-------------------AYHNGRSF\n",
      "STFDK------------NCALSYKGAFWYRNCHRVNLMGRYGDNNHSQ----------GV\n",
      "NWFHWKGHEHSIQFAEMKLRPS----\n",
      ">_seed_Q15485\n",
      "------------------------------------------------TGPRTCKDLLDR\n",
      "GHFLSGWHTIYLPDCR--P--LTVLCDMDTDGGGWTVFQRRVDGSVD---FYRDWATYKQ\n",
      "GFGSRL---------------GEFWLGNDNIHALTAQGT--SELRVDLVDFEDNYQFAKY\n",
      "RS-----FKVADEAEKYNLVLGAFVEGSAGDSL-------------------TFHNNQSF\n",
      "STKDQDNDLN-----TGNCAVMFQGAWWYKNCHVSNLNGRYLRGTHGS-------FANGI\n",
      "NWKSGKGYNYSYKVSEMKVRPA----\n",
      ">_seed_O00602\n",
      "------------------------------------------------TGPRNCKDLLDR\n",
      "GYFLSGWHTIYLPDCR--P--LTVLCDMDTDGGGWTVFQRRMDGSVD---FYRDWAAYKQ\n",
      "GFGSQL---------------GEFWLGNDNIHALTAQGS--SELRVDLVDFEGNHQFAKY\n",
      "KS-----FKVADEAEKYKLVLGAFVGGSAGNSL-------------------TGHNNNFF\n",
      "STKDQDNDVS-----SSNCAEKFQGAWWYADCHASNLNGLYLMGPHES-------FANGI\n",
      "NWSAAKGYKYSYKVSEMKVRPA----\n",
      ">_seed_O75636\n",
      "-------------------------------------------------GPRNCRELLSQ\n",
      "GATLSGWYHLCLPEGR--A--LPVFCDMDTEGGGWLVFQRRQDGSVD---FFRSWSSYRA\n",
      "GFGNQE---------------SEFWLGNENLHQLTLQGN--WELRVELEDFNGNRTFAHY\n",
      "AT-----FRLLGEVDHYQLALGKFSEGTAGDSL-------------------SLHSGRPF\n",
      "TTYDADHDSS-----NSNCAVIVHGAWWYASCYRSNLNGRYAVSEAAA-------HKYGI\n",
      "DWASGRGVGHPYRRVRMMLR------\n",
      ">_seed_P24821\n",
      "------------------------------------------------PFPKDCSQAMLN\n",
      "GDTTSGLYTIYLDKAQ--A--LEVFCDMTSDGGGWIVFLRRKNGREN---FYQNWKAYAA\n",
      "GFGDRR---------------EEFWLGLDNLNKITAQGQ--YELRVDLRD-HGETAFAVY\n",
      "DK-----FSVGDAKTRYKLKVEGYS-GTAGDSM-------------------AYHNGRSF\n",
      "STFDK------------NCALSYKGAFWYRNCHRVNLMGRYGDNNHSQ----------GV\n",
      "NWFHWKGHEHSIQFAEMKLRPS----\n",
      ">_seed_Q15389\n",
      "--------------------------------------------------FRDCADVYQA\n",
      "GFNKSGIYTIYINNMP-EP--KKVFCNMDVNGGGWTVIQHREDGSLD---FQRGWKEYKM\n",
      "GFGNPS---------------GEYWLGNEFIFAITSQRQ--YMLRIELMDWEGNRAYSQY\n",
      "DR-----FHIGNEKQNYRLYLKGHT-GTAGKQS------------------SLILHGADF\n",
      "STKDADNDNC-----MCKCALMLTGGWWFDACGPSNLNGMFYTAGQNH------GKLNGI\n",
      "KWHYFKGPSYSLRSTTMMIRPLD--F\n",
      ">_seed_O15123\n",
      "--------------------------------------------------FRDCAEVFKS\n",
      "GHTTNGIYTLTFPNST-EE--IKAYCDMEAGGGGWTIIQRREDGSVD---FQRTWKEYKV\n",
      "GFGNPS---------------GEYWLGNEFVSQLTNQQR--YVLKIHLKDWEGNEAYSLY\n",
      "EH-----FYLSSEELNYRIHLKGLT-GTAGKIS------------------SISQPGNDF\n",
      "STKDGDNDKC-----ICKCSQMLTGGWWFDACGPSNLNGMYYPQRQNT------NKFNGI\n",
      "KWYYWKGSGYSLKATTMMIRPAD--F\n",
      ">_seed_Q9BY76\n",
      "--------------------------------------------------PRDCQELFQV\n",
      "GERQSGLFEIQ-PQGS-PP--FLVNCKMTSD-GGWTVIQRRHDGSVD---FNRPWEAYKA\n",
      "GFGDPH---------------GEFWLGLEKVHSITGDRN--SRLAVQLRDWDGNAELLQF\n",
      "S------VHLGGEDTAYSLQLTAPVAGQLGATTVPPSG---------------------F\n",
      "STWDQDHDLR-----RDKCAKSLSGGWWFGTCSHSNLNGQYFPQRQ--------KLKKGI\n",
      "FWKTWRGRYYPLQATTMLIQPM----\n",
      ">_seed_P02679\n",
      "HDS---SIRYLQEIYNSNNQKIVNLKEKVAQLEAQCQEPCKDTVQIHDITGKDCQDIANK\n",
      "GAKQSGLYFIK-PLKANQQ--FLVYCEIDGSGNGWTVFQKRLDGSVD---FKKNWIQYKE\n",
      "GFGHLS-----------PTGTTEFWLGNEKIHLISTQSAIPYALRVELEDWNGRTSTADY\n",
      "AM-----FKVGPEADKYRLTYAYFAGGDAGDAF---DGFD--FGDDPSDKFFTSHKGMQF\n",
      "STWDNDNDKF-----EGNCAEQDGSGWWMNKCHAGHLNGVYYQGGTYSKASTPNGYDNGI\n",
      "IWATWKTRWYSMKKTTMKIIPFNRLT\n",
      ">_seed_O93568\n",
      "HEN---TIQQLTDMHIMNSNKITQLKQKIAQLESHCQEPCKDTAEIQETTGRDCQDIANK\n",
      "GARKSGLYFIK-PQKAKQS--FLVYCEIDTYGNGWTVLQRRLDGSED---FRRNWVQYKE\n",
      "GFGHLS-----------PDDTTEFWLGNEKIHLITTQSTLPYALRIELEDWSGKKGTADY\n",
      "AV-----FKVGTEEDKYRLTYAYFIGGEAGDAF---DGFN--FGDDPSDKSYTYHNGMRF\n",
      "STFDNDNDNF-----EGNCAEQDGSGWWMNRCHAGHLNGPYYIGGVYSRDTGTNSYDNGI\n",
      "IWATWRDRWYSMKKTTMKIIPFNRLS\n",
      ">_seed_P04115\n",
      "HDA---QIQELSEMWRVNQQFVTRLQQQLVDIRQTCSRSCQDTTKISPITGKDCQQVVDN\n",
      "GGKDSGLYYIK-PLKAKQP--FLVFCEIE-NGNGWTVIQHRHDGSVN---FTRDWVSYRE\n",
      "GFGYLA-----------PTLTTEFWLGNEKIHLLTGQQA--YRLRIDLTDWENTHRYADY\n",
      "GH-----FKLTPESDEYRLFYSMYLDGDAGNAF---DGFD--FGDDPQDKFYTTHLGMLF\n",
      "STPERDNDKY-----EGSCAEQDGSGWWMNRCHAGHLNGKYYFGGNYRKTDVEFPYDDGI\n",
      "IWATWHDRWYSLKMTTMKLLPMGRLS\n",
      ">P02675\n",
      "NSNIPTNLRVLRSILENLRSKIQKLESDVSAQMEYCRTPCTVSCNIPVVSGKECEEIIRK\n",
      "GGETSEMYLIQ-PDSSVKP--YRVYCDMNTENGGWTVIQNRQDGSVD---FGRKWDPYKQ\n",
      "GFGNVATNTDGKNYCGLP---GEYWLGNDKISQLTRMGP--TELLIEMEDWKGDKVKAHY\n",
      "GG-----FTVQNEANKYQISVNKYR-GTAGNALM--DGASQLMGE---NRTMTIHNGMFF\n",
      "STYDRDNDGWLTSDPRKQCSKEDGGGWWYNRCHAANPNGRYYWGGQYTWDMAKHGTDDGV\n",
      "VWMNWKGSWYSMRKMSMKIRPF---F\n",
      ">Q02020\n",
      "DNNIPSSLRVLRAVIDSLHKKIQKLENAIATQTDYCRSPCVASCNIPVVSGRECEDIYRK\n",
      "GGETSEMYIIQ-PDPFTTP--YRVYCDMETDNGGWTLIQNRQDGSVN---FGRAWDEYKR\n",
      "GFGNIA-KSGGKKYCDTP---GEYWLGNDKISQLTKIGP--TKVLIEMEDWNGDKVSALY\n",
      "GG-----FTIHNEGNKYQLSVSNYK-GNAGNALM--EGASQLYGE---NRTMTIHNGMYF\n",
      "STYDRDNDGWLTTDPRKQCSKEDGGGWWYNRCHAANPNGRYYWGGTYSWDMAKHGTDDGI\n",
      "VWMNWKGSWYSMKKMSMKIKPY---F\n",
      ">P02678\n",
      "ESTVAGSLRSMKSVLEHLRAKMQRMEEAIKTQKELCSAPCTVNCRVPVVSGMHCEDIYRN\n",
      "GGRTSEAYYIQ-PDLFSEP--YKVFCDMESHGGGWTVVQNRVDGSSN---FARDWNTYKA\n",
      "EFGNIA-FGNGKSICNIP---GEYWLGTKTVHQLTKQHT--QQVLFDMSDWEGSSVYAQY\n",
      "AS-----FRPENEAQGYRLWVEDYS-GNAGNALL--EGATQLMGD---NRTMTIHNGMQF\n",
      "STFDRDNDNWNPGDPTKHCSREDAGGWWYNRCHAANPNGRYYWGGIYTKEQADYGTDDGV\n",
      "VWMNWKGSWYSMRQMAMKLRPK----\n",
      ">P02671\n",
      "------------------------------------------------------------\n",
      "--------------------------------GGWLLIQQRMDGSLN---FNRTWQDYKR\n",
      "GFGSLN--DEGE---------GEFWLGNDYLHLLTQRGSV---LRVELEDWAGNEAYAEY\n",
      "H------FRVGSEAEGYALQVSSYE-GTAGDALI--EGSVEEGAE------YTSHNNMQF\n",
      "STFDRDADQW-----EENCAEVYGGGWWYNNCQAANLNGIYYPGGSYDRNNSPYEIENGV\n",
      "VWVSFRGADYSLRAVRMKIRPL---V\n",
      "\n",
      ">_seed__seed__seed_Q8N539\n",
      "--------------------------------------------------AQFQQSVVVV\n",
      "PDQDWAKDWHAFPVR--RVTDIFTWFCPPPRTRWTWFFKFAAQPDQ---FQDDDVCQQCK\n",
      "DDDR------------VHGIGNHLVVQQSNCQR--AFKKKKKWFAAPPRDIKIKIWRGWH\n",
      "FCNPPPDRVVQQRAIATHDI-DMDQDDWPVV--------------LHRAHKFHPPRAREC\n",
      "DP-----DGVCVQLVHIGSAHNGGQAGQSFHQPHFFDDD-------FQNAGDNCSHRHGG\n",
      "THTNMMIMIMDHD----\n",
      ">_seed__seed__seed_Q9U8W8\n",
      "--------------------------------------------------AQALLSVVVV\n",
      "PDQDWAWDWHDQVVCVPVNIDIFTWDCPPPRTSWTFFFKFAPDPDALLVQLDDPVCQQCK\n",
      "DDDN------------VHGIGNHLVVQQRSCPS--AWKKKKKWFAAPVRDIWMKIWGRWH\n",
      "F-----HDVVQQRATATHDM-DIPQDPFCVV--------------LHGAHWAHPVGARED\n",
      "AP-----DRQCVQQRHTGRAYVGGQARQSFGDPRFFDDR-------FQNAGHNPSHPHGG\n",
      "THTNIMIMIMHHP----\n",
      ">_seed__seed_Q15485\n",
      "------------------------------------------------DFAAFLQVVVVV\n",
      "PPQDWAWDWHAFPVR---DIDIFTWANPPPNTGWTWQWKFAAQPDQ---QQDAPVDQQCK\n",
      "DDDS------------VHTIGNHLVVQQRSQVP--AFKKKKKWFAAPVGDIKIWIFRGWH\n",
      "F-----HDVVLQRAIATHDTDDMQQDDFCVV--------------LHRAHKDHQVRAREQ\n",
      "DP-----DRQCVLQVHIGSAHNGGQAGQSFGCPHFFAPR-------FQNARHHPSHPHRG\n",
      "THTNMIIMIIHGD----\n",
      ">_seed__seed_O00602\n",
      "------------------------------------------------PFAAFQLSVVVV\n",
      "PPQCWAWDWHAFPVR---DIDIFTWDNPPPNTRWTWQWKFAAQPDQ---QQDAPVCQQCK\n",
      "DDDS------------VHTIGNHLVVQQRSCVP--AFKKKKKWFAAPVGDIKIWIFRGWH\n",
      "F-----DDVVLQRAIATHDTDDMLLDDWCVV--------------LHRAHKHHQVRAREQ\n",
      "DP-----DRQCVQLVHIGSAHNGGQAGQRFHPPHFFADR-------FQRARDNDSHVHGG\n",
      "THTNMIIMIIHHD----\n",
      ">_seed__seed_O75636\n",
      "------------------------------------------------D-ALFQVVVVVV\n",
      "PPQDWAWDWHQFPVR---DIAIFTWDCDPPNTGKTWQWKFADQPDQ---FQDDLVCQQCK\n",
      "DDDR------------VHTIGRHLVVQQRSPVP--HFKKKKKWFAADVGDIKIWIFRTWH\n",
      "F-----HDVVQQRAIATHDTDDMPLDDWCVV--------------LHRAHKDHQVRAREQ\n",
      "DP-----DRQCVVQVHIGSAHNHTQAGQSGHDDPHVVRW-------DQHGRHHCSHCHRR\n",
      "GHTRIMIMIID------\n",
      ">_seed__seed_P24821\n",
      "------------------------------------------------DAAAFQLSVVVV\n",
      "VDQDWAWDWHAQPVR----IDIFTWQNPDPRTRKTWFFKFAAQPDA---QQDDPVCQQCK\n",
      "DDDR------------VYGIGRHLVVQQRSCVN--AFKKKKKWFDAP-HDIKMKIFLGWG\n",
      "F-----HDVVLLRFIAGHDI-DMDQDPFCVV--------------LHGAHKDHPVND---\n",
      "---------PCVVLSHIGSADVGGQAGQRFHDP---AQD-------FQNAGHNDSRPRRR\n",
      "DHTNIMIMIMDHS----\n",
      ">_seed_Q15485\n",
      "------------------------------------------------DFAAFLQVVVVV\n",
      "PPQDWAWDWHAFPVR---DIDIFTWANPPPNTGWTWQWKFAAQPDQ---QQDAPVDQQCK\n",
      "DDDS------------VHTIGNHLVVQQRSQVP--AFKKKKKWFAAPVGDIKIWIFRGWH\n",
      "F-----HDVVLQRAIATHDTDDMQQDDFCVV--------------LHRAHKDHQVRAREQ\n",
      "DP-----DRQCVLQVHIGSAHNGGQAGQSFGCPHFFAPR-------FQNARHHPSHPHRG\n",
      "THTNMIIMIIHGD----\n",
      ">_seed_O00602\n",
      "------------------------------------------------PFAAFQLSVVVV\n",
      "PPQCWAWDWHAFPVR---DIDIFTWDNPPPNTRWTWQWKFAAQPDQ---QQDAPVCQQCK\n",
      "DDDS------------VHTIGNHLVVQQRSCVP--AFKKKKKWFAAPVGDIKIWIFRGWH\n",
      "F-----DDVVLQRAIATHDTDDMLLDDWCVV--------------LHRAHKHHQVRAREQ\n",
      "DP-----DRQCVQLVHIGSAHNGGQAGQRFHPPHFFADR-------FQRARDNDSHVHGG\n",
      "THTNMIIMIIHHD----\n",
      ">_seed_O75636\n",
      "------------------------------------------------D-ALFQVVVVVV\n",
      "PPQDWAWDWHQFPVR---DIAIFTWDCDPPNTGKTWQWKFADQPDQ---FQDDLVCQQCK\n",
      "DDDR------------VHTIGRHLVVQQRSPVP--HFKKKKKWFAADVGDIKIWIFRTWH\n",
      "F-----HDVVQQRAIATHDTDDMPLDDWCVV--------------LHRAHKDHQVRAREQ\n",
      "DP-----DRQCVVQVHIGSAHNHTQAGQSGHDDPHVVRW-------DQHGRHHCSHCHRR\n",
      "GHTRIMIMIID------\n",
      ">_seed_P24821\n",
      "------------------------------------------------DAAAFQLSVVVV\n",
      "VDQDWAWDWHAQPVR----IDIFTWQNPDPRTRKTWFFKFAAQPDA---QQDDPVCQQCK\n",
      "DDDR------------VYGIGRHLVVQQRSCVN--AFKKKKKWFDAP-HDIKMKIFLGWG\n",
      "F-----HDVVLLRFIAGHDI-DMDQDPFCVV--------------LHGAHKDHPVND---\n",
      "---------PCVVLSHIGSADVGGQAGQRFHDP---AQD-------FQNAGHNDSRPRRR\n",
      "DHTNIMIMIMDHS----\n",
      ">_seed_Q15389\n",
      "--------------------------------------------------FQFLLVVVVV\n",
      "PPQDWDWDWGDFPPD--PDIAIFTWDCPPPNTRKTWQFKAAPQPDA---QQDDPVCQACW\n",
      "DDDR------------VHTIGNHLVSQQRSCVV--DWKKKKKWFAAQPGDIKMKIFLTWA\n",
      "F-----DHVVSLRKIATHDI-DIDQADDFLAD-------------RPIAHKDHPVRQGEQ\n",
      "AP-----DRVCVVQNGIGRAHVHTRFDQSFGADDDPRQAQ------QQNAGGHDRRPHRR\n",
      "GHTNIMIMIMDDSVD--\n",
      ">_seed_O15123\n",
      "--------------------------------------------------DLFLLRVVVV\n",
      "PNLDWAWDWHAQPPG--RDIFIFTWDNPPPRTRKTWQFKDAAQPDQ---QQDDLVCQQCK\n",
      "DDDN------------VHGIGRHLVSQQRVLVV--AFKKKKKWFAALVGDIKIKIFRGWH\n",
      "F-----HHVVQQRWIPTHDI-DMDLAPDFQQD-------------PPTAHKAHPHRQREC\n",
      "AP-----DRCCVVQNHIDRAHVHTQADQSFGDDDNPRQAQ------ASNAGGHDRRPHPR\n",
      "GHTNMMIMIMHGSPR--\n",
      ">_seed_Q9BY76\n",
      "--------------------------------------------------AQAQQSVQVV\n",
      "PDADFAWDWHHFPPG---DIDIFTWDQDPV-GIKTWQFKAAAQPDA---QQDALVCQACW\n",
      "DDDR------------VHTIGHHLVVVCSNQPQ--FWKKKWKWFAAPVGDIDIDID-TWG\n",
      "F-----DDVVQQRWT----------WAAPVRVVVLPDP----DADRDIFHKDHQVRADPP\n",
      "DP-----DGLNVQAGHIDRCPDSQAFDQSFADDDDDPVSN------LSRAGDGDVPPHRS\n",
      "GHTRMMIMIMDGD----\n",
      ">_seed_P02679\n",
      "---CVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVPPDDDDDPADEDDQEEAAQLRSLVV\n",
      "PDAAWFKYWYDFPPD--PDIDIFTWDHDPVSFTKTWFFKFAAQPDF---FLDAPVCQACW\n",
      "DWDGGRVR--------PTGIGRHLVVQQRSQPSPDFGKKKKKWFAAPVGDIKIWIFARWH\n",
      "F-----DDVVLQTATAGDDTPDMPLDPFLQADDPPPDPCRSVQAHRHRAHKDAPVRAGEP\n",
      "DP-----HHPRPLLRHIDSAHNHHQAAQSFHDDPNFADDLVRGGPRAQTARHRVSHDPSR\n",
      "GHTNMIIIIMDGVVPPD\n",
      ">_seed_O93568\n",
      "---VVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVPDDDDDDQQDADEFEELAFLRSLVV\n",
      "PDAQWFKYWYAHAVR--PDIDIFTWDHHPVSFIWTWFWKFAALPDA---FLDDDVCQACK\n",
      "DWDGGGVP--------PITIGRHLVVQQRSLVSDPFKKKKKKKFAAPVGDIKIWIFIRWH\n",
      "F-----DDVVQQRATATDDTDDMPLDQFLQFDPPVPDPCRRVQAHQHGAHKDHQVRAREC\n",
      "DP-----DRQNVLLSHIGRGHVHHQAAQSGGADPPFADDWDPDPSTDLNAGDHPVHDDSR\n",
      "GHTNIIIMIMGGVVVVD\n",
      ">_seed_P04115\n",
      "---VVVVVVVVVVVVVVVVVVVVVVVVVVVVVVPVPPDDDDDVVPADPAEEQAQLVSLVS\n",
      "PDDQWFWYFYAHDVR--PDTGIFTW-HDDPSFTWTFFWKDAAQPDQ---QLDDDQCQACK\n",
      "DWDGGRDP--------RTGIGNHLVVQQRSLVR--AKKKKKKWFAAVVRDIKIWIFTSWH\n",
      "F-----HDCVLPTQTATPGTDDMDAADFLQFDPPPPDPCRSVQAHQHRAHWAHLVGAREC\n",
      "DP-----DRLRVLQSHIGRHHVHHQAAQSFHDDPPFADADPDDPNRFQRAGDHNVHPDSR\n",
      "GHTNMMTIIIDDDDPPD\n",
      ">P02675\n",
      "DDPVPVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVDDDDAQADEDPQEEQALLVSVVV\n",
      "PPQAFAKHWYDNDPV--DDIDIWTWFNPPPNTSWTWQFKFAAQPDA---LQDDPVCQAQK\n",
      "DDFRFHDPVVDNHGRGGTGIHNHLPVQQVVLVS--AFKKKKKKFAAPVGDIWIKIFRGWH\n",
      "F-----HDVVQLGWIATHDI-DIDLADPQAQHADPDDDPVRVQAHLHGAHKDFLVGAGEQ\n",
      "DPDDDNCDDQHNQAGGIDRAHNGHQFAQRFHDDPPFFDAQVVGPRSFLHEGDHNSHPHSG\n",
      "THTNIIIMIMDHND---\n",
      ">Q02020\n",
      "PPNVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVDDDDAQADEDPQEELAQLVSVVV\n",
      "PPQAWAKHWYPVDPP--DDIGIFTWDCPPPNTSWTWAFKFAAQPDA---FQDAPQCQAQK\n",
      "DWFQFHPP-DDSHTRGGTGIGRHLVVVQSNCVS--AQKKKKKKFAAPVGDIKIKIFSGWH\n",
      "F-----HDVVQLRQTATHPI-DIDLDPFQAQAAPPDDDPVRVQAHQHRAGKDHLVGAREQ\n",
      "CDDPPNCPGQRNQLRHIGRHHNHHQAAQSFGDDPPFFDALVVGPNSFLNAGDNCNHPHSG\n",
      "GHTNIITMIMHHND---\n",
      ">P02678\n",
      "CPVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVPVDPDDAQADADPQEEAALLRSVVV\n",
      "PPQDWFKHWHNVDPP--DDTDIFTWHNPPPNTRWTWQFKDAAQPDA---FQDAPVCQQQK\n",
      "DWDQFHQP-PVRTTRHGTGIGHHLQVVQVNCVV--FWKKKKKKFAAPVGDIWIKIFRTWH\n",
      "F-----HHVVVLRATATHDI-DIDQDQCQQFNLVVDDDPVGVQAHLHRAHKHHLPGARAQ\n",
      "QDVPDNPQRLNVQQRHIGRHHSHHRFAASGRDDPDFFDALVVGPVSFLHAGDNCSHRHSG\n",
      "THTNIMIMIIHRD----\n",
      ">P02671\n",
      "------------------------------------------------------------\n",
      "-------------------------------DDKDFFFKDAALPDA---AQDDPVCLQCK\n",
      "DWDADPVR--------DTGIGRHLVVQQVVDV---QKWKKKKWFAAPVGDIWIKIW-TWH\n",
      "F-----DDVVLLSFIATDDI-DIDQDPCAQQN---DDVVGRQFAHQHRAHKAHQVGDDEC\n",
      "DP-----DGCNVVLRHIGRAHNGHQFDQNFGDDPDFFDALCPDPPSALRAHDRDSHPHSG\n",
      "THTNIMTMIIDHVD---\n",
      "\n",
      "final aln\n",
      "childalnsAA {<toytree.TreeNode.TreeNode object at 0x7f18d57b3d60>: {'fasta': '../../testcat/4.10.530/alnscratch/28_inter.fasta'}, <toytree.TreeNode.TreeNode object at 0x7f18d57c4790>: {'fasta': '../../testcat/4.10.530/alnscratch/27_inter.fasta'}, <toytree.TreeNode.TreeNode object at 0x7f18d57c5c30>: {'fasta': '../../testcat/4.10.530/alnscratch/26_inter.fasta'}}\n",
      "childalns3di {<toytree.TreeNode.TreeNode object at 0x7f18d57b3d60>: {'fasta': '../../testcat/4.10.530/alnscratch/28_inter.3di.fasta'}, <toytree.TreeNode.TreeNode object at 0x7f18d57c4790>: {'fasta': '../../testcat/4.10.530/alnscratch/27_inter3di.fasta'}, <toytree.TreeNode.TreeNode object at 0x7f18d57c5c30>: {'fasta': '../../testcat/4.10.530/alnscratch/26_inter3di.fasta'}}\n",
      "finalaln ../../testcat/4.10.530/alnscratch/29_root.fasta.iter.iter\n",
      "nsequences 16\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " identical.   \n",
      "Converged.\n",
      "\n",
      "done\n",
      "dvtditr (text) Version 7.520\n",
      "alg=A, model=Extended, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " NW-INS-i (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'../../testcat/4.10.530/aln3difoldtree.fasta'"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#prepare tree attributes\n",
    "import shutil\n",
    "tre = toytree.tree(infolder + 'struct_tree.PP.nwk.rooted.final'  )\n",
    "for i,n in enumerate(tre.treenode.traverse()):\n",
    "    n.aln = None\n",
    "    n.aln3di = None\n",
    "    n.leafset = None\n",
    "    if len(n.name) == 0:\n",
    "        n.name = 'internal_'+str(i)\n",
    "\n",
    "alnfolder = infolder+'alnscratch/'\n",
    "if not os.path.exists(alnfolder):\n",
    "    os.mkdir(infolder+'alnscratch/')\n",
    "\n",
    "#clear all files in aln scratch\n",
    "cleanup(alnfolder)\n",
    "print(tre.treenode)\n",
    "submat = '../mafftmat/3diHEXmat.txt'\n",
    "finalaln, finalaln3di = traverse_tree_merge_mafft( tre.treenode.get_tree_root(), get_leafset(tre.treenode.get_tree_root()) , alndf , infolder+'alnscratch/' , submat = submat , verbose = True ) \n",
    "print('finalaln',finalaln)\n",
    "#print the final alignments\n",
    "print('nsequences' , len(tre.get_tip_labels()))\n",
    "finalaln = remove_seeds(finalaln)\n",
    "finalaln3di = remove_seeds(finalaln3di)\n",
    "\n",
    "finalaln = remove_redundant(finalaln)\n",
    "finalaln3di = remove_redundant(finalaln3di)\n",
    "\n",
    "#copy to root folder\n",
    "shutil.copy(finalaln, infolder + 'alnAAfoldtree.fasta')\n",
    "shutil.copy(finalaln3di, infolder + 'aln3difoldtree.fasta')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "finalaln = remove_seeds(finalaln)\n",
    "finalaln3di = remove_seeds(finalaln3di)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['../../testcat/4.10.530/alnAA_AA.fasta', '../../testcat/4.10.530/alnAAfoldtree.fasta', '../../testcat/4.10.530/aln3di_3di.fasta', '../../testcat/4.10.530/aln3difoldtree.fasta', '../../testcat/4.10.530/alnAA_3di.fasta', '../../testcat/4.10.530/aln3di_AA.fasta']\n"
     ]
    }
   ],
   "source": [
    "print( glob.glob(infolder + 'aln*fasta'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "snamk\n",
      "16\n",
      ">Q8N539\n",
      "--------------------------------------------------PRDCLDVLLSGQQDDGVYSVF-PTHY--PAGFQVYCDMRTDGGGWTVFQRREDGSVN---FFRGWDAYRDGFGRLT---------------GEHWLGLKRIHALTTQAA--YELHVDLEDFENGTAYARYGSFGVGLFSVDPEEDGYPLTVADYS-GTAGDSL-------------------LKHSGMRFTTKDRDSDHS-----ENNCAAFYRGAWWYRNCHTSNLNGQYLRGAHAS-------YADGVEWSSWTGWQYSLKFSEMKIRPV----\n",
      ">Q9U8W8\n",
      "--------------------------------------------------PTDCADILLNGYRSSGGYRIW-PKSWMTVGTLNVYCDMETDGGGWTVIQRRGNYGNPSDYFYKPWKNYKLGFGNIE---------------KDFWLGNDRIFALTNQRN--YMIRFDLKDKENDTRYAIYQDFWI-----ENEDYLYCLHIGNYS-GDAGNSF-------------------GRHNGHNFSTIDKDHDTH-----ETHCAQTYKGGWWYDRCHESNLNGLYLNGEHNS-------YADGIEWRAWKGYHYSLPQVEMKIRPV----\n",
      ">Q15485\n",
      "------------------------------------------------TGPRTCKDLLDRGHFLSGWHTIYLPDCR--P--LTVLCDMDTDGGGWTVFQRRVDGSVD---FYRDWATYKQGFGSRL---------------GEFWLGNDNIHALTAQGT--SELRVDLVDFEDNYQFAKYRS-----FKVADEAEKYNLVLGAFVEGSAGDSL-------------------TFHNNQSFSTKDQDNDLN-----TGNCAVMFQGAWWYKNCHVSNLNGRYLRGTHGS-------FANGINWKSGKGYNYSYKVSEMKVRPA----\n",
      ">O00602\n",
      "------------------------------------------------TGPRNCKDLLDRGYFLSGWHTIYLPDCR--P--LTVLCDMDTDGGGWTVFQRRMDGSVD---FYRDWAAYKQGFGSQL---------------GEFWLGNDNIHALTAQGS--SELRVDLVDFEGNHQFAKYKS-----FKVADEAEKYKLVLGAFVGGSAGNSL-------------------TGHNNNFFSTKDQDNDVS-----SSNCAEKFQGAWWYADCHASNLNGLYLMGPHES-------FANGINWSAAKGYKYSYKVSEMKVRPA----\n",
      ">O75636\n",
      "-------------------------------------------------GPRNCRELLSQGATLSGWYHLCLPEGR--A--LPVFCDMDTEGGGWLVFQRRQDGSVD---FFRSWSSYRAGFGNQE---------------SEFWLGNENLHQLTLQGN--WELRVELEDFNGNRTFAHYAT-----FRLLGEVDHYQLALGKFSEGTAGDSL-------------------SLHSGRPFTTYDADHDSS-----NSNCAVIVHGAWWYASCYRSNLNGRYAVSEAAA-------HKYGIDWASGRGVGHPYRRVRMMLR------\n",
      ">P24821\n",
      "------------------------------------------------PFPKDCSQAMLNGDTTSGLYTIYLDKAQ--A--LEVFCDMTSDGGGWIVFLRRKNGREN---FYQNWKAYAAGFGDRR---------------EEFWLGLDNLNKITAQGQ--YELRVDLRD-HGETAFAVYDK-----FSVGDAKTRYKLKVEGYS-GTAGDSM-------------------AYHNGRSFSTFDK------------NCALSYKGAFWYRNCHRVNLMGRYGDNNHSQ----------GVNWFHWKGHEHSIQFAEMKLRPS----\n",
      ">Q15389\n",
      "--------------------------------------------------FRDCADVYQAGFNKSGIYTIYINNMP-EP--KKVFCNMDVNGGGWTVIQHREDGSLD---FQRGWKEYKMGFGNPS---------------GEYWLGNEFIFAITSQRQ--YMLRIELMDWEGNRAYSQYDR-----FHIGNEKQNYRLYLKGHT-GTAGKQS------------------SLILHGADFSTKDADNDNC-----MCKCALMLTGGWWFDACGPSNLNGMFYTAGQNH------GKLNGIKWHYFKGPSYSLRSTTMMIRPLD--F\n",
      ">O15123\n",
      "--------------------------------------------------FRDCAEVFKSGHTTNGIYTLTFPNST-EE--IKAYCDMEAGGGGWTIIQRREDGSVD---FQRTWKEYKVGFGNPS---------------GEYWLGNEFVSQLTNQQR--YVLKIHLKDWEGNEAYSLYEH-----FYLSSEELNYRIHLKGLT-GTAGKIS------------------SISQPGNDFSTKDGDNDKC-----ICKCSQMLTGGWWFDACGPSNLNGMYYPQRQNT------NKFNGIKWYYWKGSGYSLKATTMMIRPAD--F\n",
      ">Q9BY76\n",
      "--------------------------------------------------PRDCQELFQVGERQSGLFEIQ-PQGS-PP--FLVNCKMTSD-GGWTVIQRRHDGSVD---FNRPWEAYKAGFGDPH---------------GEFWLGLEKVHSITGDRN--SRLAVQLRDWDGNAELLQFS------VHLGGEDTAYSLQLTAPVAGQLGATTVPPSG---------------------FSTWDQDHDLR-----RDKCAKSLSGGWWFGTCSHSNLNGQYFPQRQ--------KLKKGIFWKTWRGRYYPLQATTMLIQPM----\n",
      ">P02679\n",
      "HDS---SIRYLQEIYNSNNQKIVNLKEKVAQLEAQCQEPCKDTVQIHDITGKDCQDIANKGAKQSGLYFIK-PLKANQQ--FLVYCEIDGSGNGWTVFQKRLDGSVD---FKKNWIQYKEGFGHLS-----------PTGTTEFWLGNEKIHLISTQSAIPYALRVELEDWNGRTSTADYAM-----FKVGPEADKYRLTYAYFAGGDAGDAF---DGFD--FGDDPSDKFFTSHKGMQFSTWDNDNDKF-----EGNCAEQDGSGWWMNKCHAGHLNGVYYQGGTYSKASTPNGYDNGIIWATWKTRWYSMKKTTMKIIPFNRLT\n",
      ">O93568\n",
      "HEN---TIQQLTDMHIMNSNKITQLKQKIAQLESHCQEPCKDTAEIQETTGRDCQDIANKGARKSGLYFIK-PQKAKQS--FLVYCEIDTYGNGWTVLQRRLDGSED---FRRNWVQYKEGFGHLS-----------PDDTTEFWLGNEKIHLITTQSTLPYALRIELEDWSGKKGTADYAV-----FKVGTEEDKYRLTYAYFIGGEAGDAF---DGFN--FGDDPSDKSYTYHNGMRFSTFDNDNDNF-----EGNCAEQDGSGWWMNRCHAGHLNGPYYIGGVYSRDTGTNSYDNGIIWATWRDRWYSMKKTTMKIIPFNRLS\n",
      ">P04115\n",
      "HDA---QIQELSEMWRVNQQFVTRLQQQLVDIRQTCSRSCQDTTKISPITGKDCQQVVDNGGKDSGLYYIK-PLKAKQP--FLVFCEIE-NGNGWTVIQHRHDGSVN---FTRDWVSYREGFGYLA-----------PTLTTEFWLGNEKIHLLTGQQA--YRLRIDLTDWENTHRYADYGH-----FKLTPESDEYRLFYSMYLDGDAGNAF---DGFD--FGDDPQDKFYTTHLGMLFSTPERDNDKY-----EGSCAEQDGSGWWMNRCHAGHLNGKYYFGGNYRKTDVEFPYDDGIIWATWHDRWYSLKMTTMKLLPMGRLS\n",
      ">P02675\n",
      "NSNIPTNLRVLRSILENLRSKIQKLESDVSAQMEYCRTPCTVSCNIPVVSGKECEEIIRKGGETSEMYLIQ-PDSSVKP--YRVYCDMNTENGGWTVIQNRQDGSVD---FGRKWDPYKQGFGNVATNTDGKNYCGLP---GEYWLGNDKISQLTRMGP--TELLIEMEDWKGDKVKAHYGG-----FTVQNEANKYQISVNKYR-GTAGNALM--DGASQLMGE---NRTMTIHNGMFFSTYDRDNDGWLTSDPRKQCSKEDGGGWWYNRCHAANPNGRYYWGGQYTWDMAKHGTDDGVVWMNWKGSWYSMRKMSMKIRPF---F\n",
      ">Q02020\n",
      "DNNIPSSLRVLRAVIDSLHKKIQKLENAIATQTDYCRSPCVASCNIPVVSGRECEDIYRKGGETSEMYIIQ-PDPFTTP--YRVYCDMETDNGGWTLIQNRQDGSVN---FGRAWDEYKRGFGNIA-KSGGKKYCDTP---GEYWLGNDKISQLTKIGP--TKVLIEMEDWNGDKVSALYGG-----FTIHNEGNKYQLSVSNYK-GNAGNALM--EGASQLYGE---NRTMTIHNGMYFSTYDRDNDGWLTTDPRKQCSKEDGGGWWYNRCHAANPNGRYYWGGTYSWDMAKHGTDDGIVWMNWKGSWYSMKKMSMKIKPY---F\n",
      ">P02678\n",
      "ESTVAGSLRSMKSVLEHLRAKMQRMEEAIKTQKELCSAPCTVNCRVPVVSGMHCEDIYRNGGRTSEAYYIQ-PDLFSEP--YKVFCDMESHGGGWTVVQNRVDGSSN---FARDWNTYKAEFGNIA-FGNGKSICNIP---GEYWLGTKTVHQLTKQHT--QQVLFDMSDWEGSSVYAQYAS-----FRPENEAQGYRLWVEDYS-GNAGNALL--EGATQLMGD---NRTMTIHNGMQFSTFDRDNDNWNPGDPTKHCSREDAGGWWYNRCHAANPNGRYYWGGIYTKEQADYGTDDGVVWMNWKGSWYSMRQMAMKLRPK----\n",
      ">P02671\n",
      "--------------------------------------------------------------------------------------------GGWLLIQQRMDGSLN---FNRTWQDYKRGFGSLN--DEGE---------GEFWLGNDYLHLLTQRGSV---LRVELEDWAGNEAYAEYH------FRVGSEAEGYALQVSSYE-GTAGDALI--EGSVEEGAE------YTSHNNMQFSTFDRDADQW-----EENCAEVYGGGWWYNNCQAANLNGIYYPGGSYDRNNSPYEIENGVVWVSFRGADYSLRAVRMKIRPL---V\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print('snamk')\n",
    "#print number of sequences in the final alignment   \n",
    "with open(infolder + 'alnAAfoldtree.fasta') as f:\n",
    "    print( f.read().count('>'))\n",
    "with open(infolder + 'alnAAfoldtree.fasta') as f:\n",
    "    print( f.read())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16\n",
      "finalaln ../../testcat/4.10.530/alnscratch/29_root.fasta.iter.iter\n",
      "16\n",
      "finalaln3di ../../testcat/4.10.530/alnscratch/29_inter3di.fasta.iter.iter\n",
      "16\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print( len(glob.glob(infolder + 'structs/*.pdb')) )\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "print('finalaln',finalaln)\n",
    "#print number of sequences in the final alignment   \n",
    "with open(finalaln) as f:\n",
    "    print( f.read().count('>'))\n",
    "\n",
    "print('finalaln3di',finalaln3di)\n",
    "with open(finalaln3di) as f:\n",
    "    print( f.read().count('>'))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "foldtree",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
