{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/projects/snake_tree/notebooks\n"
     ]
    }
   ],
   "source": [
    "cd projects/snake_tree/notebooks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "import glob\n",
    "import os\n",
    "import numpy as np\n",
    "import sys\n",
    "\n",
    "sys.path.append( '../src/')\n",
    "from matplotlib import pyplot as plt\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from compile_treeset import compile_folder , compare_treesets\n",
    "import plotly.express as px\n",
    "import plotly.figure_factory as ff\n",
    "import plotly.express as px\n",
    "from scipy.stats import wilcoxon, ks_2samp , describe\n",
    "import treescore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#sample_hogs = glob.glob('../OMA_data_unfiltered/OMA_data/LUCA/*/')\n",
    "#sample_hogs = sample_hogs[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "test directory already exists\n"
     ]
    }
   ],
   "source": [
    "#make a test directory and copy the files there\n",
    "over_write = False\n",
    "\n",
    "if not os.path.exists('test'):\n",
    "    os.makedirs('test')\n",
    "    for hog in sample_hogs:\n",
    "        os.system('cp -r '+hog+' test/')\n",
    "    print('copied files')\n",
    "else:\n",
    "    if over_write:\n",
    "        #clear files\n",
    "        os.system('rm -r test')\n",
    "        os.makedirs('test')\n",
    "        for hog in sample_hogs:\n",
    "            os.system('cp -r '+hog+' test/')\n",
    "        print('copied files')\n",
    "    else:\n",
    "        print('test directory already exists')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['test/HOG1077/', 'test/HOG748/', 'test/HOG458/', 'test/HOG411/', 'test/HOG569/', 'test/HOG1118/', 'test/HOG423/', 'test/HOG600/', 'test/HOG475/', 'test/HOG1231/']\n",
      "['test/HOG1231/3di_xaln_tree.nwk', 'test/HOG1231/fident_1_raw_struct_tree.PP.nwk.rooted', 'test/HOG1231/alntmscore_1_raw_struct_tree.PP.nwk', 'test/HOG1231/alntmscore_0_raw_treescores_struct_tree.json', 'test/HOG1231/alntmscore_0_exp_treescores_struct_tree.json', 'test/HOG1231/alnfident_dist_1_raw_.json', 'test/HOG1231/templateXfoldtree.nx.treefile.rooted', 'test/HOG1231/lddt_distmat.npy', 'test/HOG1231/templateXAA.nx.log', 'test/HOG1231/3di_xaln_tree.PP.nwk.rooted', 'test/HOG1231/logs', 'test/HOG1231/templateXfoldtree.nx.treefile.rooted.final', 'test/HOG1231/3di_xaln_tree.PP.nwk.rooted.final.treescore', 'test/HOG1231/templateX3di.nx.iqtree', 'test/HOG1231/alnfident_dist_3di_AA.json', 'test/HOG1231/lddt_1_exp_struct_tree.PP.nwk', 'test/HOG1231/sequences.aln.muscle.fst.log', 'test/HOG1231/rejected', 'test/HOG1231/fident_0_raw_struct_tree.PP.nwk', 'test/HOG1231/alntmscore_1_raw_struct_tree.PP.nwk.rooted', 'test/HOG1231/lddt_1_exp_struct_tree.nwk', 'test/HOG1231/sequences.aln.clustalo.fst', 'test/HOG1231/ML_astral_support.tre', 'test/HOG1231/lddt_0_raw_struct_tree.PP.nwk', 'test/HOG1231/lddt_0_raw_struct_tree.nwk', 'test/HOG1231/alnfident_dist_AA_AA.json', 'test/HOG1231/plddt.json', 'test/HOG1231/alntmscore_0_raw_struct_tree.PP.nwk.rooted', 'test/HOG1231/sequences.aln.clustalo.fst.nwk', 'test/HOG1231/alntmscore_0_exp_struct_tree.nwk', 'test/HOG1231/allvall_1.csv', 'test/HOG1231/outdb_ss', 'test/HOG1231/fident_0_exp_struct_tree.PP.nwk', 'test/HOG1231/alntmscore_1_exp_treescores_struct_tree.json', 'test/HOG1231/alnAA_3di.fasta.treefile.rooted.final.treescore', 'test/HOG1231/templateX3di.nx', 'test/HOG1231/templateXAA.nx.treefile', 'test/HOG1231/fident_1_raw_struct_tree.nwk', 'test/HOG1231/AA_xaln_fastmemat.txt', 'test/HOG1231/alnAA_foldtree.fasta', 'test/HOG1231/sequences.fst', 'test/HOG1231/alnAA_3di.fasta.treefile.rooted', 'test/HOG1231/sequences.aln.clustalo.fst.treefile.rooted', 'test/HOG1231/alnAA_3di.fasta.ckp.gz', 'test/HOG1231/fident_0_raw_fastmemat.txt', 'test/HOG1231/tmp', 'test/HOG1231/alntmscore_0_raw_struct_tree.PP.nwk', 'test/HOG1231/lddt_0_raw_fastmemat.txt', 'test/HOG1231/sequences.aln.clustalo.fst.iqtree', 'test/HOG1231/alntmscore_1_exp_struct_tree.nwk', 'test/HOG1231/aln3di_3di.fasta', 'test/HOG1231/alnfident_dist_AA_3dcoffee.json', 'test/HOG1231/outdb_ss.index', 'test/HOG1231/alnscratch', 'test/HOG1231/alnAA_3di.fasta.iqtree', 'test/HOG1231/AA_xaln_tree.PP.nwk.rooted', 'test/HOG1231/alntmscore_distmat.npy', 'test/HOG1231/alnfident_dist_0_raw_.json', 'test/HOG1231/fident_1_exp_struct_tree.nwk', 'test/HOG1231/sequences.aln.muscle.fst', 'test/HOG1231/outdb.dbtype', 'test/HOG1231/sequence_dataset.csv', 'test/HOG1231/templateXAA.nx.bionj', 'test/HOG1231/alnfident_dist_AA_foldtree.json', 'test/HOG1231/alnAA_3di.fasta.log', 'test/HOG1231/fident_0_raw_treescores_struct_tree.json', 'test/HOG1231/lddt_1_exp_struct_tree.PP.nwk.rooted', 'test/HOG1231/templateXAA.nx.mldist', 'test/HOG1231/ML_astral_support.tre.log', 'test/HOG1231/AA_xaln_tree.PP.nwk.rooted.final', 'test/HOG1231/alnAA_3di.fasta.treefile.rooted.final', 'test/HOG1231/outdb', 'test/HOG1231/outdb.lookup', 'test/HOG1231/lddt_0_exp_treescores_struct_tree.json', 'test/HOG1231/outdb.index', 'test/HOG1231/fident_1_exp_fastmemat.txt', 'test/HOG1231/templateXfoldtree.nx.best_model.nex', 'test/HOG1231/templateX3di.nx.treefile.rooted', 'test/HOG1231/sequences.aln.muscle.fst.ckp.gz', 'test/HOG1231/templateXAA.nx.best_model.nex', 'test/HOG1231/templateXAA.nx.ckp.gz', 'test/HOG1231/templateX3di.nx.log', 'test/HOG1231/alntmscore_1_exp_struct_tree.PP.nwk', 'test/HOG1231/aln3di_foldtree.fasta', 'test/HOG1231/treescores_sequences.clustalo.json', 'test/HOG1231/sequences.aln.clustalo.fst.ckp.gz', 'test/HOG1231/treescores_sequences.muscle.json', 'test/HOG1231/3di_xaln_fastmemat.txt', 'test/HOG1231/outdb_ca.dbtype', 'test/HOG1231/fident_0_exp_fastmemat.txt', 'test/HOG1231/lddt_0_raw_struct_tree.PP.nwk.rooted', 'test/HOG1231/lddt_1_exp_treescores_struct_tree.json', 'test/HOG1231/AA_xaln_tree.PP.nwk.rooted.final.treescore', 'test/HOG1231/fident_0_exp_treescores_struct_tree.json', 'test/HOG1231/lddt_0_exp_struct_tree.PP.nwk.rooted', 'test/HOG1231/alntmscore_0_raw_struct_tree.nwk', 'test/HOG1231/sequences.aln.muscle.fst.treefile.rooted', 'test/HOG1231/templateXfoldtree.nx', 'test/HOG1231/templateX3di.nx.bionj', 'test/HOG1231/3di_xaln_tree.PP.nwk.rooted.final', 'test/HOG1231/outdb_h.dbtype', 'test/HOG1231/templateX3di.nx.treefile.rooted.final.treescore', 'test/HOG1231/lddt_1_raw_fastmemat.txt', 'test/HOG1231/lddt_1_raw_struct_tree.nwk', 'test/HOG1231/templateXfoldtree.nx.mldist', 'test/HOG1231/astral_support.tre', 'test/HOG1231/sequences.aln.clustalo.fst.treefile', 'test/HOG1231/alnAA_3di.fasta.treefile', 'test/HOG1231/FT_astral_support.tre', 'test/HOG1231/lddt_0_exp_struct_tree.nwk', 'test/HOG1231/outdb_ss.dbtype', 'test/HOG1231/templateX3di.nx.treefile.rooted.final', 'test/HOG1231/alntmscore_0_raw_fastmemat.txt', 'test/HOG1231/speciesmap.txt', 'test/HOG1231/alntmscore_1_raw_fastmemat.txt', 'test/HOG1231/alnfident_dist_3di_foldtree.json', 'test/HOG1231/FT_astral_support.tre.log', 'test/HOG1231/fident_0_raw_struct_tree.PP.nwk.rooted', 'test/HOG1231/lddt_1_raw_struct_tree.PP.nwk', 'test/HOG1231/templateXfoldtree.nx.treefile.rooted.final.treescore', 'test/HOG1231/outdb_h.index', 'test/HOG1231/templateXAA.nx.treefile.rooted', 'test/HOG1231/sequences.aln.clustalo.fst.log', 'test/HOG1231/outdb_h', 'test/HOG1231/lddt_1_raw_treescores_struct_tree.json', 'test/HOG1231/templateXfoldtree.nx.treefile', 'test/HOG1231/finalset.csv', 'test/HOG1231/allvall_0.csv', 'test/HOG1231/alnfident_dist_3di_3di.json', 'test/HOG1231/lddt_1_exp_fastmemat.txt', 'test/HOG1231/tcoffee_templates.txt', 'test/HOG1231/alnAA_3di.fasta.mldist', 'test/HOG1231/alntmscore_1_exp_fastmemat.txt', 'test/HOG1231/sequences.aln.clustalo.fst.nwk.rooted', 'test/HOG1231/outdb_ca.index', 'test/HOG1231/sequences.aln.muscle.fst.mldist', 'test/HOG1231/fident_1_raw_treescores_struct_tree.json', 'test/HOG1231/templateXfoldtree.nx.log', 'test/HOG1231/fident_0_raw_struct_tree.nwk', 'test/HOG1231/sequences.aln.muscle.fst.nwk.rooted', 'test/HOG1231/lddt_0_exp_fastmemat.txt', 'test/HOG1231/fident_1_exp_treescores_struct_tree.json', 'test/HOG1231/templateXAA.nx.treefile.rooted.final', 'test/HOG1231/finalset_speciesmap.txt', 'test/HOG1231/lddt_1_raw_struct_tree.PP.nwk.rooted', 'test/HOG1231/templateX3di.nx.mldist', 'test/HOG1231/sequences.aln.muscle.fst.treefile', 'test/HOG1231/sequences.aln.muscle.fst.iqtree', 'test/HOG1231/outdb_ca', 'test/HOG1231/templateXfoldtree.nx.ckp.gz', 'test/HOG1231/fident_0_exp_struct_tree.nwk', 'test/HOG1231/alntmscore_1_exp_struct_tree.PP.nwk.rooted', 'test/HOG1231/alnfident_dist_AA_3di.json', 'test/HOG1231/3di_xaln_tree.PP.nwk', 'test/HOG1231/lddt_0_exp_struct_tree.PP.nwk', 'test/HOG1231/alntmscore_0_exp_struct_tree.PP.nwk.rooted', 'test/HOG1231/alnAA_AA.fasta', 'test/HOG1231/alntmscore_1_raw_struct_tree.nwk', 'test/HOG1231/templateXAA.nx.iqtree', 'test/HOG1231/treescores_sequences_iq.muscle.json', 'test/HOG1231/treescores_sequences_iq.clustalo.json', 'test/HOG1231/AA_xaln_tree.PP.nwk', 'test/HOG1231/alntmscore_1_raw_treescores_struct_tree.json', 'test/HOG1231/seq3di.fasta', 'test/HOG1231/templateX3di.nx.treefile', 'test/HOG1231/templateXAA.nx.treefile.rooted.final.treescore', 'test/HOG1231/alntmscore_0_exp_struct_tree.PP.nwk', 'test/HOG1231/fident_1_raw_struct_tree.PP.nwk', 'test/HOG1231/templateX3di.nx.best_model.nex', 'test/HOG1231/outdb.source', 'test/HOG1231/templateXAA.nx', 'test/HOG1231/fident_1_raw_fastmemat.txt', 'test/HOG1231/aln3di_AA.fasta', 'test/HOG1231/sequences.aln.muscle.fst.nwk', 'test/HOG1231/alnAA_3di.fasta', 'test/HOG1231/fident_distmat.npy', 'test/HOG1231/finalset_ncbi_tree.nwk', 'test/HOG1231/sequences.aln.muscle.fst.bionj', 'test/HOG1231/fident_1_exp_struct_tree.PP.nwk.rooted', 'test/HOG1231/identifiers.txt', 'test/HOG1231/fident_0_exp_struct_tree.PP.nwk.rooted', 'test/HOG1231/templateXfoldtree.nx.iqtree', 'test/HOG1231/sequences.aln.clustalo.fst.bionj', 'test/HOG1231/fident_1_exp_struct_tree.PP.nwk', 'test/HOG1231/sequences.aln.clustalo.fst.mldist', 'test/HOG1231/alntmscore_0_exp_fastmemat.txt', 'test/HOG1231/AA_xaln_tree.nwk', 'test/HOG1231/alnAA_3di.fasta.bionj', 'test/HOG1231/templateX3di.nx.ckp.gz', 'test/HOG1231/structs', 'test/HOG1231/templateXfoldtree.nx.bionj', 'test/HOG1231/lddt_0_raw_treescores_struct_tree.json']\n"
     ]
    }
   ],
   "source": [
    "testhogs = glob.glob('test/*/')\n",
    "print(testhogs)\n",
    "sample1 = testhogs[9]\n",
    "samplefiles = glob.glob(sample1+'*')\n",
    "print(samplefiles)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "sample1 = testhogs[9]\n",
    "treeML = sample1+'sequences.aln.muscle.fst.treefile.rooted'\n",
    "treeFT = sample1+'fident_1_raw_struct_tree.PP.nwk.rooted'\n",
    "finalset = sample1 +'finalset.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import toytree\n",
    "import ete3\n",
    "from ete3 import PhyloTree\n",
    "from ete3 import NCBITaxa\n",
    "\n",
    "ncbi = NCBITaxa()\n",
    "#ncbi.update_taxonomy_database()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def cut_tree(t ):\n",
    "    #select one of the children of a duplication node\n",
    "    recon_dups = t.search_nodes(evoltype=\"D\")\n",
    "    print(recon_dups)\n",
    "    for n in t.traverse():\n",
    "        if n in recon_dups:\n",
    "            delete = np.random.randint(0,2)\n",
    "            children = n.children\n",
    "            if len(children) == 1:\n",
    "                n = children[0]\n",
    "                cut_tree(t)\n",
    "            else:\n",
    "                if delete == 1:\n",
    "                    children[0].delete()\n",
    "                    n = children[0]\n",
    "                else:\n",
    "                    children[1].delete()\n",
    "                    n = children[0]\n",
    "                cut_tree(t)\n",
    "    return t\n",
    "\n",
    "def rf2species(recon_tree, ot_samples = 10):\n",
    "    for sample in range(ot_samples):    \n",
    "        ot = recon_tree.copy()\n",
    "\n",
    "        ot = cut_tree(ot )\n",
    "        #get the leaf names\n",
    "        leaves = ot.get_leaf_names()\n",
    "        lineages = ot.get_leaf_names()\n",
    "        lineages = [ int(x.split('_')[0]) for x in lineages]\n",
    "        species_tree = ncbi.get_topology(lineages, intermediate_nodes=False)\n",
    "        for l in ot.iter_leaves():\n",
    "            l.name = l.name.split('_')[0]\n",
    "        RFdist = ot.robinson_foulds(species_tree , unrooted_trees=True)\n",
    "        rfs.append(RFdist)\n",
    "    return rfs\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import copy\n",
    "\n",
    "def calc_scores(t , uniprot_csv):\n",
    "    uniprot_df = pd.read_csv(uniprot_csv)\n",
    "    tree = toytree.tree(t)\n",
    "    lineages = treescore.make_lineages(uniprot_df)\n",
    "    species = treescore.get_species(uniprot_df)\n",
    "    tree = treescore.label_leaves( tree , lineages , species)\n",
    "    #tree = treescore.labelwRED(tree.treenode)\n",
    "    overlap = treescore.getTaxOverlap(tree.treenode)\n",
    "    taxscore = tree.treenode.score\n",
    "\n",
    "    #calc descriptive stats on normalized branch lens to see if trees are balanced  \n",
    "    lengths = np.array([node.dist for node in tree.treenode.traverse()])\n",
    "    lengths /= np.sum(lengths)\n",
    "    #calc the root first taxscore\n",
    "\n",
    "    tree4 = copy.deepcopy(tree)\n",
    "    treescore.getTaxOverlap_root(tree4.treenode)\n",
    "    root_score , root_score_nr = treescore.sum_rootscore(tree4.treenode)\n",
    "\n",
    "    species_set = set()\n",
    "    #label the leaves with species and number\n",
    "    for l in tree.treenode.get_leaves():\n",
    "        if l.sp_num:\n",
    "            l.name = l.sp_num\n",
    "            species_set.add(l.sp_num.split('_')[0])\n",
    "\n",
    "    #get ncbi tree of species set\n",
    "    species_set = list(species_set)\n",
    "    species_set = [int(s) for s in species_set]\n",
    "    \n",
    "    ncbitree = ncbi.get_topology(species_set, intermediate_nodes=True)\n",
    "    ncbitree.write(outfile=uniprot_csv.replace('.csv','_ncbi_tree.nwk' ) , format=0)\n",
    "    #ncbitree = PhyloTree( ncbitree.write() , sp_naming_function=None)\n",
    "    etetree = PhyloTree( tree.treenode.write( format = 0) , sp_naming_function=None)\n",
    "\n",
    "    \n",
    "    #label the leaves with species\n",
    "    for l in etetree.get_leaves():\n",
    "        l.species   = l.name.split('_')[0]\n",
    "    \n",
    "    \n",
    "    recon_tree, events = etetree.reconcile(ncbitree)\n",
    "    recon_dups = recon_tree.search_nodes(evoltype=\"D\")\n",
    "    recon_losses = recon_tree.search_nodes(evoltype=\"L\")\n",
    "    recon_speciations = recon_tree.search_nodes(evoltype=\"S\")\n",
    "    print( 'algo 1' )\n",
    "    print('dups:', len(recon_dups))\n",
    "    print('losses:', len(recon_losses))\n",
    "    print('speciations:', len(recon_speciations))\n",
    "\n",
    "    #rfs = rf2species(etetree , ot_samples = 10)\n",
    "    #print('RFs:', rfs)\n",
    "\n",
    "\n",
    "    print( 'algo 2' )\n",
    "    events = etetree.get_descendant_evol_events()\n",
    "    dups = etetree.search_nodes(evoltype=\"D\")\n",
    "    losses = etetree.search_nodes(evoltype=\"L\")\n",
    "    speciations = etetree.search_nodes(evoltype=\"S\")    \n",
    "    print('dups:', len(dups))\n",
    "    print('losses:', len(losses))\n",
    "    print('speciations:', len(speciations))\n",
    "\n",
    "    scores = {}\n",
    "    #measure the distances of leaves to root\n",
    "    distances = np.array([ node.get_distance(tree.treenode) for node in tree.treenode.get_leaves() ])\n",
    "    distances_norm = distances / np.mean(distances)\n",
    "    scores[t] = {'score': taxscore, 'stats': describe(lengths) , 'ultrametricity':  describe(distances), \n",
    "                    'ultrametricity_norm':  describe(distances_norm) , 'root_score': root_score , 'root_score_nr': root_score_nr  , \n",
    "                    'RECON_speciations': len(recon_speciations) , 'RECON_dups': len(recon_dups) , 'RECON_losses': len(recon_losses) ,\n",
    "                    'SO_speciations':len(speciations) ,'SO_dups': len(dups) , 'SO_losses': len(losses)  }\n",
    "    return scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "test/HOG1077/\n",
      "ML\n",
      "root scoring\n",
      "root score 4623\n",
      "root score_nr 3909\n",
      "algo 1\n",
      "dups: 23\n",
      "losses: 3389\n",
      "speciations: 35\n",
      "algo 2\n",
      "dups: 12\n",
      "losses: 0\n",
      "speciations: 46\n",
      "{'test/HOG1077/sequences.aln.muscle.fst.treefile.rooted': {'score': 626, 'stats': DescribeResult(nobs=117, minmax=(2.991111384783379e-08, 0.0726053954572658), mean=0.00854700854700855, variance=0.00017982310920735504, skewness=2.671951856873734, kurtosis=7.301067678776121), 'ultrametricity': DescribeResult(nobs=59, minmax=(1.707441299, 4.3434943268), mean=2.139202316576271, variance=0.16713058024887115, skewness=3.168209235115583, kurtosis=13.302032524293164), 'ultrametricity_norm': DescribeResult(nobs=59, minmax=(0.7981672821543633, 2.030427086369106), mean=1.0, variance=0.03652180224239135, skewness=3.1682092351155826, kurtosis=13.302032524293164), 'root_score': 4623, 'root_score_nr': 3909, 'RECON_speciations': 35, 'RECON_dups': 23, 'RECON_losses': 3389, 'SO_speciations': 46, 'SO_dups': 12, 'SO_losses': 0}}\n",
      "FT\n",
      "root scoring\n",
      "root score 4673\n",
      "root score_nr 3871\n",
      "algo 1\n",
      "dups: 24\n",
      "losses: 3915\n",
      "speciations: 34\n",
      "algo 2\n",
      "dups: 11\n",
      "losses: 0\n",
      "speciations: 47\n",
      "{'test/HOG1077/fident_1_exp_struct_tree.PP.nwk.rooted': {'score': 598, 'stats': DescribeResult(nobs=117, minmax=(0.0, 0.07133576512676505), mean=0.008547008547008546, variance=0.00014689123631819967, skewness=2.5645555439636056, kurtosis=7.619932273280028), 'ultrametricity': DescribeResult(nobs=59, minmax=(0.4569437, 0.9896036999999999), mean=0.5861682576271187, variance=0.010348762225819373, skewness=2.2766300381634377, kurtosis=6.112437106906205), 'ultrametricity_norm': DescribeResult(nobs=59, minmax=(0.7795435765317017, 1.6882587672113765), mean=0.9999999999999998, variance=0.03011922655978452, skewness=2.2766300381634417, kurtosis=6.112437106906217), 'root_score': 4673, 'root_score_nr': 3871, 'RECON_speciations': 34, 'RECON_dups': 24, 'RECON_losses': 3915, 'SO_speciations': 47, 'SO_dups': 11, 'SO_losses': 0}}\n",
      "test/HOG748/\n",
      "ML\n",
      "root scoring\n",
      "root score 6616\n",
      "root score_nr 5772\n",
      "algo 1\n",
      "dups: 54\n",
      "losses: 3533\n",
      "speciations: 31\n",
      "algo 2\n",
      "dups: 34\n",
      "losses: 0\n",
      "speciations: 51\n",
      "{'test/HOG748/sequences.aln.muscle.fst.treefile.rooted': {'score': 1003, 'stats': DescribeResult(nobs=171, minmax=(1.9125109876945515e-08, 0.09134175578051827), mean=0.005847953216374269, variance=0.00010888839006397293, skewness=5.206253235378204, kurtosis=33.602964484779775), 'ultrametricity': DescribeResult(nobs=86, minmax=(0.9525909739999999, 7.9462350232), mean=1.998825808355814, variance=0.8719684500642272, skewness=4.20270040591009, kurtosis=21.945812908236356), 'ultrametricity_norm': DescribeResult(nobs=86, minmax=(0.47657528235718466, 3.9754514825563425), mean=1.0, variance=0.2182483026232015, skewness=4.202700405910089, kurtosis=21.94581290823635), 'root_score': 6616, 'root_score_nr': 5772, 'RECON_speciations': 31, 'RECON_dups': 54, 'RECON_losses': 3533, 'SO_speciations': 51, 'SO_dups': 34, 'SO_losses': 0}}\n",
      "FT\n",
      "root scoring\n",
      "root score 7953\n",
      "root score_nr 6787\n",
      "algo 1\n",
      "dups: 54\n",
      "losses: 3205\n",
      "speciations: 31\n",
      "algo 2\n",
      "dups: 34\n",
      "losses: 0\n",
      "speciations: 51\n",
      "{'test/HOG748/fident_1_exp_struct_tree.PP.nwk.rooted': {'score': 958, 'stats': DescribeResult(nobs=171, minmax=(0.0, 0.057140310317597275), mean=0.005847953216374269, variance=6.69726033637575e-05, skewness=3.6643521602802407, kurtosis=17.65390190804322), 'ultrametricity': DescribeResult(nobs=86, minmax=(0.73846, 1.2600900000000004), mean=0.8871412790697676, variance=0.007445972764227089, skewness=1.5781169293579047, kurtosis=5.008626918398431), 'ultrametricity_norm': DescribeResult(nobs=86, minmax=(0.8324040571918029, 1.420393830981799), mean=1.0, variance=0.00946097441583391, skewness=1.5781169293579047, kurtosis=5.008626918398427), 'root_score': 7953, 'root_score_nr': 6787, 'RECON_speciations': 31, 'RECON_dups': 54, 'RECON_losses': 3205, 'SO_speciations': 51, 'SO_dups': 34, 'SO_losses': 0}}\n",
      "test/HOG458/\n",
      "ML\n",
      "root scoring\n",
      "root score 210\n",
      "root score_nr 137\n",
      "algo 1\n",
      "dups: 8\n",
      "losses: 533\n",
      "speciations: 5\n",
      "algo 2\n",
      "dups: 2\n",
      "losses: 0\n",
      "speciations: 11\n",
      "{'test/HOG458/sequences.aln.muscle.fst.treefile.rooted': {'score': 21, 'stats': DescribeResult(nobs=27, minmax=(0.0059972505397157955, 0.12917520785904119), mean=0.037037037037037035, variance=0.0010250521549931163, skewness=1.387734270543774, kurtosis=1.2422408747653684), 'ultrametricity': DescribeResult(nobs=14, minmax=(0.604292819, 1.5848094895), mean=1.0697201607857143, variance=0.1183996493911888, skewness=0.12495370583822032, kurtosis=-1.42843304737579), 'ultrametricity_norm': DescribeResult(nobs=14, minmax=(0.5649073852699421, 1.4815178283037596), mean=1.0, variance=0.10346895299451978, skewness=0.12495370583822002, kurtosis=-1.4284330473757896), 'root_score': 210, 'root_score_nr': 137, 'RECON_speciations': 5, 'RECON_dups': 8, 'RECON_losses': 533, 'SO_speciations': 11, 'SO_dups': 2, 'SO_losses': 0}}\n",
      "FT\n",
      "root scoring\n",
      "root score 212\n",
      "root score_nr 139\n",
      "algo 1\n",
      "dups: 7\n",
      "losses: 464\n",
      "speciations: 6\n",
      "algo 2\n",
      "dups: 2\n",
      "losses: 0\n",
      "speciations: 11\n",
      "{'test/HOG458/fident_1_exp_struct_tree.PP.nwk.rooted': {'score': 22, 'stats': DescribeResult(nobs=27, minmax=(0.0020370674052665864, 0.15044810969472575), mean=0.037037037037037035, variance=0.0011947176526748778, skewness=1.480904809914163, kurtosis=2.412827715279657), 'ultrametricity': DescribeResult(nobs=14, minmax=(0.361757, 0.749123), mean=0.5451411428571429, variance=0.01493732250982417, skewness=0.08642599732840896, kurtosis=-0.9460798378033166), 'ultrametricity_norm': DescribeResult(nobs=14, minmax=(0.6636024536764791, 1.3741817322276695), mean=0.9999999999999999, variance=0.050263744287054025, skewness=0.08642599732841093, kurtosis=-0.946079837803318), 'root_score': 212, 'root_score_nr': 139, 'RECON_speciations': 6, 'RECON_dups': 7, 'RECON_losses': 464, 'SO_speciations': 11, 'SO_dups': 2, 'SO_losses': 0}}\n",
      "test/HOG411/\n",
      "ML\n",
      "root scoring\n",
      "root score 336\n",
      "root score_nr 251\n",
      "algo 1\n",
      "dups: 5\n",
      "losses: 444\n",
      "speciations: 8\n",
      "algo 2\n",
      "dups: 2\n",
      "losses: 0\n",
      "speciations: 11\n",
      "{'test/HOG411/sequences.aln.muscle.fst.treefile.rooted': {'score': 54, 'stats': DescribeResult(nobs=27, minmax=(0.005678702670629255, 0.16112529017674118), mean=0.037037037037037035, variance=0.0011975955114128232, skewness=2.054807242386695, kurtosis=4.335583884812022), 'ultrametricity': DescribeResult(nobs=14, minmax=(1.075574145, 2.309023222), mean=1.6438420627714285, variance=0.12742360012560752, skewness=-0.01053943993687412, kurtosis=-0.9213459652548708), 'ultrametricity_norm': DescribeResult(nobs=14, minmax=(0.6543050390051709, 1.404650285020151), mean=1.0, variance=0.04715521106673023, skewness=-0.01053943993687383, kurtosis=-0.9213459652548717), 'root_score': 336, 'root_score_nr': 251, 'RECON_speciations': 8, 'RECON_dups': 5, 'RECON_losses': 444, 'SO_speciations': 11, 'SO_dups': 2, 'SO_losses': 0}}\n",
      "FT\n",
      "root scoring\n",
      "root score 349\n",
      "root score_nr 250\n",
      "algo 1\n",
      "dups: 6\n",
      "losses: 538\n",
      "speciations: 7\n",
      "algo 2\n",
      "dups: 3\n",
      "losses: 0\n",
      "speciations: 10\n",
      "{'test/HOG411/fident_1_exp_struct_tree.PP.nwk.rooted': {'score': 53, 'stats': DescribeResult(nobs=27, minmax=(0.003051918200287693, 0.14829534500912014), mean=0.03703703703703703, variance=0.0012416591125058737, skewness=1.6687288961414521, kurtosis=2.4332409649753775), 'ultrametricity': DescribeResult(nobs=14, minmax=(0.6298699999999999, 0.9568599999999999), mean=0.8092278571428569, variance=0.0072481254027472524, skewness=-0.44485648130286065, kurtosis=-0.010259685885915193), 'ultrametricity_norm': DescribeResult(nobs=14, minmax=(0.7783592648724226, 1.1824358140343663), mean=1.0000000000000002, variance=0.011068379703795931, skewness=-0.44485648130286815, kurtosis=-0.010259685885909864), 'root_score': 349, 'root_score_nr': 250, 'RECON_speciations': 7, 'RECON_dups': 6, 'RECON_losses': 538, 'SO_speciations': 10, 'SO_dups': 3, 'SO_losses': 0}}\n",
      "test/HOG569/\n",
      "ML\n",
      "root scoring\n",
      "root score 425\n",
      "root score_nr 308\n",
      "algo 1\n",
      "dups: 8\n",
      "losses: 818\n",
      "speciations: 8\n",
      "algo 2\n",
      "dups: 4\n",
      "losses: 0\n",
      "speciations: 12\n",
      "{'test/HOG569/sequences.aln.muscle.fst.treefile.rooted': {'score': 52, 'stats': DescribeResult(nobs=33, minmax=(0.003361628522695759, 0.09593448630499905), mean=0.030303030303030304, variance=0.0006593417285188652, skewness=0.9201777951803927, kurtosis=-0.06713012741668489), 'ultrametricity': DescribeResult(nobs=17, minmax=(1.6127939939999998, 3.0933647998000002), mean=2.0386776204, variance=0.20256399627188992, skewness=1.469604517496269, kurtosis=1.0863596465705472), 'ultrametricity_norm': DescribeResult(nobs=17, minmax=(0.7910981009756513, 1.5173388714558336), mean=1.0, variance=0.04873771291201753, skewness=1.4696045174962689, kurtosis=1.0863596465705472), 'root_score': 425, 'root_score_nr': 308, 'RECON_speciations': 8, 'RECON_dups': 8, 'RECON_losses': 818, 'SO_speciations': 12, 'SO_dups': 4, 'SO_losses': 0}}\n",
      "FT\n",
      "root scoring\n",
      "root score 433\n",
      "root score_nr 318\n",
      "algo 1\n",
      "dups: 7\n",
      "losses: 719\n",
      "speciations: 9\n",
      "algo 2\n",
      "dups: 4\n",
      "losses: 0\n",
      "speciations: 12\n",
      "{'test/HOG569/fident_1_exp_struct_tree.PP.nwk.rooted': {'score': 57, 'stats': DescribeResult(nobs=33, minmax=(0.002255773138403261, 0.09854841146366365), mean=0.030303030303030297, variance=0.0007575236243871292, skewness=0.7012775717788547, kurtosis=-0.6900859142596749), 'ultrametricity': DescribeResult(nobs=17, minmax=(0.678257, 0.942907), mean=0.7847736470588235, variance=0.0068310408586176475, skewness=0.4516655632836373, kurtosis=-0.9400370504198956), 'ultrametricity_norm': DescribeResult(nobs=17, minmax=(0.864270866563847, 1.2015018643067708), mean=1.0, variance=0.011091698637581651, skewness=0.4516655632836379, kurtosis=-0.9400370504198943), 'root_score': 433, 'root_score_nr': 318, 'RECON_speciations': 9, 'RECON_dups': 7, 'RECON_losses': 719, 'SO_speciations': 12, 'SO_dups': 4, 'SO_losses': 0}}\n",
      "test/HOG1118/\n",
      "ML\n",
      "root scoring\n",
      "root score 836\n",
      "root score_nr 790\n",
      "algo 1\n",
      "dups: 1\n",
      "losses: 75\n",
      "speciations: 7\n",
      "algo 2\n",
      "dups: 0\n",
      "losses: 0\n",
      "speciations: 8\n",
      "{'test/HOG1118/sequences.aln.muscle.fst.treefile.rooted': {'score': 162, 'stats': DescribeResult(nobs=17, minmax=(1.5390224572589244e-05, 0.23155828229766695), mean=0.058823529411764705, variance=0.004419010525093272, skewness=1.2785122221755565, kurtosis=0.7359287320910739), 'ultrametricity': DescribeResult(nobs=9, minmax=(1.334570371, 1.669072913), mean=1.458943301811111, variance=0.011166778933836965, skewness=0.8470046051674676, kurtosis=-0.31431205732288126), 'ultrametricity_norm': DescribeResult(nobs=9, minmax=(0.9147513610318397, 1.1440286342368733), mean=0.9999999999999998, variance=0.005246275458214987, skewness=0.8470046051674766, kurtosis=-0.3143120573228728), 'root_score': 836, 'root_score_nr': 790, 'RECON_speciations': 7, 'RECON_dups': 1, 'RECON_losses': 75, 'SO_speciations': 8, 'SO_dups': 0, 'SO_losses': 0}}\n",
      "FT\n",
      "root scoring\n",
      "root score 836\n",
      "root score_nr 790\n",
      "algo 1\n",
      "dups: 1\n",
      "losses: 75\n",
      "speciations: 7\n",
      "algo 2\n",
      "dups: 0\n",
      "losses: 0\n",
      "speciations: 8\n",
      "{'test/HOG1118/fident_1_exp_struct_tree.PP.nwk.rooted': {'score': 162, 'stats': DescribeResult(nobs=17, minmax=(0.002060253184566543, 0.26345948651746076), mean=0.0588235294117647, variance=0.0048551412992872515, skewness=1.808971750676322, kurtosis=2.703781973743851), 'ultrametricity': DescribeResult(nobs=9, minmax=(0.60451, 0.73939), mean=0.6889555555555555, variance=0.0017057356777777782, skewness=-0.7988485888645112, kurtosis=0.10350518422868404), 'ultrametricity_norm': DescribeResult(nobs=9, minmax=(0.8774296035867497, 1.0732042060445763), mean=1.0, variance=0.0035935965725235586, skewness=-0.798848588864512, kurtosis=0.10350518422868404), 'root_score': 836, 'root_score_nr': 790, 'RECON_speciations': 7, 'RECON_dups': 1, 'RECON_losses': 75, 'SO_speciations': 8, 'SO_dups': 0, 'SO_losses': 0}}\n",
      "test/HOG423/\n",
      "ML\n",
      "root scoring\n",
      "root score 240\n",
      "root score_nr 179\n",
      "algo 1\n",
      "dups: 8\n",
      "losses: 157\n",
      "speciations: 3\n",
      "algo 2\n",
      "dups: 6\n",
      "losses: 0\n",
      "speciations: 5\n",
      "{'test/HOG423/sequences.aln.muscle.fst.treefile.rooted': {'score': 43, 'stats': DescribeResult(nobs=23, minmax=(0.0004167458616684485, 0.16473503273103732), mean=0.043478260869565216, variance=0.0023578561748345372, skewness=1.320286326394552, kurtosis=0.6294874007240647), 'ultrametricity': DescribeResult(nobs=12, minmax=(0.5381114336, 1.2939798146), mean=0.8063970663750001, variance=0.05577957634881248, skewness=1.0904986394378136, kurtosis=0.20001420017812466), 'ultrametricity_norm': DescribeResult(nobs=12, minmax=(0.6673033125219248, 1.6046435044919403), mean=0.9999999999999999, variance=0.08577827989177653, skewness=1.0904986394378147, kurtosis=0.2000142001781251), 'root_score': 240, 'root_score_nr': 179, 'RECON_speciations': 3, 'RECON_dups': 8, 'RECON_losses': 157, 'SO_speciations': 5, 'SO_dups': 6, 'SO_losses': 0}}\n",
      "FT\n",
      "root scoring\n",
      "root score 256\n",
      "root score_nr 181\n",
      "algo 1\n",
      "dups: 8\n",
      "losses: 173\n",
      "speciations: 3\n",
      "algo 2\n",
      "dups: 6\n",
      "losses: 0\n",
      "speciations: 5\n",
      "{'test/HOG423/fident_1_exp_struct_tree.PP.nwk.rooted': {'score': 45, 'stats': DescribeResult(nobs=23, minmax=(0.0009906195772604292, 0.22565366224611144), mean=0.043478260869565216, variance=0.0028129716542458004, skewness=1.9740377141085685, kurtosis=4.120324865590082), 'ultrametricity': DescribeResult(nobs=12, minmax=(0.48671999999999993, 0.72573), mean=0.5704983333333334, variance=0.004241401033333332, skewness=0.9737180997321918, kurtosis=0.7280437510250191), 'ultrametricity_norm': DescribeResult(nobs=12, minmax=(0.8531488552405935, 1.2720983701383874), mean=0.9999999999999999, variance=0.013031685123965523, skewness=0.9737180997321944, kurtosis=0.7280437510250235), 'root_score': 256, 'root_score_nr': 181, 'RECON_speciations': 3, 'RECON_dups': 8, 'RECON_losses': 173, 'SO_speciations': 5, 'SO_dups': 6, 'SO_losses': 0}}\n",
      "test/HOG600/\n",
      "ML\n",
      "root scoring\n",
      "root score 13827\n",
      "root score_nr 12348\n",
      "algo 1\n",
      "dups: 56\n",
      "losses: 4684\n",
      "speciations: 73\n",
      "algo 2\n",
      "dups: 36\n",
      "losses: 0\n",
      "speciations: 93\n",
      "{'test/HOG600/sequences.aln.muscle.fst.treefile.rooted': {'score': 1895, 'stats': DescribeResult(nobs=259, minmax=(1.2916577498312821e-08, 0.020572011462258736), mean=0.003861003861003861, variance=2.3078508839278938e-05, skewness=1.5880757869388469, kurtosis=1.6831168098554787), 'ultrametricity': DescribeResult(nobs=130, minmax=(1.275871982, 3.1839201508999997), mean=1.7774851215307692, variance=0.1478616734515561, skewness=1.3781026127479368, kurtosis=1.3896034599152651), 'ultrametricity_norm': DescribeResult(nobs=130, minmax=(0.7177961528596198, 1.7912499589071156), mean=1.0000000000000002, variance=0.04679976462119233, skewness=1.3781026127479332, kurtosis=1.389603459915258), 'root_score': 13827, 'root_score_nr': 12348, 'RECON_speciations': 73, 'RECON_dups': 56, 'RECON_losses': 4684, 'SO_speciations': 93, 'SO_dups': 36, 'SO_losses': 0}}\n",
      "FT\n",
      "root scoring\n",
      "root score 15405\n",
      "root score_nr 13779\n",
      "algo 1\n",
      "dups: 64\n",
      "losses: 4731\n",
      "speciations: 65\n",
      "algo 2\n",
      "dups: 36\n",
      "losses: 0\n",
      "speciations: 93\n",
      "{'test/HOG600/fident_1_exp_struct_tree.PP.nwk.rooted': {'score': 1928, 'stats': DescribeResult(nobs=259, minmax=(0.0, 0.024571624442377697), mean=0.003861003861003861, variance=2.370881365937881e-05, skewness=1.5544723817295334, kurtosis=1.6468058728302681), 'ultrametricity': DescribeResult(nobs=130, minmax=(0.583806, 1.098896), mean=0.739332, variance=0.00900331998108527, skewness=1.3539509133377343, kurtosis=2.1316342960849974), 'ultrametricity_norm': DescribeResult(nobs=130, minmax=(0.7896398370420867, 1.4863363144027313), mean=1.0, variance=0.01647114072284735, skewness=1.3539509133377339, kurtosis=2.1316342960849957), 'root_score': 15405, 'root_score_nr': 13779, 'RECON_speciations': 65, 'RECON_dups': 64, 'RECON_losses': 4731, 'SO_speciations': 93, 'SO_dups': 36, 'SO_losses': 0}}\n",
      "test/HOG475/\n",
      "ML\n",
      "root scoring\n",
      "root score 3021\n",
      "root score_nr 2558\n",
      "algo 1\n",
      "dups: 21\n",
      "losses: 4116\n",
      "speciations: 25\n",
      "algo 2\n",
      "dups: 9\n",
      "losses: 0\n",
      "speciations: 37\n",
      "{'test/HOG475/sequences.aln.muscle.fst.treefile.rooted': {'score': 436, 'stats': DescribeResult(nobs=93, minmax=(2.83314223416679e-08, 0.06526093844657502), mean=0.010752688172043013, variance=0.0002025415621498512, skewness=2.039858919600559, kurtosis=3.7144237617457776), 'ultrametricity': DescribeResult(nobs=47, minmax=(1.2635825576, 3.6384618846000003), mean=2.1773751068319154, variance=0.22456956710107748, skewness=0.9664233186756652, kurtosis=1.1969229565548734), 'ultrametricity_norm': DescribeResult(nobs=47, minmax=(0.5803237823539348, 1.6710312675035441), mean=0.9999999999999999, variance=0.04736792875478408, skewness=0.9664233186756679, kurtosis=1.1969229565548796), 'root_score': 3021, 'root_score_nr': 2558, 'RECON_speciations': 25, 'RECON_dups': 21, 'RECON_losses': 4116, 'SO_speciations': 37, 'SO_dups': 9, 'SO_losses': 0}}\n",
      "FT\n",
      "root scoring\n",
      "root score 2813\n",
      "root score_nr 2311\n",
      "algo 1\n",
      "dups: 24\n",
      "losses: 4640\n",
      "speciations: 22\n",
      "algo 2\n",
      "dups: 10\n",
      "losses: 0\n",
      "speciations: 36\n",
      "{'test/HOG475/fident_1_exp_struct_tree.PP.nwk.rooted': {'score': 410, 'stats': DescribeResult(nobs=93, minmax=(0.0, 0.06337203886226911), mean=0.010752688172043012, variance=0.00017913048749718363, skewness=1.7112878741330697, kurtosis=2.5118921591565115), 'ultrametricity': DescribeResult(nobs=47, minmax=(0.550548, 0.977178), mean=0.7271045957446808, variance=0.0097543836359852, skewness=0.6826055302582074, kurtosis=-0.18999809100181642), 'ultrametricity_norm': DescribeResult(nobs=47, minmax=(0.7571785451804822, 1.3439304409831172), mean=1.0000000000000002, variance=0.018450411339301725, skewness=0.682605530258202, kurtosis=-0.18999809100182086), 'root_score': 2813, 'root_score_nr': 2311, 'RECON_speciations': 22, 'RECON_dups': 24, 'RECON_losses': 4640, 'SO_speciations': 36, 'SO_dups': 10, 'SO_losses': 0}}\n",
      "test/HOG1231/\n",
      "ML\n",
      "root scoring\n",
      "root score 744\n",
      "root score_nr 639\n",
      "algo 1\n",
      "dups: 11\n",
      "losses: 384\n",
      "speciations: 6\n",
      "algo 2\n",
      "dups: 9\n",
      "losses: 0\n",
      "speciations: 8\n",
      "{'test/HOG1231/sequences.aln.muscle.fst.treefile.rooted': {'score': 143, 'stats': DescribeResult(nobs=35, minmax=(0.0034049105971058024, 0.11241875978306243), mean=0.028571428571428567, variance=0.0006097868621712291, skewness=1.470273612667945, kurtosis=2.1881883511687947), 'ultrametricity': DescribeResult(nobs=18, minmax=(0.8511997070000001, 2.3324786239999997), mean=1.7005136213111112, variance=0.15804953265854188, skewness=-0.2016442979455516, kurtosis=-0.4860690607399696), 'ultrametricity_norm': DescribeResult(nobs=18, minmax=(0.5005544773841433, 1.3716318380335217), mean=1.0, variance=0.05465538853086497, skewness=-0.20164429794555142, kurtosis=-0.48606906073996825), 'root_score': 744, 'root_score_nr': 639, 'RECON_speciations': 6, 'RECON_dups': 11, 'RECON_losses': 384, 'SO_speciations': 8, 'SO_dups': 9, 'SO_losses': 0}}\n",
      "FT\n",
      "root scoring\n",
      "root score 740\n",
      "root score_nr 623\n",
      "algo 1\n",
      "dups: 10\n",
      "losses: 316\n",
      "speciations: 7\n",
      "algo 2\n",
      "dups: 9\n",
      "losses: 0\n",
      "speciations: 8\n",
      "{'test/HOG1231/fident_1_exp_struct_tree.PP.nwk.rooted': {'score': 139, 'stats': DescribeResult(nobs=35, minmax=(0.0018279944472697708, 0.10511756453535198), mean=0.028571428571428567, variance=0.0005956428217754908, skewness=1.1872276646870867, kurtosis=1.2098931797194448), 'ultrametricity': DescribeResult(nobs=18, minmax=(0.535658, 0.9676979999999998), mean=0.7868008888888888, variance=0.008528468154928102, skewness=-0.75885563116389, kurtosis=1.6255664880170304), 'ultrametricity_norm': DescribeResult(nobs=18, minmax=(0.6808050264870062, 1.2299147264139112), mean=1.0, variance=0.01377657777588684, skewness=-0.7588556311638898, kurtosis=1.6255664880170304), 'root_score': 740, 'root_score_nr': 623, 'RECON_speciations': 7, 'RECON_dups': 10, 'RECON_losses': 316, 'SO_speciations': 8, 'SO_dups': 9, 'SO_losses': 0}}\n"
     ]
    }
   ],
   "source": [
    "for sample in testhogs:\n",
    "    print( sample)\n",
    "    treeML = sample+'sequences.aln.muscle.fst.treefile.rooted'\n",
    "    treeFT = sample+'fident_1_exp_struct_tree.PP.nwk.rooted'\n",
    "    finalset = sample +'finalset.csv'\n",
    "\n",
    "    print('ML')\n",
    "    scores = calc_scores(treeML , finalset)\n",
    "    print(scores)\n",
    "    print('FT')\n",
    "    scores = calc_scores(treeFT , finalset)\n",
    "    print(scores)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "#running astral\n",
    "def run_astral( phylo , st,  mapping, output ,root  ,outfmt = '2', astralpath = '/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro3' ):\n",
    "    logfile = output+'.log'\n",
    "    cmd = astralpath + ' -c {st} -a {mapping} -u {outfmt} -i {phylo} -o {output} -C -T -E --root {root} 2>{logfile}'.format(phylo = phylo , st=st, mapping = mapping , output = output , root=root, logfile = logfile , outfmt = outfmt)\n",
    "    print(cmd)\n",
    "    subprocess.run(cmd , shell = True)\n",
    "    return output , logfile\n",
    "\n",
    "def prepare_astral_input(uniprot_df , output):\n",
    "    finalset = pd.read_csv(uniprot_df)\n",
    "    finalset['species'] = finalset['Taxonomic lineage (Ids)'].map(lambda x: x.split(',')[-1].split('(')[0].strip())\n",
    "    finalset['species'] = finalset['species'].map(lambda x: x.split('_')[0])\n",
    "    mapper = dict(zip(finalset['Entry'], finalset['species']))\n",
    "    with open(uniprot_df.replace('.csv','_speciesmap.txt' ) , 'w') as f:\n",
    "        for k,v in mapper.items():\n",
    "            f.write('{}\\t{}\\n'.format(k,int(v)))\n",
    "    \n",
    "    #get ncbi tree of species set\n",
    "    species_set = list(finalset.species.unique())\n",
    "    species_set = [int(s) for s in species_set]\n",
    "    st = ncbi.get_topology(species_set, intermediate_nodes=False)\n",
    "    st.name = 'root'\n",
    "    #write with internal node names\n",
    "\n",
    "    st.write(outfile=uniprot_df.replace('.csv','_ncbi_tree.nwk' ) , format= 1 )\n",
    "    return mapper , uniprot_df.replace('.csv','_speciesmap.txt' )  , uniprot_df.replace('.csv','_ncbi_tree.nwk' ) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FT\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro3 -c test/HOG1077/finalset_ncbi_tree.nwk -a test/HOG1077/finalset_speciesmap.txt -u 2 -i test/HOG1077/fident_1_exp_struct_tree.PP.nwk.rooted -o test/HOG1077/FT_astral_support.tre -C -T -E --root root 2>test/HOG1077/FT_astral_support.tre.log\n",
      "{'mean': 0.32970871428571424, 'std': 0.0005210136075714289, 'skewness': -1.015143816812902, 'nancount': 0, 'duploss': 14}\n",
      "ML\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro3 -c test/HOG1077/finalset_ncbi_tree.nwk -a test/HOG1077/finalset_speciesmap.txt -u 2 -i test/HOG1077/sequences.aln.muscle.fst.treefile.rooted -o test/HOG1077/ML_astral_support.tre -C -T -E --root root 2>test/HOG1077/ML_astral_support.tre.log\n",
      "{'mean': 0.32563671428571433, 'std': 0.003455933421904763, 'skewness': -1.5878599953867245, 'nancount': 0, 'duploss': 15}\n",
      "FT\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro3 -c test/HOG748/finalset_ncbi_tree.nwk -a test/HOG748/finalset_speciesmap.txt -u 2 -i test/HOG748/fident_1_exp_struct_tree.PP.nwk.rooted -o test/HOG748/FT_astral_support.tre -C -T -E --root root 2>test/HOG748/FT_astral_support.tre.log\n",
      "{'mean': 0.2675749166666667, 'std': 0.005004633700992423, 'skewness': 0.4872296582648715, 'nancount': 0, 'duploss': 40}\n",
      "ML\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro3 -c test/HOG748/finalset_ncbi_tree.nwk -a test/HOG748/finalset_speciesmap.txt -u 2 -i test/HOG748/sequences.aln.muscle.fst.treefile.rooted -o test/HOG748/ML_astral_support.tre -C -T -E --root root 2>test/HOG748/ML_astral_support.tre.log\n",
      "{'mean': 0.29459808333333337, 'std': 0.002034144492446969, 'skewness': 0.24254937271079152, 'nancount': 0, 'duploss': 40}\n",
      "FT\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro3 -c test/HOG458/finalset_ncbi_tree.nwk -a test/HOG458/finalset_speciesmap.txt -u 2 -i test/HOG458/fident_1_exp_struct_tree.PP.nwk.rooted -o test/HOG458/FT_astral_support.tre -C -T -E --root root 2>test/HOG458/FT_astral_support.tre.log\n",
      "{'mean': 0.460360375, 'std': 0.02908595994398214, 'skewness': -0.3519592730688657, 'nancount': 0, 'duploss': 4}\n",
      "ML\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro3 -c test/HOG458/finalset_ncbi_tree.nwk -a test/HOG458/finalset_speciesmap.txt -u 2 -i test/HOG458/sequences.aln.muscle.fst.treefile.rooted -o test/HOG458/ML_astral_support.tre -C -T -E --root root 2>test/HOG458/ML_astral_support.tre.log\n",
      "{'mean': 0.460360375, 'std': 0.02908595994398214, 'skewness': -0.3519592730688657, 'nancount': 0, 'duploss': 4}\n",
      "FT\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro3 -c test/HOG411/finalset_ncbi_tree.nwk -a test/HOG411/finalset_speciesmap.txt -u 2 -i test/HOG411/fident_1_exp_struct_tree.PP.nwk.rooted -o test/HOG411/FT_astral_support.tre -C -T -E --root root 2>test/HOG411/FT_astral_support.tre.log\n",
      "{'mean': 0.46231988888888886, 'std': 0.02979182637861111, 'skewness': -0.5617962545307191, 'nancount': 0, 'duploss': 3}\n",
      "ML\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro3 -c test/HOG411/finalset_ncbi_tree.nwk -a test/HOG411/finalset_speciesmap.txt -u 2 -i test/HOG411/sequences.aln.muscle.fst.treefile.rooted -o test/HOG411/ML_astral_support.tre -C -T -E --root root 2>test/HOG411/ML_astral_support.tre.log\n",
      "{'mean': 0.46801066666666663, 'std': 0.028891600208250007, 'skewness': -0.3186609887808574, 'nancount': 0, 'duploss': 3}\n",
      "FT\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro3 -c test/HOG569/finalset_ncbi_tree.nwk -a test/HOG569/finalset_speciesmap.txt -u 2 -i test/HOG569/fident_1_exp_struct_tree.PP.nwk.rooted -o test/HOG569/FT_astral_support.tre -C -T -E --root root 2>test/HOG569/FT_astral_support.tre.log\n",
      "{'mean': 0.35153566666666664, 'std': 0.0009940112213333331, 'skewness': 0.7071067811865496, 'nancount': 1, 'duploss': 5}\n",
      "ML\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro3 -c test/HOG569/finalset_ncbi_tree.nwk -a test/HOG569/finalset_speciesmap.txt -u 2 -i test/HOG569/sequences.aln.muscle.fst.treefile.rooted -o test/HOG569/ML_astral_support.tre -C -T -E --root root 2>test/HOG569/ML_astral_support.tre.log\n",
      "{'mean': 0.34420599999999996, 'std': 0.0003546663870000002, 'skewness': 0.7071067811865547, 'nancount': 1, 'duploss': 7}\n",
      "FT\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro3 -c test/HOG1118/finalset_ncbi_tree.nwk -a test/HOG1118/finalset_speciesmap.txt -u 2 -i test/HOG1118/fident_1_exp_struct_tree.PP.nwk.rooted -o test/HOG1118/FT_astral_support.tre -C -T -E --root root 2>test/HOG1118/FT_astral_support.tre.log\n",
      "{'mean': 0.40173300000000006, 'std': 0.014389046470000005, 'skewness': 0.8384740648867577, 'nancount': 0, 'duploss': 0}\n",
      "ML\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro3 -c test/HOG1118/finalset_ncbi_tree.nwk -a test/HOG1118/finalset_speciesmap.txt -u 2 -i test/HOG1118/sequences.aln.muscle.fst.treefile.rooted -o test/HOG1118/ML_astral_support.tre -C -T -E --root root 2>test/HOG1118/ML_astral_support.tre.log\n",
      "{'mean': 0.40173300000000006, 'std': 0.014389046470000005, 'skewness': 0.8384740648867577, 'nancount': 0, 'duploss': 0}\n",
      "FT\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro3 -c test/HOG423/finalset_ncbi_tree.nwk -a test/HOG423/finalset_speciesmap.txt -u 2 -i test/HOG423/fident_1_exp_struct_tree.PP.nwk.rooted -o test/HOG423/FT_astral_support.tre -C -T -E --root root 2>test/HOG423/FT_astral_support.tre.log\n",
      "{'mean': 0.41379450000000007, 'std': 0.029109875161666677, 'skewness': -0.1134483388432332, 'nancount': 0, 'duploss': 8}\n",
      "ML\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro3 -c test/HOG423/finalset_ncbi_tree.nwk -a test/HOG423/finalset_speciesmap.txt -u 2 -i test/HOG423/sequences.aln.muscle.fst.treefile.rooted -o test/HOG423/ML_astral_support.tre -C -T -E --root root 2>test/HOG423/ML_astral_support.tre.log\n",
      "{'mean': 0.36687349999999996, 'std': 0.02563271625500001, 'skewness': 0.7606699006515871, 'nancount': 0, 'duploss': 8}\n",
      "FT\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro3 -c test/HOG600/finalset_ncbi_tree.nwk -a test/HOG600/finalset_speciesmap.txt -u 2 -i test/HOG600/fident_1_exp_struct_tree.PP.nwk.rooted -o test/HOG600/FT_astral_support.tre -C -T -E --root root 2>test/HOG600/FT_astral_support.tre.log\n",
      "{'mean': 0.303561, 'std': 0.002659115951999999, 'skewness': -0.7071067811865495, 'nancount': 0, 'duploss': 55}\n",
      "ML\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro3 -c test/HOG600/finalset_ncbi_tree.nwk -a test/HOG600/finalset_speciesmap.txt -u 2 -i test/HOG600/sequences.aln.muscle.fst.treefile.rooted -o test/HOG600/ML_astral_support.tre -C -T -E --root root 2>test/HOG600/ML_astral_support.tre.log\n",
      "{'mean': 0.303561, 'std': 0.002659115951999999, 'skewness': -0.7071067811865495, 'nancount': 0, 'duploss': 58}\n",
      "FT\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro3 -c test/HOG475/finalset_ncbi_tree.nwk -a test/HOG475/finalset_speciesmap.txt -u 2 -i test/HOG475/fident_1_exp_struct_tree.PP.nwk.rooted -o test/HOG475/FT_astral_support.tre -C -T -E --root root 2>test/HOG475/FT_astral_support.tre.log\n",
      "{'mean': 0.3463514285714286, 'std': 0.009002415384952384, 'skewness': 0.8119145818310718, 'nancount': 0, 'duploss': 16}\n",
      "ML\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro3 -c test/HOG475/finalset_ncbi_tree.nwk -a test/HOG475/finalset_speciesmap.txt -u 2 -i test/HOG475/sequences.aln.muscle.fst.treefile.rooted -o test/HOG475/ML_astral_support.tre -C -T -E --root root 2>test/HOG475/ML_astral_support.tre.log\n",
      "{'mean': 0.33979662499999996, 'std': 0.00010860256426785712, 'skewness': 1.7984503858595897, 'nancount': 1, 'duploss': 14}\n",
      "FT\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro3 -c test/HOG1231/finalset_ncbi_tree.nwk -a test/HOG1231/finalset_speciesmap.txt -u 2 -i test/HOG1231/fident_1_exp_struct_tree.PP.nwk.rooted -o test/HOG1231/FT_astral_support.tre -C -T -E --root root 2>test/HOG1231/FT_astral_support.tre.log\n",
      "{'mean': 0.40351, 'std': 0.02171362458233333, 'skewness': 0.04708330200246432, 'nancount': 0, 'duploss': 11}\n",
      "ML\n",
      "/work/FAC/FBM/DBC/cdessim2/default/dmoi/software/ASTER-Linux/bin/astral-pro3 -c test/HOG1231/finalset_ncbi_tree.nwk -a test/HOG1231/finalset_speciesmap.txt -u 2 -i test/HOG1231/sequences.aln.muscle.fst.treefile.rooted -o test/HOG1231/ML_astral_support.tre -C -T -E --root root 2>test/HOG1231/ML_astral_support.tre.log\n",
      "{'mean': 0.46620614285714285, 'std': 0.01756194869847619, 'skewness': 0.033794526796969405, 'nancount': 0, 'duploss': 11}\n"
     ]
    }
   ],
   "source": [
    "from Bio import Phylo\n",
    "import re\n",
    "from scipy.stats import describe\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "q1,q2,q3: these three values show quartet support (as defined in the description of -t 1) for the main topology, the first alternative, and the second alternative, respectively.\n",
    "f1, f2, f3: these three values show the total number of quartet trees in all the gene trees that support the main topology, the first alternative, and the second alternative, respectively.\n",
    "pp1, pp2, pp3: these three show the local posterior probabilities (as defined in the description of -t 4) for the main topology, the first alternative, and the second alternative, respectively.\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "def print_tree(tree):\n",
    "    tree = ete3.Tree(tree)\n",
    "    print(tree)\n",
    "\n",
    "def parse_astral(string , index = 0):\n",
    "    #'[pp1=0.284784;pp2=0.426968;pp3=0.288248;f1=0.000000;f2=0.263158;f3=0.008772;q1=0.000000;q2=0.967742;q3=0.032258]'\n",
    "    string = string.replace('[','').replace(']','')\n",
    "    string = string.split(';')\n",
    "    string = {s.split('=')[0]:float(s.split('=')[1]) for s in string}\n",
    "    #to dataframe\n",
    "    string = pd.DataFrame(string , index = [index])\n",
    "    return string\n",
    "\n",
    "def retastral_support( astral_file):\n",
    "    with open(astral_file , 'r') as f:\n",
    "        nstring = f.read()\n",
    "    #find all quoted strings with single quotes\n",
    "    quoted = re.findall(r\"'(.*?)'\", nstring)\n",
    "    dfs = [ parse_astral(q,i) for i,q in enumerate(quoted)]\n",
    "    dfs = pd.concat(dfs)\n",
    "    #make sure pp1 does not contain nan\n",
    "    return dfs\n",
    "\n",
    "def verify_map(mapping, tree):\n",
    "    t = ete3.Tree(tree)\n",
    "    for l in t.get_leaves():\n",
    "        if l.name not in mapping.keys():\n",
    "            print('missing', l.name)\n",
    "            return False\n",
    "    return True \n",
    "\n",
    "def return_astral_score(astrolout, logfile):\n",
    "    #get dup loss and speciation events from astral logfile\n",
    "    with open(logfile , 'r') as f:\n",
    "        for l in f:\n",
    "            if '#Duploss' in l:\n",
    "                duploss = int(l.split(' ')[-1])\n",
    "    astraldf = retastral_support(astrolout)\n",
    "    #count nans \n",
    "    nancount = astraldf['pp1'].isna().sum()\n",
    "    #get summary stats\n",
    "    astral_stas = describe(astraldf['pp1'].dropna())\n",
    "    return { 'mean': astral_stas.mean , 'std': astral_stas.variance ,'skewness':astral_stas.skewness, 'nancount': nancount , 'duploss': duploss }\n",
    "\n",
    "for sample in testhogs:\n",
    "    speciesmap, speciesmap_file , ncbitree = prepare_astral_input(sample+'finalset.csv' , sample+'speciesmap.txt')\n",
    "    FT = sample+'fident_1_exp_struct_tree.PP.nwk.rooted'\n",
    "    ML = sample+'sequences.aln.muscle.fst.treefile.rooted'\n",
    "    t = PhyloTree(ncbitree , sp_naming_function=None)\n",
    "    #find root node name\n",
    "    root = t.get_tree_root()\n",
    "    root.name = 'root'\n",
    "    root.write(outfile=ncbitree , quoted_node_names = False, format=1)\n",
    "    #open ncbi tree\n",
    "    #t = PhyloTree(ncbitree )\n",
    "    #find root node name\n",
    "    #t.write(outfile=ncbitree , quoted_node_names = False, format=1)\n",
    "    print('FT')\n",
    "    output , log = run_astral(  FT , ncbitree, speciesmap_file , sample+'FT_astral_support.tre' , 'root')\n",
    "    print( return_astral_score(output, log))\n",
    "    print('ML')\n",
    "    output , log = run_astral( ML , ncbitree,speciesmap_file , sample+'ML_astral_support.tre', 'root')\n",
    "    print( return_astral_score(output, log))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mFailed to connect to the remote Jupyter Server 'http://localhost:38699/'. Verify the server is running and reachable. (Failed to connect to the remote Jupyter Server 'http://localhost:38699/'. Verify the server is running and reachable. (request to http://localhost:38699/api/kernels?1725285606960 failed, reason: read ECONNRESET).)."
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
